---
title: "CCAT_survival analysis"
author: "Masachika Ikegami"
date: "08/28/2023"
output:
  html_document: default
  pdf_document:
    latex_engine: xelatex
mainfont: Hiragino Kaku Gothic Pro
---
```{r report_setting, include=FALSE, warning=FALSE}
# Output source codes in the report?
# 1: Yes, 0: No.
Source_code = 0
```

```{r setting_up, echo = (Source_code == 1), include=FALSE, warning=FALSE}
## Setting up
# Create directories under the Working directory, and download files from C-CAT and https://github.com/MANO-B/CCAT  

# Change working directory
if(file.exists("~/Desktop/length_bias/Ver.20230220/")){
  Working = "~/Desktop/length_bias/Ver.20230220"
} else{
  Working = "please set your own working directory"
}

# Select the organs of interest
# Cancer = c("Adrenal", "Ampulla_of_Vater", "Biliary", "Bladder", "Bone",
#            "Breast", "Cervix", "Colorectal", "CNS", "Esophagus_Stomach",
#            "Eye", "Head_Neck", "Kidney", "Liver", "Lung", "Other",
#            "Ovary", "Pancreas", "Penis", "Peripheral_Nervous_System",
#            "Peritoneum", "Pleura", "Prostate", "Skin", "Soft_tissue",
#            "Testis", "Thymus", "Thyroid",
#            "Uterus_carcinoma", "Uterus_sarcoma", "Vagina", "NEC")
#
## "Lymph" and "Penis" are excluded because patients < 10. 

## Examples
#Cancer = c("Adrenal")
#Cancer = c("Ampulla_of_Vater")
#Cancer = c("Biliary")
#Cancer = c("Bladder")
#Cancer = c("Bone", "Soft_tissue")
#Cancer = c("Bone")
#Cancer = c("Breast")
#Cancer = c("Cervix")
#Cancer = c("CNS")
#Cancer = c("Colorectal")
#Cancer = c("Esophagus_Stomach")
#Cancer = c("Eye")
#Cancer = c("Head_Neck")
#Cancer = c("Kidney")
#Cancer = c("Liver")
#Cancer = c("Lung")
#Cancer = c("Other")
#Cancer = c("Ovary")
#Cancer = c("Pleura")
#Cancer = c("Peripheral_Nervous_System")
#Cancer = c("Peritoneum")
Cancer = c("Skin")
#Cancer = c("Soft_tissue")
#Cancer = c("Pancreas")
#Cancer = c("Prostate")
#Cancer = c("Testis")
#Cancer = c("Thymus")
#Cancer = c("Thyroid")
#Cancer = c("Uterus_carcinoma", "Uterus_sarcoma")
#Cancer = c("Vagina")

#Cancer = c("Cervix", "Ovary", "Uterus_carcinoma", "Uterus_sarcoma")


# Choose genes of interest
# Patients with the 1st target_gene mutation listed will be analyzed in "survival_gene" chunk.
# If no corresponding gene is available, the most frequent gene can be used.
# See following explanations.
#
## Examples
## target_gene = list(c("BRCA1", "F"), c("BRCA2", "F"))
## gene_pattern = list(c(1, 2))
#target_gene = list(c("EGFR", "F"), c("CCNE1", "F"), c("SMARCA4", "F"), c("KEAP1", "F"), c("STK11", "F"))
#gene_pattern = list(c(1), c(2), c(3), c(4), c(8), c(16))
#target_gene = list(c("Gene_of_interest", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("TP53", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("TERT", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("PTPN11", "F"))
#gene_pattern = list(c(1))
# target_gene = list(c("MDM2", "F"), c("CDK4", "F"))
# gene_pattern = list(c(1,2))
#target_gene = list(c("BRAF", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("AKT1", "E17K"))
#gene_pattern = list(c(1))
#target_gene = list(c("FGFR2", "F"))
#gene_pattern = list(c(1))
target_gene = list(c("CDKN2A", "F"))
gene_pattern = list(c(1))
#target_gene = list(c("JUN", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("PTPN11", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("EGFR", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("ERBB2", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("BRAF", "F"), c("MET", "F"))
#gene_pattern = list(c(1), c(2), c(3))
#target_gene = list(c("MSI", "F"), c("TMB", "F"), c("NTRK1", "F"), c("NTRK2", "F"), c("NTRK3", "F"))
#gene_pattern = list(c(1), c(2), c(1,2), c(4,8,16))
#target_gene = list(c("TMB", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("VEGFA", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("RB1", "F"))
#gene_pattern = list(c(1))
#target_gene = list(c("TP53", "F"), c("RB1", "F"), c("ERBB2", "F"), c("EGFR", "F"), c("MSI", "F"), c("PTEN", "F"), c("PIK3CA", "F"), c("ATRX", "F"))
#gene_pattern = list(c(1), c(2), c(4), c(8), c(16), c(32), c(64), c(128))
#target_gene = list(c("MSI", "F"), c("VEGFA", "F"), c("EGFR", "F"))
#gene_pattern = list(c(3), c(4))
#target_gene = list(c("TP53", "F"), c("CDKN2A", "F"), c("TERT", "F"), c("KRAS", "F"))
#gene_pattern = list(c(1,2,4,8))
#target_gene = list(c("TP53", "F"), c("KMT2D", "F"), c("CDKN2A", "F"), c("KDR", "F"), c("MYC", "F"))
#gene_pattern = list(c(1,2,4,8,16))
#target_gene = list(c("TP53", "F"), c("BCL6", "F"), c("STK11", "F"), c("CCNE1", "F"))
#gene_pattern = list(c(1,2,4,8))
#target_gene = list(c("EGFR", "F"), c("ALK", "F"), c("ROS1", "F"), c("BRAF", "F"), c("MET", "F"), c("NTRK1", "F"), c("NTRK2", "F"), c("NTRK3", "F"))
#gene_pattern = list(c(1,2,4,8,16,32,64,128))

# Select a gene to be analyzed double-hit in prognosis?
double_hit_gene = "TP53"

# Choose genes with mutation not annotated as oncogenic
Analyze_VUS = NULL
#Analyze_VUS = c("PTPN11", "IDH1", "IDH2")
#Analyze_VUS = c("PTPN11")

# Similar drugs are considered the same drug
# 1: Yes, 0: No.
Drug_mix = 0

# Choose drugs of interest
# If no corresponding drug is available, the most frequent drug will be used.
# See following explanations.


# Drugs for sarcoma
#selected_drug = list("Doxorubicin")
#selected_drug = list("Pazopanib")
#selected_drug = list("Eribulin")
#selected_drug = list("Trabectedin")
#selected_drug = list("Gemcitabine", "Docetaxel")
#selected_drug = list("Ifosfamide")
#selected_drug = list("Doxorubicin", "Ifosfamide")

#selected_drug = list("Temozolomide")
#selected_drug = list("Bevacizumab")
#selected_drug = list("Gemcitabine", "Cisplatin")
#selected_drug = list("Gemcitabine")
selected_drug = list("Docetaxel")
#selected_drug = list("Cabazitaxel")
#selected_drug = list("Atezolizumab")
#selected_drug = list("Ipilimumab")
#selected_drug = list("Durvalumab")
#selected_drug = list("Pembrolizumab")
#selected_drug = list("Pemigatinib")
#selected_drug = list("Tegafur")

## drugs for breast cancer
#selected_drug = list("Eribulin")
#selected_drug = list("Paclitaxel")
#selected_drug = list("Carboplatin")

## drugs for stomach cancer
#selected_drug = list("Nivolumab")
#selected_drug = list("Platinum")
#selected_drug = list("Taxane")
#selected_drug = list("F_pyrimidine")
#selected_drug = list("TAS-102")
#selected_drug = list("Irinotecan")
#selected_drug = list("Trastuzumab")
#selected_drug = list("trastuzumab Deruxtecan")
#selected_drug = list("Ramucirumab")
#selected_drug = list("Investigational Agent")
#selected_drug = list("Olaparib")

drug_pattern = list(c(1))
#drug_pattern = list(c(3))

# Choose pathways of interest (MAX 5 pathways)
# See the following pathway list.
## Examples　
#target_pathway = c("frequent_genes") # most frequent genes
#target_pathway = c("TP53_genes")
#target_pathway = c("BRAF_MET_mut")
#target_pathway = c("Breast_mut")
#target_pathway = c("PTPN11_mut")
target_pathway = c("RTK", "DNA_damage_control", "PI3K_AKT_mTOR", "Cell_cycle", "TP53_genes")
#target_pathway = c("Epigenetic_modification", "Transcriptional_regulation", "MAPK", "MYC_genes", "DNA_damage_response")
#target_pathway = c("Epigenetic_modification", "Transcriptional_regulation", "WNT", "MYC_genes")

#target_pathway = c("ERBB2_mut", "MSI_mut", "TMB_mut", "MSI_TMB_mut")
#target_pathway = c("TP53_mut", "CCNE1_mut", "ARID1A_mut", "FGFR2_mut", "CDH1_mut")
#target_pathway = c("KRAS_mut", "APC_mut")
#target_pathway = c("EGFR_mut")
#target_pathway = c("VEGFA_mut")
#target_pathway = c("Biliary_mut")

#target_pathway = c("DNA_damage_response", "TP53_genes", "DNA_damage_control", "Cell_cycle", "PI3K_AKT_mTOR")
#target_pathway = c("TP53_genes", "Cell_cycle", "RB1_mut", "TP53_mut", "MDM2_CDK4_mut")
#target_pathway = c("TP53_genes", "Cell_cycle", "RB1_mut", "TP53_mut", "ERBB2_mut")
#target_pathway = c("RB1_mut", "TP53_mut", "PTEN_mut")
#target_pathway = c("TP53_genes")
#target_pathway = c("DNA_damage_response", "RTK", "DNA_damage_control", "Cell_cycle", "PI3K_AKT_mTOR")
#target_pathway = c("DNA_damage_response", "TP53_genes", "DNA_damage_control", "Prostate_driver", "PI3K_AKT_mTOR")
#target_pathway = c("DNA_damage_response")
#target_pathway = c("frequent_genes")
#target_pathway = c("URS_1")
#target_pathway = c("CNS_PTPN11")
#target_pathway = c("URS_1", "URS_2", "URS_3", "URS_4", "URS_5")
#target_pathway = c("URS_1")
#target_pathway = c("STS_1", "STS_2", "STS_3", "STS_4", "STS_5")
#target_pathway = c("STS_2")
#target_pathway = c("STS_1", "URS_2", "URS_3", "URS_4", "URS_5")

# Entry period
# 1: 2019-06-01 to 2021-06-30
# 2: 2019-06-01 to 2021-12-31
# 3: 2019-06-01 to 2022-06-30
# 4: 2019-06-01 to 2023-02-20
Entry_period = 4

# survival rate at k-year
k_year = 2

# target gene number (must be <= 30)
gene_number = 30

# target drug number
drug_number = 1

# Definition of neuroendocrine tumor
# Small cell carcinoma is a neuroendocrine carcinoma?
# 1: Yes, 0: No.
SCC_is_NEC = 1

# Analyze all cancer types?
# 1: Yes, 0: No.
All_cancer_type = 0

# Combine resembling diseases?
# 1: Yes, 0: No.
Disease_simple = 1

# Diseases to analyze selected automatically?
# 1: Yes, 0: No.
Auto_disease_selection = 1

# Exclude patients with unknown PS?
# 1: Yes, 0: No.
Ex_PS9 = 0

# Exclude patients with multi-cancer?
# 1: Yes, 0: No.
Ex_multi = 0

# select only active multi-cancer?
# exclude inactive multi-cancer?
# 1: Yes, 0: No.
Ex_inactive_multi = 0

# Analyze survival from CGP test?
# 1: Yes, 0: No.
Analyze_enroll = 1

# Analyze survival from the first palliative chemotherapy, Bayesian?
# 2: 2nd method, 1: 1st method, 0: No.
Analyze_chemo = 2

# Analyze survival from the diagnosis, Bayesian?
# 2: 2nd method, 1: 1st method, 0: No.
Analyze_diagnosis = 0

# Analyze mutation-based clustering and drug efficiency?
# 1: Yes, 0: No.
Analyze_clustering = 1

# Analyze accuracy of Kosugi's criteria for PGPV?
# 1: Yes, 0: No.
PGPV_analysis = 0

# Is SD regareded as effective?
# 2: Only long-SD is effective, 1: SD is effective, 0: No.
SD_is_effective = 2

# Is NE regareded as ineffective?
# 1: NE is ineffective, 0: Omit NE.
Evaluate_NE_as_ineffective = 0

# Is long-SD regareded as effective?
LongSD_days = 180

# Analyze with genes shared by NOP and F1?
# 1: Yes, 0: No.
Only_shared_genes = 0

# Perform subanalysis?
# 1: Yes, 0: No.
Subanalysis = 0

# Depict lolliplot in the drug effect analysis?
# 1: Yes, 0: No.
Figure_lolliplot = 0
lolliplot_gene = "TP53"

# Analyze only ER positive breast cancer?
# 1: Yes, 0: No.
ER_positive_BRCA = 0

# Analyze only patients with BRCA1/2 pathogenic mutation?
# 1: Yes, 0: No.
BRCA_mutated = 0

# Analyze only EGFR oncogenic mutation positive lung adenocarcinoma?
# 1: Yes, 0: No.
EGFR_positive_LUAD = 0

# Analyze only EGFR oncogenic mutation other than amplification?
# 1: Yes, 0: No.
EGFR_non_amp = 0

# Analyze Stomach cancer classified as differentiated or not?
# 1: Yes, 0: No.
STAD_diff = 1

# Analyze cervical cancer considering HPV infection?
# 1: Yes, 0: No.
Cervix_HPV = 0

# Analyze cervical cancer classified as adeno/SCC/NEC?
# 1: Yes, 0: No.
Cervix_analysis_adeno_SCC_NEC = 1

# Analyze cervical adenocarcinoma classified as gastric type or not?
# 1: Yes, 0: No.
Cervix_analysis_gastric = 0

# Analyze only IDH wild-type glioblastoma patients?
# 1: Yes, 0: No.
CNS_Glioblastoma_IDH = 0

# Analyze patients with ultra-rare sarcoma?
# 1: Yes, 0: No.
Ultra_rare_sarcoma = 0

# Combine Ewing sarcomas of bone and soft tissue?
# 1: Yes, 0: No.
EWS_soft_bone_combine = 1

#Perihilar Cholangiocarcinoma is a subset of Extrahepatic Cholangiocarcinoma?
# 1: Yes, 0: No.
Perihilar_is_Extrahepatic = 1

# Analyze relation between drug duration and mutation?
# 1: Yes, 0: No.
Drug_analysis = 0

# Bisect patients using a specific age as a threshold?
# If not, use median age as a threshold.
# 1: Yes, 0: No.
# If Define_median_age == 1, define Median_age
Define_median_age = 0
Median_age = 39

# Analyze only patitents treated with designated drug?
# 1: Yes, 0: No.
Drug_treated_cases_only = 0

# Select oncopanels
# "NCC oncopanel", "FoundationOne CDx", "FoundationOne Liquid CDx"
#Panel = c("NCC OncoPanel", "FoundationOne CDx", "F1Liquid CDx")
#Panel = c("NCC OncoPanel", "FoundationOne CDx")
#Panel = c("FoundationOne CDx", "F1Liquid CDx")
Panel = c("FoundationOne CDx")
#Panel = c("NCC OncoPanel")
#Panel = c("F1Liquid CDx")

# Select diseases manually, if Auto_disease_selection == 0
# See the disease list below.
#Diseases = c("Lung Adenocarcinoma (LUAD)")
#Diseases = c("Lung Squamous Cell Carcinoma (LUSC)")

# Use only organ-based classification and do not use classification by histology
# 1: Yes, 0: No.
No_histology = 0

# Exclude patients underwent CGP before "time_10" days
time_10 = 0


## Pathways and genes based on KEGG (Kyoto Encyclopedia of Genes and Genomes)
## You can add genesets as you like.
PATH = NULL
PATH$Cell_cycle = c("BTG2", "CCND1", "CCND2", "CCND3", "CCNE1", "CDK4", "CDK6", "CDKN1A", "CDKN1B", "CDKN2A", "CDKN2B", "CDKN2C", "RB1", "SKP2")
PATH$Cell_division = c("AURKA", "AURKB", "KNSTRN", "RAD21", "STAG2")
PATH$Apoptosis = c("BAX", "BCL2", "BCL2L1", "BCL2L11", "BCL2L2", "CASP8", "DAXX", "FAS", "MCL1")
PATH$GPCR = c("CXCR4", "GNA11", "GNA13", "GNAQ", "GNAS", "GRM3", "P2RY8", "S1PR3", "TSHR")
PATH$Hippo = c("NF2")
PATH$JAK_STAT = c("CALR", "CRLF2", "CSF3R", "JAK1", "JAK2", "JAK3", "MPL", "JAK1", "JAK2", "JAK3", "MPL", "PIM1", "SOCS1", "STAT3")
PATH$MAPK = c("ARAF", "BRAF", "CIC", "HRAS", "KRAS", "LZTR1", "MAP2K1", "MAP2K4", "MAP3K1", "MAP3K13", "MAP3K4", "MAPK1", "MKNK1", "NF1", "NRAS", "PTPN11", "RAC1", "RAC2", "RAF1", "RHOA", "RIT1", "RAF1", "RHOA", "RIT1", "RRAS2", "SHOC2", "SOS1", "SRC")
PATH$MYC_genes = c("MAX", "MXI1", "MYC", "MYCL", "MYCN")
PATH$NFKB = c("BCL10","CARD11","MYD88","NFKBIA","REL","TNFAIP3")
PATH$Neuclear_receptor = c("AR","ESR1","PPARG","RARA")
PATH$PI3K_AKT_mTOR = c("AKT1","AKT2","AKT3","CRKL","EPAS1","FLCN","INPP4B","MTOR","PIK3C2B","PIK3C2G","PIK3CA","PIK3CB","PIK3CG","PIK3R1","PIK3R2","PPP2R1A","PPP2R2A","PRKAR1A","PTEN","RHEB","RICTOR","RPTOR","SGK1","STK11","TMEM127","TSC1","TSC2","VHL")
PATH$RTK = c("ABL1","ABL2","ALK","AXL","CBL","DDR1","DDR2","EGFR","EPHA3","EPHB1","EPHB4","ERBB2","ERBB3","ERBB4","ERRFI1","FGF10","FGF12","FGF14","FGF19","FGF23","FGF3","FGF4","FGF6","FGFR1","FGFR2","FGFR3","FGFR4", "FLT1","FLT3","HGF","IGF1R","IGF2","IRS2","KDR","KIT","MET","MST1R","NRG1","NTRK1","NTRK2","NTRK3","PDGFRA","PDGFRB","PTPRK","PTPRO","PTPRT","RET","ROS1","TEK","TNK2","TYRO3","VEGFA")
PATH$TGFB = c("ACVR1B","BMPR1A","ENG","SMAD2","SMAD4","TGFBR1","TGFBR2")
PATH$Hedgehog = c("PRKCI","PTCH1","SMO","SUFU")
PATH$NOTCH = c("HDAC2","HEY1","MAML2","NOTCH1","NOTCH2","NOTCH3")
PATH$WNT = c("AMER1","APC","AXIN1","CDH1","CTNNA1","CTNNB1","FAT1","GSK3B","LRP5","RNF43","RSPO2","RSPO3","SOX9","TCF7L2")
PATH$DNA_damage_response = c("BARD1","BLM","BRCA1","BRCA2","BRIP1","ERCC2","ERCC4","FANCA","FANCC","FANCL","MLH1","MRE11","MSH2","MSH3","MSH6","NBN","PALB2","PARP1","PMS1","PMS2","POLE","RAD50","RAD51","RAD52","UBE2T","XPA","XRCC2")
PATH$DNA_damage_control = c("FANCG","MUTYH","PARP2","PARP3","POLD1","POLH","RAD51B","RAD51C","RAD51D","RAD54L","RECQL4","SLX4","TERC","TERT")
PATH$TP53_genes = c("ATM","ATR","CHEK1","CHEK2","MDM2","MDM4","TP53")
PATH$Immune = c("ALOX12B","B2M","BCL6","BTK","CD22","CD274","CD70","CD74","CD79A","CD79B","CSF1R","CTLA4","IKBKE","IL7R","IRF2","IRF4","LYN","PDCD1","PDCD1LG2","SH2D1A","SYK","TNFRSF14")
PATH$Drug_metabolism = c("ABCB1","ABCG2","ALDH2","CDA","COMT","CYP1A2","CYP2A6","CYP2B6","CYP2C19","CYP2C9","CYP2D6","CYP2E1","CYP3A4","CYP3A5","DPYD","MTHFR","NAT2","TPMT","UGT1A1")
PATH$Metabolic_pathway = c("ADH1B","CYP17A1","CYP3A43","EXT1","EXT2","FH","G6PD","GALNT12","HSD3B1","MTAP","MTRR","NT5C2","PDE11A","PDGFB","PDK1","SDHA","SDHAF2","SDHB","SDHC","SDHD","SLC22A18","TIPARP")
PATH$KEAP1_NRF2 = c("CUL3","KEAP1","NFE2L2")
PATH$Protein_homeostasis = c("BAP1","CUL4A","CYLD","ENO1","FBXW7","MKRN1","PRKN","RBBP6","SPOP","TRAF7")
PATH$Epigenetic_modification = c("ARID1A","ARID1B","ARID2","ASXL1","ATRX","BCOR","BCORL1","CREBBP","CTCF","DNMT1","DNMT3A","DOT1L","EED","EP300","EZH2","H3","HDAC1","IDH1","IDH2","KDM5A","KDM5C", "KDM6A","KMT2A","KMT2C","KMT2D","MEN1","NCOA2","NCOA3","NCOR1","NPM1","PBRM1","PPP6C","SETBP1","SETD2","SMARCA4","SMARCB1","TET2","NSD2","NSD3")
PATH$RNA_metabolism = c("BTG1","DICER1","DIS3","EIF3E","EWSR1","TENT5C","QKI","RBM10","SF3B1","TACC3","U2AF1","XPO1")
PATH$Transcriptional_regulation = c("ATF1","BRD4","CBFB","CDC73","CDK8","CDK12","CEBPA","CRTC3","DDIT3","ERG","ETV4","ETV5","ETV6","FOXL2","FUBP1","FUS","GATA3","GATA4","GATA6","HMGA2","HNF1A","HOXB13","ID3","IKZF1","JUN","KLF4","LMO1", "MAF","MED12","MEF2B","MITF","MYB","NCOA4","NFIB","NKX2-1","PAX5","PHOX2B","PLAG1","PRDM1","RUNX1","SALL4","SOX2","SPEN","SS18","SSX1","TBX3","TMPRSS2","TP63","WT1","ZNF217","ZNF703")
PATH$miscellaneous = c("ACTN4","AIP","ARFRP1","BCR","CCDC6","CDH23","CHCHD7","COL1A1","DNAAF1","EML4","EMSY","EPCAM","EZR","GABRA6","GID4","KEL","KIAA1549","KIF1B","KIF5B","KLHL6", "LTK","MC1R","MERTK","NCF2","NUTM1","RUNDC1","SAMD9","SDC4","SLC34A2","SNCAIP","SPINK1","STRN","TPM3","VTI1A")
PATH$CNS_PTPN11 = c("PTPN11", "SOS1", "MYC")
PATH$Prostate_driver = c("ERG", "ETV1", "ETV6", "SPOP", "IDH1")
PATH$CCNE1_mut = c("CCNE1")
PATH$ARID1A_mut = c("ARID1A")
PATH$FGFR2_mut = c("FGFR2")
PATH$CDH1_mut = c("CDH1")
PATH$KRAS_mut = c("KRAS")
PATH$APC_mut = c("APC")
PATH$PTEN_mut = c("PTEN")
PATH$ERBB2_mut = c("ERBB2")
PATH$VEGFA_mut = c("VEGFA")
PATH$RB1_mut = c("RB1")
PATH$EGFR_mut = c("EGFR")
PATH$TP53_mut = c("TP53")
PATH$MSI_mut = c("MSI")
PATH$TMB_mut = c("TMB")
PATH$PTPN11_mut = c("PTPN11")
PATH$MSI_TMB_mut = c("MSI", "TMB")
PATH$NTRK_mut = c("NTRK1", "NTRK2", "NTRK3")
PATH$MDM2_CDK4_mut = c("MDM2", "CDK4")
PATH$CCNE1_mut = c("CCNE1")
PATH$ARID1A_mut = c("ARID1A")
PATH$FGFR2_mut = c("FGFR2")
PATH$CDH1_mut = c("CDH1")
PATH$BRAF_MET_mut = c("BRAF", "MET")
PATH$KRAS_mut = c("KRAS")
PATH$APC_mut = c("APC")
PATH$ERBB2_mut = c("ERBB2")
PATH$RB1_mut = c("RB1")
PATH$TP53_mut = c("TP53")
PATH$MSI_mut = c("MSI")
PATH$TMB_mut = c("TMB")
PATH$MSI_TMB_mut = c("MSI", "TMB")
PATH$NTRK_mut = c("NTRK1", "NTRK2", "NTRK3")
#PATH$URS_1 = c("RB1", "CDK4", "CCND2", "PTEN", "ATRX", "RICTOR", "JUN", "NF1", "CCNE1", "KRAS", "DDR2", "BRAF", "RAC1", "JAK2", "CARD11", "KDM5A", "FGF19", "ERBB3", "EGFR", "NOTCH1", "IDH1") #eribulin resistant
PATH$URS_1 = c("RB1", "CCND2", "JUN", "CDKN2A", "MDM2", "PIK3CA")
PATH$URS_2 = c("RB1", "TP53", "TERT", "PTEN", "ATRX")
PATH$URS_3 = c("TP53")
PATH$URS_4 = c("RB1")
PATH$URS_5 = c("TERT")
PATH$STS_1 = c("PTEN")
PATH$STS_2 = c("CCNE1")
PATH$STS_3 = c("ATRX")
PATH$STS_4 = c("ATRX")
PATH$STS_5 = c("JUN")
PATH$Breast_mut = c("TP53","PTEN","CDKN2A","RB1","MYC")
PATH$Biliary_mut = c("FGFR2","IDH1","IDH2","EPHA2","BAP1",
                     "PRKACA","PRKACB","ELF3","ARID1B","EGFR",
                     "ERBB2","ERBB3","PTEN","ARID2","KMT2D",
                     "KMT2C","TERT","KRAS","SMAD2","GNAS",
                     "TP53","BRCA1","BRCA2","PIK3CA","MSI","TMB",
                     "BRAF","NTRK1","NTRK2","NTRK3","MET")
# Automatic disease selecion
# Select the diseases of interest
if(No_histology == 1){
    cancer_select = function(x){
    return(switch(x,
    "Uterus_carcinoma" = "Uterus",
    "Uterus_sarcoma" = "Uterus",
    "NEC" = c("Bladder","Cervix","Esophagus_Stomach", "Pancreas", "Breast",
              "Lung", "Ovary", "Uterus", "Colorectal", "Prostate",
              "Kidney"),
    x
  ))}
    
  Cancer_tmp = NULL
  for(i in 1:length(Cancer)){
    Cancer_tmp = c(Cancer_tmp, cancer_select(Cancer[i]))
  }
  Cancer_tmp = unique(Cancer_tmp)
  Cancer = Cancer_tmp
  disease_select = function(x) {
    return(x)
  }
  Diseases = NULL
  for(i in 1:length(Cancer)){
    Diseases = c(Diseases, disease_select(Cancer[i]))
  }
  Diseases = unique(Diseases)
} else if(Auto_disease_selection == 1){
  disease_select = function(x) {
    return(switch(x,
    "Adrenal" = "Adrenal",
    "Ampulla_of_Vater" = "Ampulla_of_Vater",
    # "Biliary" = c("Cholangiocarcinoma (CHOL)_胆管癌",
    #               "Gallbladder Cancer (GBC)_胆嚢癌"#,
    #               #"NEC_Bile"
    #               ),
    "Biliary" = c("Extrahepatic Cholangiocarcinoma (EHCH)_肝外胆管癌",
                  "Intrahepatic Cholangiocarcinoma (IHCH)_肝内胆管癌",
                  "Perihilar Cholangiocarcinoma (PHCH)_肝門部胆管癌",
                  "Gallbladder Cancer (GBC)_胆嚢癌"#,
                  #"NEC_Bile"
                  ),
    "Bladder" = c("BLAD_UTUC", "BLAD_UC", "BLAD_BUC"),
    "Bone" = c("Chondrosarcoma (CHS)_軟骨肉腫",
               "Ewing Sarcoma (ES)_ユーイング肉腫",
               "Chordoma (CHDM)_脊索腫",
               "Osteosarcoma (OS)_骨肉腫",
               "Other_sarcoma"
                ),
     # "Bone" = c("Chondroblastic Osteosarcoma (CHOS)_軟骨芽細胞型骨肉腫", #Major sarcomas
     #            "Chondrosarcoma (CHS)_軟骨肉腫",
     #            "Ewing Sarcoma (ES)_ユーイング肉腫",
     #            "Fibroblastic Osteosarcoma (FIOS)_線維芽細胞型骨肉腫",
     #            "Osteoblastic Osteosarcoma (OSOS)_骨芽細胞型骨肉腫",
     #            "Osteosarcoma (OS)_骨肉腫",
     #            "Secondary Osteosarcoma (SECOS)_二次性骨肉腫",
     #            "Small Cell Osteosarcoma (SCOS)_小細胞型骨肉腫"),
    # "Bone" = c("Angiosarcoma of bone", #Ultra-rare sarcomas
    #            "CIC-rearranged sarcoma",
    #            "Conventional Type Chordoma (CCHDM)_通常型脊索腫",
    #            "Chordoma_NOS (CHDM)_脊索腫",
    #            "Dedifferentiated Chondrosarcoma (DDCHS)_脱分化型軟骨肉腫",
    #            "Dedifferentiated Chordoma (DDCHDM)_脱分化型脊索腫",
    #            "Extraskeletal myxoid chondrosarcoma",
    #            "extraskeletal osteosarcoma",
    #            "Glomangiosarcoma (GS)_悪性グロムス腫瘍",
    #            "High-Grade Surface Osteosarcoma (HGSOS)_高悪性度表在性骨肉腫",
    #            "Leiomyosarcoma of bone",
    #            "Malignancy in giant cell tumor of bone",
    #            "mesenchymal chondrosarcoma",
    #            "NTRK-rearranged spindle cell sarcoma",
    #            "Parosteal Osteosarcoma (PAOS)_傍骨性骨肉腫",
    #            "Periosteal Osteosarcoma (PEOS)",
    #            "Sarcoma with BCOR genetic alterations"),
    "Breast" = c("breast invasive ductal carcinoma"),#,
                 #"breast invasive lobular carcinoma"),#,
                 #"Invasive Breast Carcinoma"),
    # "Cervix" = c("Cervix_NEC",
    #              "Cervical Squamous Cell Carcinoma_CESC_扁平上皮癌",
    #              "Gastric Type Mucinous Carcinoma_GCEMU_胃型粘液性癌",
    #              "Cervix_non_SCC"
    #              ),
    #"Cervix" = c(#"Cervix_NEC"#,
    #             #"HPV-associated adeno"#,
    #             #"HPV-independent adeno"#,
    #             "SCC"
    #             ),
    "Cervix" = c("Cervix_NEC",
                "Cervix_adeno",
                "Cervix_SCC"#,
                #"Other"
                ),
    #"CNS" = c("Astrocytoma", "AT/RT", "CNS embryonal tumour", "Diffuse midline glioma",
    #          "Ependymoma", "Glioblastoma", "High-grade glioma",
    #          "Medulloblastoma", "Meningioma", "Neuroblastoma",
    #          "Oligodendroglioma"),
    "CNS" = c("Glioblastoma"#,
              #"Astrocytoma",
              #"Diffuse midline glioma",
              #"High-grade glioma"
              ),
    "Colorectal" = "Colorectal Adenocarcinoma (COADREAD)_結腸直腸腺癌",
    "Esophagus_Stomach" = c(#"Esophageal Squamous Cell Carcinoma (ESCC)"#,
              #"Stomach Adenocarcinoma (STAD)"#,
              "Stomach Adenocarcinoma, NOS (STAD_NOS)",
              "Stomach Adenocarcinoma, undifferentiated (STAD_undif)",
              "Stomach Adenocarcinoma, differentiated (STAD_dif)"#,
              #"Gastrointestinal Neuroendocrine Carcinoma of the Esophagus/Stomach"
               ),
    "Eye" = "Eye",
    "Head_Neck" = c("Head and Neck Squamous Cell Carcinoma (HNSC)_頭頸部扁平上皮癌"#,
                    #"Head and Neck Neuroendocrine Carcinoma (HNNE)_頭頸部神経内分泌癌",
                    #"Head and Neck Mucosal Melanoma (HNMUCM)_頭頸部粘膜黒色腫",
                    #"Salivary Carcinoma (SACA)_唾液腺癌"
                    ),
    "Kidney" = c("CC"
                 #, "Ch", "Pap"
                 ),
    "Liver" = c("Hepatocellular Carcinoma (HCC)_肝細胞癌",
                "Intrahepatic Cholangiocarcinoma"#,
                #"Hepatoblastoma (LIHB)_肝芽腫"
                ),
    "Lung" = c("Lung Adenocarcinoma (LUAD)",
               "Lung Squamous Cell Carcinoma (LUSC)"#,
               #"Lung_NEC"
               ),
    "Lymph" = "Lymph",
    "Other" = "Other",
    # "Ovary" = c("Clear Cell Ovarian Cancer_CCOV_明細胞癌",
    #            "Endometrioid Ovarian Cancer_EOV_類内膜癌",
    #            "High-Grade Serous Cancer",
    #            "Mucinous Ovarian Cancer_MOV_粘液性癌",
    #            "Ovarian Carcinosarcoma",
    #            "Ovarian Epithelial Tumor, Other"#,
    #            #"Other"
    #            ),

    "Ovary" = c("Ovary_Clear Cell Cancer",
               "Ovary_Endometrioid Cancer",
               "Ovary_High-Grade Serous Cancer",
               "Ovary_Mucinous Cancer",
               "Ovary_Carcinosarcoma"#,
               #"Ovary_Epithelial Tumor, Other"#,
               #"Other"
               ),
    # "Pancreas" = c("PAAD_腺癌", "PANEC_膵神経内分泌癌"),
    "Pancreas" = "PAAD_腺癌",
    "Penis" = "Penis",
    "Peripheral_Nervous_System" = "Peripheral_Nervous_System",
    "Peritoneum" = "Peritoneum",
    "Pleura" = "Pleura",
    #"Prostate" = c("PRAD", "PRNEC")
    "Prostate" = "PRAD",
    "Skin" = c("Skin Adnexal Carcinoma (SKAC)_皮膚付属器癌",
               "Melanoma (MEL)_悪性黒色腫",
               "Cutaneous Squamous Cell Carcinoma (CSCC)_有棘細胞癌",
               "Extramammary Paget Disease (EMPD)_乳房外パジェット病"
               ),
    # "Soft_tissue" = c("Undifferentiated Pleomorphic Sarcoma(UPS)_未分化多形肉腫",
    #                   "Myxofibrosarcoma (MFS)_粘液線維肉腫",
    #                   "Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫",
    #                   "Leiomyosarcoma (LMS)_平滑筋肉腫"
    #                   ),
    "Soft_tissue" = c("Rhabdomyosarcoma (RMS)_横紋筋肉腫",
                      "Gastrointestinal Stromal Tumor (GIST)_消化管間質腫瘍",
                      "Undifferentiated Pleomorphic Sarcoma(UPS)_未分化多形肉腫",
                      "Myxofibrosarcoma (MFS)_粘液線維肉腫",
                      "Myxoid/Round-Cell Liposarcoma (MRLS)_粘液型/円形細胞型脂肪肉腫",
                      "Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
                      "Synovial Sarcoma (SYNS)_滑膜肉腫",
                      "Ewing Sarcoma of Soft Tissue (ESST)_骨外性ユーイング肉腫",
                      "Ewing Sarcoma (ES)_ユーイング肉腫",
                      "Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫",
                      "Leiomyosarcoma (LMS)_平滑筋肉腫",
                      "Angiosarcoma (ANGS)_血管肉腫",
                      "Alveolar Soft Part Sarcoma (ASPS)_胞巣状軟部肉腫",
                      "Intimal Sarcoma (INTS)_血管内膜肉腫",
                      "Other_sarcoma"
                      ),
    # "Soft_tissue" = c("Angiosarcoma (ANGS)_血管肉腫", #Major sarcoma
    #                   "Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫",
    #                   "Gastrointestinal Stromal Tumor (GIST)_消化管間質腫瘍",
    #                   "Leiomyosarcoma (LMS)_平滑筋肉腫",
    #                   "Myxofibrosarcoma (MFS)_粘液線維肉腫",
    #                   "Myxoid/Round-Cell Liposarcoma (MRLS)_粘液型/円形細胞型脂肪肉腫",
    #                   "Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
    #                   "Synovial Sarcoma (SYNS)_滑膜肉腫",
    #                   "Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma (MFH)_未分化多形肉腫/悪性線維性組織球腫/高悪性度紡錐細胞肉腫",
    #                   "Well-Differentiated Liposarcoma (WDLS)_高分化型脂肪肉腫"
    #                  ),
    # "Soft_tissue" = c("Alveolar Rhabdomyosarcoma (ARMS)_胞巣型横紋筋肉腫", #Ultra-rare sarcoma
    #                   "Alveolar Soft Part Sarcoma (ASPS)_胞巣状軟部肉腫",
    #                   "Angiomatoid Fibrous Histiocytoma (AFH)_類血管腫型線維性組織球腫",
    #                   "CIC-rearranged sarcoma",
    #                   "Clear Cell Sarcoma (CCS)_明細胞肉腫",
    #                   "Dendritic Cell Sarcoma (DCS)_樹状細胞肉腫",
    #                   "Desmoid/Aggressive Fibromatosis (DES)_デスモイド/侵襲性線維腫症",
    #                   "Desmoplastic Small-Round-Cell Tumor (DSRCT)_線維形成性小円形細胞腫瘍",
    #                   "Proximal-Type Epithelioid Sarcoma (PTES)_近位型類上皮肉腫",
    #                   "Embryonal Rhabdomyosarcoma (ERMS)_胎児型横紋筋肉腫",
    #                   "Epithelioid Hemangioendothelioma (EHAE)_類上皮血管内皮腫",
    #                   "Epithelioid Sarcoma_NOS (EPIS)_類上皮肉腫",
    #                   "Ewing Sarcoma of Soft Tissue (ESST)_骨外性ユーイング肉腫",
    #                   "Extraskeletal myxoid chondrosarcoma",
    #                   "extraskeletal osteosarcoma",
    #                   "Adult fibrosarcoma",
    #                   "Glomangiosarcoma (GS)_悪性グロムス腫瘍",
    #                   "Histiocytic Dendritic Cell Sarcoma (HDCS)_組織球性樹状細胞肉腫",
    #                   "Infantile Fibrosarcoma (IFS)_乳児型線維肉腫",
    #                   "Inflammatory Myofibroblastic Tumor (IMT)_炎症性筋線維芽細胞腫瘍",
    #                   "Intimal Sarcoma (INTS)_血管内膜肉腫",
    #                   "Low-Grade Fibromyxoid Sarcoma (LGFMS)_低悪性度線維粘液肉腫",
    #                   "malignant granular cell tumor",
    #                   "malignant phyllodes tumor",
    #                   "malignant rhabdoid tumor",
    #                   "mesenchymal chondrosarcoma",
    #                   "MPNST",
    #                   "NTRK-rearranged spindle cell sarcoma",
    #                   "Ossifying Fibromyxoid Tumor (OFMT)_骨化性線維粘液性腫瘍",
    #                   "Paraganglioma (PGNG)_傍神経節腫 パラガングリオーマ",
    #                   "Perivascular Epithelioid Cell Tumor (PECOMA)_血管周囲性類上皮細胞性腫瘍",
    #                   "Pleomorphic Liposarcoma (PLLS)_多形型脂肪肉腫",
    #                   "Pleomorphic Rhabdomyosarcoma (PLRMS)_多形型横紋筋肉腫",
    #                   "Pseudomyogenic Hemangioendothelioma (PMHE)_偽筋原性血管内皮腫",
    #                   "Rhabdomyosarcoma (RMS)_横紋筋肉腫",
    #                   "Round Cell Sarcoma, NOS (RCSNOS)_円形細胞肉腫、特定不能",
    #                   "Sarcoma with BCOR genetic alterations",
    #                   "Sclerosing Epithelioid Fibrosarcoma (SEF)_硬化性類上皮線維肉腫",
    #                   "Soft Tissue Myoepithelial Carcinoma (STMYEC)_軟部の筋上皮癌",
    #                   #"Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
    #                   "Spindle Cell/Sclerosing Rhabdomyosarcoma (SCSRMS)_紡錘形細胞型/硬化型横紋筋肉腫"
    #                    ),
    # "Soft_tissue" = c("Angiosarcoma (ANGS)_血管肉腫", #Non-translocation-related sarcoma
    #                   "Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫",
    #                   "Gastrointestinal Stromal Tumor (GIST)_消化管間質腫瘍",
    #                   "Leiomyosarcoma (LMS)_平滑筋肉腫",
    #                   "Myxofibrosarcoma (MFS)_粘液線維肉腫",
    #                   #"Myxoid/Round-Cell Liposarcoma (MRLS)_粘液型/円形細胞型脂肪肉腫",
    #                   #"Synovial Sarcoma (SYNS)_滑膜肉腫",
    #                   "Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma (MFH)_未分化多形肉腫/悪性線維性組織球腫/高悪性度紡錐細胞肉腫",
    #                   "Well-Differentiated Liposarcoma (WDLS)_高分化型脂肪肉腫",
    #                   #"Alveolar Rhabdomyosarcoma (ARMS)_胞巣型横紋筋肉腫",
    #                   #"Alveolar Soft Part Sarcoma (ASPS)_胞巣状軟部肉腫",
    #                   #"Angiomatoid Fibrous Histiocytoma (AFH)_類血管腫型線維性組織球腫",
    #                   #"CIC-rearranged sarcoma",
    #                   #"Clear Cell Sarcoma (CCS)_明細胞肉腫",
    #                   "Dendritic Cell Sarcoma (DCS)_樹状細胞肉腫",
    #                   #"Dermatofibrosarcoma Protuberance",
    #                   "Desmoid/Aggressive Fibromatosis (DES)_デスモイド/侵襲性線維腫症",
    #                   #"Desmoplastic Small-Round-Cell Tumor (DSRCT)_線維形成性小円形細胞腫瘍",
    #                   "Proximal-Type Epithelioid Sarcoma (PTES)_近位型類上皮肉腫",
    #                   "Embryonal Rhabdomyosarcoma (ERMS)_胎児型横紋筋肉腫",
    #                   #"Epithelioid Hemangioendothelioma (EHAE)_類上皮血管内皮腫",
    #                   "Epithelioid Sarcoma_NOS (EPIS)_類上皮肉腫",
    #                   #"Ewing Sarcoma of Soft Tissue (ESST)_骨外性ユーイング肉腫",
    #                   #"Extraskeletal myxoid chondrosarcoma",
    #                   "extraskeletal osteosarcoma",
    #                   "Adult fibrosarcoma",
    #                   "MPNST",
    #                   #"Glomangiosarcoma (GS)_悪性グロムス腫瘍",
    #                   "Histiocytic Dendritic Cell Sarcoma (HDCS)_組織球性樹状細胞肉腫",
    #                   #"Infantile Fibrosarcoma (IFS)_乳児型線維肉腫",
    #                   #"Inflammatory Myofibroblastic Tumor (IMT)_炎症性筋線維芽細胞腫瘍",
    #                   "Intimal Sarcoma (INTS)_血管内膜肉腫",
    #                   "Liposarcoma (LIPO)_脂肪肉腫",
    #                   "Low-Grade Fibromyxoid Sarcoma (LGFMS)_低悪性度線維粘液肉腫",
    #                   "malignant granular cell tumor",
    #                   "malignant phyllodes tumor",
    #                   "malignant rhabdoid tumor",
    #                   #"mesenchymal chondrosarcoma",
    #                   #"NTRK-rearranged spindle cell sarcoma",
    #                   #"Ossifying Fibromyxoid Tumor (OFMT)_骨化性線維粘液性腫瘍",
    #                   "Paraganglioma (PGNG)_傍神経節腫 パラガングリオーマ",
    #                   #"Perivascular Epithelioid Cell Tumor (PECOMA)_血管周囲性類上皮細胞性腫瘍",
    #                   "Pleomorphic Liposarcoma (PLLS)_多形型脂肪肉腫",
    #                   "Pleomorphic Rhabdomyosarcoma (PLRMS)_多形型横紋筋肉腫",
    #                   #"Pseudomyogenic Hemangioendothelioma (PMHE)_偽筋原性血管内皮腫",
    #                   #"Rhabdomyosarcoma (RMS)_横紋筋肉腫",
    #                   #"Round Cell Sarcoma, NOS (RCSNOS)_円形細胞肉腫、特定不能",
    #                   #"Sarcoma with BCOR genetic alterations",
    #                   "Sarcoma, NOS (SARCNOS)_肉腫、特定不能",
    #                   #"Sclerosing Epithelioid Fibrosarcoma (SEF)_硬化性類上皮線維肉腫",
    #                   #"Soft Tissue Myoepithelial Carcinoma (STMYEC)_軟部の筋上皮癌",
    #                   #"Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
    #                   "Spindle Cell/Sclerosing Rhabdomyosarcoma (SCSRMS)_紡錘形細胞型/硬化型横紋筋肉腫"
    #                    ),
    # "Soft_tissue" = c(#"Angiosarcoma (ANGS)_血管肉腫", #Translocation-related sarcoma
    #                   #"Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫",
    #                   #"Gastrointestinal Stromal Tumor (GIST)_消化管間質腫瘍",
    #                   #"Leiomyosarcoma (LMS)_平滑筋肉腫",
    #                   #"Myxofibrosarcoma (MFS)_粘液線維肉腫",
    #                   "Myxoid/Round-Cell Liposarcoma (MRLS)_粘液型/円形細胞型脂肪肉腫",
    #                   "Synovial Sarcoma (SYNS)_滑膜肉腫",
    #                   #"Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma (MFH)_未分化多形肉腫/悪性線維性組織球腫/高悪性度紡錐細胞肉腫",
    #                   #"Well-Differentiated Liposarcoma (WDLS)_高分化型脂肪肉腫",
    #                   "Alveolar Rhabdomyosarcoma (ARMS)_胞巣型横紋筋肉腫",
    #                   "Alveolar Soft Part Sarcoma (ASPS)_胞巣状軟部肉腫",
    #                   "Angiomatoid Fibrous Histiocytoma (AFH)_類血管腫型線維性組織球腫",
    #                   "CIC-rearranged sarcoma",
    #                   "Clear Cell Sarcoma (CCS)_明細胞肉腫",
    #                   #"Dendritic Cell Sarcoma (DCS)_樹状細胞肉腫",
    #                   "Dermatofibrosarcoma Protuberance",
    #                   #"Desmoid/Aggressive Fibromatosis (DES)_デスモイド/侵襲性線維腫症",
    #                   "Desmoplastic Small-Round-Cell Tumor (DSRCT)_線維形成性小円形細胞腫瘍",
    #                   #"Proximal-Type Epithelioid Sarcoma (PTES)_近位型類上皮肉腫",
    #                   #"Embryonal Rhabdomyosarcoma (ERMS)_胎児型横紋筋肉腫",
    #                   "Epithelioid Hemangioendothelioma (EHAE)_類上皮血管内皮腫",
    #                   #"Epithelioid Sarcoma_NOS (EPIS)_類上皮肉腫",
    #                   "Ewing Sarcoma of Soft Tissue (ESST)_骨外性ユーイング肉腫",
    #                   "Extraskeletal myxoid chondrosarcoma",
    #                   #"extraskeletal osteosarcoma",
    #                   #"Adult fibrosarcoma",
    #                   "Glomangiosarcoma (GS)_悪性グロムス腫瘍",
    #                   #"Histiocytic Dendritic Cell Sarcoma (HDCS)_組織球性樹状細胞肉腫",
    #                   "Infantile Fibrosarcoma (IFS)_乳児型線維肉腫",
    #                   "Inflammatory Myofibroblastic Tumor (IMT)_炎症性筋線維芽細胞腫瘍",
    #                   #"Intimal Sarcoma (INTS)_血管内膜肉腫",
    #                   "Low-Grade Fibromyxoid Sarcoma (LGFMS)_低悪性度線維粘液肉腫",
    #                   #"Liposarcoma (LIPO)_脂肪肉腫",
    #                   #"malignant granular cell tumor",
    #                   #"malignant phyllodes tumor",
    #                   #"malignant rhabdoid tumor",
    #                   "mesenchymal chondrosarcoma",
    #                   "NTRK-rearranged spindle cell sarcoma",
    #                   "Ossifying Fibromyxoid Tumor (OFMT)_骨化性線維粘液性腫瘍",
    #                   #"Paraganglioma (PGNG)_傍神経節腫 パラガングリオーマ",
    #                   "Perivascular Epithelioid Cell Tumor (PECOMA)_血管周囲性類上皮細胞性腫瘍",
    #                   #"Pleomorphic Liposarcoma (PLLS)_多形型脂肪肉腫",
    #                   #"Pleomorphic Rhabdomyosarcoma (PLRMS)_多形型横紋筋肉腫",
    #                   "Pseudomyogenic Hemangioendothelioma (PMHE)_偽筋原性血管内皮腫",
    #                   #"Rhabdomyosarcoma (RMS)_横紋筋肉腫",
    #                   #"MPNST",
    #                   #"Round Cell Sarcoma, NOS (RCSNOS)_円形細胞肉腫、特定不能",
    #                   "Sarcoma with BCOR genetic alterations",
    #                   "Sclerosing Epithelioid Fibrosarcoma (SEF)_硬化性類上皮線維肉腫",
    #                   "Soft Tissue Myoepithelial Carcinoma (STMYEC)_軟部の筋上皮癌"#,
    #                   "Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
    #                   #"Spindle Cell/Sclerosing Rhabdomyosarcoma (SCSRMS)_紡錘形細胞型/硬化型横紋筋肉腫"
    #                    ),
    # "Soft_tissue" = c("Angiosarcoma (ANGS)_血管肉腫", #All sarcoma
    #                   "Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫",
    #                   "Gastrointestinal Stromal Tumor (GIST)_消化管間質腫瘍",
    #                   "Leiomyosarcoma (LMS)_平滑筋肉腫",
    #                   "Myxofibrosarcoma (MFS)_粘液線維肉腫",
    #                   "Myxoid/Round-Cell Liposarcoma (MRLS)_粘液型/円形細胞型脂肪肉腫",
    #                   "Synovial Sarcoma (SYNS)_滑膜肉腫",
    #                   "Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma (MFH)_未分化多形肉腫/悪性線維性組織球腫/高悪性度紡錐細胞肉腫",
    #                   "Well-Differentiated Liposarcoma (WDLS)_高分化型脂肪肉腫",
    #                   "Alveolar Rhabdomyosarcoma (ARMS)_胞巣型横紋筋肉腫",
    #                   "Alveolar Soft Part Sarcoma (ASPS)_胞巣状軟部肉腫",
    #                   "Angiomatoid Fibrous Histiocytoma (AFH)_類血管腫型線維性組織球腫",
    #                   "CIC-rearranged sarcoma",
    #                   "Clear Cell Sarcoma (CCS)_明細胞肉腫",
    #                   "Dendritic Cell Sarcoma (DCS)_樹状細胞肉腫",
    #                   "Dermatofibrosarcoma Protuberance",
    #                   "Desmoid/Aggressive Fibromatosis (DES)_デスモイド/侵襲性線維腫症",
    #                   "Desmoplastic Small-Round-Cell Tumor (DSRCT)_線維形成性小円形細胞腫瘍",
    #                   "Proximal-Type Epithelioid Sarcoma (PTES)_近位型類上皮肉腫",
    #                   "Embryonal Rhabdomyosarcoma (ERMS)_胎児型横紋筋肉腫",
    #                   "Epithelioid Hemangioendothelioma (EHAE)_類上皮血管内皮腫",
    #                   "Epithelioid Sarcoma_NOS (EPIS)_類上皮肉腫",
    #                   "Ewing Sarcoma of Soft Tissue (ESST)_骨外性ユーイング肉腫",
    #                   "Extraskeletal myxoid chondrosarcoma",
    #                   "extraskeletal osteosarcoma",
    #                   "Adult fibrosarcoma",
    #                   "Glomangiosarcoma (GS)_悪性グロムス腫瘍",
    #                   "Histiocytic Dendritic Cell Sarcoma (HDCS)_組織球性樹状細胞肉腫",
    #                   "Infantile Fibrosarcoma (IFS)_乳児型線維肉腫",
    #                   "Inflammatory Myofibroblastic Tumor (IMT)_炎症性筋線維芽細胞腫瘍",
    #                   "Intimal Sarcoma (INTS)_血管内膜肉腫",
    #                   "Liposarcoma (LIPO)_脂肪肉腫",
    #                   "Low-Grade Fibromyxoid Sarcoma (LGFMS)_低悪性度線維粘液肉腫",
    #                   "MPNST",
    #                   "malignant granular cell tumor",
    #                   "malignant phyllodes tumor",
    #                   "malignant rhabdoid tumor",
    #                   "mesenchymal chondrosarcoma",
    #                   "NTRK-rearranged spindle cell sarcoma",
    #                   "Ossifying Fibromyxoid Tumor (OFMT)_骨化性線維粘液性腫瘍",
    #                   "Paraganglioma (PGNG)_傍神経節腫 パラガングリオーマ",
    #                   "Perivascular Epithelioid Cell Tumor (PECOMA)_血管周囲性類上皮細胞性腫瘍",
    #                   "Pleomorphic Liposarcoma (PLLS)_多形型脂肪肉腫",
    #                   "Pleomorphic Rhabdomyosarcoma (PLRMS)_多形型横紋筋肉腫",
    #                   "Pseudomyogenic Hemangioendothelioma (PMHE)_偽筋原性血管内皮腫",
    #                   "Rhabdomyosarcoma (RMS)_横紋筋肉腫",
    #                   "Round Cell Sarcoma, NOS (RCSNOS)_円形細胞肉腫、特定不能",
    #                   "Sarcoma with BCOR genetic alterations",
    #                   "Sarcoma, NOS (SARCNOS)_肉腫、特定不能",
    #                   "Sclerosing Epithelioid Fibrosarcoma (SEF)_硬化性類上皮線維肉腫",
    #                   "Soft Tissue Myoepithelial Carcinoma (STMYEC)_軟部の筋上皮癌",
    #                   "Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
    #                   "Spindle Cell/Sclerosing Rhabdomyosarcoma (SCSRMS)_紡錘形細胞型/硬化型横紋筋肉腫"
    #                    ),
    # # "Soft_tissue" = c("Angiosarcoma (ANGS)_血管肉腫", #All sarcoma but not L-sarcoma
    #                   #"Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫",
    #                   "Gastrointestinal Stromal Tumor (GIST)_消化管間質腫瘍",
    #                   #"Leiomyosarcoma (LMS)_平滑筋肉腫",
    #                   "Myxofibrosarcoma (MFS)_粘液線維肉腫",
    #                   #"Myxoid/Round-Cell Liposarcoma (MRLS)_粘液型/円形細胞型脂肪肉腫",
    #                   "Synovial Sarcoma (SYNS)_滑膜肉腫",
    #                   "Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma (MFH)_未分化多形肉腫/悪性線維性組織球腫/高悪性度紡錐細胞肉腫",
    #                   #"Well-Differentiated Liposarcoma (WDLS)_高分化型脂肪肉腫",
    #                   "Alveolar Rhabdomyosarcoma (ARMS)_胞巣型横紋筋肉腫",
    #                   "Alveolar Soft Part Sarcoma (ASPS)_胞巣状軟部肉腫",
    #                   "Angiomatoid Fibrous Histiocytoma (AFH)_類血管腫型線維性組織球腫",
    #                   "CIC-rearranged sarcoma",
    #                   "Clear Cell Sarcoma (CCS)_明細胞肉腫",
    #                   "Dendritic Cell Sarcoma (DCS)_樹状細胞肉腫",
    #                   "Dermatofibrosarcoma Protuberance",
    #                   "Desmoid/Aggressive Fibromatosis (DES)_デスモイド/侵襲性線維腫症",
    #                   "Desmoplastic Small-Round-Cell Tumor (DSRCT)_線維形成性小円形細胞腫瘍",
    #                   "Proximal-Type Epithelioid Sarcoma (PTES)_近位型類上皮肉腫",
    #                   "Embryonal Rhabdomyosarcoma (ERMS)_胎児型横紋筋肉腫",
    #                   "Epithelioid Hemangioendothelioma (EHAE)_類上皮血管内皮腫",
    #                   "Epithelioid Sarcoma_NOS (EPIS)_類上皮肉腫",
    #                   "Ewing Sarcoma of Soft Tissue (ESST)_骨外性ユーイング肉腫",
    #                   "Extraskeletal myxoid chondrosarcoma",
    #                   "extraskeletal osteosarcoma",
    #                   "Adult fibrosarcoma",
    #                   "Glomangiosarcoma (GS)_悪性グロムス腫瘍",
    #                   "Histiocytic Dendritic Cell Sarcoma (HDCS)_組織球性樹状細胞肉腫",
    #                   "Infantile Fibrosarcoma (IFS)_乳児型線維肉腫",
    #                   "Inflammatory Myofibroblastic Tumor (IMT)_炎症性筋線維芽細胞腫瘍",
    #                   "Intimal Sarcoma (INTS)_血管内膜肉腫",
    #                   #"Liposarcoma (LIPO)_脂肪肉腫",
    #                   "Low-Grade Fibromyxoid Sarcoma (LGFMS)_低悪性度線維粘液肉腫",
    #                   "malignant granular cell tumor",
    #                   "malignant phyllodes tumor",
    #                   "mesenchymal chondrosarcoma",
    #                   "malignant rhabdoid tumor",
    #                   "NTRK-rearranged spindle cell sarcoma",
    #                   "Ossifying Fibromyxoid Tumor (OFMT)_骨化性線維粘液性腫瘍",
    #                   "Paraganglioma (PGNG)_傍神経節腫 パラガングリオーマ",
    #                   "Perivascular Epithelioid Cell Tumor (PECOMA)_血管周囲性類上皮細胞性腫瘍",
    #                   "Pleomorphic Liposarcoma (PLLS)_多形型脂肪肉腫",
    #                   "Pleomorphic Rhabdomyosarcoma (PLRMS)_多形型横紋筋肉腫",
    #                   "Pseudomyogenic Hemangioendothelioma (PMHE)_偽筋原性血管内皮腫",
    #                   "Rhabdomyosarcoma (RMS)_横紋筋肉腫",
    #                   "Round Cell Sarcoma, NOS (RCSNOS)_円形細胞肉腫、特定不能",
    #                   "Sarcoma with BCOR genetic alterations",
    #                   "Sarcoma, NOS (SARCNOS)_肉腫、特定不能",
    #                   "Sclerosing Epithelioid Fibrosarcoma (SEF)_硬化性類上皮線維肉腫",
    #                   "Soft Tissue Myoepithelial Carcinoma (STMYEC)_軟部の筋上皮癌",
    #                   "Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
    #                   "Spindle Cell/Sclerosing Rhabdomyosarcoma (SCSRMS)_紡錘形細胞型/硬化型横紋筋肉腫"
    #                    ),
    # "Soft_tissue" = c(#L-sarcoma
    #                   "Dedifferentiated Liposarcoma (DDLS)_脱分化型脂肪肉腫"#,
    #                   #"Leiomyosarcoma (LMS)_平滑筋肉腫"#,
    #                   #"Myxoid/Round-Cell Liposarcoma (MRLS)_粘液型/円形細胞型脂肪肉腫",
    #                   #"Well-Differentiated Liposarcoma (WDLS)_高分化型脂肪肉腫",
    #                   #"Liposarcoma (LIPO)_脂肪肉腫",
    #                   #"Pleomorphic Liposarcoma (PLLS)_多形型脂肪肉腫",
    #                   #"Synovial Sarcoma (SYNS)_滑膜肉腫",
    #                   #"Solitary Fibrous Tumor/Hemangiopericytoma (SFT)_孤立性線維性腫瘍/血管周皮種",
    #                   #"Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma (MFH)_未分化多形肉腫/悪性線維性組織球腫/高悪性度紡錐細胞肉腫"#,
    #                    ),
    "Testis" = "Testis",
    "Thymus" = "Thymus",
    "Thyroid" = "Thyroid",
    "Uterus_carcinoma" = c("Uterine Endometrioid Carcinoma_UEC_子宮類内膜癌",
               "Uterine Clear Cell Carcinoma_UCCC_子宮明細胞癌",
               "Uterine Carcinosarcoma",
               "Uterine Serous Carcinoma"#,
               #"Uterine other carcinoma"
               ),
    "Uterus_sarcoma" = c("Uterine Endometrial Stromal Sarcoma",
               "Uterine Leiomyosarcoma_ULMS_子宮平滑筋肉腫",
               "Uterine Adenosarcoma_UAS_子宮腺肉腫"#,
               #"Undifferentiated Uterine Sarcoma_UUS_子宮未分化肉腫"#,
               #"Uterine Sarcoma, Other_OUSARC_子宮肉腫その他"
               ),
    "Vagina" = "Vagina",
    "NEC" = c("Gastrointestinal Neuroendocrine Carcinoma of the Esophagus/Stomach",
               "Neuroendocrine Carcinoma of the Bladder",
               "NEC_Ovary",
               "Uterine Neuroendocrine Carcinoma_UNEC_子宮神経内分泌腫瘍",
               "Cervix_NEC",
               "Lung_NEC",
               "NEC_BOWEL",
               "PANEC_膵神経内分泌癌",
               "Breast_NEC",
               "PRNEC",
               "NEC_Kidney",
               "Head and Neck Neuroendocrine Carcinoma (HNNE)_頭頸部神経内分泌癌",
               "NEC_Bile",
               "NEC_liver",
               "NEC_skin")
    ))}
  Diseases = NULL
  for(i in 1:length(Cancer)){
    Diseases = c(Diseases, disease_select(Cancer[i]))
  }
  Diseases = unique(Diseases)
}
Diseases = sort(Diseases)
cancer_select = function(x){
  return(switch(x,
  "Uterus_carcinoma" = "Uterus",
  "Uterus_sarcoma" = "Uterus",
  "NEC" = c("Bladder","Cervix","Esophagus_Stomach", "Pancreas", "Breast",
            "Lung", "Ovary", "Uterus", "Colorectal", "Prostate",
            "Kidney", "Biliary", "Liver", "Skin"),
  x
))}
  
Cancer_tmp = NULL
for(i in 1:length(Cancer)){
  Cancer_tmp = c(Cancer_tmp, cancer_select(Cancer[i]))
}
Cancer_tmp = unique(Cancer_tmp)
Cancer = Cancer_tmp

## Target drug selection
## format
# selected_drug = list("Oxaliplatin","Oteracil", "Tipiracil")
# search pattern
# Describe search criteria in binary form (sum of 1, 2, 4, 8, 16,,,)
# Patients administered Oteracil before EP: c(2)
# Patients administered Oxaliplatin and Tipiracil before EP: c(5)
# Patients administered Oxaliplatin or Tipiracil before EP: c(1, 4)
# Patients administered all drugs before EP: c(7)
# Patients administered any drugs before EP: c(1, 2, 4)

# examples
# selected_drug = list("Oxaliplatin","Oteracil", "Tipiracil")
# drug_pattern = list(c(1), c(3,5))
# selected_drug = list("Nivolumab", "Regorafenib")
# drug_pattern = list(c(1), c(2), c(1, 2), c(3))
# selected_drug = list("Cisplatin")
# drug_pattern = list(c(1))

## Target gene selection
## format
# target_gene = list(c("KRAS", "G12V"))
# target_gene = list(c("KRAS", "G12V"), c("KRAS", "G13D"), c("KRAS", "Q61H"), c("FGFR2", "R450H"))
# target_gene = list(c("TMB", "high"), c("MSI", "high"))
# F means all oncogenic mutations
# WT means without oncogenic mutations
# target_gene = list(c("TP53", "F"), c("BRCA1", "F"), c("BRCA2", "F"))
# target_gene = list(c("MDM2", "Amplification"), c("CDKN2A", "Deletion"))

# search pattern
# Example; target_gene = list(c("KRAS", "G12V"), c("KRAS", "G13D"), c("KRAS", "Q61H"), c("FGFR2", "R450H"))
# Describe search criteria in binary form (sum of 1, 2, 4, 8, 16,,,)
# All KRAS mutations: c(1, 2, 3, 4, 5, 6, 7)
# FGFR R450H: c(8)
# KRAS G12V or KRAS G13D: c(1, 2)
# KRAS G12V and KRAS G13D: c(3)
# Any KRAS mutations and FGFR R450H: c(9, 10, 11, 12, 13, 14, 15)

# examples
# target_gene = list(c("KRAS", "G12V"), c("KRAS", "G13D"), c("KRAS", "Q61H"), c("FGFR2", "R450H"))
# gene_pattern = list(c(1, 2, 3, 4, 5, 6, 7), c(8), c(9, 10, 11, 12, 13, 14, 15))
#
# target_gene = list(c("BRAF", "V600E"), c("MSI", "high"), c("TMB", "high"))
# gene_pattern = list(c(1), c(2), c(1, 2), c(3), c(4))

shannon.entropy = function(x, x_length){ 
    x = x[(!is.na(x))]
    x = x[(x != 0)]
    if(sum(x) <= 0) {
        return(log(x_length, base=2)) 
    } else {
        p = x/sum(x)
        score = -sum(p * log(p, base=2))
        return(score)
    }
} 

odds.ratio <- function(a, b, c, d, correct=FALSE){
  cl <- function(x){
    or*exp(c(1,-1)*qnorm(x)*sqrt(1/a+1/b+1/c+1/d))
  }
  if (correct || a*b*c*d==0) {
    a <- a+0.5
    b <- b+0.5
    c <- c+0.5
    d <- d+0.5
  }
  or <- a*d/(b*c)
  conf <- rbind(cl90=cl(0.05), cl95=cl(0.025), cl99=cl(0.005), cl999=cl(0.0005))
  conf <- data.frame(conf)
  colnames(conf) <- paste(c("lower","upper"), " limit" , sep="")
  rownames(conf) <- paste(c(90, 95, 99, 99.9), "%CI" , sep="")
  list(or=or, conf=conf)
}


knitr::opts_chunk$set(
	error = FALSE,
	fig.align = "center",
	fig.height = 12,
	fig.pos = "H",
	fig.width = 12,
	message = FALSE,
	warning = FALSE,
	out.extra = "",
	tidy = TRUE,
	tidy.opts = list(width.cutoff = 45)
)
knitr::opts_chunk$set(tidy=TRUE)
knitr::opts_chunk$set(warning=FALSE)
knitr::opts_chunk$set(error=FALSE)
knitr::opts_chunk$set(fig.width=12, fig.height=12, fig.align='center', 
                      fig.pos='H', out.extra='', 
                      tidy.opts = list(width.cutoff = 45))
knitr::opts_knit$set(root.dir=Working)
setwd(Working)
Working = getwd()
set.seed(1)

options(scipen=100)
options(max.print=1000) 

library(gtsummary)
library(stringr)
library(ggplot2)
library(tidyverse)
library(RColorBrewer)
library(survival)
library(DT)
library(WriteXLS)
library(openxlsx)
library(broom)
library(ComplexHeatmap)
library(circlize)
library(future)
library(ggiraph)
library(factoextra)
library(Matching)
library(gridExtra)
library(KMsurv)
library(survminer)
library(tranSurv)
library(exactRankTests)
library(ggsci)
library(grid)
library(forestploter)
library(gt)
library(patchwork)
library(gtools)
library(Rediscover)
library(readr)

if(Analyze_chemo >= 1 | Analyze_enroll == 1 | Analyze_diagnosis >= 1 | Subanalysis == 1){
  library(rstan)
  library(bayesplot)
  library(tidybayes)
  rstan_options(auto_write = TRUE)
  options(mc.cores = parallel::detectCores())
  plan(multisession)
}
if(Analyze_clustering == 1){
  #library(ConsensusClusterPlus)
  library(umap)
  library(biomaRt)
  library(GenVisR)
  library(dbscan)
}

```

```{r preprocessing, echo = (Source_code == 1), warning=FALSE}
## Data loading, Ver.20230220   , patients ~ 2023.02.20, downloaded @ 2023.03.01
List_case = NULL
List_report = NULL
Source_dir = paste(Working, "/source", sep="")

if(All_cancer_type == 1){
  #for(i in Cancer){
  for(i in (list.dirs(Source_dir))){
    if(i != paste(Working, "/source", sep="") &
       i != paste(Working, "/source/PGPV", sep="")){
    List_case = c(List_case, paste(i, "/", list.files(i, pattern="case"), sep=""))
  #  List_case = c(List_case, paste("source/", i, "/", list.files(paste(Working, "/source/", i, sep=""), pattern="case"), sep=""))
    List_report = c(List_report, paste(i, "/", list.files(i, pattern="report"), sep=""))
  #  List_report = c(List_report, paste("source/", i, "/", list.files(paste(Working, "/source/", i, sep=""), pattern="report"), sep=""))
    }
  }
  Data_case = NULL
  Data_report = NULL
  for(i in List_case){
    Encode = guess_encoding(i)[[1]][[1]]
    if(Encode == "Shift_JIS"){
      Encode = "CP932"
    }
    tmp = read.csv(file(i, encoding=Encode))
    tmp$directory = str_split(i, pattern = "/", simplify = TRUE)[8]
  #  tmp$directory = str_split(i, pattern = "/", simplify = TRUE)[2]
    Data_case <- rbind(Data_case, tmp)
  }
  for(i in List_report){
    Encode = guess_encoding(i)[[1]][[1]]
    if(Encode == "Shift_JIS"){
      Encode = "CP932"
    }
    tmp = read.csv(file(i, encoding=Encode))
    tmp$directory = str_split(i, pattern = "/", simplify = TRUE)[2]
    Data_report <- rbind(Data_report, tmp)
  }
  Data_case_raw = Data_case
  Data_case = Data_case %>% dplyr::filter(症例.検体情報.パネル.名称. %in% Panel)
} else{
  for(i in Cancer){
    if(i != paste(Working, "/source", sep="") &
       i != paste(Working, "/source/PGPV", sep="")){
      List_case = c(List_case, paste("source/", i, "/", list.files(paste(Working, "/source/", i, sep=""), pattern="case"), sep=""))
      List_report = c(List_report, paste("source/", i, "/", list.files(paste(Working, "/source/", i, sep=""), pattern="report"), sep=""))
    }
  }
  Data_case = NULL
  Data_report = NULL
  for(i in List_case){
    Encode = guess_encoding(i)[[1]][[1]]
    if(Encode == "Shift_JIS"){
      Encode = "CP932"
    }
    tmp = read.csv(file(i, encoding=Encode))
    tmp$directory = str_split(i, pattern = "/", simplify = TRUE)[2]
    Data_case <- rbind(Data_case, tmp)
  }
  for(i in List_report){
    Encode = guess_encoding(i)[[1]][[1]]
    if(Encode == "Shift_JIS"){
      Encode = "CP932"
    }
    tmp = read.csv(file(i, encoding=Encode))
    tmp$directory = str_split(i, pattern = "/", simplify = TRUE)[2]
    Data_report <- rbind(Data_report, tmp)
  }
  Data_case_raw = Data_case
  Data_case = Data_case %>% dplyr::filter(症例.検体情報.パネル.名称. %in% Panel)
}

if(length(Cancer) > 1){
  Cancer = paste(Cancer[[1]], "etc.")
}

if(length(Diseases) == 1){
  if(Cancer == "Esophagus_Stomach"){
    if(Diseases == "Esophageal Squamous Cell Carcinoma (ESCC)"){
      Cancer = "Esophagus"
    }
    else{
      Cancer = "Stomach"
    }
  }
  if(Cancer == "Uterus"){
    if(Diseases == "Uterine Endometrioid Carcinoma_UEC_子宮類内膜癌"){
      Cancer = "Uterus_carcinoma"
    }
    if(Diseases == "Uterine Leiomyosarcoma_ULMS_子宮平滑筋肉腫"){
      Cancer = "Uterus_sarcoma"
    }
  }
  if(Cancer == "Lung"){
    if(Diseases == "Lung Adenocarcinoma (LUAD)"){
      Cancer = "LUAD"
    }
    if(Diseases == "Lung Squamous Cell Carcinoma (LUSC)"){
      Cancer = "LUSC"
    }
  }
}

if(all(Diseases %in% c("Stomach Adenocarcinoma, differentiated (STAD_dif)",
                     "Stomach Adenocarcinoma, undifferentiated (STAD_undif)",
                     "Stomach Adenocarcinoma, NOS (STAD_NOS)"))){
  Cancer = "Stomach"
}

if(ER_positive_BRCA == 1){
  Cancer = "ER_positive_BRCA"
}
if(!file.exists(paste("output/", Cancer, "_", Sys.Date(), sep=""))){
  dir.create(paste("output/", Cancer, "_", Sys.Date(), sep=""))
}
OutDirName = paste("output/", Cancer, "_", Sys.Date(), "/", sep="")
annotation_list = read.csv(file = "source/annotation_table.csv", skip = 0)
annotation_list[is.na(annotation_list$Annotation_correction),]$Annotation_correction = 0

annotation_list_1 = annotation_list[annotation_list$Annotation_correction == 1,]
Data_report$tmp = paste(Data_report$C.CAT調査結果.変異情報.マーカー,
                        Data_report$C.CAT調査結果.変異情報.変異内容,
                        sep = " _ ")
annotation_list_1$tmp = paste(annotation_list_1$Gene,
                              annotation_list_1$Variant,
                              sep = " _ ")

fusion_list = read.csv(file = "source/fusion_table.csv", skip = 0)
fusion_list$tmp = paste(fusion_list$Gene_Ref,
                              "fusion",
                              sep = " _ ")
if(Drug_treated_cases_only == 1){
  ID_list = read.csv(file = paste("source/ID_table_", Cancer, ".csv", sep=""), skip = 0)
  Data_case = Data_case %>% dplyr::filter(
  C.CAT調査結果.基本項目.ハッシュID %in% ID_list$ID
  )
}

Data_report = Data_report %>% 
  dplyr::mutate(C.CAT調査結果.エビデンス.エビデンスレベル = case_when(
    tmp %in% annotation_list_1$tmp ~ "G",
    TRUE ~ C.CAT調査結果.エビデンス.エビデンスレベル))
Data_report$tmp2 = unlist(lapply(list(Data_report$tmp), function(x) {
  as.vector(fusion_list$Gene_Alt[match(x, fusion_list$tmp)])}))

Data_report = Data_report %>% 
  dplyr::mutate(C.CAT調査結果.変異情報.マーカー = case_when(
    is.na(tmp2) ~ C.CAT調査結果.変異情報.マーカー,
    TRUE ~ tmp2))

for(i in Analyze_VUS){
  Data_report = Data_report %>% 
    dplyr::mutate(C.CAT調査結果.エビデンス.エビデンスレベル = case_when(
      C.CAT調査結果.変異情報.マーカー == i ~ "F",
      TRUE ~ C.CAT調査結果.エビデンス.エビデンスレベル))
}
Data_report = Data_report %>% 
  dplyr::mutate(C.CAT調査結果.エビデンス.エビデンスレベル = case_when(
    C.CAT調査結果.変異情報.マーカー == "STK11" &
      C.CAT調査結果.変異情報.変異内容 == "F354L" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "PALB2" &
      C.CAT調査結果.変異情報.変異内容 == "E837K" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "TP53" &
      C.CAT調査結果.変異情報.変異内容 == "E11Q" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "STK11" &
      C.CAT調査結果.変異情報.変異内容 == "P281L" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "MSH6" &
      C.CAT調査結果.変異情報.変異内容 == "K1358fs*2" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "MUTYH" &
      C.CAT調査結果.変異情報.変異内容 == "c.934-2A>G" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "RAD51" &
      C.CAT調査結果.変異情報.変異内容 == "M1fs*" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "TP53" &
      C.CAT調査結果.変異情報.変異内容 == "A189V" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "MLH1" &
      C.CAT調査結果.変異情報.変異内容 == "R385C" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "CDKN2A" &
      C.CAT調査結果.変異情報.変異内容 == "H83Y" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "CHEK2" &
      C.CAT調査結果.変異情報.変異内容 == "R181H" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "SDHA" &
      C.CAT調査結果.変異情報.変異内容 == "A236T" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "APC" &
      C.CAT調査結果.変異情報.変異内容 == "T1292M" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "ATM" &
      C.CAT調査結果.変異情報.変異内容 == "A220V" ~ "G",
    C.CAT調査結果.変異情報.マーカー == "TMB" &
      C.CAT調査結果.変異情報.変異内容 == "high" ~ "F",
    C.CAT調査結果.変異情報.マーカー == "MSI" &
      C.CAT調査結果.変異情報.変異内容 == "high" ~ "F",
    TRUE ~ C.CAT調査結果.エビデンス.エビデンスレベル))

if(EGFR_non_amp == 1){
  Data_report = Data_report %>% 
    dplyr::mutate(C.CAT調査結果.エビデンス.エビデンスレベル = case_when(
      C.CAT調査結果.変異情報.マーカー == "EGFR" &
        C.CAT調査結果.変異情報.変異内容 == "amplification" ~ "G",
      TRUE ~ C.CAT調査結果.エビデンス.エビデンスレベル))
}

Data_case$症例.管理情報.登録日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.管理情報.登録日)
Data_case$症例.転帰情報.最終生存確認日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.転帰情報.最終生存確認日)
Data_case$症例.転帰情報.死亡日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.転帰情報.死亡日)
Data_case$症例.検体情報.検体採取日.腫瘍組織. = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.検体情報.検体採取日.腫瘍組織.)
Data_case$症例.EP前レジメン情報.投与開始日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.EP前レジメン情報.投与開始日)
Data_case$症例.EP後レジメン情報.投与開始日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.EP後レジメン情報.投与開始日)
Data_case$症例.EP前レジメン情報.投与終了日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.EP前レジメン情報.投与終了日)
Data_case$症例.EP後レジメン情報.投与終了日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.EP後レジメン情報.投与終了日)
Data_case$症例.背景情報.診断日 = str_replace_all(replacement = "-", pattern = "/", string = Data_case$症例.背景情報.診断日)
Data_case$sampling_to_test = floor(difftime(as.Date(Data_case$症例.管理情報.登録日), as.Date(Data_case$症例.検体情報.検体採取日.腫瘍組織.), units = "days") / 365.25)
Data_case$sampling_to_test[is.na(Data_case$sampling_to_test)] = 0
Data_case$症例.基本情報.年齢 = as.integer(Data_case$症例.基本情報.年齢)
Data_case$sampling_age = as.integer(Data_case$症例.基本情報.年齢) - Data_case$sampling_to_test
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "（Genetical Recombination） ", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "（genetical recombination） ", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "（Genetical Recombination）", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "\\(Genetical Recombination\\)" , "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "（genetical recombination）", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Amrubicin", "Amurubicin")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "tegafur", "Tegafur")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "\\[Bevacizumab Biosimilar 1\\]", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "［Bevacizumab Biosimilar 2］", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "・", ",")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Hydrochloride", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Hydrate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Acetate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Hydrate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Sulfate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Maleate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " maleate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Mesilate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " ditartrate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Ditartrate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " gimeracil oteracil", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Sodium", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Acetonate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Phosphate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Chloride", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "（223Ra）chloride", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "-233chloride", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "-234chloride", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Calcium ", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Malate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Potassium", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Citrate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "nab-Paclitaxel", "nab-paclitaxel")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Liposomal Doxorubicin", "Liposomal doxorubicin")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Gemcitabine", "Gemci_tabine")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Gemcitabin", "Gemcitabine")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Gemci_tabine", "Gemcitabine")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Sulfoxide", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Dimethyl", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., " Tosilate", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Atezolizumab ", "Atezolizumab")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "［Trastuzumab Biosimilar 1］", "")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Trastuzumab Emtansine", "trastuzumab Emtansine")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Trastuzumab Deruxtecan", "trastuzumab Deruxtecan")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "治験", "Investigational Agent")
Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., "Liposomal doxorubicin", "Doxorubicin")

Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "（Genetical Recombination） ", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "（genetical recombination） ", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "（Genetical Recombination）", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "\\(Genetical Recombination\\)" , "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "（genetical recombination）", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Amrubicin", "Amurubicin")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "tegafur", "Tegafur")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "\\[Bevacizumab Biosimilar 1\\]", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "［Bevacizumab Biosimilar 2］", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "・", ",")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Hydrochloride", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Hydrate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Acetate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Hydrate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Sulfate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Maleate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " maleate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Mesilate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " ditartrate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Ditartrate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " gimeracil oteracil", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Sodium", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Acetonate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Phosphate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Chloride", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "（223Ra）chloride", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "-233chloride", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "-234chloride", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Calcium ", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Malate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Potassium", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Citrate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Tiperacil", "Tipiracil")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "nab-Paclitaxel", "nab-paclitaxel")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Liposomal Doxorubicin", "Liposomal doxorubicin")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Gemcitabine", "Gemci_tabine")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Gemcitabin", "Gemcitabine")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Gemci_tabine", "Gemcitabine")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Sulfoxide", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Dimethyl", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., " Tosilate", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Atezolizumab ", "Atezolizumab")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "［Trastuzumab Biosimilar 1］", "")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Trastuzumab Emtansine", "trastuzumab Emtansine")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Trastuzumab Deruxtecan", "trastuzumab Deruxtecan")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "治験", "Investigational Agent")
Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(Data_case$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., "Liposomal doxorubicin", "Doxorubicin")

Data_case$症例.転帰情報.最終生存確認日 = as.character(ymd(Data_case$症例.転帰情報.最終生存確認日, quiet = TRUE))
Data_case$症例.転帰情報.最終生存確認日 = tidyr::replace_na(Data_case$症例.転帰情報.最終生存確認日, "")
Data_case$症例.転帰情報.死亡日 = as.character(ymd(Data_case$症例.転帰情報.死亡日, quiet = TRUE))
Data_case$症例.転帰情報.死亡日 = tidyr::replace_na(Data_case$症例.転帰情報.死亡日, "")

## Preprocessing
diagnoses_raw = Data_case %>% dplyr::distinct(
  症例.背景情報.病理診断名, 症例.基本情報.がん種.OncoTree..名称., directory)
directories = unique(diagnoses_raw$directory)
diag_tmp = NULL
for(i in directories){
  if(file.exists(paste("source", i, "diagnoses_cor_table.csv", sep="/"))){
    tmp = read.csv(file = paste("source", i, "diagnoses_cor_table.csv", sep="/"), skip = 0)
    tmp$directory = i
    tmp = tmp[, colnames(tmp) != "z"]
  } else{
    tmp = diagnoses_raw %>% dplyr::filter(directory == i)
    tmp$y = i
  }
  diag_tmp = rbind(diag_tmp, tmp)
  diag_tmp = diag_tmp %>%
    dplyr::distinct(症例.基本情報.がん種.OncoTree..名称.,
                    症例.背景情報.病理診断名, .keep_all = TRUE)
}
diagnoses_raw$症例.基本情報.がん種.OncoTree..名称.[is.na(diagnoses_raw$症例.基本情報.がん種.OncoTree..名称.)] = ""
diagnoses_raw$y = ""
diagnoses_raw$z = paste(
  diagnoses_raw$症例.基本情報.がん種.OncoTree..名称., "_",
  diagnoses_raw$症例.背景情報.病理診断名, sep="")
diag_tmp$tmp = paste(
  diag_tmp$症例.基本情報.がん種.OncoTree..名称., "_",
  diag_tmp$症例.背景情報.病理診断名, sep="")
diagnoses_raw$y = unlist(lapply(diagnoses_raw$z, function(k) {
as.vector(diag_tmp$y[match(k,
          diag_tmp$tmp)])
}))
if(!file.exists(paste("output/disease_name_table/diagnoses_", Cancer, "_raw.xls", sep=""))){
  WriteXLS(diagnoses_raw, ExcelFileName=paste("output/disease_name_table/diagnoses_", Cancer, "_raw.xls", sep=""))
}

# Make correspondence table as diagnoses_cor_table.csv in UTF-8 format.  
# The 3rd column should be named as "y".
# Input curated disease names in the 3rd column.
# Save the table as "diagnoses_cor_table.csv".

## Explanation for Japanese users
#(1) diagnoses_raw.xlsをエクセルで読み込みます。  
#(2) 3列目の列名をyにします。
#(3) 3列目に標準化された病名を記載していきます。
#(4) csvファイルを"diagnoses_cor_table.csv"の名前でUTF-8形式で保存します。
#(5) 解析したい病名を複数あれば, # select the disease of interest において  
#　　c("病名1", "病名2", "病名3")という書式で記載します。

correspondence_table = NULL
for(i in directories){
  if(file.exists(paste("source", i, "diagnoses_cor_table.csv", sep="/")) & No_histology == 0){
    tmp = read.csv(file = paste("source", i, "diagnoses_cor_table.csv", sep="/"), skip = 0)
    tmp$directory = i
    tmp = tmp[, colnames(tmp) != "z"]
  } else{
    tmp = diagnoses_raw %>% dplyr::filter(directory == i)
    tmp$y = i
    tmp = tmp[, colnames(tmp) != "z"]
  }
  correspondence_table = rbind(correspondence_table, tmp)
}
Data_case$症例.基本情報.がん種.OncoTree..名称.[is.na(Data_case$症例.基本情報.がん種.OncoTree..名称.)] = ""
correspondence_table$z = paste(
  correspondence_table$症例.基本情報.がん種.OncoTree..名称., "_",
  correspondence_table$症例.背景情報.病理診断名, sep="")
Data_case$tmp = paste(
  Data_case$症例.基本情報.がん種.OncoTree..名称., "_",
  Data_case$症例.背景情報.病理診断名, sep="")
Data_case$diagnosis = lapply(list(Data_case$tmp),
                             function(k) {
  as.vector(correspondence_table$y[match(k,
            correspondence_table$z)])
})[[1]]

EP_option_ID = Data_case %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.) %>%
  dplyr::filter(症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. == "はい")
EP_option_ID = unique(EP_option_ID$C.CAT調査結果.基本項目.ハッシュID)
Data_case = Data_case %>%
  dplyr::mutate(症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. =
                  case_when(
                    C.CAT調査結果.基本項目.ハッシュID %in% EP_option_ID ~ "はい",
                    TRUE ~ "いいえ"))
EP_treat_ID = Data_case %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 症例.EP後レジメン情報.提示された治療薬を投与した.名称.) %>%
  dplyr::filter(症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した")
EP_treat_ID = unique(EP_treat_ID$C.CAT調査結果.基本項目.ハッシュID)
Data_case = Data_case %>%
  dplyr::mutate(症例.EP後レジメン情報.提示された治療薬を投与した.名称. =
                  case_when(
                    C.CAT調査結果.基本項目.ハッシュID %in% EP_treat_ID ~ "投与した",
                    TRUE ~ "投与しなかった"))

# combine disease names (optional)
Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Colon Adenocarcinoma (COAD)_大腸腺癌（直腸を除く）"),
        replacement = "Colorectal Adenocarcinoma (COADREAD)_結腸直腸腺癌"))
Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Rectal Adenocarcinoma (READ)_直腸腺癌"),
        replacement = "Colorectal Adenocarcinoma (COADREAD)_結腸直腸腺癌"))
Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Mucinous Adenocarcinoma of the Colon and Rectum (MACR)_結腸直腸粘液腺癌"),
        replacement = "Colorectal Adenocarcinoma (COADREAD)_結腸直腸腺癌"))
Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Signet Ring Cell Adenocarcinoma of the Colon and Rectum (SRCCR)_結腸直腸印環細胞腺癌"),
        replacement = "Colorectal Adenocarcinoma (COADREAD)_結腸直腸腺癌"))


Data_case = Data_case %>% dplyr::filter(症例.基本情報.性別.名称. == "男" | directory != "Prostate")

if(length(Data_case[is.na(Data_case$症例.転帰情報.最終生存確認日),]$症例.転帰情報.最終生存確認日) != 0){
  Data_case[is.na(Data_case$症例.転帰情報.最終生存確認日),]$症例.転帰情報.最終生存確認日 = ""
}
if(length(Data_case[is.na(Data_case$症例.転帰情報.死亡日),]$症例.転帰情報.死亡日) != 0){
  Data_case[is.na(Data_case$症例.転帰情報.死亡日),]$症例.転帰情報.死亡日 = ""
}
if(length(Data_case[is.na(Data_case$症例.背景情報.ECOG.PS),]$症例.背景情報.ECOG.PS) != 0){
  Data_case[is.na(Data_case$症例.背景情報.ECOG.PS),]$症例.背景情報.ECOG.PS = 9
}
if(length(Data_case[is.na(Data_case$症例.背景情報.喫煙歴有無.名称),]$症例.背景情報.喫煙歴有無.名称.) != 0){
  Data_case[is.na(Data_case$症例.背景情報.喫煙歴有無.名称),]$症例.背景情報.喫煙歴有無.名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.喫煙歴有無.名称. == "",]$症例.背景情報.喫煙歴有無.名称.) != 0){
  Data_case[Data_case$症例.背景情報.喫煙歴有無.名称. == "",]$症例.背景情報.喫煙歴有無.名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.喫煙歴有無.名称. == "不明",]$症例.背景情報.喫煙歴有無.名称.) != 0){
  Data_case[Data_case$症例.背景情報.喫煙歴有無.名称. == "不明",]$症例.背景情報.喫煙歴有無.名称. = NA
}
if(length(Data_case[is.na(Data_case$症例.背景情報.アルコール多飲有無.名称.),]$症例.背景情報.アルコール多飲有無.名称.) != 0){
  Data_case[is.na(Data_case$症例.背景情報.アルコール多飲有無.名称.),]$症例.背景情報.アルコール多飲有無.名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.アルコール多飲有無.名称. == "",]$症例.背景情報.アルコール多飲有無.名称.) != 0){
  Data_case[Data_case$症例.背景情報.アルコール多飲有無.名称. == "",]$症例.背景情報.アルコール多飲有無.名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.アルコール多飲有無.名称. == "不明",]$症例.背景情報.アルコール多飲有無.名称.) != 0){
  Data_case[Data_case$症例.背景情報.アルコール多飲有無.名称. == "不明",]$症例.背景情報.アルコール多飲有無.名称. = NA
}
if(length(Data_case[is.na(Data_case$症例.基本情報.性別.名称.),]$症例.基本情報.性別.名称.) != 0){
  Data_case[is.na(Data_case$症例.基本情報.性別.名称.),]$症例.基本情報.性別.名称. = "未入力・不明"
}
if(length(Data_case[Data_case$症例.基本情報.性別.名称. == "未入力・不明",]$症例.基本情報.性別.名称.) != 0){
  Data_case[Data_case$症例.基本情報.性別.名称. == "未入力・不明",]$症例.基本情報.性別.名称. = NA
}
if(length(Data_case[is.na(Data_case$症例.背景情報.多発がん有無.同一臓器..名称.),]$症例.背景情報.多発がん有無.同一臓器..名称.) != 0){
  Data_case[is.na(Data_case$症例.背景情報.多発がん有無.同一臓器..名称.),]$症例.背景情報.多発がん有無.同一臓器..名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.多発がん有無.同一臓器..名称. == "",]$症例.背景情報.多発がん有無.同一臓器..名称.) != 0){
  Data_case[Data_case$症例.背景情報.多発がん有無.同一臓器..名称. == "",]$症例.背景情報.多発がん有無.同一臓器..名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.多発がん有無.同一臓器..名称. == "不明",]$症例.背景情報.多発がん有無.同一臓器..名称.) != 0){
  Data_case[Data_case$症例.背景情報.多発がん有無.同一臓器..名称. == "不明",]$症例.背景情報.多発がん有無.同一臓器..名称. = NA
}
if(length(Data_case[is.na(Data_case$症例.背景情報.重複がん有無.異なる臓器..名称.),]$症例.背景情報.重複がん有無.異なる臓器..名称.) != 0){
  Data_case[is.na(Data_case$症例.背景情報.重複がん有無.異なる臓器..名称.),]$症例.背景情報.重複がん有無.異なる臓器..名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.重複がん有無.異なる臓器..名称. == "",]$症例.背景情報.重複がん有無.異なる臓器..名称.) != 0){
  Data_case[Data_case$症例.背景情報.重複がん有無.異なる臓器..名称. == "",]$症例.背景情報.重複がん有無.異なる臓器..名称. = "不明"
}
if(length(Data_case[Data_case$症例.背景情報.重複がん有無.異なる臓器..名称. == "不明",]$症例.背景情報.重複がん有無.異なる臓器..名称.) != 0){
  Data_case[Data_case$症例.背景情報.重複がん有無.異なる臓器..名称. == "不明",]$症例.背景情報.重複がん有無.異なる臓器..名称. = NA
}
if(length(Data_case[is.na(Data_case$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.),]$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.) != 0){
  Data_case[is.na(Data_case$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.),]$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. = "いいえ"
}
if(length(Data_case[Data_case$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. == "",]$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.) != 0){
  Data_case[Data_case$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. == "",]$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. = "いいえ"
}
if(length(Data_case[is.na(Data_case$症例.EP後レジメン情報.提示された治療薬を投与した.名称.),]$症例.EP後レジメン情報.提示された治療薬を投与した.名称.) != 0){
  Data_case[is.na(Data_case$症例.EP後レジメン情報.提示された治療薬を投与した.名称.),]$症例.EP後レジメン情報.提示された治療薬を投与した.名称. = "投与しなかった"
}
if(length(Data_case[Data_case$症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "",]$症例.EP後レジメン情報.提示された治療薬を投与した.名称.) != 0){
  Data_case[Data_case$症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "",]$症例.EP後レジメン情報.提示された治療薬を投与した.名称. = "投与しなかった"
}
if(length(Data_case[Data_case$症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "不明",]$症例.EP後レジメン情報.提示された治療薬を投与した.名称.) != 0){
  Data_case[Data_case$症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "不明",]$症例.EP後レジメン情報.提示された治療薬を投与した.名称. = "投与しなかった"
}

#TMB_ID = unique((Data_report %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー == "TMB" & C.CAT調査結果.エビデンス.エビデンスレベル =="F"))$C.CAT調査結果.基本項目.ハッシュID)
#Data_case =  Data_case %>% 
#    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% TMB_ID)
#Data_report =  Data_report %>% 
#    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% TMB_ID)


Data_summary =  Data_case %>% 
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1)) %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
      TRUE ~ 0))


if(Ex_inactive_multi == 1){
  Data_summary =  Data_summary %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}
Data_summary_all = Data_summary 
Data_summary = Data_summary %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.,
  症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
  diagnosis) 

theme_gtsummary_journal(journal = c("jama"), set_theme = TRUE)
theme_gtsummary_compact()

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table 1. All patients Characteristics**") %>%
  bold_labels()
table_tmp

# Germline mutations, only NCC oncopanel patients
# Only used NCC OncoPanel version >2
PGPV_gene = c("APC", "ATM", "BAP1", "BARD1", "BMPR1A", "BRCA1", "BRCA2", "BRIP1", "CDH1", "CDK4", "CDKN2A", "CHEK2", "EPCAM", "FH", "FLCN", "HNF1A", "MAX", "MEN1", "MET", "MLH1", "MSH2", "MSH6", "MUTYH", "NBN", "NF1", "NF2", "PALB2", "PMS2", "POLD1", "POLE", "POT1", "PTEN", "RAD51C", "RAD51D", "RB1", "RET", "SDHA", "SDHAF2", "SDHB", "SDHC", "SDHD", "SMAD3", "SMAD4", "SMARCB1", "STK11", "TERF2IP", "TERT", "TGFBR1", "TGFBR2", "TMEM127", "TP53", "TSC1", "TSC2", "VHL", "WT1") 
BRCA_gene = c("BRCA1", "BRCA2")
MSI_gene = c("MLH1", "MSH2", "MSH6", "PMS2")
U30_gene = c("APC", "NF1", "RB1", "TP53")
Data_report$REF = nchar(Data_report$C.CAT調査結果.変異情報.リファレンス塩基)
Data_report$ALT = nchar(Data_report$C.CAT調査結果.変異情報.塩基変化)
Data_report$VAF = as.numeric(str_split(Data_report$C.CAT調査結果.変異情報.変異アレル頻度, " ",simplify = TRUE)[,1])
Data_report[is.na(Data_report$VAF),]$VAF= 0

ID_MSI_H = (Data_report %>%
              dplyr::filter(C.CAT調査結果.変異情報.マーカー == "MSI" &
                            C.CAT調査結果.変異情報.変異内容 == "high")
            )$C.CAT調査結果.基本項目.ハッシュID

Data_case_age = Data_case_raw %>%
  dplyr::select(症例.基本情報.年齢,
                C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::distinct()
Data_report$age = unlist(lapply(list(Data_report$C.CAT調査結果.基本項目.ハッシュID),
                           function(k) {
as.vector(Data_case_age$症例.基本情報.年齢[match(k,
          Data_case_age$C.CAT調査結果.基本項目.ハッシュID)])
})[[1]])

Data_report$age = as.integer(Data_report$age)

Data_case_T_ratio = Data_case_raw %>%
  dplyr::select(症例.検体情報.腫瘍細胞含有割合,
                C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::distinct()
Data_report$T_ratio = unlist(lapply(list(Data_report$C.CAT調査結果.基本項目.ハッシュID),
                           function(k) {
as.vector(Data_case_T_ratio$症例.検体情報.腫瘍細胞含有割合[match(k,
          Data_case_T_ratio$C.CAT調査結果.基本項目.ハッシュID)])
})[[1]])

Data_report$T_ratio = as.integer(Data_report$T_ratio)

Data_report = Data_report %>%
  dplyr::mutate(MSI_H = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% ID_MSI_H ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(SNV = case_when(
    REF == 1 & ALT == 1 ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(BRCA = case_when(
    C.CAT調査結果.変異情報.マーカー %in% BRCA_gene ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(MSI = case_when(
    C.CAT調査結果.変異情報.マーカー %in% MSI_gene ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(PGPV = case_when(
    C.CAT調査結果.変異情報.マーカー %in% PGPV_gene ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(U30 = case_when(
    C.CAT調査結果.変異情報.マーカー %in% U30_gene ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(age_U30 = case_when(
    age < 30 ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(SNV = case_when(
    REF == 1 & ALT == 1 ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(VAF_threshold = case_when(
    VAF <0.10 ~ 0,
    VAF <0.20 ~ 1,
    VAF <0.30 ~ 2,
    TRUE ~ 3))

Data_report = Data_report %>%
  dplyr::mutate(PGPV_plus = case_when(
    Data_report$VAF_threshold > 0 & Data_report$BRCA == 1 &
      Data_report$C.CAT調査結果.エビデンス.エビデンスレベル == "F" ~ 1,
    Data_report$VAF_threshold > 0 & Data_report$MSI == 1 &
      Data_report$MSI_H == 1 &
      Data_report$C.CAT調査結果.エビデンス.エビデンスレベル == "F" ~ 1,
    Data_report$VAF_threshold > 1 & Data_report$U30 == 1 &
      Data_report$SNV == 0 &
      Data_report$age_U30 == 0 &
      Data_report$C.CAT調査結果.エビデンス.エビデンスレベル == "F" ~ 0,
    Data_report$VAF_threshold > 2 & Data_report$U30 == 1 &
      Data_report$SNV == 1 &
      Data_report$age_U30 == 0 &
      Data_report$C.CAT調査結果.エビデンス.エビデンスレベル == "F" ~ 0,
    Data_report$VAF_threshold > 1 & Data_report$PGPV == 1 &
      Data_report$SNV == 0 &
      Data_report$C.CAT調査結果.エビデンス.エビデンスレベル == "F" ~ 1,
    Data_report$VAF_threshold > 2 & Data_report$PGPV == 1 &
      Data_report$SNV == 1 &
      Data_report$C.CAT調査結果.エビデンス.エビデンスレベル == "F" ~ 1,
    TRUE ~ 0))

Data_report_PGPV = Data_report %>% dplyr::filter(PGPV_plus == 1)
# 
# c = c %>%
#   dplyr::mutate(PGPV_plus = case_when(
#     c$VAF_threshold > 0 & c$BRCA == 1 &
#       c$C.CAT調査結果.エビデンス.エビデンスレベル_WT == "F" ~ 1,
#     c$VAF_threshold > 0 & c$MSI == 1 &
#       c$MSI_H == 1 &
#       c$C.CAT調査結果.エビデンス.エビデンスレベル_WT == "F" ~ 1,
#     c$VAF_threshold > 1 & c$U30 == 1 &
#       c$SNV == 0 &
#       c$age_U30 == 0 &
#       c$C.CAT調査結果.エビデンス.エビデンスレベル_WT == "F" ~ 0,
#     c$VAF_threshold > 2 & c$U30 == 1 &
#       c$SNV == 1 &
#       c$age_U30 == 0 &
#       c$C.CAT調査結果.エビデンス.エビデンスレベル_WT == "F" ~ 0,
#     c$VAF_threshold > 1 & c$PGPV == 1 &
#       c$SNV == 0 &
#       c$C.CAT調査結果.エビデンス.エビデンスレベル_WT == "F" ~ 1,
#     c$VAF_threshold > 2 & c$PGPV == 1 &
#       c$SNV == 1 &
#       c$C.CAT調査結果.エビデンス.エビデンスレベル_WT == "F" ~ 1,
#     TRUE ~ 0))

#NOP_germline_gene = c("APC", "BRCA1", "BRCA2", "MLH1", "MSH2", "PTEN", "RB1", "RET", "SMAD4", "STK11", "TP53", "TSC1", "VHL")
Data_PGPV = Data_report %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  PGPV_plus,
                  .keep_all=TRUE) %>%
  dplyr::mutate(gene_mut = paste(C.CAT調査結果.変異情報.マーカー,
                                 C.CAT調査結果.変異情報.変異内容,
                                 sep=" _ ")) %>%
  dplyr::mutate(gene_patho = C.CAT調査結果.変異情報.マーカー) %>%
  dplyr::filter(str_detect(C.CAT調査結果.基本項目.パネル名,
                 pattern="NCC オンコパネル システム ver.2"))# %>%
  #dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% NOP_germline_gene)
PGPV_mutations = Data_PGPV %>%
                dplyr::count(gene_mut) %>%
                dplyr::arrange(-n)
PGPV_mutations$NCC_germline = 0
for(i in 1:length(PGPV_mutations$n)){
  PGPV_mutations$NCC_germline[i] = sum((Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_mut == PGPV_mutations$gene_mut[i])
}
PGPV_mutations = PGPV_mutations %>%
                dplyr::arrange(-NCC_germline) %>%
                dplyr::arrange(-n)
PGPV_mutations$NCC_PGPV = 0
for(i in 1:length(PGPV_mutations$n)){
  PGPV_mutations$NCC_PGPV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1))$gene_mut == PGPV_mutations$gene_mut[i])
}
PGPV_mutations = PGPV_mutations %>%
                dplyr::arrange(-NCC_PGPV) %>%
                dplyr::arrange(-n)
PGPV_mutations$NCC_PGPV_PGV  = 0
for(i in 1:length(PGPV_mutations$n)){
  PGPV_mutations$NCC_PGPV_PGV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1 & C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_mut == PGPV_mutations$gene_mut[i])
}
PGPV_mutations = PGPV_mutations %>%
                dplyr::arrange(-NCC_PGPV_PGV) %>%
                dplyr::arrange(-n)
colnames(PGPV_mutations) = c("Mutation", "mutation counts", "NCC validated PGV", "NCC PGPV", "NCC PGPV & PGV")
DT::datatable(PGPV_mutations, filter = 'top', extensions = 'Scroller', 
             caption = paste("All", Cancer ,"patients with NCC oncopanel version >=2",length(unique((Data_report %>% dplyr::filter(str_detect(C.CAT調査結果.基本項目.パネル名, pattern="NCC オンコパネル システム ver.2")))$C.CAT調査結果.基本項目.ハッシュID)), "patients,", length((unique(Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$C.CAT調査結果.基本項目.ハッシュID)), "patients had validated germline mutations."))
WriteXLS(PGPV_mutations, ExcelFileName = paste(OutDirName, Cancer, "_NOP_PGPV_mutation.xls", sep=""))

PGPV_mutations = Data_PGPV %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
PGPV_mutations$NCC_germline = 0
for(i in 1:length(PGPV_mutations$n)){
  PGPV_mutations$NCC_germline[i] = sum((Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_patho == PGPV_mutations$gene_patho[i])
}
PGPV_mutations = PGPV_mutations %>%
                dplyr::arrange(-NCC_germline) %>%
                dplyr::arrange(-n)
PGPV_mutations$NCC_PGPV = 0
for(i in 1:length(PGPV_mutations$n)){
  PGPV_mutations$NCC_PGPV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1))$gene_patho == PGPV_mutations$gene_patho[i])
}
PGPV_mutations = PGPV_mutations %>%
                dplyr::arrange(-NCC_PGPV) %>%
                dplyr::arrange(-n)
PGPV_mutations$NCC_PGPV_PGV  = 0
for(i in 1:length(PGPV_mutations$n)){
  PGPV_mutations$NCC_PGPV_PGV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1 & C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_patho == PGPV_mutations$gene_patho[i])
}
PGPV_mutations = PGPV_mutations %>%
                dplyr::arrange(-NCC_PGPV_PGV) %>%
                dplyr::arrange(-n)
colnames(PGPV_mutations) = c("Mutation", "mutation counts", "NCC validated PGV", "NCC PGPV", "NCC PGPV & PGV")
DT::datatable(PGPV_mutations, filter = 'top', extensions = 'Scroller', 
             caption = paste("All", Cancer,"patients with target disease and NCC oncopanel version >=2",length(unique((Data_report %>% dplyr::filter(str_detect(C.CAT調査結果.基本項目.パネル名, pattern="NCC オンコパネル システム ver.2")))$C.CAT調査結果.基本項目.ハッシュID)), "patients,", length((unique(Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$C.CAT調査結果.基本項目.ハッシュID)), "patients had validated germline mutations."))

WriteXLS(PGPV_mutations, ExcelFileName = paste(OutDirName, Cancer, "_NOP_PGPV_gene.xls", sep=""))

Data_summary =  Data_case %>% 
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ "")) %>%
  dplyr::arrange(final_observe) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
 # dplyr::filter(症例.転帰情報.転帰.名称. != "") %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
 # dplyr::filter(症例.背景情報.診断日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))
  Data_summary = Data_summary %>%
  dplyr::mutate(enrollment = case_when(
    Data_summary$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_summary$症例.管理情報.登録日 >= "2020-07-01" &
    Data_summary$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_summary$症例.管理情報.登録日 >= "2021-07-01" &
    Data_summary$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_summary$症例.管理情報.登録日 >= "2022-01-01" &
    Data_summary$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))

  Data_summary = Data_summary %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_summary =  Data_summary %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_summary = Data_summary %>%
  dplyr::mutate(final_observe = as.Date(Data_summary$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_summary$症例.管理情報.登録日))# %>%
Data_summary = Data_summary %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_summary$final_observe,
                                      Data_summary$症例.管理情報.登録日,
                                      units = "days")))
Data_summary = Data_summary %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))
Data_summary = Data_summary %>%
  dplyr::filter(time_enroll_final > 0)

ID_prog = unique(Data_summary$C.CAT調査結果.基本項目.ハッシュID) 

Data_summary_prog = Data_summary
Data_summary = Data_summary %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.,
  症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
  症例.転帰情報.転帰.名称.,
  enrollment,
#  time_diag_final,
  time_enroll_final,
#  time_diag_enroll,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table 2. Patients with prognostic information, Characteristics**") %>%
  bold_labels()
table_tmp


# SCC is renamed as NEC
if(SCC_is_NEC == 1){
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Small Cell Carcinoma of the Stomach (STSC)"),
          replacement = "Gastrointestinal Neuroendocrine Carcinoma of the Esophagus/Stomach"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="BLAD_SCC"),
          replacement = "Neuroendocrine Carcinoma of the Bladder"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="BLAD_NEC"),
          replacement = "Neuroendocrine Carcinoma of the Bladder"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Small Cell Carcinoma of the Ovary_SCCO_卵巣小細胞癌"),
          replacement = "NEC_Ovary"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Small Cell Carcinoma of the Cervix_SCCE_小細胞癌"),
          replacement = "Cervix_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mixed Cervical Carcinoma_MCCE_混合癌"),
          replacement = "Cervix_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Neuroendocrine Tumor_CENE_神経内分泌腫瘍"),
          replacement = "Cervix_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Neuroendocrine Tumor_CENEC_神経内分泌癌"),
          replacement = "Cervix_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Small Cell Lung Cancer (SCLC)"),
          replacement = "Lung_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="PRSCC"),
          replacement = "PRNEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="PANET_膵神経内分泌腫瘍"),
          replacement = "PANEC_膵神経内分泌癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="SCC_Kidney"),
          replacement = "NEC_Kidney"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Small Cell Gallbladder Carcinoma (SCGBC)_胆嚢小細胞癌"),
          replacement = "NEC_Bile"))
}

# Combining diseases
if(Disease_simple == 1){
  if(STAD_diff == 1){
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(Data_case$diagnosis,
            fixed(pattern="Diffuse Type Stomach Adenocarcinoma (DSTAD)"),
            replacement = "Stomach Adenocarcinoma, undifferentiated (STAD_undif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(Data_case$diagnosis,
            fixed(pattern="Tubular Stomach Adenocarcinoma (TSTAD)"),
            replacement = "Stomach Adenocarcinoma, differentiated (STAD_dif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Papillary Stomach Adenocarcinoma (PSTAD)"),
            replacement = "Stomach Adenocarcinoma, differentiated (STAD_dif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Intestinal Type Stomach Adenocarcinoma (ISTAD)"),
            replacement = "Stomach Adenocarcinoma, differentiated (STAD_dif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Mucinous Stomach Adenocarcinoma (MSTAD)"),
            replacement = "Stomach Adenocarcinoma, undifferentiated (STAD_undif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Signet Ring Cell Carcinoma of the Stomach (SSRCC)"),
            replacement = "Stomach Adenocarcinoma, undifferentiated (STAD_undif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Poorly Differentiated Carcinoma of the Stomach (SPDAC)"),
            replacement = "Stomach Adenocarcinoma, undifferentiated (STAD_undif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Undifferentiated Stomach Adenocarcinoma (USTAD)"),
            replacement = "Stomach Adenocarcinoma, undifferentiated (STAD_undif)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Gastric Remnant Adenocarcinoma (GRC)"),
            replacement = "Stomach Adenocarcinoma, NOS (STAD_NOS)"))
    Data_case = Data_case %>% dplyr::mutate(
      diagnosis=str_replace_all(diagnosis,
            fixed(pattern="Stomach Adenocarcinoma (STAD)"),
            replacement = "Stomach Adenocarcinoma, NOS (STAD_NOS)"))
  }
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(Data_case$diagnosis,
          fixed(pattern="Diffuse Type Stomach Adenocarcinoma (DSTAD)"),
          replacement = "Stomach Adenocarcinoma (STAD)"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(Data_case$diagnosis,
          fixed(pattern="Tubular Stomach Adenocarcinoma (TSTAD)"),
          replacement = "Stomach Adenocarcinoma (STAD)"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Neuroendocrine Carcinoma of the Esophagus (HGNEE)"),
          replacement = "Gastrointestinal Neuroendocrine Carcinoma of the Esophagus/Stomach"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Tubular Stomach Adenocarcinoma (TSTAD)"),
          replacement = "Stomach Adenocarcinoma (STAD)"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Neuroendocrine Carcinoma of the Stomach (HGNES)"),
          replacement = "Gastrointestinal Neuroendocrine Carcinoma of the Esophagus/Stomach"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Papillary Stomach Adenocarcinoma (PSTAD)"),
          replacement = "Stomach Adenocarcinoma (STAD)"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Intestinal Type Stomach Adenocarcinoma (ISTAD)"),
          replacement = "Stomach Adenocarcinoma (STAD)"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mucinous Stomach Adenocarcinoma (MSTAD)"),
          replacement = "Stomach Adenocarcinoma (STAD)"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Low-Grade Endometrial Stromal Sarcoma_LGESS_低異型度子宮内膜間質肉腫"),
          replacement = "Endometrial Stromal Sarcoma_ESS_子宮内膜間質肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Endometrial Stromal Sarcoma_HGESS_高異型度子宮内膜間質肉腫"),
          replacement = "Endometrial Stromal Sarcoma_ESS_子宮内膜間質肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Endometrial Stromal Sarcoma_ESS_子宮内膜間質肉腫"),
          replacement = "Uterine Endometrial Stromal Sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Dedifferentiated Carcinoma_UDDC_子宮脱分化癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Poorly Differentiated Carcinoma of the Uterus_UPDC_子宮低分化癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Adenosquamous Carcinoma_UASC_子宮腺扁平上皮癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Epithelioid Leiomyosarcoma_UELMS_類上皮平滑筋肉腫"),
          replacement = "Uterine Leiomyosarcoma_ULMS_子宮平滑筋肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Mesonephric Carcinoma_UMNC_子宮中腎癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Mixed Endometrial Carcinoma_UMEC"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Mucinous Carcinoma_UMC_子宮粘液性癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Myxoid Leiomyosarcoma_UMLMS_子宮粘液様平滑筋肉腫"),
          replacement = "Uterine Leiomyosarcoma_ULMS_子宮平滑筋肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Perivascular Epithelioid Cell Tumor_UPECOMA_子宮血管周囲性類上皮細胞性腫瘍"),
          replacement = "Uterine other sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Undifferentiated Carcinoma_UUC_子宮未分化癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Placental Site Trophoblastic Tumor (PSTT)_胎盤部トロホブラスト腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Undifferentiated Uterine Sarcoma_UUS_子宮未分化肉腫"),
          replacement = "Uterine other sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Leiomyoma (ULM)_子宮平滑筋腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Carcinosarcoma/Uterine Malignant Mixed Mullerian Tumor_UCS_子宮癌肉腫"),
          replacement = "Uterine Carcinosarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Leiomyoma_ULM_子宮平滑筋腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Mesonephric Carcinoma (UMNC)_子宮中腎癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Neuroendocrine Carcinoma_UNEC_子宮神経内分泌腫瘍"),
          replacement = "Uterine_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Sarcoma, Other_OUSARC_子宮肉腫その他"),
          replacement = "Uterine other sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Serous Carcinoma/Uterine Papillary Serous Carcinoma_USC_子宮体部漿液性癌	"),
          replacement = "Uterine Serous Carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Uterine Smooth Muscle Tumor of Uncertain Malignant Potential_USTUMP_悪性度不明な子宮平滑筋腫瘍"),
          replacement = "Uterine other sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Choriocarcinoma_UCCA_絨毛癌"),
          replacement = "Uterine other carcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="BLAD_SCC"),
          replacement = "BLAD_Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="BLAD_AD"),
          replacement = "BLAD_Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="BLAD_Melanoma"),
          replacement = "BLAD_Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="BLAD_Unknown"),
          replacement = "BLAD_Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="BLAD_Uret"),
          replacement = "BLAD_Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Neuroendocrine Carcinoma of the Ovary_HGONEC_卵巣神経内分泌腫瘍"),
          replacement = "NEC_Ovary"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Brenner Tumor, Malignant_BTMOV_悪性ブレンナー腫瘍"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Brenner Tumor (BTOV)_ブレンナー腫瘍"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Carcinoid tumor_カルチノイド腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Clear Cell Borderline Ovarian Tumor (CCBOV)_明細胞境界悪性卵巣腫瘍"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Dysgerminoma_ODYS_未分化胚細胞腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Granulosa Cell Tumor_GRCT_顆粒膜細胞腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Serous Fallopian Tube Cancer_HGSFT_高悪性度漿液性卵管癌"),
          replacement = "High-Grade Serous Cancer"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Serous Ovarian Cancer_HGSOC_高異型度漿液性癌"),
          replacement = "High-Grade Serous Cancer"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Neuroendocrine Carcinoma of the Ovary_HGONEC_卵巣神経内分泌腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Immature Teratoma_OIMT_未熟奇形腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Juvenile Granulosa Cell Tumor_JGRCT_若年型顆粒膜細胞腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Leiomyosarcoma_卵巣平滑筋肉腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mature Teratoma malignant transformation_OMT_成熟奇形腫悪性転化"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mixed Germ Cell Tumor_OMGCT_卵巣混合性胚細胞腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mixed Ovarian Carcinoma_MXOV_混合癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mucinous Borderline Ovarian Tumor_MBOV_粘液性境界悪性腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Adenosarcoma_卵巣腺肉腫"),
          replacement = "Ovarian Carcinosarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Ovarian Carcinosarcoma/Malignant Mixed Mesodermal Tumor_OCS_卵巣癌肉腫"),
          replacement = "Ovarian Carcinosarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mucinous Carcinoid_粘液性カルチノイド"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mucinous Ovarian Cancer_MOV_粘液性癌"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Ovarian Choriocarcinoma, NOS_OCNOS_卵巣絨毛癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Ovarian Germ Cell Tumor_OGCT_卵巣胚細胞腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Ovarian Seromucinous Borderline Tumor_OSMBT_漿液粘液性境界悪性腫瘍"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Serous Borderline Ovarian Tumor_SBOV_漿液性境界悪性腫瘍"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Sertoli-Leydig Cell Tumor_SLCT_卵巣セルトリー・ライデッグ細胞腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="NEC_Ovary"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Steroid Cell Tumor, NOS_SCT_悪性ステロイド細胞腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Struma Ovarii_卵巣甲状腺腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Yolk Sac Tumor_OYST_卵黄嚢腫瘍"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Adenocarcinoma Other_その他の腺癌"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Low-Grade Serous Ovarian Cancer_LGSOC_低異型度漿液性癌"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mixed Ovarian Carcinoma_MXOV_混合癌"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Ovarian Seromucinous Carcinoma_OSMCA_卵巣漿液粘液性癌"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Serous Ovarian Cancer SOC_漿液性卵巣癌"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Squamous Cell Carcinoma_SCC_扁平上皮癌"),
          replacement = "Ovarian Epithelial Tumor, Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Clear Cell Ovarian Cancer_CCOV_明細胞癌"),
          replacement = "Ovary_Clear Cell Cancer"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Endometrioid Ovarian Cancer_EOV_類内膜癌"),
          replacement = "Ovary_Endometrioid Cancer"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Serous Cancer"),
          replacement = "Ovary_High-Grade Serous Cancer"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mucinous Ovarian Cancer_MOV_粘液性癌"),
          replacement = "Ovary_Mucinous Cancer"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Ovarian Carcinosarcoma"),
          replacement = "Ovary_Carcinosarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Ovarian Epithelial Tumor, Other"),
          replacement = "Ovary_Epithelial Tumor, Other"))
  if(Cervix_HPV == 1){
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Clear Cell Carcinoma_CECC_明細胞癌"),
          replacement = "HPV-independent adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="ervical Adenosquamous Carcinoma_CEAS_腺扁平上皮癌"),
          replacement = "HPV-associated adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Basaloid Squamous Sell Carcinoma_BSCC_類基底細胞扁平上皮癌"),
          replacement = "SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Squamous Cell Carcinoma_CESC_扁平上皮癌"),
          replacement = "SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Endocervical Adenocarcinoma_ECAD_腺癌"),
          replacement = "HPV-associated adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Gastric Type Mucinous Carcinoma_GCEMU_胃型粘液性癌"),
          replacement = "HPV-independent adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Endometrioid Carcinoma_CEEN_類内膜癌"),
          replacement = "HPV-independent adeno"))
  }
  if(Cervix_analysis_adeno_SCC_NEC == 1){
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Clear Cell Carcinoma_CECC_明細胞癌"),
          replacement = "Cervix_adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Adenosquamous Carcinoma_CEAS_腺扁平上皮癌"),
          replacement = "Cervix_adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Basaloid Squamous Sell Carcinoma_BSCC_類基底細胞扁平上皮癌"),
          replacement = "Cervix_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Squamous Cell Carcinoma_CESC_扁平上皮癌"),
          replacement = "Cervix_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Endocervical Adenocarcinoma_ECAD_腺癌"),
          replacement = "Cervix_adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Gastric Type Mucinous Carcinoma_GCEMU_胃型粘液性癌"),
          replacement = "Cervix_adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Endometrioid Carcinoma_CEEN_類内膜癌"),
          replacement = "Cervix_adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Adenocarcinoma In Situ_CEAIS_上皮内腺癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Adenoid Cystic Carcinoma_CACC_腺様嚢胞癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical poorly Differentiated Carcinoma_低分化癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Serous Carcinoma_CESE_漿液性癌"),
          replacement = "Cervix_adeno"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical undifferentiated Carcinoma_未分化腺癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical_Carcinosarcoma_癌肉腫"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Glassy Cell Carcinoma of the Cervix_CEGCC_すりガラス細胞癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mixed Cervical Carcinoma_MCCE_混合癌"),
          replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Signet Ring Mucinous Carcinoma_SCEMU_印環細胞癌"),
          replacement = "Cervix_adeno"))
  }
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Adenocarcinoma In Situ_CEAIS_上皮内腺癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Adenoid Cystic Carcinoma_CACC_腺様嚢胞癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Adenosquamous Carcinoma_CEAS_腺扁平上皮癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Basaloid Squamous Sell Carcinoma_BSCC_類基底細胞扁平上皮癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Neuroendocrine Tumor_CENE_神経内分泌腫瘍"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical poorly Differentiated Carcinoma_低分化癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical undifferentiated Carcinoma_未分化腺癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical_Carcinosarcoma_癌肉腫"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Endocervical Adenocarcinoma_ECAD_腺癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Glassy Cell Carcinoma of the Cervix_CEGCC_すりガラス細胞癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mixed Cervical Carcinoma_MCCE_混合癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Signet Ring Mucinous Carcinoma_SCEMU_印環細胞癌"),
          replacement = "Cervix_non_SCC"))

  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Clear Cell Carcinoma_CECC_明細胞癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Endometrioid Carcinoma_CEEN_類内膜癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Cervical Serous Carcinoma_CESE_漿液性癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Endocervical Adenocarcinoma_ECAD_腺癌"),
          replacement = "Cervix_non_SCC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Large Cell Neuroendocrine Carcinoma (LUNE)"),
          replacement = "Lung_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Combined Small Cell Lung Carcinoma (CSCLC)"),
          replacement = "Lung_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Large Cell Neuroendocrine Carcinoma (LUNE)"),
          replacement = "Lung_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Lung Neuroendocrine Carcinoma (LNEC)"),
          replacement = "Lung_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Neuroendocrine Carcinoma of the Colon and Rectum (HGNEC)_結腸直腸高悪性度神経内分泌腫瘍"),
          replacement = "NEC_BOWEL"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Small Bowel Well-Differentiated Neuroendocrine Carcinoma (SBWDNEC)_小腸高分化神経内分泌癌"),
          replacement = "NEC_BOWEL"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Gastrointestinal Neuroendocrine Carcinoma (GINEC)_消化管神経内分泌癌"),
          replacement = "NEC_BOWEL"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="carcinoma with neuroendocrine features"),
          replacement = "Breast_NEC"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Pilocytic astrocytoma"),
          replacement = "Astrocytoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Acinic Cell Carcinoma (ACCC)_腺房細胞癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Adenoid Cystic Carcinoma (ACYC)_腺様嚢胞癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Basal Cell Adenocarcinoma (BCAC)_基底細胞がん"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Carcinoma ex Pleomorphic Adenoma (CAEXPA)_悪性多形性腺腫"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Sinonasal Squamous Cell Carcinoma (SNSC)_副鼻腔扁平上皮癌"),
          replacement = "Head and Neck Squamous Cell Carcinoma (HNSC)_頭頸部扁平上皮癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Nasopharyngeal Carcinoma (NPC)_上咽頭癌"),
          replacement = "Head and Neck Squamous Cell Carcinoma (HNSC)_頭頸部扁平上皮癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mammary Analogue Secretory Carcinoma of Salivary Gland Origin (HNMASC)_唾液腺起源の乳腺相似分泌癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Myoepithelial Carcinoma (MYEC)_筋上皮癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mucoepidermoid Carcinoma (MUCC)_粘表皮癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Oral Cavity Squamous Cell Carcinoma (OCSC)_口腔内扁平上皮癌"),
          replacement = "Head and Neck Squamous Cell Carcinoma (HNSC)_頭頸部扁平上皮癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Salivary Adenocarcinoma (SAAD)_唾液腺腺癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Pleomorphic Adenoma (PADA)_多形腺腫"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Salivary Carcinoma, Other (OSACA)_その他の唾液癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Polymorphous Adenocarcinoma (PAC)_多型腺がん"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Salivary Duct Carcinoma (SDCA)_唾液腺導管癌"),
          replacement = "Salivary Carcinoma (SACA)_唾液腺癌"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Adenosquamous Carcinoma of the Gallbladder (GBASC)_胆嚢腺扁平上皮癌"),
        replacement = "Gallbladder Cancer (GBC)_胆嚢癌"))
  #Data_case = Data_case %>% dplyr::mutate(
  #diagnosis=str_replace_all(diagnosis,
  #      fixed(pattern="Perihilar Cholangiocarcinoma (PHCH)_肝門部胆管癌"),
  #      replacement = "Cholangiocarcinoma (CHOL)_胆管癌"))
  #Data_case = Data_case %>% dplyr::mutate(
  #diagnosis=str_replace_all(diagnosis,
  #      fixed(pattern="Extrahepatic Cholangiocarcinoma (EHCH)_肝外胆管癌"),
  #      replacement = "Cholangiocarcinoma (CHOL)_胆管癌"))
  #Data_case = Data_case %>% dplyr::mutate(
  #diagnosis=str_replace_all(diagnosis,
  #      fixed(pattern="Intrahepatic Cholangiocarcinoma (IHCH)_肝内胆管癌"),
  #      replacement = "Cholangiocarcinoma (CHOL)_胆管癌"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Gallbladder Adenocarcinoma, NOS (GBAD)_胆嚢腺癌、特定不能"),
        replacement = "Gallbladder Cancer (GBC)_胆嚢癌"))
  #Data_case = Data_case %>% dplyr::mutate(
  #diagnosis=str_replace_all(diagnosis,
  #      fixed(pattern="Intracholecystic Papillary Neoplasm (ICPN)_胆嚢内乳頭状腫瘍"),
  #      replacement = "Gallbladder Cancer (GBC)_胆嚢癌"))
  #Data_case = Data_case %>% dplyr::mutate(
  #diagnosis=str_replace_all(diagnosis,
  #      fixed(pattern="Hepatocellular Carcinoma plus Intrahepatic Cholangiocarcinoma (HCCIHCH)_混合型肝細胞癌"),
  #      replacement = "Intrahepatic Cholangiocarcinoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Acral Melanoma (ACRM)_先端黒色腫"),
        replacement = "Melanoma (MEL)_悪性黒色腫"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Cutaneous Melanoma (SKCM)_皮膚黒色腫"),
        replacement = "Melanoma (MEL)_悪性黒色腫"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Lentigo Maligna Melanoma (SKLMM)_悪性黒子黒色腫"),
        replacement = "Melanoma (MEL)_悪性黒色腫"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Melanoma of Unknown Primary (MUP)_原発不明の黒色腫"),
        replacement = "Melanoma (MEL)_悪性黒色腫"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Spitzoid Melanoma (SPZM)_Spitz 母斑様黒色腫"),
        replacement = "Melanoma (MEL)_悪性黒色腫"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Porocarcinoma/Spiroadenocarcinoma (POCA)_汗孔癌/らせん腺癌"),
        replacement = "Skin Adnexal Carcinoma (SKAC)_皮膚付属器癌"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Sebaceous Carcinoma (SEBA)_皮脂腺癌"),
        replacement = "Skin Adnexal Carcinoma (SKAC)_皮膚付属器癌"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Sweat Gland Carcinoma/Apocrine Eccrine Carcinoma (AECA)_汗腺癌/アポクリン・エクリン癌"),
        replacement = "Skin Adnexal Carcinoma (SKAC)_皮膚付属器癌"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Sweat Gland Adenocarcinoma (SGAD)_汗腺腺癌"),
        replacement = "Skin Adnexal Carcinoma (SKAC)_皮膚付属器癌"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Merkel Cell Carcinoma (MCC)_メルケル細胞癌"),
        replacement = "NEC_skin"))
  if(Ultra_rare_sarcoma == 0){
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Chondroblastic Osteosarcoma (CHOS)_軟骨芽細胞型骨肉腫"),
          replacement = "Osteosarcoma (OS)_骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Telangiectatic Osteosarcoma (TEOS)_血管拡張型骨肉腫"),
          replacement = "Osteosarcoma (OS)_骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Secondary Osteosarcoma (SECOS)_二次性骨肉腫"),
          replacement = "Osteosarcoma (OS)_骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Fibroblastic Osteosarcoma (FIOS)_線維芽細胞型骨肉腫"),
          replacement = "Osteosarcoma (OS)_骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Osteoblastic Osteosarcoma (OSOS)_骨芽細胞型骨肉腫"),
          replacement = "Osteosarcoma (OS)_骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Small Cell Osteosarcoma (SCOS)_小細胞型骨肉腫"),
          replacement = "Osteosarcoma (OS)_骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="High-Grade Surface Osteosarcoma (HGSOS)_高悪性度表在性骨肉腫"),
          replacement = "Osteosarcoma (OS)_骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Dedifferentiated Chordoma (DDCHDM)_脱分化型脊索腫"),
          replacement = "Chordoma (CHDM)_脊索腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Conventional Type Chordoma (CCHDM)_通常型脊索腫"),
          replacement = "Chordoma (CHDM)_脊索腫"))
  
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Angiosarcoma of bone"),
          replacement = "Malignancy in giant cell tumor of bone"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Angiosarcoma of bone"),
          replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Leiomyosarcoma of bone"),
          replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Dedifferentiated Chondrosarcoma (DDCHS)_脱分化型軟骨肉腫"),
          replacement = "Chondrosarcoma (CHS)_軟骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mesenchymal Chondrosarcoma (MCHS)_間葉型軟骨肉腫"),
          replacement = "Mesenchymal_Chondrosarcoma (MCHS)_間葉型軟骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="mesenchymal chondrosarcoma"),
          replacement = "Mesenchymal Chondrosarcoma (MCHS)_間葉型軟骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mesenchymal_Chondrosarcoma (MCHS)_間葉型軟骨肉腫"),
          replacement = "Mesenchymal Chondrosarcoma (MCHS)_間葉型軟骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Extraskeletal Myxoid Chondrosarcoma (EMCHS)_骨外性粘液性軟骨肉腫"),
          replacement = "Extraskeletal_Myxoid Chondrosarcoma (EMCHS)_骨外性粘液性軟骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Extraskeletal myxoid chondrosarcoma"),
          replacement = "Extraskeletal Myxoid Chondrosarcoma (EMCHS)_骨外性粘液性軟骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Extraskeletal_Myxoid Chondrosarcoma (EMCHS)_骨外性粘液性軟骨肉腫"),
          replacement = "Extraskeletal Myxoid Chondrosarcoma (EMCHS)_骨外性粘液性軟骨肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Spindle Cell/Sclerosing Rhabdomyosarcoma (SCSRMS)_紡錘形細胞型/硬化型横紋筋肉腫"),
          replacement = "Rhabdomyosarcoma (RMS)_横紋筋肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Pleomorphic Rhabdomyosarcoma (PLRMS)_多形型横紋筋肉腫"),
          replacement = "Rhabdomyosarcoma (RMS)_横紋筋肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Spindle Cell Rhabdomyosarcoma (SCRMS)_紡錘形細胞型横紋筋肉腫"),
          replacement = "Rhabdomyosarcoma (RMS)_横紋筋肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma (MFH)_未分化多形肉腫/悪性線維性組織球腫/高悪性度紡錐細胞肉腫"),
          replacement = "Undifferentiated Pleomorphic Sarcoma(UPS)_未分化多形肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Embryonal Rhabdomyosarcoma (ERMS)_胎児型横紋筋肉腫"),
          replacement = "Rhabdomyosarcoma (RMS)_横紋筋肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Alveolar Rhabdomyosarcoma (ARMS)_胞巣型横紋筋肉腫"),
          replacement = "Rhabdomyosarcoma (RMS)_横紋筋肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Well-Differentiated Liposarcoma (WDLS)_高分化型脂肪肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Soft Tissue Myoepithelial Carcinoma (STMYEC)_軟部の筋上皮癌"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Sclerosing Epithelioid Fibrosarcoma (SEF)_硬化性類上皮線維肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Sarcoma, NOS (SARCNOS)_肉腫、特定不能"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Round Cell Sarcoma, NOS (RCSNOS)_円形細胞肉腫、特定不能"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Pleomorphic Liposarcoma (PLLS)_多形型脂肪肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Perivascular Epithelioid Cell Tumor (PECOMA)_血管周囲性類上皮細胞性腫瘍"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Parosteal Osteosarcoma (PAOS)"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Periosteal Osteosarcoma (PEOS)"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Paraganglioma (PGNG)_傍神経節腫 パラガングリオーマ"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="MPNST"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Mesenchymal Chondrosarcoma (MCHS)_間葉型軟骨肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="malignant phyllodes tumor"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="malignant granular cell tumor"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Liposarcoma (LIPO)_脂肪肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Low-Grade Fibromyxoid Sarcoma (LGFMS)_低悪性度線維粘液肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Infantile Fibrosarcoma (IFS)_乳児型線維肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Histiocytic Dendritic Cell Sarcoma (HDCS)_組織球性樹状細胞肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Hemangioma (HEMA)_血管腫"),
        replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Glomangiosarcoma (GS)_悪性グロムス腫瘍"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Giant Cell Tumor of Bone (GCTB)_骨巨細胞腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Fibrosarcoma (FIBS)_線維肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Extraskeletal Myxoid Chondrosarcoma (EMCHS)_骨外性粘液性軟骨肉腫"),
        replacement = "Other_sarcoma"))
  if(EWS_soft_bone_combine == 1){
    Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
         fixed(pattern="Ewing Sarcoma of Soft Tissue (ESST)_骨外性ユーイング肉腫"),
         replacement = "Ewing Sarcoma (ES)_ユーイング肉腫"))
  }
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Desmoplastic Small-Round-Cell Tumor (DSRCT)_線維形成性小円形細胞腫瘍"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Desmoid/Aggressive Fibromatosis (DES)_デスモイド/侵襲性線維腫症"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Dermatofibrosarcoma Protuberance"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Clear Cell Sarcoma (CCS)_明細胞肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="chondroblastoma"),
        replacement = "Other"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Dendritic Cell Sarcoma (DCS)_樹状細胞肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Epithelioid Hemangioendothelioma (EHAE)_類上皮血管内皮腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="extraskeletal osteosarcoma"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="malignant rhabdoid tumor"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Ossifying Fibromyxoid Tumor (OFMT)_骨化性線維粘液性腫瘍"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Parosteal Osteosarcoma (PAOS)_傍骨性骨肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Periosteal Osteosarcoma (PEOS)"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Proximal-Type Epithelioid Sarcoma (PTES)_近位型類上皮肉腫"),
        replacement = "Epithelioid Sarcoma (EPIS)_類上皮肉腫"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Pseudomyogenic Hemangioendothelioma (PMHE)_偽筋原性血管内皮腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Angiomatoid Fibrous Histiocytoma (AFH)_類血管腫型線維性組織球腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Inflammatory Myofibroblastic Tumor (IMT)_炎症性筋線維芽細胞腫瘍"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Adult fibrosarcoma"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="CIC-rearranged sarcoma"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Epithelioid Sarcoma (EPIS)_類上皮肉腫"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="NTRK-rearranged spindle cell sarcoma"),
        replacement = "Other_sarcoma"))
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Sarcoma with BCOR genetic alterations"),
        replacement = "Other_sarcoma"))
  }
  if(Ultra_rare_sarcoma == 1){
    Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Mesenchymal Chondrosarcoma (MCHS)_間葉型軟骨肉腫"),
          replacement = "mesenchymal chondrosarcoma"))
    Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Extraskeletal Myxoid Chondrosarcoma (EMCHS)_骨外性粘液性軟骨肉腫"),
          replacement = "Extraskeletal myxoid chondrosarcoma"))
    Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Spindle Cell Rhabdomyosarcoma (SCRMS)_紡錘形細胞型横紋筋肉腫"),
          replacement = "Spindle Cell/Sclerosing Rhabdomyosarcoma (SCSRMS)_紡錘形細胞型/硬化型横紋筋肉腫"))
    Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Chordoma (CHDM)_脊索腫"),
          replacement = "Chordoma_NOS (CHDM)_脊索腫"))
    Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Epithelioid Sarcoma (EPIS)_類上皮肉腫"),
          replacement = "Epithelioid Sarcoma_NOS (EPIS)_類上皮肉腫"))
    Data_case = Data_case %>% dplyr::mutate(
    diagnosis=str_replace_all(diagnosis,
          fixed(pattern="Rhabdomyosarcoma (RMS)_横紋筋肉腫"),
          replacement = "Rhabdomyosarcoma_NOS (RMS)_横紋筋肉腫"))
  }
  # Data_case = Data_case %>% dplyr::mutate(
  # diagnosis=str_replace_all(diagnosis,
  #       fixed(pattern=""),
  #       replacement = ""))
}

# select the disease of interest
Data_case_target = Data_case %>%
                     dplyr::filter(diagnosis %in% Diseases)
Data_case_target$diagnosis = str_split(string = Data_case_target$diagnosis, pattern = "\\ \\(", simplify = TRUE)[,1]
Data_case_target$diagnosis = str_split(string = Data_case_target$diagnosis, pattern = "\\(", simplify = TRUE)[,1]
Data_case_target$diagnosis = str_split(string = Data_case_target$diagnosis, pattern = "/Hemangiopericytoma", simplify = TRUE)[,1]
Diseases = unique(Data_case_target$diagnosis)
Diseases = sort(Diseases)
Data_case_target = Data_case_target %>%
                     dplyr::filter(症例.検体情報.パネル.名称. %in% Panel)

if(Ex_PS9 == 1){
  Data_case_target = Data_case_target %>%
                dplyr::filter(症例.背景情報.ECOG.PS != 9)
}

if(Ex_multi == 1){
  Data_case_target = Data_case_target %>%
                dplyr::filter(症例.背景情報.重複がん有無.異なる臓器..名称. != "あり")
}


if(CNS_Glioblastoma_IDH == 1 & Cancer == "CNS"){
  IDH1_mut_ID = unique((Data_report %>% dplyr::filter(
    C.CAT調査結果.変異情報.マーカー == "IDH1" &
      C.CAT調査結果.エビデンス.エビデンスレベル == "F" |
    C.CAT調査結果.変異情報.マーカー == "IDH2" &
      C.CAT調査結果.エビデンス.エビデンスレベル == "F"
        ))$C.CAT調査結果.基本項目.ハッシュID)
  Data_case_target = Data_case_target %>%
                dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% IDH1_mut_ID)
}

if(BRCA_mutated == 1){
  BRCA_mut_ID = unique((Data_report %>% dplyr::filter(
    C.CAT調査結果.変異情報.マーカー == "BRCA1" &
      C.CAT調査結果.エビデンス.エビデンスレベル == "F" |
    C.CAT調査結果.変異情報.マーカー == "BRCA2" &
      C.CAT調査結果.エビデンス.エビデンスレベル == "F"
        ))$C.CAT調査結果.基本項目.ハッシュID)
  Data_case_target = Data_case_target %>%
                dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% BRCA_mut_ID)
}

if(ER_positive_BRCA == 1){
  Data_case_target = Data_case_target %>%
                dplyr::filter(症例.がん種情報.乳.ER.名称. == "陽性")
  if(sum(is.na(Data_case_target$症例.がん種情報.乳.ER.名称.))>0){
  Data_case_target = Data_case_target %>%
                dplyr::filter(!is.na(症例.がん種情報.乳.ER.名称.))
  }
  Cancer = "ER_positive_BRCA"
}

if(EGFR_positive_LUAD == 1){
  Data_case_target_original = Data_case_target
  EGFR_table = read.csv(file = "source/EGFR_oncogenic_mutations.csv", skip = 0)
  EGFR_ID = unique((Data_report %>%
              dplyr::filter(C.CAT調査結果.変異情報.マーカー == "EGFR" &
                              C.CAT調査結果.変異情報.変異内容 %in%
                              EGFR_table$C.CAT調査結果.変異情報.変異内容))$
                C.CAT調査結果.基本項目.ハッシュID)
  Data_case_target = Data_case_target_original
  Data_case_target_EGFR_before = Data_case_target %>%
#                dplyr::filter(症例.がん種情報.肺.EGFR.名称. %in% c("陽性"))
#                dplyr::filter(症例.がん種情報.肺.EGFR.名称. %in% c("陰性"))
                dplyr::filter(症例.がん種情報.肺.EGFR.名称. %in% c("陽性", "陰性"))
  ID_EGFR_pos_before = unique(Data_case_target_EGFR_before$C.CAT調査結果.基本項目.ハッシュID)
  Data_case_target = Data_case_target_EGFR_before %>%
                dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_EGFR_pos_before)
  Data_case_target$EGFR_exon = ""
  Data_case_target$EGFR_mutation = ""
  for(i in unique(EGFR_table$C.CAT調査結果.変異情報.変異内容)){
      EGFR_exon_tmp = unique((EGFR_table %>%
              dplyr::filter(C.CAT調査結果.変異情報.変異内容 == i))$
                EGFR_group)
      EGFR_ID_tmp = unique((Data_report %>%
              dplyr::filter(C.CAT調査結果.変異情報.マーカー == "EGFR" &
                              C.CAT調査結果.変異情報.変異内容 ==
                              i))$
                C.CAT調査結果.基本項目.ハッシュID)
      if(length(Data_case_target[
        Data_case_target$C.CAT調査結果.基本項目.ハッシュID %in% EGFR_ID_tmp,]$
         EGFR_mutation) > 0){
        Data_case_target[
        Data_case_target$C.CAT調査結果.基本項目.ハッシュID %in% EGFR_ID_tmp,]$
         EGFR_mutation = paste(Data_case_target[
        Data_case_target$C.CAT調査結果.基本項目.ハッシュID %in% EGFR_ID_tmp,]$
         EGFR_mutation, i, "(+) ", sep="")
      }
      if(!i %in% c("amplification", "C797S", "T790M", "E709G", "E709A", "L833V", "L718Q", "L747S", "V843I")){
      if(length(Data_case_target[
        Data_case_target$C.CAT調査結果.基本項目.ハッシュID %in% EGFR_ID_tmp,]$
         EGFR_mutation) > 0){
        Data_case_target[
          Data_case_target$C.CAT調査結果.基本項目.ハッシュID %in% EGFR_ID_tmp,]$
           EGFR_exon = paste(Data_case_target[
          Data_case_target$C.CAT調査結果.基本項目.ハッシュID %in% EGFR_ID_tmp,]$
           EGFR_exon, EGFR_exon_tmp, "(+) ", sep=""
          )
      }
      }
  }
  Data_case_target = Data_case_target %>% 
    dplyr::mutate(EGFR_exon = case_when(
      EGFR_mutation == "" ~ "No EGFR mut",
      EGFR_mutation == "amplification(+) " ~ "Only amp",
      EGFR_exon == "" ~ "Only resistant mut",
      TRUE ~ EGFR_exon
    ))
  Data_case_target$EGFR_mutation = str_replace_all(
    string = Data_case_target$EGFR_mutation,
    pattern = "H773L\\(\\+\\) V774M\\(\\+\\)",
    replacement = "H773_V774delinsLM\\(\\+\\)")
#  Data_case_target = Data_case_target %>%
#    dplyr::filter(EGFR_mutation != "amplification(+) ") %>%
#    dplyr::filter(症例.がん種情報.肺.EGFR.名称. %in% c("陽性", "陰性"))
  Data_case_simple = Data_case_target %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.管理情報.登録日,
                  症例.背景情報.診断日,
                  症例.背景情報.喫煙歴有無.名称.,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.ECOG.PS,
                  症例.検体情報.パネル.名称.,
                  症例.検体情報.検体採取日.腫瘍組織.,
                  症例.検体情報.検体採取方法.その他.,
                  症例.検体情報.検体採取部位.名称.,
                  症例.検体情報.具体的な採取部位.名称.,
                  症例.検体情報.具体的な採取部位その他,
                  症例.がん種情報.肺.EGFR.名称.,
                  症例.がん種情報.肺.EGFR.type.名称.,
                  症例.がん種情報.肺.EGFR.検査方法.名称.,
                  症例.がん種情報.肺.EGFR.TKI耐性後EGFR.T790M.名称.,
                  症例.がん種情報.肺.ALK融合.名称.,
                  症例.がん種情報.肺.ALK.検査方法.名称.,
                  症例.がん種情報.肺.ROS1.名称.,
                  症例.がん種情報.肺.BRAF.V600..名称.,
                  症例.がん種情報.肺.PD.L1.IHC..名称.,
                  症例.がん種情報.肺.PD.L1.IHC..検査方法.名称.,
                  症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                  症例.EP前レジメン情報.投与開始日,
                  症例.EP前レジメン情報.投与終了日,
                  症例.EP前レジメン情報.終了理由.名称.,
                  症例.EP前レジメン情報.最良総合効果.名称.,
                  症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.,
                  症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                  症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                  症例.EP後レジメン情報.投与開始日,
                  症例.EP後レジメン情報.投与終了日,
                  症例.EP後レジメン情報.終了理由.名称.,
                  症例.EP後レジメン情報.最良総合効果.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis,
                  EGFR_exon,
                  EGFR_mutation
                  ) %>%
    dplyr::mutate(final_observe = case_when(
      症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
      症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
      症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
      TRUE ~ "")) %>%
    dplyr::arrange(final_observe) %>%
    dplyr::distinct()
  Data_case_unique = Data_case_simple %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.管理情報.登録日,
                  症例.背景情報.診断日,
                  症例.背景情報.喫煙歴有無.名称.,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.ECOG.PS,
                  症例.検体情報.パネル.名称.,
                  症例.検体情報.検体採取日.腫瘍組織.,
                  症例.検体情報.検体採取方法.その他.,
                  症例.検体情報.検体採取部位.名称.,
                  症例.検体情報.具体的な採取部位.名称.,
                  症例.検体情報.具体的な採取部位その他,
                  症例.がん種情報.肺.EGFR.名称.,
                  症例.がん種情報.肺.EGFR.type.名称.,
                  症例.がん種情報.肺.EGFR.検査方法.名称.,
                  症例.がん種情報.肺.EGFR.TKI耐性後EGFR.T790M.名称.,
                  症例.がん種情報.肺.ALK融合.名称.,
                  症例.がん種情報.肺.ALK.検査方法.名称.,
                  症例.がん種情報.肺.ROS1.名称.,
                  症例.がん種情報.肺.BRAF.V600..名称.,
                  症例.がん種情報.肺.PD.L1.IHC..名称.,
                  症例.がん種情報.肺.PD.L1.IHC..検査方法.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis,
                  EGFR_exon,
                  EGFR_mutation
                  ) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)
  Data_case_unique %>% dplyr::filter(症例.がん種情報.肺.EGFR.名称. == "陰性")
  tmp_1 = Data_case_simple %>% dplyr::select(
                  C.CAT調査結果.基本項目.ハッシュID,
                  症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                  症例.EP前レジメン情報.投与開始日,
                  症例.EP前レジメン情報.投与終了日,
                  症例.EP前レジメン情報.最良総合効果.名称.)
  tmp_1$treat = paste(tmp_1$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                     tmp_1$症例.EP前レジメン情報.投与開始日,
                     "_",
                     tmp_1$症例.EP前レジメン情報.投与終了日,
                     tmp_1$症例.EP前レジメン情報.最良総合効果.名称.)
  tmp_2 = Data_case_simple %>% dplyr::select(
                  C.CAT調査結果.基本項目.ハッシュID,
                  症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                  症例.EP後レジメン情報.投与開始日,
                  症例.EP後レジメン情報.投与終了日,
                  症例.EP後レジメン情報.最良総合効果.名称.)
  tmp_2$treat = paste(tmp_2$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                     tmp_2$症例.EP後レジメン情報.投与開始日,
                     "_",
                     tmp_2$症例.EP後レジメン情報.投与終了日,
                     tmp_2$症例.EP後レジメン情報.最良総合効果.名称.)
  Data_case_unique$treat_before_CGP = ""
  Data_case_unique$treat_after_CGP = ""
  for(i in 1:length(Data_case_unique$C.CAT調査結果.基本項目.ハッシュID)){
    tmp_3 = tmp_1 %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID ==
                              Data_case_unique$C.CAT調査結果.基本項目.ハッシュID[i])
    Data_case_unique$treat_before_CGP[i] = str_flatten(unique(tmp_3$treat), collapse = " ______ ")
    tmp_4 = tmp_2 %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID ==
                              Data_case_unique$C.CAT調査結果.基本項目.ハッシュID[i])
    Data_case_unique$treat_after_CGP[i] = str_flatten(unique(tmp_4$treat), collapse = " ______ ")
  }
  WriteXLS(Data_case_unique, ExcelFileName=paste(OutDirName, "EGFR_mut_exon.xls", sep=""))
}

Data_report_tmp = Data_report %>% dplyr::mutate(
    C.CAT調査結果.エビデンス.エビデンスレベル_WT = case_when(
      C.CAT調査結果.エビデンス.エビデンスレベル == "F" ~ "F",
      TRUE ~ "WT"
    ))
Data_report_tmp = Data_report_tmp[order(Data_report_tmp$C.CAT調査結果.エビデンス.エビデンスレベル_WT), ]
Data_report_tmp = Data_report_tmp %>% dplyr::distinct(
  C.CAT調査結果.基本項目.ハッシュID,
  C.CAT調査結果.変異情報.変異種類,
  C.CAT調査結果.変異情報.マーカー,
  C.CAT調査結果.変異情報.変異内容,
  .keep_all = TRUE) %>% dplyr::select(
  C.CAT調査結果.基本項目.ハッシュID,
  C.CAT調査結果.変異情報.変異種類,
  C.CAT調査結果.変異情報.マーカー,
  C.CAT調査結果.変異情報.変異内容,
  C.CAT調査結果.エビデンス.エビデンスレベル_WT
  )

Data_report = left_join(Data_report, Data_report_tmp, by = c(
  "C.CAT調査結果.基本項目.ハッシュID",
  "C.CAT調査結果.変異情報.変異種類",
  "C.CAT調査結果.変異情報.マーカー",
  "C.CAT調査結果.変異情報.変異内容"))

ID_target = unique(Data_case_target$C.CAT調査結果.基本項目.ハッシュID) 
Data_PGPV = Data_PGPV %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target)
if(length(Data_PGPV[,1]) >1){
  PGPV_mutations = Data_PGPV %>%
                  dplyr::count(gene_mut) %>%
                  dplyr::arrange(-n)
  PGPV_mutations$NCC_germline = 0
  for(i in 1:length(PGPV_mutations$n)){
    PGPV_mutations$NCC_germline[i] = sum((Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_mut == PGPV_mutations$gene_mut[i])
  }
  PGPV_mutations = PGPV_mutations %>%
                  dplyr::arrange(-NCC_germline) %>%
                  dplyr::arrange(-n)
  PGPV_mutations$NCC_PGPV = 0
  for(i in 1:length(PGPV_mutations$n)){
    PGPV_mutations$NCC_PGPV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1))$gene_mut == PGPV_mutations$gene_mut[i])
  }
  PGPV_mutations = PGPV_mutations %>%
                  dplyr::arrange(-NCC_PGPV) %>%
                  dplyr::arrange(-n)
  PGPV_mutations$NCC_PGPV_PGV  = 0
  for(i in 1:length(PGPV_mutations$n)){
    PGPV_mutations$NCC_PGPV_PGV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1 & C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_mut == PGPV_mutations$gene_mut[i])
  }
  PGPV_mutations = PGPV_mutations %>%
                  dplyr::arrange(-NCC_PGPV_PGV) %>%
                  dplyr::arrange(-n)
  colnames(PGPV_mutations) = c("Mutation", "mutation counts", "NCC validated PGV", "NCC PGPV", "NCC PGPV & PGV")
  DT::datatable(PGPV_mutations, filter = 'top', extensions = 'Scroller', 
               caption = paste("All", Cancer ,"patients with NCC oncopanel version >=2",length(unique((Data_report %>% dplyr::filter(str_detect(C.CAT調査結果.基本項目.パネル名, pattern="NCC オンコパネル システム ver.2")) %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target))$C.CAT調査結果.基本項目.ハッシュID)), "patients,", length((unique(Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$C.CAT調査結果.基本項目.ハッシュID)), "patients had validated germline mutations."))
  WriteXLS(PGPV_mutations, ExcelFileName = paste(OutDirName, Cancer, "_NOP_PGPV_mutation_selected_disease.xls", sep=""))
  
  PGPV_mutations = Data_PGPV %>%
                  dplyr::count(gene_patho) %>%
                  dplyr::arrange(-n)
  PGPV_mutations$NCC_germline = 0
  for(i in 1:length(PGPV_mutations$n)){
    PGPV_mutations$NCC_germline[i] = sum((Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_patho == PGPV_mutations$gene_patho[i])
  }
  PGPV_mutations = PGPV_mutations %>%
                  dplyr::arrange(-NCC_germline) %>%
                  dplyr::arrange(-n)
  PGPV_mutations$NCC_PGPV = 0
  for(i in 1:length(PGPV_mutations$n)){
    PGPV_mutations$NCC_PGPV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1))$gene_patho == PGPV_mutations$gene_patho[i])
  }
  PGPV_mutations = PGPV_mutations %>%
                  dplyr::arrange(-NCC_PGPV) %>%
                  dplyr::arrange(-n)
  PGPV_mutations$NCC_PGPV_PGV  = 0
  for(i in 1:length(PGPV_mutations$n)){
    PGPV_mutations$NCC_PGPV_PGV[i] = sum((Data_PGPV %>% dplyr::filter(PGPV_plus == 1 & C.CAT調査結果.変異情報.変異由来 == "Germline"))$gene_patho == PGPV_mutations$gene_patho[i])
  }
  PGPV_mutations = PGPV_mutations %>%
                  dplyr::arrange(-NCC_PGPV_PGV) %>%
                  dplyr::arrange(-n)
  colnames(PGPV_mutations) = c("Mutation", "mutation counts", "NCC validated PGV", "NCC PGPV", "NCC PGPV & PGV")
  DT::datatable(PGPV_mutations, filter = 'top', extensions = 'Scroller', 
               caption = paste("All", Cancer ,"patients with NCC oncopanel version >=2",length(unique((Data_report %>% dplyr::filter(str_detect(C.CAT調査結果.基本項目.パネル名, pattern="NCC オンコパネル システム ver.2")) %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target))$C.CAT調査結果.基本項目.ハッシュID)), "patients,", length((unique(Data_PGPV %>% dplyr::filter(C.CAT調査結果.変異情報.変異由来 == "Germline"))$C.CAT調査結果.基本項目.ハッシュID)), "patients had validated germline mutations."))
  
  WriteXLS(PGPV_mutations, ExcelFileName = paste(OutDirName, Cancer, "_NOP_PGPV_gene_selected_disease.xls", sep=""))
}


Data_case_drug = Data_case_target %>%
  dplyr::select(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                C.CAT調査結果.基本項目.ハッシュID)
drug_list = as.vector(str_split(unique(Data_case_target$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
drug_list = drug_list[-which(drug_list %in% c("", "Unknown"))]
if (length(drug_list[-which(is.na(drug_list))] > 0)){
  drug_list = drug_list[-which(is.na(drug_list))]
}
drug_list = unique(drug_list)

drug_list = data.frame(drug_list)
drug_list$patients = 0

for(i in 1:length(drug_list[,1])){
  drug_list[i,2] = length(unique((Data_case_target %>% dplyr::filter(
    str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                 pattern=drug_list[i,1]))
   )$C.CAT調査結果.基本項目.ハッシュID))
}

c_name_drug = c("drug_list", "all_patients")
for(i in 1:length(Diseases)){
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = drug_list[,2]
  for(j in 1:length(tmp)){
   tmp[j] = length(unique((Data_case_tmp %>% dplyr::filter(
      str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list[j,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  drug_list = cbind(drug_list, tmp)
  c_name_drug = c(c_name_drug, Diseases[[i]])
}  
colnames(drug_list) =c_name_drug

DT::datatable(drug_list %>% dplyr::arrange(-all_patients),
             caption = "Drugs used before expert panel (patients), selected disease only", filter = 'top', extensions = 'Scroller')

# combine drug names (optional)
if(Drug_mix == 1){
  if(Cancer == "Stomach"){
    Data_case = Data_case %>% dplyr::mutate(
      症例.EP前レジメン情報.薬剤名.YJ一般名.EN. =case_when(
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Cisplatin" ~ "Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Carboplatin" ~ "Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Oxaliplatin" ~ "Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Etoposide,Cisplatin" ~ "Etoposide,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Cisplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Capecitabine,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel,Cisplatin,Fluorouracil" ~ "Taxane,F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Gimeracil,Oteracil" ~ "F_pyrimidine,Gimeracil,Oteracil",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Uracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Cisplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Paclitaxel" ~ "Taxane",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "nab-paclitaxel" ~ "Taxane",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel" ~ "Taxane",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tipiracil" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tiperacil" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tiperacil" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Ramucirumab,Paclitaxel" ~ "Ramucirumab,Taxane",
        TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
      )
    )
    Data_case_target = Data_case_target %>% dplyr::mutate(
      症例.EP前レジメン情報.薬剤名.YJ一般名.EN. =case_when(
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Cisplatin" ~ "Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Carboplatin" ~ "Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Oxaliplatin" ~ "Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Etoposide,Cisplatin" ~ "Etoposide,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Cisplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Capecitabine" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Capecitabine,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel,Cisplatin,Fluorouracil" ~ "Taxane,F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Gimeracil,Oteracil" ~ "F_pyrimidine,Gimeracil,Oteracil",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Uracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Cisplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Paclitaxel" ~ "Taxane",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "nab-paclitaxel" ~ "Taxane",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel" ~ "Taxane",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tipiracil" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Tiperacil" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tiperacil" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine" ~ "TAS-102",
        症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Ramucirumab,Paclitaxel" ~ "Ramucirumab,Taxane",
        TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
      )
    )
  }
  Data_case_drug = Data_case_target %>%
    dplyr::select(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                  C.CAT調査結果.基本項目.ハッシュID)
  drug_list = as.vector(str_split(unique(Data_case_target$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
  drug_list = drug_list[-which(drug_list %in% c("", "Unknown"))]
  if (length(drug_list[-which(is.na(drug_list))] > 0)){
    drug_list = drug_list[-which(is.na(drug_list))]
  }
  drug_list = unique(drug_list)
  
  drug_list = data.frame(drug_list)
  drug_list$patients = 0
  
  for(i in 1:length(drug_list[,1])){
    drug_list[i,2] = length(unique((Data_case_target %>% dplyr::filter(
      str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list[i,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  
  c_name_drug = c("drug_list", "all_patients")
  for(i in 1:length(Diseases)){
    Data_case_tmp = Data_case_target %>%
                       dplyr::filter(diagnosis %in% Diseases[[i]])
    ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
    tmp = drug_list[,2]
    for(j in 1:length(tmp)){
     tmp[j] = length(unique((Data_case_tmp %>% dplyr::filter(
        str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                     pattern=drug_list[j,1]))
       )$C.CAT調査結果.基本項目.ハッシュID))
    }
    drug_list = cbind(drug_list, tmp)
    c_name_drug = c(c_name_drug, Diseases[[i]])
  }  
  colnames(drug_list) =c_name_drug
  
  DT::datatable(drug_list %>% dplyr::arrange(-all_patients),
               caption = "Drugs (names combined) used before expert panel (patients), selected disease only", filter = 'top', extensions = 'Scroller')
}

Data_tmp = Data_case_target %>% 
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
drug_list_EP = as.vector(str_split(unique(Data_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
drug_list_EP = drug_list_EP[-which(drug_list_EP %in% c("", "Unknown"))]
if (length(drug_list_EP[-which(is.na(drug_list_EP))] > 0)){
  drug_list_EP = drug_list_EP[-which(is.na(drug_list_EP))]
}
drug_list_EP = unique(drug_list_EP)

drug_list_EP = data.frame(drug_list_EP)
drug_list_EP$patients = 0

for(i in 1:length(drug_list_EP[,1])){
  drug_list_EP[i,2] = length(unique((Data_tmp %>% dplyr::filter(
    str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                 pattern=drug_list_EP[i,1]))
   )$C.CAT調査結果.基本項目.ハッシュID))
}

c_name_drug = c("drug_list", "all_patients")
for(i in 1:length(Diseases)){
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = drug_list_EP[,2]
  for(j in 1:length(tmp)){
   tmp[j] = length(unique((Data_case_tmp %>% dplyr::filter(
      str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list_EP[j,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  drug_list_EP = cbind(drug_list_EP, tmp)
  c_name_drug = c(c_name_drug, Diseases[[i]])
}  
colnames(drug_list_EP) =c_name_drug

DT::datatable(drug_list_EP %>% dplyr::arrange(-all_patients),
             caption = "Drugs used after expert panel, selected disease only", filter = 'top', extensions = 'Scroller')

Data_tmp = Data_case_target %>% 
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                症例.EP後レジメン情報.提示された治療薬を投与した.名称.) %>%
  dplyr::distinct()
drug_list_EP = as.vector(str_split(unique(Data_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
drug_list_EP = drug_list_EP[-which(drug_list_EP %in% c("", "Unknown"))]
if (length(drug_list_EP[-which(is.na(drug_list_EP))] > 0)){
  drug_list_EP = drug_list_EP[-which(is.na(drug_list_EP))]
}
drug_list_EP = unique(drug_list_EP)

drug_list_EP = data.frame(drug_list_EP)
drug_list_EP$patients = 0

for(i in 1:length(drug_list_EP[,1])){
  drug_list_EP[i,2] = length(unique(((Data_tmp %>%dplyr::filter(
        症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した")) %>% dplyr::filter(
    str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                 pattern=drug_list_EP[i,1]))
   )$C.CAT調査結果.基本項目.ハッシュID))
}

c_name_drug = c("drug_list", "all_patients")
for(i in 1:length(Diseases)){
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = drug_list_EP[,2]
  for(j in 1:length(tmp)){
   tmp[j] = length(unique(((Data_case_tmp %>% dplyr::filter(
      症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した")) %>% dplyr::filter(
      str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list_EP[j,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  drug_list_EP = cbind(drug_list_EP, tmp)
  c_name_drug = c(c_name_drug, Diseases[[i]])
}  
colnames(drug_list_EP) =c_name_drug

DT::datatable(drug_list_EP %>% dplyr::arrange(-all_patients),
             caption = "Drugs used after and recommended by expert panel, selected disease only", filter = 'top', extensions = 'Scroller')

if(Drug_mix == 1){
  if(Cancer == "Stomach"){
    Data_case = Data_case %>% dplyr::mutate(
      症例.EP後レジメン情報.薬剤名.YJ一般名.EN. =case_when(
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Cisplatin" ~ "Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Oxaliplatin" ~ "Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Etoposide,Cisplatin" ~ "Etoposide,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Capecitabine,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Capecitabine" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel,Cisplatin,Fluorouracil" ~ "Taxane,F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Gimeracil,Oteracil" ~ "F_pyrimidine,Gimeracil,Oteracil",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Uracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Cisplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Paclitaxel" ~ "Taxane",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "nab-paclitaxel" ~ "Taxane",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel" ~ "Taxane",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tipiracil" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tiperacil" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tiperacil" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Ramucirumab,Paclitaxel" ~ "Ramucirumab,Taxane",
        TRUE ~ 症例.EP後レジメン情報.薬剤名.YJ一般名.EN.
      )
    )
    Data_case_target = Data_case_target %>% dplyr::mutate(
      症例.EP後レジメン情報.薬剤名.YJ一般名.EN. =case_when(
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Cisplatin" ~ "Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Oxaliplatin" ~ "Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Etoposide,Cisplatin" ~ "Etoposide,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Capecitabine,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Capecitabine" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel,Cisplatin,Fluorouracil" ~ "Taxane,F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Gimeracil,Oteracil" ~ "F_pyrimidine,Gimeracil,Oteracil",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tegafur,Uracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Cisplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Paclitaxel" ~ "Taxane",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "nab-paclitaxel" ~ "Taxane",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Docetaxel" ~ "Taxane",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil,Oxaliplatin" ~ "F_pyrimidine,Platinum",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Fluorouracil" ~ "F_pyrimidine",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tipiracil" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Tiperacil" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine,Tiperacil" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trifluridine" ~ "TAS-102",
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Ramucirumab,Paclitaxel" ~ "Ramucirumab,Taxane",
        TRUE ~ 症例.EP後レジメン情報.薬剤名.YJ一般名.EN.
      )
    )
  }
  Data_tmp = Data_case_target %>% 
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
    dplyr::distinct()
  drug_list_EP = as.vector(str_split(unique(Data_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
  drug_list_EP = drug_list_EP[-which(drug_list_EP %in% c("", "Unknown"))]
  if (length(drug_list_EP[-which(is.na(drug_list_EP))] > 0)){
    drug_list_EP = drug_list_EP[-which(is.na(drug_list_EP))]
  }
  drug_list_EP = unique(drug_list_EP)
  
  drug_list_EP = data.frame(drug_list_EP)
  drug_list_EP$patients = 0
  
  for(i in 1:length(drug_list_EP[,1])){
    drug_list_EP[i,2] = length(unique((Data_tmp %>% dplyr::filter(
      str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list_EP[i,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  
  c_name_drug = c("drug_list", "all_patients")
  for(i in 1:length(Diseases)){
    Data_case_tmp = Data_case_target %>%
                       dplyr::filter(diagnosis %in% Diseases[[i]])
    ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
    tmp = drug_list_EP[,2]
    for(j in 1:length(tmp)){
     tmp[j] = length(unique((Data_case_tmp %>% dplyr::filter(
        str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                     pattern=drug_list_EP[j,1]))
       )$C.CAT調査結果.基本項目.ハッシュID))
    }
    drug_list_EP = cbind(drug_list_EP, tmp)
    c_name_drug = c(c_name_drug, Diseases[[i]])
  }  
  colnames(drug_list_EP) =c_name_drug
  
  DT::datatable(drug_list_EP %>% dplyr::arrange(-all_patients),
               caption = "Drugs (names combined) used after expert panel, selected disease only", filter = 'top', extensions = 'Scroller')
}
if(Drug_mix == 1){
  Data_tmp = Data_case_target %>% 
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                  症例.EP後レジメン情報.提示された治療薬を投与した.名称.) %>%
    dplyr::distinct()
  drug_list_EP = as.vector(str_split(unique(Data_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
  drug_list_EP = drug_list_EP[-which(drug_list_EP %in% c("", "Unknown"))]
  if (length(drug_list_EP[-which(is.na(drug_list_EP))] > 0)){
    drug_list_EP = drug_list_EP[-which(is.na(drug_list_EP))]
  }
  drug_list_EP = unique(drug_list_EP)
  
  drug_list_EP = data.frame(drug_list_EP)
  drug_list_EP$patients = 0
  
  for(i in 1:length(drug_list_EP[,1])){
    drug_list_EP[i,2] = length(unique(((Data_tmp %>%dplyr::filter(
          症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した")) %>% dplyr::filter(
      str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list_EP[i,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  
  c_name_drug = c("drug_list", "all_patients")
  for(i in 1:length(Diseases)){
    Data_case_tmp = Data_case_target %>%
                       dplyr::filter(diagnosis %in% Diseases[[i]])
    ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
    tmp = drug_list_EP[,2]
    for(j in 1:length(tmp)){
     tmp[j] = length(unique(((Data_case_tmp %>% dplyr::filter(
        症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した")) %>% dplyr::filter(
        str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                     pattern=drug_list_EP[j,1]))
       )$C.CAT調査結果.基本項目.ハッシュID))
    }
    drug_list_EP = cbind(drug_list_EP, tmp)
    c_name_drug = c(c_name_drug, Diseases[[i]])
  }  
  colnames(drug_list_EP) =c_name_drug
  
  DT::datatable(drug_list_EP %>% dplyr::arrange(-all_patients),
               caption = "Drugs (names combined) used after and recommended by expert panel, selected disease only", filter = 'top', extensions = 'Scroller')
}


Data_summary =  Data_case_target %>% 
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1)) %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))
if(Ex_inactive_multi == 1){
  Data_summary =  Data_summary %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_summary = Data_summary %>%
  dplyr::filter(diagnosis %in%
# "Pancreas") %>%
# "Colorectal") %>%
# "Biliary") %>%
# "Lung") %>%
# "Soft_tissue") %>%
# "Ovary") %>%
# "Breast") %>%
            unique(Data_case_target$diagnosis))# %>%
#  dplyr::filter(症例.検体情報.パネル.名称. %in%
#                 "F1Liquid CDx")
#                  "NCC OncoPanel")
#                  "FoundationOne CDx")
Data_summary = Data_summary %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.,
  症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table 3. Patients with selected disease, Characteristics**") %>%
  bold_labels()
table_tmp

text_patient = paste(Cancer, ", Total patients: ", length((Data_case_target %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n", sep="") 

Data_survival = Data_case_target %>%
  dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
text_patient = paste(text_patient, "patients underwent survival prolonging therapy: ", length((Data_survival %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n", sep="") 

Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日))
text_patient = paste(text_patient, "patients underwent survival prolonging therapy with start date: ", length((Data_survival %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n", sep="") 
Data_survival = Data_survival %>%
  dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::filter(症例.管理情報.登録日 != "")
text_patient = paste(text_patient, "patients underwent survival prolonging therapy with start date with entry date: ", length((Data_survival %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n", sep="") 
Data_survival = Data_survival %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))
text_patient = paste(text_patient, "patients underwent survival prolonging therapy with start date with entry date with last observation date: ", length((Data_survival %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n", sep="") 

Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
  Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日)) %>%
  dplyr::mutate(症例.背景情報.診断日 =
                    as.Date(Data_survival$症例.背景情報.診断日))

Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

text_patient = paste(text_patient, "patients underwent survival prolonging therapy with start date with entry date with last observation date with follow-up after CGP: ", length((Data_survival %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n", sep="") 

text_patient = paste(text_patient, "patients underwent survival prolonging therapy with start date with entry date with last observation date with follow-up after CGP without input errors: ", length((Data_survival %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n\n", sep="") 

write (text_patient, file = "output/patient_selection_process.txt", append = TRUE)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_summary = Data_survival %>%
  dplyr::filter(diagnosis %in%
            unique(Data_case_target$diagnosis)) %>%
  dplyr::filter(症例.検体情報.パネル.名称. %in% Panel)
ID_prog = unique(Data_summary$C.CAT調査結果.基本項目.ハッシュID) 

Data_case_drug = Data_case_target %>%
  dplyr::select(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog)
drug_list = as.vector(str_split(unique(Data_case_drug$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
drug_list = drug_list[-which(drug_list %in% c("", "Unknown"))]
if (length(drug_list[-which(is.na(drug_list))] > 0)){
  drug_list = drug_list[-which(is.na(drug_list))]
}
drug_list = unique(drug_list)

drug_list = data.frame(drug_list)
drug_list$patients = 0


for(i in 1:length(drug_list[,1])){
  drug_list[i,2] = length(unique((Data_case_drug %>% dplyr::filter(
    str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                 pattern=drug_list[i,1]))
   )$C.CAT調査結果.基本項目.ハッシュID))
}

c_name_drug = c("drug_list", "all_patients")
for(i in 1:length(Diseases)){
  Data_case_tmp = Data_case_target %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = drug_list[,2]
  for(j in 1:length(tmp)){
   tmp[j] = length(unique((Data_case_tmp %>% dplyr::filter(
      str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list[j,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  drug_list = cbind(drug_list, tmp)
  c_name_drug = c(c_name_drug, Diseases[[i]])
}  
colnames(drug_list) =c_name_drug

DT::datatable(drug_list %>% dplyr::arrange(-all_patients),
             caption = "Drugs used before expert panel, selected disease with prognostic information only", filter = 'top', extensions = 'Scroller')

for(i in unlist(selected_drug)){
  if(!i %in% drug_list$drug_list){
    selected_drug = list(drug_list$drug_list[[1]])
    drug_pattern = list(c(1))
  }
}

target_drug = (drug_list %>% dplyr::arrange(-all_patients))[1:drug_number,1]

Data_tmp = Data_case_target %>% 
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
  dplyr::distinct()
drug_list_EP = as.vector(str_split(unique(Data_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
drug_list_EP = drug_list_EP[-which(drug_list_EP %in% c("", "Unknown"))]
if (length(drug_list_EP[-which(is.na(drug_list_EP))] > 0)){
  drug_list_EP = drug_list_EP[-which(is.na(drug_list_EP))]
}
drug_list_EP = unique(drug_list_EP)

drug_list_EP = data.frame(drug_list_EP)
if(length(drug_list_EP$drug_list_EP) > 0){
  drug_list_EP$patients = 0
  
  for(i in 1:length(drug_list_EP[,1])){
    drug_list_EP[i,2] = length(unique((Data_tmp %>% dplyr::filter(
      str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list_EP[i,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  c_name_drug = c("drug_list", "all_patients")
  for(i in 1:length(Diseases)){
    Data_case_tmp = Data_case_target %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
                       dplyr::filter(diagnosis %in% Diseases[[i]])
    ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
    tmp = drug_list_EP[,2]
    for(j in 1:length(tmp)){
     tmp[j] = length(unique((Data_case_tmp %>% dplyr::filter(
        str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                     pattern=drug_list_EP[j,1]))
       )$C.CAT調査結果.基本項目.ハッシュID))
    }
    drug_list_EP = cbind(drug_list_EP, tmp)
    c_name_drug = c(c_name_drug, Diseases[[i]])
  }  
  colnames(drug_list_EP) =c_name_drug

  DT::datatable(drug_list_EP %>% dplyr::arrange(-all_patients),
               caption = "Drugs used after expert panel, selected disease with prognostic information only", filter = 'top', extensions = 'Scroller')
} else{
  print("No drugs were used after expert panel")
}

Data_tmp = Data_case_target %>% 
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                症例.EP後レジメン情報.提示された治療薬を投与した.名称.) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
  dplyr::distinct()
drug_list_EP = as.vector(str_split(unique(Data_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.), ",", simplify = TRUE))
drug_list_EP = drug_list_EP[-which(drug_list_EP %in% c("", "Unknown"))]
if (length(drug_list_EP[-which(is.na(drug_list_EP))] > 0)){
  drug_list_EP = drug_list_EP[-which(is.na(drug_list_EP))]
}
drug_list_EP = unique(drug_list_EP)

drug_list_EP = data.frame(drug_list_EP)
if(length(drug_list_EP$drug_list_EP) > 0){
  drug_list_EP$patients = 0
  
  for(i in 1:length(drug_list_EP[,1])){
    drug_list_EP[i,2] = length(unique(((Data_tmp %>%dplyr::filter(
          症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した")) %>% dplyr::filter(
      str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                   pattern=drug_list_EP[i,1]))
     )$C.CAT調査結果.基本項目.ハッシュID))
  }
  
  c_name_drug = c("drug_list", "all_patients")
  for(i in 1:length(Diseases)){
    Data_case_tmp = Data_case_target %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
                       dplyr::filter(diagnosis %in% Diseases[[i]])
    ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
    tmp = drug_list_EP[,2]
    for(j in 1:length(tmp)){
     tmp[j] = length(unique(((Data_case_tmp %>% dplyr::filter(
        症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した")) %>% dplyr::filter(
        str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                     pattern=drug_list_EP[j,1]))
       )$C.CAT調査結果.基本項目.ハッシュID))
    }
    drug_list_EP = cbind(drug_list_EP, tmp)
    c_name_drug = c(c_name_drug, Diseases[[i]])
  }  
  colnames(drug_list_EP) =c_name_drug
  
  DT::datatable(drug_list_EP %>% dplyr::arrange(-all_patients),
               caption = "Drugs used after and recommended by expert panel, selected disease with prognostic information only", filter = 'top', extensions = 'Scroller')
} else{
  print("No drugs were used after expert panel")
}

## Target gene selection
Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.エビデンス.エビデンスレベル,
                  .keep_all=TRUE) %>%
  dplyr::mutate(gene_mut = paste(C.CAT調査結果.変異情報.マーカー,
                                 C.CAT調査結果.変異情報.変異内容,
                                 sep=" _ ")) %>%
  dplyr::mutate(gene_patho = C.CAT調査結果.変異情報.マーカー)

oncogenic_mutations = Data_major_gene %>%
                dplyr::count(gene_mut) %>%
                dplyr::arrange(-n)

c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_mutations
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_mut)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_mut)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_mutations = cbind(oncogenic_mutations, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_mutations) =c_name_gene

DT::datatable(oncogenic_mutations, filter = 'top', extensions = 'Scroller',
             caption = "Oncogenic mutations in selected diseases")

oncogenic_genes = Data_major_gene %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations in selected diseases (counts)")

oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
  dplyr::distinct() %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
      dplyr::distinct() %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations in selected diseases (patients)")

Gene_names = oncogenic_genes[1:min(40, length(oncogenic_genes[,1])),]$gene_mutation
Patient_names = unique(Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
Mutation_matrix = matrix(rep(0, length(Gene_names) * length(Patient_names)),
                         nrow = length(Gene_names),
                         ncol = length(Patient_names))
rownames(Mutation_matrix) = Gene_names
colnames(Mutation_matrix) = Patient_names

oncogenic_genes_save = oncogenic_genes
# Mutually exclusive or co-occurrence
Mutation_list_for_matrix = Data_major_gene %>%
  dplyr::filter(gene_patho %in% Gene_names)
for(i in 1:length(Mutation_list_for_matrix$C.CAT調査結果.基本項目.ハッシュID)){
  Mutation_matrix[Mutation_list_for_matrix$gene_patho[i],
                  Mutation_list_for_matrix$C.CAT調査結果.基本項目.ハッシュID[i]] = 1
}
PMA <- getPM(Mutation_matrix)
mutually_exclusive <- getMutex(A=Mutation_matrix,PM=PMA)
co_occurrence <- getMutex(A=Mutation_matrix,PM=PMA, lower.tail = FALSE)
colnames(mutually_exclusive) = Gene_names
rownames(mutually_exclusive) = Gene_names
mutually_exclusive[lower.tri(mutually_exclusive,diag=TRUE)] <- NA
Data_mut_ex = expand.grid(Gene_names, Gene_names)
colnames(Data_mut_ex) = c("Gene_1", "Gene_2")
Data_mut_ex$P_value = mutually_exclusive@x
Data_mut_ex = Data_mut_ex %>%
  dplyr::mutate(P_co_or_ex = case_when(
    P_value > 0.5 ~ log(1-P_value,base=10),
    TRUE ~ -log(P_value,base=10)
  ))
Data_mut_ex = Data_mut_ex %>%
  dplyr::mutate(P_figure = case_when(
    P_co_or_ex > 3 ~ 3,
    P_co_or_ex < -3 ~ -3,
    TRUE ~ P_co_or_ex
  ))
Data_mut_ex = Data_mut_ex %>%
  dplyr::mutate(P_signif = case_when(
    P_value <= 0.001 | P_value >= 0.999 ~ "*",
    TRUE ~ ""
  ))

Data_mut_ex$Gene_1 <- factor(Data_mut_ex$Gene_1, levels = Gene_names)
Data_mut_ex$Gene_2 <- factor(Data_mut_ex$Gene_2, levels = Gene_names)
 
g = Data_mut_ex %>% 
  ggplot(aes(x = Gene_1, y = Gene_2, fill = P_figure)) + 
  geom_tile(color = "white") + 
  scale_fill_gradient2(low = "red", high = "blue", mid = "white", 
                       midpoint = 0, limit = c(-3, 3), 
                       name = "-log10(P-value)", space = "Lab",
                       na.value = "white")+
  geom_text(aes(label = P_signif),
            color = "black", size = 4)+
  scale_y_discrete(limits = rev(levels(Data_mut_ex$Gene_2)))+
  theme_classic() +
  ggtitle("Mutually exclusive (blue) or co-occurrence (red)\n*: p<0.001, estimated with Rediscover package") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
pdf(file = paste(OutDirName, Cancer, "_mutually_exclusive.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

Data_mut_ex_table = Data_mut_ex %>%
  dplyr::select(Gene_1, Gene_2, P_co_or_ex)
Data_mut_ex_table = Data_mut_ex_table %>%
  dplyr::mutate(P_value = 10^-abs(P_co_or_ex))
Data_mut_ex_table$G1mutG2mut = 0
Data_mut_ex_table$G1mutG2wt = 0
Data_mut_ex_table$G1wtG2mut = 0
Data_mut_ex_table$G1wtG2wt = 0
Data_mut_ex_table$OddsRatio = 0
Data_mut_ex_table$OR_95CI_LL = 0
Data_mut_ex_table$OR_95CI_UL = 0
Gene_num = length(Gene_names)
Single_mut = rowSums(Mutation_matrix)
Total_patients = length(Mutation_matrix[1,])
for(i in 1:Gene_num){
  for(j in 1:Gene_num){
    tmp = sum(Mutation_matrix[i,] & Mutation_matrix[j,])
    Data_mut_ex_table$G1mutG2mut[(i-1)*Gene_num+j] = tmp
    Data_mut_ex_table$G1mutG2wt[(i-1)*Gene_num+j] = Single_mut[i] - tmp
    Data_mut_ex_table$G1wtG2mut[(i-1)*Gene_num+j] = Single_mut[j] - tmp
    Data_mut_ex_table$G1wtG2wt[(i-1)*Gene_num+j] = Total_patients - Single_mut[i] - Single_mut[j] + tmp
    tmp = fisher.test(matrix(c(tmp, Single_mut[i] - tmp, Single_mut[j] - tmp, Total_patients - Single_mut[i] - Single_mut[j] + tmp), nrow = 2))
    Data_mut_ex_table$OddsRatio[(i-1)*Gene_num+j] = tmp$estimate[[1]]
    Data_mut_ex_table$OR_95CI_LL[(i-1)*Gene_num+j] = tmp$conf.int[[1]]
    Data_mut_ex_table$OR_95CI_UL[(i-1)*Gene_num+j] = tmp$conf.int[[2]]
  }
}
Data_mut_ex_table = Data_mut_ex_table %>%
  dplyr::filter(!is.na(P_value))
Data_mut_ex_table = Data_mut_ex_table %>%
  dplyr::select(-P_co_or_ex)
DT::datatable(Data_mut_ex_table,
             caption = "Mutually exclusive or co-occurence",
             filter = 'top',
             extensions = 'Scroller')


Data_mut_ex$OddsRatio = NA
for(i in 1:length(Data_mut_ex[,1])){
  if(i %% 40 != 0 & i %% 40 <= floor(i/40)){
    Data_mut_ex[i,"OddsRatio"] = (Data_mut_ex_table %>% dplyr::filter(Gene_1 == Data_mut_ex[i,"Gene_1"] & Gene_2 == Data_mut_ex[i,"Gene_2"]))$OddsRatio
  }
}
Data_mut_ex$OddsRatio = -log10(Data_mut_ex$OddsRatio)
Data_mut_ex = Data_mut_ex %>%
  dplyr::mutate(OddsRatio_figure = case_when(
    OddsRatio > 3 ~ 3,
    OddsRatio < -3 ~ -3,
    TRUE ~ OddsRatio
  ))

g = Data_mut_ex %>% 
  ggplot(aes(x = Gene_1, y = Gene_2, fill = OddsRatio_figure)) + 
  geom_tile(color = "white") + 
  scale_fill_gradient2(low = "red", high = "blue", mid = "white", 
                       midpoint = 0, limit = c(-3, 3), 
                       name = "-log10(OddsRatio)", space = "Lab",
                       na.value = "white")+
  geom_text(aes(label = P_signif),
            color = "black", size = 4)+
  scale_y_discrete(limits = rev(levels(Data_mut_ex$Gene_2)))+
  theme_classic() +
  ggtitle("Mutually exclusive (blue) or co-occurrence (red)\n*: p<0.001, estimated with Rediscover package") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
pdf(file = paste(OutDirName, "mutually_exclusive_OddsRatio.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()


Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.エビデンス.エビデンスレベル,
                  .keep_all=TRUE) %>%
  dplyr::mutate(gene_mut = paste(C.CAT調査結果.変異情報.マーカー,
                                 C.CAT調査結果.変異情報.変異内容,
                                 sep=" _ ")) %>%
  dplyr::mutate(gene_patho = C.CAT調査結果.変異情報.マーカー)

oncogenic_mutations = Data_major_gene %>%
                dplyr::count(gene_mut) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_mutations
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_mut)
  tmp_no$n = lapply(tmp_no[,1], function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_mut)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_mutations = cbind(oncogenic_mutations, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_mutations) =c_name_gene

DT::datatable(oncogenic_mutations,
             caption = "Oncogenic mutations in selected diseases with prognostic information", filter = 'top', extensions = 'Scroller')
oncogenic_genes = Data_major_gene %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene
DT::datatable(oncogenic_genes,
             caption = "Genes with oncogenic mutations in selected diseases with prognostic information (counts)", filter = 'top', extensions = 'Scroller')
oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
  dplyr::distinct() %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
      dplyr::distinct() %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene
DT::datatable(oncogenic_genes,
             caption = "Genes with oncogenic mutations in selected diseases with prognostic information (patients)", filter = 'top', extensions = 'Scroller')
Data_major_gene = Data_major_gene %>%
  dplyr::filter(C.CAT調査結果.変異情報.マーカー %in%
                  oncogenic_genes[1:min(length(oncogenic_genes$all_patients), gene_number),]$gene_mutation)
  
## OncoPrint
oncogenic_genes = oncogenic_genes_save
PATH$frequent_genes = oncogenic_genes$gene_mutation[1:min(gene_number, 30)]
ht_opt$message = FALSE

Data_oncoprint = Data_case_target %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis) %>%
  dplyr::distinct() %>%
  dplyr::arrange(diagnosis)

mat_oncoprint = matrix("",
            nrow = length(Data_oncoprint$diagnosis),
            ncol = length(PATH$frequent_genes))
rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
colnames(mat_oncoprint) = PATH$frequent_genes
Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH$frequent_genes) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.変異種類,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  .keep_all=FALSE)

for(i in 1:length(Data_major_gene$C.CAT調査結果.基本項目.ハッシュID)){
  if(Data_major_gene[i,2] == "rearrangement"){
    mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
      paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
            Data_major_gene[i,4],
            sep=";")
  } else if(Data_major_gene[i,2] == "copy_number_alteration"){
    mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
      paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
            Data_major_gene[i,4],
            sep=";")
  } else if(Data_major_gene[i,2] == "small_scale_variant"){
    mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
      paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
            "small_scale_variant",
            sep=";")
  } else if(Data_major_gene[i,2] == "other_biomarker"){
    mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
      paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
            "TMB_high",
            sep=";")
  }
}

rownames(mat_oncoprint) = Data_oncoprint$diagnosis
mat_oncoprint = t(mat_oncoprint)

col = c("amplification" = "red",
        "loss" = "blue",
        "TMB_high" = "orange",
        "small_scale_variant" = "green",
        "rearrangement" = "purple",
        "fusion" = "gray",
        "truncation" = "black",
        "rearrangements" = "yellow")
alter_fun = list(
    background = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = "#CCCCCC", col = NA))
    },
    amplification = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = col["amplification"], col = NA))
    },
    loss = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
            gp = gpar(fill = col["loss"], col = NA))
    },
    TMB_high = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
            gp = gpar(fill = col["TMB_high"], col = NA))
    },
    small_scale_variant = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
            gp = gpar(fill = col["small_scale_variant"], col = NA))
    },
    rearrangement = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
            gp = gpar(fill = col["rearrangement"], col = NA))
    },
    fusion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
            gp = gpar(fill = col["fusion"], col = NA))
    },
    truncation = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
            gp = gpar(fill = col["truncation"], col = NA))
    },
    rearrangements = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
            gp = gpar(fill = col["rearrangements"], col = NA))
    }
)  

column_title = "OncoPrint_frequent_genes_detail"
tmp = mat_oncoprint
tmp[tmp != ""] = 1
tmp[tmp == ""] = 0
tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
tmp = data.frame(tmp)
colnames(tmp) = 1:ncol(tmp)
sample_order = 1:ncol(tmp)
colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
anno_diag = colnames(mat_oncoprint)
diag_list = unique(anno_diag)
diag_col = colPal3(length(diag_list))
names(diag_col) = diag_list
for(l in tmp2){
  if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
    sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
    tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
  }
}
sample_order_final = NULL
for(l in diag_list){
  sample_order_final = c(sample_order_final,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
}
ha = HeatmapAnnotation(Diagnosis = anno_diag,
	col = list(Diagnosis = diag_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis")))

ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
    column_order = sample_order, row_order = rev(tmp2),
    alter_fun = alter_fun, col = col, 
    column_title = column_title,
    remove_empty_columns = FALSE, remove_empty_rows = FALSE,
    bottom_annotation = ha,
    use_raster = FALSE,
    alter_fun_is_vectorized = TRUE)
#draw(ht)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_detail.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()

ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
    column_order = sample_order_final, row_order = rev(tmp2),
    alter_fun = alter_fun, col = col, 
    column_title = column_title,
    remove_empty_columns = FALSE, remove_empty_rows = FALSE,
    bottom_annotation = ha,
    use_raster = FALSE,
    alter_fun_is_vectorized = TRUE)
#draw(ht)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_detail_2.pdf",sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()

# young or old, threshold: median age
if(Define_median_age == 0){
  Median_age = median((Data_case_target %>%
          dplyr::select(sampling_age, C.CAT調査結果.基本項目.ハッシュID)
          %>% dplyr::distinct())$sampling_age)
}
Data_case_target = Data_case_target %>% dplyr::mutate(YoungOld = case_when(
  sampling_age <= Median_age ~ "Younger",
  TRUE ~ "Older"))
Data_survival = Data_case_target %>% dplyr::arrange(YoungOld)
Data_oncoprint = Data_survival %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis, YoungOld) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)

ID_target_drug = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)
mat_oncoprint = matrix("",
            nrow = length(Data_oncoprint$diagnosis),
            ncol = length(PATH$frequent_genes))
rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
colnames(mat_oncoprint) = PATH$frequent_genes
Data_major_gene_tmp = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_drug) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH$frequent_genes) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                  .keep_all=FALSE)

for(i in 1:length(Data_major_gene_tmp$C.CAT調査結果.基本項目.ハッシュID)){
  if(Data_major_gene_tmp[i,2] == "rearrangement"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            Data_major_gene_tmp[i,4],
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "copy_number_alteration"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            Data_major_gene_tmp[i,4],
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "small_scale_variant"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            "small_scale_variant",
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "other_biomarker"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            "TMB_high",
            sep=";")
  }
}

rownames(mat_oncoprint) = Data_oncoprint$diagnosis
mat_oncoprint = t(mat_oncoprint)
col = c("amplification" = "red",
        "loss" = "blue",
        "TMB_high" = "orange",
        "small_scale_variant" = "green",
        "rearrangement" = "purple",
        "fusion" = "gray",
        "truncation" = "black",
        "rearrangements" = "yellow")
alter_fun = list(
    background = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = "#CCCCCC", col = NA))
    },
    amplification = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = col["amplification"], col = NA))
    },
    loss = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
            gp = gpar(fill = col["loss"], col = NA))
    },
    TMB_high = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
            gp = gpar(fill = col["TMB_high"], col = NA))
    },
    small_scale_variant = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
            gp = gpar(fill = col["small_scale_variant"], col = NA))
    },
    rearrangement = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
            gp = gpar(fill = col["rearrangement"], col = NA))
    },
    fusion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
            gp = gpar(fill = col["fusion"], col = NA))
    },
    truncation = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
            gp = gpar(fill = col["truncation"], col = NA))
    },
    rearrangements = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
            gp = gpar(fill = col["rearrangements"], col = NA))
    }
)

column_title = paste("OncoPrint for selected diseases, frequently mutated genes, threshold: age <=", Median_age)

tmp = mat_oncoprint
tmp[tmp != ""] = 1
tmp[tmp == ""] = 0
tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
tmp = data.frame(tmp)
colnames(tmp) = 1:ncol(tmp)
sample_order = 1:ncol(tmp)
colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))

anno_diag = colnames(mat_oncoprint)
anno_RECIST = Data_oncoprint$YoungOld
diag_list = unique(anno_diag)
diag_col = colPal3(length(diag_list))
RECIST_list = unique(Data_oncoprint$YoungOld)
RECIST_col = colPal3(length(RECIST_list))
names(diag_col) = diag_list
names(RECIST_col) = RECIST_list

for(l in tmp2){
  if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
    sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
    tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
  }
}
sample_order_final = NULL
for(l in diag_list){
  sample_order_final = c(sample_order_final,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
}
sample_order = sample_order_final
sample_order_final = NULL
for(l in RECIST_list){
  sample_order_final = c(sample_order_final,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_RECIST == l)]])
}

heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
        labels = c("Oncogenic mut"))
ha = HeatmapAnnotation(Diagnosis = anno_diag,
  Age = anno_RECIST,
	col = list(Diagnosis = diag_col, Age = RECIST_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis"),
                                 Age = list(title = paste(Median_age, "<= Age or", Median_age, "> Age"))))
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, pct_gp = gpar(fontsize = 5),
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_Age_1.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_Age_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()

# Primary sample or metastatic site sample
Data_case_target = Data_case_target %>% dplyr::mutate(sampling_site = case_when(
  症例.検体情報.検体採取部位.名称. == "転移巣" ~ "Metastasis",
  TRUE ~ "Primary"))
Data_survival = Data_case_target %>% dplyr::arrange(sampling_site)
Data_oncoprint = Data_survival %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis, sampling_site) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)

ID_target_drug = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)
mat_oncoprint = matrix("",
            nrow = length(Data_oncoprint$diagnosis),
            ncol = length(PATH$frequent_genes))
rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
colnames(mat_oncoprint) = PATH$frequent_genes
Data_major_gene_tmp = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_drug) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH$frequent_genes) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                  .keep_all=FALSE)

for(i in 1:length(Data_major_gene_tmp$C.CAT調査結果.基本項目.ハッシュID)){
  if(Data_major_gene_tmp[i,2] == "rearrangement"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            Data_major_gene_tmp[i,4],
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "copy_number_alteration"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            Data_major_gene_tmp[i,4],
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "small_scale_variant"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            "small_scale_variant",
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "other_biomarker"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            "TMB_high",
            sep=";")
  }
}

rownames(mat_oncoprint) = Data_oncoprint$diagnosis
mat_oncoprint = t(mat_oncoprint)
col = c("amplification" = "red",
        "loss" = "blue",
        "TMB_high" = "orange",
        "small_scale_variant" = "green",
        "rearrangement" = "purple",
        "fusion" = "gray",
        "truncation" = "black",
        "rearrangements" = "yellow")
alter_fun = list(
    background = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = "#CCCCCC", col = NA))
    },
    amplification = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = col["amplification"], col = NA))
    },
    loss = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
            gp = gpar(fill = col["loss"], col = NA))
    },
    TMB_high = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
            gp = gpar(fill = col["TMB_high"], col = NA))
    },
    small_scale_variant = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
            gp = gpar(fill = col["small_scale_variant"], col = NA))
    },
    rearrangement = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
            gp = gpar(fill = col["rearrangement"], col = NA))
    },
    fusion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
            gp = gpar(fill = col["fusion"], col = NA))
    },
    truncation = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
            gp = gpar(fill = col["truncation"], col = NA))
    },
    rearrangements = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
            gp = gpar(fill = col["rearrangements"], col = NA))
    }
)

column_title = "OncoPrint for selected diseases, frequently mutated genes, sampling site"

tmp = mat_oncoprint
tmp[tmp != ""] = 1
tmp[tmp == ""] = 0
tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
tmp = data.frame(tmp)
colnames(tmp) = 1:ncol(tmp)
sample_order = 1:ncol(tmp)
colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))

anno_diag = colnames(mat_oncoprint)
anno_RECIST = Data_oncoprint$sampling_site
diag_list = unique(anno_diag)
diag_col = colPal3(length(diag_list))
RECIST_list = unique(Data_oncoprint$sampling_site)
RECIST_col = colPal3(length(RECIST_list))
names(diag_col) = diag_list
names(RECIST_col) = RECIST_list

for(l in tmp2){
  if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
    sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
    tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
  }
}
sample_order_final = NULL
for(l in diag_list){
  sample_order_final = c(sample_order_final,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
}
sample_order = sample_order_final
sample_order_final = NULL
for(l in RECIST_list){
  sample_order_final = c(sample_order_final,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_RECIST == l)]])
}

heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
        labels = c("Oncogenic mut"))
ha = HeatmapAnnotation(Diagnosis = anno_diag,
  Sampling_site = anno_RECIST,
	col = list(Diagnosis = diag_col, Sampling_site = RECIST_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis"),
                                 Sampling_site = list(title = "Sampling site")))
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, pct_gp = gpar(fontsize = 5),
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_Sampling_site_1.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_Sampling_site_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()

# Oncoprint with pathway alteration
if(length(target_pathway) > 0){
  Data_oncoprint = Data_case_target %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)
  
  mat_oncoprint = matrix("",
              nrow = length(Data_oncoprint$diagnosis),
              ncol = length(PATH[target_pathway[1]][[1]]))
  rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
  colnames(mat_oncoprint) = PATH[target_pathway[1]][[1]]
  Data_major_gene = Data_report %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[1]][[1]]) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=FALSE)

  for(i in 1:length(Data_major_gene$C.CAT調査結果.基本項目.ハッシュID)){
    if(Data_major_gene[i,2] == "rearrangement"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "copy_number_alteration"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "small_scale_variant"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "small_scale_variant",
              sep=";")
    } else if(Data_major_gene[i,2] == "other_biomarker"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "TMB_high",
              sep=";")
    }
  }

  rownames(mat_oncoprint) = Data_oncoprint$diagnosis
  mat_oncoprint = t(mat_oncoprint)

  col = c("amplification" = "red",
          "loss" = "blue",
          "TMB_high" = "orange",
          "small_scale_variant" = "green",
          "rearrangement" = "purple",
          "fusion" = "gray",
          "truncation" = "black",
          "rearrangements" = "yellow")
  alter_fun = list(
      background = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = "#CCCCCC", col = NA))
      },
      amplification = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = col["amplification"], col = NA))
      },
      loss = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
              gp = gpar(fill = col["loss"], col = NA))
      },
      TMB_high = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
              gp = gpar(fill = col["TMB_high"], col = NA))
      },
      small_scale_variant = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
              gp = gpar(fill = col["small_scale_variant"], col = NA))
      },
      rearrangement = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
              gp = gpar(fill = col["rearrangement"], col = NA))
      },
      fusion = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
              gp = gpar(fill = col["fusion"], col = NA))
      },
      truncation = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
              gp = gpar(fill = col["truncation"], col = NA))
      },
      rearrangements = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
              gp = gpar(fill = col["rearrangements"], col = NA))
      }
  )  
      
  column_title = paste("OncoPrint for", target_pathway[1], "pathway")
  tmp = mat_oncoprint
  tmp[tmp != ""] = 1
  tmp[tmp == ""] = 0
  tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
  tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
  tmp = data.frame(tmp)
  colnames(tmp) = 1:ncol(tmp)
  sample_order = 1:ncol(tmp)
  colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
  anno_diag = colnames(mat_oncoprint)
  diag_list = unique(anno_diag)
  diag_col = colPal3(length(diag_list))
  names(diag_col) = diag_list
  for(l in tmp2){
    if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
      sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
      tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
    }
  }
  sample_order_final = NULL
  for(l in diag_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
  }
  heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
          labels = c("Oncogenic mut"))
  ha = HeatmapAnnotation(Diagnosis = anno_diag,
	col = list(Diagnosis = diag_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis")))
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[1], ".pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  #draw(ht)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[1], "_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
}
if(length(target_pathway) > 1){
  Data_oncoprint = Data_case_target %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)
  
  mat_oncoprint = matrix("",
              nrow = length(Data_oncoprint$diagnosis),
              ncol = length(PATH[target_pathway[2]][[1]]))
  rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
  colnames(mat_oncoprint) = PATH[target_pathway[2]][[1]]
  Data_major_gene = Data_report %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[2]][[1]]) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=FALSE)

  for(i in 1:length(Data_major_gene$C.CAT調査結果.基本項目.ハッシュID)){
    if(Data_major_gene[i,2] == "rearrangement"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "copy_number_alteration"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "small_scale_variant"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "small_scale_variant",
              sep=";")
    } else if(Data_major_gene[i,2] == "other_biomarker"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "TMB_high",
              sep=";")
    }
  }

  rownames(mat_oncoprint) = Data_oncoprint$diagnosis
  mat_oncoprint = t(mat_oncoprint)

  col = c("amplification" = "red",
          "loss" = "blue",
          "TMB_high" = "orange",
          "small_scale_variant" = "green",
          "rearrangement" = "purple",
          "fusion" = "gray",
          "truncation" = "black",
          "rearrangements" = "yellow")
  alter_fun = list(
      background = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = "#CCCCCC", col = NA))
      },
      amplification = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = col["amplification"], col = NA))
      },
      loss = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
              gp = gpar(fill = col["loss"], col = NA))
      },
      TMB_high = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
              gp = gpar(fill = col["TMB_high"], col = NA))
      },
      small_scale_variant = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
              gp = gpar(fill = col["small_scale_variant"], col = NA))
      },
      rearrangement = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
              gp = gpar(fill = col["rearrangement"], col = NA))
      },
      fusion = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
              gp = gpar(fill = col["fusion"], col = NA))
      },
      truncation = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
              gp = gpar(fill = col["truncation"], col = NA))
      },
      rearrangements = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
            gp = gpar(fill = col["rearrangements"], col = NA))
      }
  )  
      
  column_title = paste("OncoPrint for", target_pathway[2], "pathway")
  tmp = mat_oncoprint
  tmp[tmp != ""] = 1
  tmp[tmp == ""] = 0
  tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
  tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
  tmp = data.frame(tmp)
  colnames(tmp) = 1:ncol(tmp)
  sample_order = 1:ncol(tmp)
  colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
  anno_diag = colnames(mat_oncoprint)
  diag_list = unique(anno_diag)
  diag_col = colPal3(length(diag_list))
  names(diag_col) = diag_list
  for(l in tmp2){
    if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
      sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
      tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
    }
  }
  sample_order_final = NULL
  for(l in diag_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
  }
  heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
          labels = c("Oncogenic mut"))
  ha = HeatmapAnnotation(Diagnosis = anno_diag,
	col = list(Diagnosis = diag_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis")))
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[2], ".pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  #draw(ht)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[2], "_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
}
if(length(target_pathway) > 2){
  Data_oncoprint = Data_case_target %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)
  
  mat_oncoprint = matrix("",
              nrow = length(Data_oncoprint$diagnosis),
              ncol = length(PATH[target_pathway[3]][[1]]))
  rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
  colnames(mat_oncoprint) = PATH[target_pathway[3]][[1]]
  Data_major_gene = Data_report %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[3]][[1]]) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=FALSE)

  for(i in 1:length(Data_major_gene$C.CAT調査結果.基本項目.ハッシュID)){
    if(Data_major_gene[i,2] == "rearrangement"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "copy_number_alteration"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "small_scale_variant"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "small_scale_variant",
              sep=";")
    } else if(Data_major_gene[i,2] == "other_biomarker"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "TMB_high",
              sep=";")
    }
  }

  rownames(mat_oncoprint) = Data_oncoprint$diagnosis
  mat_oncoprint = t(mat_oncoprint)

  col = c("amplification" = "red",
          "loss" = "blue",
          "TMB_high" = "orange",
          "small_scale_variant" = "green",
          "rearrangement" = "purple",
          "fusion" = "gray",
          "truncation" = "black",
          "rearrangements" = "yellow")
  alter_fun = list(
      background = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = "#CCCCCC", col = NA))
      },
      amplification = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = col["amplification"], col = NA))
      },
      loss = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
              gp = gpar(fill = col["loss"], col = NA))
      },
      TMB_high = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
              gp = gpar(fill = col["TMB_high"], col = NA))
      },
      small_scale_variant = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
              gp = gpar(fill = col["small_scale_variant"], col = NA))
      },
      rearrangement = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
              gp = gpar(fill = col["rearrangement"], col = NA))
      },
      fusion = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
              gp = gpar(fill = col["fusion"], col = NA))
      },
      truncation = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
              gp = gpar(fill = col["truncation"], col = NA))
      },
      rearrangements = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
              gp = gpar(fill = col["rearrangements"], col = NA))
      }
  )  
      
  column_title = paste("OncoPrint for", target_pathway[3], "pathway")
  tmp = mat_oncoprint
  tmp[tmp != ""] = 1
  tmp[tmp == ""] = 0
  tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
  tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
  tmp = data.frame(tmp)
  colnames(tmp) = 1:ncol(tmp)
  sample_order = 1:ncol(tmp)
  colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
  anno_diag = colnames(mat_oncoprint)
  diag_list = unique(anno_diag)
  diag_col = colPal3(length(diag_list))
  names(diag_col) = diag_list
  for(l in tmp2){
    if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
      sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
      tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
    }
  }
  sample_order_final = NULL
  for(l in diag_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
  }
  heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
          labels = c("Oncogenic mut"))
  ha = HeatmapAnnotation(Diagnosis = anno_diag,
	col = list(Diagnosis = diag_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis")))
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[3], ".pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  #draw(ht)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[3], "_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
}
if(length(target_pathway) > 3){
  Data_oncoprint = Data_case_target %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)
  
  mat_oncoprint = matrix("",
              nrow = length(Data_oncoprint$diagnosis),
              ncol = length(PATH[target_pathway[4]][[1]]))
  rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
  colnames(mat_oncoprint) = PATH[target_pathway[4]][[1]]
  Data_major_gene = Data_report %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[4]][[1]]) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=FALSE)

  for(i in 1:length(Data_major_gene$C.CAT調査結果.基本項目.ハッシュID)){
    if(Data_major_gene[i,2] == "rearrangement"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "copy_number_alteration"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "small_scale_variant"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "small_scale_variant",
              sep=";")
    } else if(Data_major_gene[i,2] == "other_biomarker"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "TMB_high",
              sep=";")
    }
  }

  rownames(mat_oncoprint) = Data_oncoprint$diagnosis
  mat_oncoprint = t(mat_oncoprint)

  col = c("amplification" = "red",
          "loss" = "blue",
          "TMB_high" = "orange",
          "small_scale_variant" = "green",
          "rearrangement" = "purple",
          "fusion" = "gray",
          "truncation" = "black",
          "rearrangements" = "yellow")
  alter_fun = list(
      background = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = "#CCCCCC", col = NA))
      },
      amplification = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = col["amplification"], col = NA))
      },
      loss = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
              gp = gpar(fill = col["loss"], col = NA))
      },
      TMB_high = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
              gp = gpar(fill = col["TMB_high"], col = NA))
      },
      small_scale_variant = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
              gp = gpar(fill = col["small_scale_variant"], col = NA))
      },
      rearrangement = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
              gp = gpar(fill = col["rearrangement"], col = NA))
      },
      fusion = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
              gp = gpar(fill = col["fusion"], col = NA))
      },
      truncation = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
              gp = gpar(fill = col["truncation"], col = NA))
      },
      rearrangements = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
              gp = gpar(fill = col["rearrangements"], col = NA))
      }
  )  
      
  column_title = paste("OncoPrint for", target_pathway[4], "pathway")
  tmp = mat_oncoprint
  tmp[tmp != ""] = 1
  tmp[tmp == ""] = 0
  tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
  tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
  tmp = data.frame(tmp)
  colnames(tmp) = 1:ncol(tmp)
  sample_order = 1:ncol(tmp)
  colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
  anno_diag = colnames(mat_oncoprint)
  diag_list = unique(anno_diag)
  diag_col = colPal3(length(diag_list))
  names(diag_col) = diag_list
  for(l in tmp2){
    if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
      sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
      tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
    }
  }
  sample_order_final = NULL
  for(l in diag_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
  }
  heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
          labels = c("Oncogenic mut"))
  ha = HeatmapAnnotation(Diagnosis = anno_diag,
	col = list(Diagnosis = diag_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis")))
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[4], ".pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  #draw(ht)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[4], "_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
}
if(length(target_pathway) > 4){
  Data_oncoprint = Data_case_target %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)
  
  mat_oncoprint = matrix("",
              nrow = length(Data_oncoprint$diagnosis),
              ncol = length(PATH[target_pathway[5]][[1]]))
  rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
  colnames(mat_oncoprint) = PATH[target_pathway[5]][[1]]
  Data_major_gene = Data_report %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[5]][[1]]) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=FALSE)

  for(i in 1:length(Data_major_gene$C.CAT調査結果.基本項目.ハッシュID)){
    if(Data_major_gene[i,2] == "rearrangement"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "copy_number_alteration"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              Data_major_gene[i,4],
              sep=";")
    } else if(Data_major_gene[i,2] == "small_scale_variant"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "small_scale_variant",
              sep=";")
    } else if(Data_major_gene[i,2] == "other_biomarker"){
      mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]] =
        paste(mat_oncoprint[Data_major_gene[i,1], Data_major_gene[i,3]],
              "TMB_high",
              sep=";")
    }
  }

  rownames(mat_oncoprint) = Data_oncoprint$diagnosis
  mat_oncoprint = t(mat_oncoprint)

  col = c("amplification" = "red",
          "loss" = "blue",
          "TMB_high" = "orange",
          "small_scale_variant" = "green",
          "rearrangement" = "purple",
          "fusion" = "gray",
          "truncation" = "black",
          "rearrangements" = "yellow")
  alter_fun = list(
      background = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = "#CCCCCC", col = NA))
      },
      amplification = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = col["amplification"], col = NA))
      },
      loss = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
              gp = gpar(fill = col["loss"], col = NA))
      },
      TMB_high = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
              gp = gpar(fill = col["TMB_high"], col = NA))
      },
      small_scale_variant = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
              gp = gpar(fill = col["small_scale_variant"], col = NA))
      },
      rearrangement = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
              gp = gpar(fill = col["rearrangement"], col = NA))
      },
      fusion = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
              gp = gpar(fill = col["fusion"], col = NA))
      },
      truncation = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
              gp = gpar(fill = col["truncation"], col = NA))
      },
      rearrangements = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
              gp = gpar(fill = col["rearrangements"], col = NA))
     }
  )  
      
  column_title = paste("OncoPrint for", target_pathway[5], "pathway")
  tmp = mat_oncoprint
  tmp[tmp != ""] = 1
  tmp[tmp == ""] = 0
  tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
  tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
  tmp = data.frame(tmp)
  colnames(tmp) = 1:ncol(tmp)
  sample_order = 1:ncol(tmp)
  colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
  anno_diag = colnames(mat_oncoprint)
  diag_list = unique(anno_diag)
  diag_col = colPal3(length(diag_list))
  names(diag_col) = diag_list
  for(l in tmp2){
    if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
      sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
      tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
    }
  }
  sample_order_final = NULL
  for(l in diag_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
  }
  heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
          labels = c("Oncogenic mut"))
  ha = HeatmapAnnotation(Diagnosis = anno_diag,
	col = list(Diagnosis = diag_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis")))
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[5], ".pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  #draw(ht)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_", target_pathway[5], "_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
}


if(EGFR_positive_LUAD == 1){
  WriteXLS(Data_case_target, ExcelFileName=paste(OutDirName, "EGFR_mut_all_drug.xls", sep=""))
  EGFR_table = read.csv(file = "source/EGFR_oncogenic_mutations.csv", skip = 0)
  for(i in unique(EGFR_table$C.CAT調査結果.変異情報.変異内容)){
    EGFR_exon_tmp = unique((EGFR_table %>%
              dplyr::filter(C.CAT調査結果.変異情報.変異内容 == i))$
                EGFR_group)
    if(length(Data_report[
        Data_report$C.CAT調査結果.変異情報.マーカー == "EGFR" &
        Data_report$C.CAT調査結果.変異情報.変異内容 == i ,]$
         C.CAT調査結果.エビデンス.エビデンスレベル) > 0){
        Data_report[
          Data_report$C.CAT調査結果.変異情報.マーカー == "EGFR" &
          Data_report$C.CAT調査結果.変異情報.変異内容 == i ,]$
           C.CAT調査結果.エビデンス.エビデンスレベル = EGFR_exon_tmp
    }
  }
}

Data_target_gene = NULL
for(i in 1:length(target_gene)){
  Data_target_gene = rbind(
    Data_report %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                  C.CAT調査結果.変異情報.変異内容 == target_gene[[i]][2]),
    Data_target_gene)
}
Data_target_gene = Data_target_gene %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  .keep_all=TRUE)

for(i in 1:length(target_gene)){
  Data_target_gene = rbind(
    Data_report %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                  C.CAT調査結果.エビデンス.エビデンスレベル == target_gene[[i]][2]) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.エビデンス.エビデンスレベル,
                  .keep_all=TRUE),
    Data_target_gene)
}

Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.エビデンス.エビデンスレベル,
                  .keep_all=TRUE) %>%
  dplyr::mutate(gene_mut = paste(C.CAT調査結果.変異情報.マーカー,
                                 C.CAT調査結果.変異情報.変異内容,
                                 sep=" _ ")) %>%
  dplyr::mutate(gene_patho = C.CAT調査結果.変異情報.マーカー)

Data_survival = Data_case_target %>%
  dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
text_patient = paste(text_patient, "patients underwent survival prolonging therapy: ", length((Data_survival %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE))$tmp), "\n", sep="") 

Data_survival = Data_case_target %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)

Data_survival$diagnosis_short = str_sub(Data_survival$diagnosis, 1, 5)
cancer_freq_order = names(sort(table(Data_survival$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival <- transform(Data_survival, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival <- transform(Data_survival, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))
g = ggplot(data = Data_survival, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title="  Tumor frequency")
pdf(file = paste(OutDirName, Cancer, "_subtype_frequency.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

if(Perihilar_is_Extrahepatic == 1){
  Data_case = Data_case %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Perihilar Cholangiocarcinoma (PHCH)_肝門部胆管癌"),
        replacement = "Extrahepatic Cholangiocarcinoma (EHCH)_肝外胆管癌"))
  Data_case_target = Data_case_target %>% dplyr::mutate(
  diagnosis=str_replace_all(diagnosis,
        fixed(pattern="Perihilar Cholangiocarcinoma"),
        replacement = "Extrahepatic Cholangiocarcinoma"))
  Diseases = unique(Data_case_target$diagnosis)
}

```

```{r clustering_drug, eval = (Analyze_clustering == 1), echo = (Source_code == 1 & Analyze_clustering == 1), warning=FALSE}
Genes_F1 = as.vector(str_split(unique((Data_report %>% dplyr::filter(C.CAT調査結果.基本項目.パネル名 %in% c("FoundationOne CDx DX1", "FoundationOne CDx DX2")))$C.CAT調査結果.変異情報.マーカー), pattern = "-", simplify = TRUE))
Genes_F1[Genes_F1 == "NKX2"] = "NKX2-1"
Genes_F1[Genes_F1 == "1"] = ""
Genes_F1 = Genes_F1[Genes_F1 != ""]
Genes_F1 = unique(Genes_F1)
Genes_NOP = as.vector(str_split(unique((Data_report %>% dplyr::filter(C.CAT調査結果.基本項目.パネル名 %in% c( "OncoGuide? NCC オンコパネル システム ver.2.01-00", "OncoGuide? NCC オンコパネル システム ver.2.01-01",
"OncoGuide? NCC オンコパネル システム ver.2.02-01", "OncoGuide? NCC オンコパネル システム ver.1.06-01",
"NCC oncopanel v4.1 ", "NCC oncopanel " )))$C.CAT調査結果.変異情報.マーカー), pattern = "-", simplify = TRUE))
Genes_NOP = unique(Genes_NOP)
Genes_both = Genes_NOP[Genes_NOP %in% Genes_F1]

Data_survival = Data_case_target %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)

ID = (Data_survival)$C.CAT調査結果.基本項目.ハッシュID
tmp_report = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.基本項目.ハッシュID %in% ID)
tmp_report_drug = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル %in% c( "E", "C", "D", "B", "A") &
                  C.CAT調査結果.基本項目.ハッシュID %in% ID) %>%
    dplyr::mutate(tmp2 = paste(tmp, C.CAT調査結果.エビデンス.薬剤, C.CAT調査結果.エビデンス.エビデンスレベル, sep="_"))
tmp_case_pre = Data_case_target %>% dplyr::filter(
                  !is.na(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) &
                  !is.na(症例.EP前レジメン情報.化学療法レジメン名称)) %>%
              dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
              dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                            症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                            症例.EP前レジメン情報.薬剤名.YJ一般名JP.,
                            症例.EP前レジメン情報.薬剤名.入力商品名.,
                            症例.EP前レジメン情報.実施目的.名称.,
                            症例.EP前レジメン情報.化学療法レジメン名称,
                            症例.EP前レジメン情報.投与開始日,
                            症例.EP前レジメン情報.治療ライン.名称.
                            ) %>%
              dplyr::distinct()
tmp_case_post = Data_case_target %>% dplyr::filter(
                  !is.na(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) &
                  !is.na(症例.EP後レジメン情報.化学療法レジメン名称)) %>%
              dplyr::arrange(症例.EP後レジメン情報.投与開始日) %>%
              dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                            症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                            症例.EP後レジメン情報.薬剤名.YJ一般名JP.,
                            症例.EP後レジメン情報.薬剤名.入力商品名.,
                            症例.EP後レジメン情報.治療方針.名称.,
                            症例.EP後レジメン情報.化学療法レジメン名称,
                            症例.EP後レジメン情報.投与開始日,
                            症例.EP後レジメン情報.治療ライン.名称.
                            ) %>%
              dplyr::distinct()

Treatment_table = data.frame(ID)
Treatment_table$histology = ""
Treatment_table$BTS_STS = ""
Treatment_table$metastasis_site = ""
Treatment_table$Treated_drug_Eng_pre = ""
Treatment_table$Treated_drug_Jap_pre = ""
Treatment_table$Treated_drug_Japanese_general_pre = ""
Treatment_table$Treated_drug_Japanese_product_pre = ""
Treatment_table$Treated_regimen_pre = ""
Treatment_table$Treatment_method_pre = ""
Treatment_table$Treatment_line_pre = ""
Treatment_table$Treatment_date_pre = ""
Treatment_table$mutation = ""
Treatment_table$matched_drug = ""
Treatment_table$Treatment_recommendation = ""
Treatment_table$Treated_with_matched_drug = ""
Treatment_table$Reason_matched_drug_unused = ""
Treatment_table$Treated_drug_Eng_post = ""
Treatment_table$Treated_drug_Jap_post = ""
Treatment_table$Treated_drug_Japanese_general_post = ""
Treatment_table$Treated_drug_Japanese_product_post = ""
Treatment_table$Treated_regimen_post = ""
Treatment_table$Treatment_method_post = ""
Treatment_table$Treatment_line_post = ""
Treatment_table$Treatment_date_post = ""

for(i in 1:length(ID)){
  Treatment_table$histology[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$diagnosis
  Treatment_table$BTS_STS[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$directory
  if(length((tmp_report %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))[,1]) > 0){
    Treatment_table$mutation[i] = paste0(as.character((tmp_report %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$tmp), collapse = " -> ")
  }
  if(length((tmp_report_drug %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))[,1]) > 0){
    Treatment_table$matched_drug[i] = paste0(as.character((tmp_report_drug %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$tmp2), collapse = " -> ")
  }
  Treatment_table$Treated_with_matched_drug[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.提示された治療薬を投与した.名称.
  Treatment_table$Treatment_recommendation[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.
  if(length(Data_case_target %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i])) > 0){
    Treatment_table$Treated_drug_Eng_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.)), collapse = " -> ")
    Treatment_table$Treated_drug_Jap_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.YJ一般名JP.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_general_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.入力一般名.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_product_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.入力商品名.)), collapse = " -> ")
    Treatment_table$Treated_regimen_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.化学療法レジメン名称)), collapse = " -> ")
  Treatment_table$Treatment_method_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.実施目的.名称.)), collapse = " -> ")
  Treatment_table$Treatment_line_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.治療ライン.名称.)), collapse = " -> ")
  Treatment_table$Treatment_date_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.投与開始日)), collapse = " -> ")
    Treatment_table$Treated_drug_Eng_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.)), collapse = " -> ")
    Treatment_table$Treated_drug_Jap_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.YJ一般名JP.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_genera_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.入力一般名.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_product_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.入力商品名.)), collapse = " -> ")
    Treatment_table$Treated_regimen_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.化学療法レジメン名称)), collapse = " -> ")
  Treatment_table$Treatment_method_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.治療方針.名称.)), collapse = " -> ")
  Treatment_table$Treatment_line_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.治療ライン.名称.)), collapse = " -> ")
  Treatment_table$Treatment_date_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.投与開始日)), collapse = " -> ")
    Treatment_table$metastasis_site[i] = paste0(as.character(unique((Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.がん種情報.登録時転移部位.名称.)), collapse = " -> ")
}
  Treatment_table$Reason_matched_drug_unused[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.提示された治療薬を投与しなかった理由.名称.
}

Data_tmp_pre = Data_case_target %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.投与開始日,
                症例.EP前レジメン情報.投与終了日,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                症例.EP前レジメン情報.最良総合効果.名称.
                ) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name",
                           "RECIST")
Data_tmp_post = Data_case_target %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.投与開始日,
                症例.EP後レジメン情報.投与終了日,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                症例.EP後レジメン情報.最良総合効果.名称.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name",
                           "RECIST")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "")

Data_drug_sequence$duration = as.Date(Data_drug_sequence$drug_end_date) - as.Date(Data_drug_sequence$drug_start_date)
Data_drug_sequence$duration = replace(Data_drug_sequence$duration, which(is.na(Data_drug_sequence$duration)), 0)

Data_drug_sequence = Data_drug_sequence %>%
  dplyr::mutate(RECIST = case_when(
    RECIST == "SD" & duration > LongSD_days ~ "LongSD",
    TRUE ~ RECIST
  ))
Data_drug_sequence$drug_group = 0
for(i in 1:length(selected_drug)){
  Data_drug_sequence = Data_drug_sequence %>%
    dplyr::mutate(drug_group = case_when(
      str_detect(drug_name, selected_drug[[i]]) ~ drug_group + 2^(i-1),
      TRUE ~ drug_group))
}
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(bitwAnd(drug_group,drug_pattern[[1]][[1]]) > 0)

Data_drug_sequence_tmp = Data_drug_sequence
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                drug_start_date,
                drug_end_date,
                RECIST) %>%
  dplyr::distinct()
Data_drug_sequence$regimen = 0
for(i in 1:length(Data_drug_sequence$RECIST)){
  Data_drug_sequence$regimen[[i]] = 
    sum((Data_drug_sequence_tmp %>% dplyr::filter(
          C.CAT調査結果.基本項目.ハッシュID == Data_drug_sequence$C.CAT調査結果.基本項目.ハッシュID[[i]] &
          drug_start_date == Data_drug_sequence$drug_start_date[[i]] &
          drug_end_date == Data_drug_sequence$drug_end_date[[i]] &
          RECIST == Data_drug_sequence$RECIST[[i]]))$drug_group)
}
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(regimen == drug_pattern[[1]][[1]]) %>%
  dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::mutate(RECIST = case_when(
    RECIST == "" ~ "NE",
    TRUE ~ RECIST
  )) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)

Data_survival = Data_survival %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_drug_sequence$C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
  
Data_survival$RECIST = Data_drug_sequence$RECIST

Data_survival_EriTraPaz = Data_survival

testResult = NULL
detail_text_d = rep("", length(unique(Data_survival$YoungOld)))
Grouping = unique(Data_survival$YoungOld)
ID_All = unique(Data_report$C.CAT調査結果.基本項目.ハッシュID)
for(i in 1:length(Grouping)){
  Data_survival_tmp = Data_survival
  Data_survival_tmp_RECIST = Data_survival
  if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "PD", "SD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else{
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  }
  EffectPathway = matrix(c(length((Data_survival_tmp %>% dplyr::filter(YoungOld == Grouping[i] & Effect == 1))$YoungOld),
                           length((Data_survival_tmp %>% dplyr::filter(YoungOld == Grouping[i] & Effect == 0))$YoungOld),
                           length((Data_survival_tmp %>% dplyr::filter(YoungOld != Grouping[i] & Effect == 1))$YoungOld),
                           length((Data_survival_tmp %>% dplyr::filter(YoungOld != Grouping[i] & Effect == 0))$YoungOld)),
                         nrow = 2,
                    dimnames = list(Effect = c("Drug_Effect +", "Drug_Effect -"),
                                  Age = c(Grouping[i], "Other")))
  testResult = c(testResult, list(EffectPathway))
  testResult = c(testResult, list(fisher.test(EffectPathway)))
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(YoungOld == Grouping[i])
  detail_text_d[i] = paste(
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
    ", ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
    sep="")
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(YoungOld != Grouping[i])
  detail_text_d[i] = paste(detail_text_d[i], " vs ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
    ", ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
    sep="")
}

testSummary = data.frame(matrix(rep(NA, length(testResult)*5), nrow=length(testResult)/2))
if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Age", "Drug", "Mut(+)Effect(+,CR/PR/LongSD)", "Mut(+)Effect(-,PD/ShortSD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Age", "Drug", "Mut(+)Effect(+,CR/PR/SD)", "Mut(+)Effect(-,PD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Age", "Drug", "Mut(+)Effect(+,CR/PR)", "Mut(+)Effect(-,PD/SD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Age", "Drug", "Mut(+)Effect(+,CR/PR/LongSD)", "Mut(+)Effect(-,PD/ShortSD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Age", "Drug", "Mut(+)Effect(+,CR/PR/SD)", "Mut(+)Effect(-,PD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else{
  colnames(testSummary) = c("Age", "Drug", "Mut(+)Effect(+,CR/PR)", "Mut(+)Effect(-,PD/SD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
}
Regimen_name = ""
for(i in 1:length(selected_drug)){
  if(bitwAnd(i, drug_pattern[[1]][[1]]) > 0){
    Regimen_name = paste(Regimen_name, selected_drug[[i]], sep="")
  }
}


for(i in 1:(length(testResult)/2)){
  testSummary[i,1] = Grouping[i]
  testSummary[i,2] = Regimen_name
  testSummary[i,3] = testResult[[i*2-1]][[1]]
  testSummary[i,4] = testResult[[i*2-1]][[2]]
  testSummary[i,5] = testResult[[i*2-1]][[3]]
  testSummary[i,6] = testResult[[i*2-1]][[4]]
  testSummary[i,7] = (testResult[[i*2]]$estimate)[[1]]
  testSummary[i,8] = (testResult[[i*2]]$conf.int)[[1]]
  testSummary[i,9] = (testResult[[i*2]]$conf.int)[[2]]
  testSummary[i,10] = (testResult[[i*2]]$p.value)[[1]]
}
testSummary$detail = detail_text_d
write.csv(testSummary, file = paste(OutDirName, Cancer, "_", Regimen_name, "_age_Summary.csv", sep=""))

testResult = NULL
detail_text_d = rep("", length(gene_pattern))
j = 0
target_gene_names = unlist(target_gene)[seq(1,length(unlist(target_gene)),2)]
ID_All = unique(Data_report$C.CAT調査結果.基本項目.ハッシュID)
mut_gene_list = NULL
for(i in 1:length(gene_pattern)){
  j = j + 1
  ID_mut = NULL
  for(k in 1:length(gene_pattern[[i]])){
    ID_tmp = ID_All
    for(l in 1:length(target_gene)){
      if(bitwAnd(2^(l-1), gene_pattern[[i]][k]) > 0){
        if(target_gene[[l]][2] == "WT"){
          Data_tmp = Data_report %>% dplyr::filter(
            C.CAT調査結果.変異情報.マーカー == target_gene[[l]][1] &
            C.CAT調査結果.エビデンス.エビデンスレベル_WT == "WT")
          ID_tmp2 = Data_tmp$C.CAT調査結果.基本項目.ハッシュID
          Data_tmp = Data_report %>% dplyr::filter(
          C.CAT調査結果.変異情報.マーカー == target_gene[[l]][1])
          ID_tmp3 = setdiff(ID_All, Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
          ID_tmp = intersect(ID_tmp, union(ID_tmp2, ID_tmp3))
        } else{
          Data_tmp = Data_report %>% dplyr::filter(
            C.CAT調査結果.変異情報.マーカー == target_gene[[l]][1] &
              (C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[l]][2] |
               C.CAT調査結果.変異情報.変異内容 == target_gene[[l]][2]))
          ID_tmp = intersect(ID_tmp, Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
        }
      }
    }
    ID_mut = union(ID_mut, ID_tmp)
  }
  
  mut_gene = ""
  for(l in 1:length(gene_pattern[[i]])){
    for(m in 1:length(target_gene_names)){
      if(bitwAnd(2^(m-1), gene_pattern[[i]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  mut_gene_list = c(mut_gene_list, mut_gene)
  Data_survival_tmp = Data_survival %>%
    dplyr::mutate(path_mut = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_mut ~ 1,
      TRUE ~ 0
    ))
  Data_survival_tmp_RECIST = Data_survival_tmp
  if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "PD", "SD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else{
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  }
  EffectPathway = matrix(c(length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 0))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 0))$path_mut)),
                         nrow = 2,
                    dimnames = list(Effect = c("Drug_Effect +", "Drug_Effect -"),
                                  Pathway = c(mut_gene, "Other")))
  testResult = c(testResult, list(EffectPathway))
  testResult = c(testResult, list(fisher.test(EffectPathway)))
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(path_mut == 1)
  detail_text_d[[j]] = paste(
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
    ", ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
    sep="")
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(path_mut == 0)
  detail_text_d[[j]] = paste(detail_text_d[[j]], " vs ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
    ", ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
    sep="")
}

testSummary = data.frame(matrix(rep(NA, length(testResult)*5), nrow=length(testResult)/2))
if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/LongSD)", "Mut(+)Effect(-,PD/ShortSD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/SD)", "Mut(+)Effect(-,PD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR)", "Mut(+)Effect(-,PD/SD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/LongSD)", "Mut(+)Effect(-,PD/ShortSD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/SD)", "Mut(+)Effect(-,PD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else{
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR)", "Mut(+)Effect(-,PD/SD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
}
Regimen_name = ""
for(i in 1:length(selected_drug)){
  if(bitwAnd(i, drug_pattern[[1]][[1]]) > 0){
    Regimen_name = paste(Regimen_name, selected_drug[[i]], sep="")
  }
}


for(i in 1:(length(testResult)/2)){
  testSummary[i,1] = mut_gene_list[i]
  testSummary[i,2] = Regimen_name
  testSummary[i,3] = testResult[[i*2-1]][[1]]
  testSummary[i,4] = testResult[[i*2-1]][[2]]
  testSummary[i,5] = testResult[[i*2-1]][[3]]
  testSummary[i,6] = testResult[[i*2-1]][[4]]
  testSummary[i,7] = (testResult[[i*2]]$estimate)[[1]]
  testSummary[i,8] = (testResult[[i*2]]$conf.int)[[1]]
  testSummary[i,9] = (testResult[[i*2]]$conf.int)[[2]]
  testSummary[i,10] = (testResult[[i*2]]$p.value)[[1]]
}
testSummary$detail = detail_text_d
write.csv(testSummary, file = paste(OutDirName, Cancer, "_", Regimen_name, "_target_mutation_Summary.csv", sep=""))

testResult = NULL
detail_text_d = rep("", length(target_pathway))
j = 0
for(i in target_pathway){
  j = j + 1
  Data_report_tmp = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.変異情報.マーカー %in% PATH[i][[1]])
  ID_path = unique(Data_report_tmp$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival_tmp = Data_survival %>%
    dplyr::mutate(path_mut = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  Data_survival_tmp_RECIST = Data_survival_tmp
  if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "PD", "SD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else{
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  }
  EffectPathway = matrix(c(length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 0))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 0))$path_mut)),
                         nrow = 2,
                    dimnames = list(Effect = c("Drug_Effect +", "Drug_Effect -"),
                                  Pathway = c(i, "No mutation")))
  testResult = c(testResult, list(EffectPathway))
  testResult = c(testResult, list(fisher.test(EffectPathway)))
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(path_mut == 1)
  detail_text_d[[j]] = paste(
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
    ", ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
    sep="")
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(path_mut == 0)
  detail_text_d[[j]] = paste(detail_text_d[[j]], " vs ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
    ", ",
    paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
    sep="")
}

testSummary = data.frame(matrix(rep(NA, length(testResult)*5), nrow=length(testResult)/2))
if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/LongSD)", "Mut(+)Effect(-,PD/ShortSD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/SD)", "Mut(+)Effect(-,PD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR)", "Mut(+)Effect(-,PD/SD/NE)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/LongSD)", "Mut(+)Effect(-,PD/ShortSD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR/SD)", "Mut(+)Effect(-,PD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
} else{
  colnames(testSummary) = c("Pathway", "Drug", "Mut(+)Effect(+,CR/PR)", "Mut(+)Effect(-,PD/SD)", "Mut(-)Effect(+)", "Mut(-)Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value")
}
Regimen_name = ""
for(i in 1:length(selected_drug)){
  if(bitwAnd(i, drug_pattern[[1]][[1]]) > 0){
    Regimen_name = paste(Regimen_name, selected_drug[[i]], sep="")
  }
}


for(i in 1:(length(testResult)/2)){
  testSummary[i,1] = colnames(testResult[[i*2-1]])[[1]]
  testSummary[i,2] = Regimen_name
  testSummary[i,3] = testResult[[i*2-1]][[1]]
  testSummary[i,4] = testResult[[i*2-1]][[2]]
  testSummary[i,5] = testResult[[i*2-1]][[3]]
  testSummary[i,6] = testResult[[i*2-1]][[4]]
  testSummary[i,7] = (testResult[[i*2]]$estimate)[[1]]
  testSummary[i,8] = (testResult[[i*2]]$conf.int)[[1]]
  testSummary[i,9] = (testResult[[i*2]]$conf.int)[[2]]
  testSummary[i,10] = (testResult[[i*2]]$p.value)[[1]]
}
testSummary$detail = detail_text_d
write.csv(testSummary, file = paste(OutDirName, Cancer, "_", Regimen_name, "_pathway_Summary.csv", sep=""))

detail_text = rep("", length(Diseases))
for(i in 1:length(Diseases)){
  ID_path = unique((Data_survival %>% dplyr::filter(
                    diagnosis == Diseases[i]))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival_tmp = Data_survival %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_path)
  if(length(Data_survival_tmp[,1]) > 0){
    detail_text[[i]] = paste(
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
      ", ",
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
      sep="")
  }
}

Data_survival = Data_case_target %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)
Data_survival$target_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_target_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:length(target_gene)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

if(sum(Data_survival$target_gene_count) == 0){
  target_gene = list(c(oncogenic_genes[1,1], "F"))
  gene_pattern = list(c(1))
  Data_target_gene = NULL
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.変異情報.変異内容 == target_gene[[i]][2]),
      Data_target_gene)
  }
  Data_target_gene = Data_target_gene %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=TRUE)
  
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.エビデンス.エビデンスレベル == target_gene[[i]][2]) %>%
      dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                      C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    C.CAT調査結果.エビデンス.エビデンスレベル,
                    .keep_all=TRUE),
      Data_target_gene)
  }

  Data_survival$target_gene_count = 0
  ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
  for(i in 1:length(ID_target_gene)){
    tmp = Data_target_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
    if(dplyr::count(tmp)[[1]] > 0){
      for(j in 1:dplyr::count(tmp)[[1]]){
        for(k in 1:length(target_gene)){
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
        }
      }
    }
  }
}

Data_survival$oncogenic_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_major_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:min(length(oncogenic_genes$all_patients), gene_number)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == oncogenic_genes$gene_mutation[[k]]){
          Data_survival$oncogenic_gene_count[i] = bitwOr(
            Data_survival$oncogenic_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

Data_report$C.CAT調査結果.変異情報TMB.TMB = as.numeric(str_split(Data_report$C.CAT調査結果.変異情報TMB.TMB, "Muts", simplify = TRUE)[,1])
Data_report_TMB = Data_report %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, C.CAT調査結果.変異情報TMB.TMB) %>%
  dplyr::distinct() %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_survival$C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
Data_survival = Data_survival %>%
  dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
Data_survival$TMB = Data_report_TMB$C.CAT調査結果.変異情報TMB.TMB


Data_report_F = Data_report %>% dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_survival$C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::filter(!C.CAT調査結果.変異情報.マーカー %in% c("TMB"))
if(Only_shared_genes == 1){
  Data_report_F = Data_report_F %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% Genes_both)
}

Gene_list = unique(Data_report_F$C.CAT調査結果.変異情報.マーカー)
Gene_list = sort(Gene_list)
Data_mutation = data.frame(matrix(0,
                                  nrow=length(Data_survival$C.CAT調査結果.基本項目.ハッシュID),
                                  ncol=length(Gene_list)))
colnames(Data_mutation) = Gene_list
rownames(Data_mutation) = Data_survival$C.CAT調査結果.基本項目.ハッシュID
Data_mutation$TMB_high_10 = as.numeric(Data_report_TMB$C.CAT調査結果.変異情報TMB.TMB >= 10)
Data_mutation[is.na(Data_mutation)] <- 0
for(i in 1:length(Data_report_F$C.CAT調査結果.変異情報.マーカー)){
  Data_mutation[Data_report_F$C.CAT調査結果.基本項目.ハッシュID[i], Data_report_F$C.CAT調査結果.変異情報.マーカー[i]] = 1
}
for(i in Gene_list){
  if(length(str_split(i, "-")[[1]]) > 1 & i != "NKX2-1"){
    for(j in str_split(i, "-")[[1]]){
      if(j %in% colnames(Data_mutation)){
        Data_mutation[,j] = as.numeric(Data_mutation[,i] | Data_mutation[,j])
      } else{
        Data_mutation[,j] = as.numeric(Data_mutation[,i])
      }
    }
    Data_mutation = Data_mutation %>% dplyr::select(- all_of(i))
  }
}
Gene_list = colnames(Data_mutation)

# Obsolete clustering method
# par(mfrow=c(1,3))
# title=tempdir()
# pdf(file = paste(OutDirName, Cancer, "_clustering_dendrogram.pdf", sep=""), width = 10, height = 10)
# ConsensusClusterPlus(as.matrix(Data_mutation),maxK=18,reps=50,pItem=0.8,pFeature=1,
#                      title=title,clusterAlg="hc",distance="pearson",seed=1)
# dev.off()

set.seed(1)
Data_mutation_umap = umap(Data_mutation, random_state=1)
Data_mutation_cord = Data_mutation_umap$layout %>% as.data.frame()
p =  ggplot(Data_mutation_cord, aes(V1,V2,color = Data_survival$diagnosis)) +
geom_point() + 
  labs(x = "UMAP 1", y = "UMAP 2",
       title = "Unsupervised clustering of oncogenic alterations") +
  theme_classic()
pdf(file = paste(OutDirName, Cancer, "_clustering_diagnosis.pdf", sep=""), width = 10, height = 10)
print(p, newpage = FALSE)
dev.off()

p =  ggplot(Data_mutation_cord, aes(V1,V2,color = Data_survival$YoungOld)) +
geom_point() + 
  labs(x = "UMAP 1", y = "UMAP 2",
       title = paste("Unsupervised clustering of oncogenic alterations, age threshold:", Median_age)) +
  theme_classic()
pdf(file = paste(OutDirName, Cancer, "_clustering_age.pdf", sep=""), width = 10, height = 10)
print(p, newpage = FALSE)
dev.off()

p =  ggplot(Data_mutation_cord, aes(V1,V2,color = Data_survival$sampling_site)) +
geom_point() + 
  labs(x = "UMAP 1", y = "UMAP 2",
       title = "Unsupervised clustering of oncogenic alterations, sampling site") +
  theme_classic()
pdf(file = paste(OutDirName, Cancer, "_clustering_sampling_site.pdf", sep=""), width = 10, height = 10)
print(p, newpage = FALSE)
dev.off()

#km_res <- kmeans(Data_mutation_cord, 10, nstart = 25)
#plot(Data_mutation_cord, col=km_res$cluster+1, main="K-means")
EPS = 1.0
MINPTS = 3
dbscan_res_changed <- dbscan(Data_mutation_cord, eps = EPS, minPts = MINPTS)

if(min(dbscan_res_changed$cluster) == 0){
  dbscan_res_changed$cluster = (dbscan_res_changed$cluster + 1)
  # plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
  #      pch = (dbscan_res_changed$cluster %% 26), main="DBSCAN clustering, cluster-1 is NOT-ASSIGNED")
  # labels <- 1:(max(dbscan_res_changed$cluster))
  # legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  pdf(file = paste(OutDirName, Cancer, "_clustering_UMAP.pdf", sep=""), width = 10, height = 10)
  plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
       pch = (dbscan_res_changed$cluster %% 26), main=paste(
         "DBSCAN clustering, cluster-1 is NOT-ASSIGNED, EPS:",
         EPS, ", minPts:", MINPTS, sep=""))
  labels <- 1:(max(dbscan_res_changed$cluster))
  legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  dev.off()
} else{
  # plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
  #      pch = (dbscan_res_changed$cluster %% 26), main="DBSCAN clustering")
  # labels <- 1:(max(dbscan_res_changed$cluster))
  # legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  pdf(file = paste(OutDirName, Cancer, "_clustering_UMAP.pdf", sep=""), width = 10, height = 10)
  plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
       pch = (dbscan_res_changed$cluster %% 26), main=paste(
         "DBSCAN clustering, EPS:",
         EPS, ", minPts:", MINPTS, sep=""))
  labels <- 1:(max(dbscan_res_changed$cluster))
  legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  dev.off()
}

Data_survival$driver_mutations = rowSums(Data_mutation)
Summary_Table = data.frame(Diseases)
Summary_Table$sample_is_primary_tumor = 0
Summary_Table$sample_is_metastatic_tumor = 0
Summary_Table$sample_is_unknown_tumor = 0
Summary_Table$tumor_purity_0_25 = 0
Summary_Table$tumor_purity_25_50 = 0
Summary_Table$tumor_purity_50_75 = 0
Summary_Table$tumor_purity_75_100 = 0
Summary_Table$tumor_purity_NA = 0
Summary_Table$age_median = 0
Summary_Table$age_lowest = 0
Summary_Table$age_highest = 0
Summary_Table$age_range_25 = 0
Summary_Table$age_range_75 = 0
Summary_Table$male = 0
Summary_Table$female = 0
Summary_Table$sex_unknown = 0
Summary_Table$oncogenic_driver_plus = 0
Summary_Table$oncogenic_driver_minus = 0
Summary_Table$TMB_median = 0
Summary_Table$TMB_lowest = 0
Summary_Table$TMB_highest = 0
Summary_Table$TMB_range_25 = 0
Summary_Table$TMB_range_75 = 0
Summary_Table$entropy = 0

Data_survival$cluster = dbscan_res_changed$cluster
Disease_cluster = data.frame(matrix(rep(0, length(Diseases) * max(Data_survival$cluster)),
                                    ncol = max(Data_survival$cluster)))
colnames(Disease_cluster) = seq(1,max(Data_survival$cluster))
rownames(Disease_cluster) = Diseases
for(i in 1:length(Diseases)){
  Summary_Table$sample_is_primary_tumor[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           症例.検体情報.検体採取部位.名称. == "原発巣"))$症例.検体情報.検体採取部位.名称.)
  Summary_Table$sample_is_metastatic_tumor[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           症例.検体情報.検体採取部位.名称. == "転移巣"))$症例.検体情報.検体採取部位.名称.)
  Summary_Table$sample_is_unknown_tumor[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           症例.検体情報.検体採取部位.名称. %in% c("不明","")))$症例.検体情報.検体採取部位.名称.)
  Summary_Table$tumor_purity_0_25[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.検体情報.腫瘍細胞含有割合) &
                           症例.検体情報.腫瘍細胞含有割合 < 25))$症例.検体情報.腫瘍細胞含有割合)
  Summary_Table$tumor_purity_25_50[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.検体情報.腫瘍細胞含有割合) &
                           症例.検体情報.腫瘍細胞含有割合 >= 25 &
                           症例.検体情報.腫瘍細胞含有割合 < 50))$症例.検体情報.腫瘍細胞含有割合)
  Summary_Table$tumor_purity_50_75[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.検体情報.腫瘍細胞含有割合) &
                           症例.検体情報.腫瘍細胞含有割合 >= 50 &
                           症例.検体情報.腫瘍細胞含有割合 < 75))$症例.検体情報.腫瘍細胞含有割合)
  Summary_Table$tumor_purity_75_100[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.検体情報.腫瘍細胞含有割合) &
                           症例.検体情報.腫瘍細胞含有割合 >= 75 &
                           症例.検体情報.腫瘍細胞含有割合 <= 100))$症例.検体情報.腫瘍細胞含有割合)
  Summary_Table$tumor_purity_NA[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           is.na(症例.検体情報.腫瘍細胞含有割合)))$症例.検体情報.腫瘍細胞含有割合)
  Summary_Table$age_median[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[3]]
  Summary_Table$age_lowest[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[1]]
  Summary_Table$age_highest[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[5]]
  Summary_Table$age_range_25[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[2]]
  Summary_Table$age_range_75[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[4]]
  Summary_Table$male[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.基本情報.性別.名称.) &
                           症例.基本情報.性別.名称. == "男"))$症例.基本情報.性別.名称.)
  Summary_Table$female[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(症例.基本情報.性別.名称.) &
                           症例.基本情報.性別.名称. == "女"))$症例.基本情報.性別.名称.)
  Summary_Table$sex_unknown[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           is.na(症例.基本情報.性別.名称.)))$症例.基本情報.性別.名称.)
  Summary_Table$oncogenic_driver_plus[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           driver_mutations > 0))$driver_mutations)
  Summary_Table$oncogenic_driver_minus[i] =
    length((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           driver_mutations == 0))$driver_mutations)
  Summary_Table$TMB_median[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(TMB)))$TMB)[[3]]
  Summary_Table$TMB_lowest[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(TMB)))$TMB)[[1]]
  Summary_Table$TMB_highest[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(TMB)))$TMB)[[5]]
  Summary_Table$TMB_range_25[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(TMB)))$TMB)[[2]]
  Summary_Table$TMB_range_75[i] =
    quantile((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i] &
                           !is.na(TMB)))$TMB)[[4]]
  tmp_cluster = as.data.frame(table((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i]))$cluster))
  Summary_Table$entropy[i] =
    shannon.entropy(as.numeric(as.character(tmp_cluster[,2])), max(Data_survival$cluster))
  for(j in 1:length(tmp_cluster[,1])){
    Disease_cluster[i,as.numeric(as.character(tmp_cluster[j,1]))] =
      as.numeric(as.character(tmp_cluster[j,2]))
  }
}
Summary_Table$detail = detail_text

Disease_cluster$Disease = rownames(Disease_cluster)
Disease_cluster <- transform(Disease_cluster, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = TRUE)))
colnames(Disease_cluster) = c(seq(1,max(Data_survival$cluster)), "Disease")
WriteXLS(Disease_cluster, row.names = TRUE,
         ExcelFileName = paste(OutDirName, Cancer, "_disease_cluster_table.xls", sep=""))

Age_group = unique(Data_survival$YoungOld)
Age_cluster = data.frame(matrix(rep(0, length(Age_group) * max(Data_survival$cluster)),
                                    ncol = max(Data_survival$cluster)))
colnames(Age_cluster) = seq(1,max(Data_survival$cluster))
rownames(Age_cluster) = Age_group
for(i in 1:length(Age_group)){
  tmp_cluster = as.data.frame(table((Data_survival %>%
             dplyr::filter(YoungOld == Age_group[i]))$cluster))
  if(length(tmp_cluster[,1]) > 0){
    for(j in 1:length(tmp_cluster[,1])){
      Age_cluster[i,as.numeric(as.character(tmp_cluster[j,1]))] =
        as.numeric(as.character(tmp_cluster[j,2]))
    }
  }
}

Age_cluster$Age = rownames(Age_cluster)
Age_cluster <- transform(Age_cluster, Age= factor(Age, levels = sort(unique(Age_cluster$Age), decreasing = TRUE)))
colnames(Age_cluster) = c(seq(1,max(Data_survival$cluster)), "Age")
WriteXLS(Age_cluster, row.names = TRUE,
         ExcelFileName = paste(OutDirName, Cancer, "_Age_cluster_table.xls", sep=""))

Data_entropy = gather(Disease_cluster, key = cluster, value = samples, -Disease)
Data_entropy$cluster = as.numeric(Data_entropy$cluster)

Summary_Table = transform(Summary_Table, Diseases= factor(Diseases, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))
Data_entropy = transform(Data_entropy, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))

g = ggplot(Summary_Table, aes(x=Diseases, y=entropy, fill="black"))
g = g + geom_point(stat = "identity") +
  coord_flip() +
  theme_classic()
#g
pdf(file = paste(OutDirName, Cancer, "_entropy_point.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggplot(Data_entropy, aes(x=Disease, y=samples, fill=cluster))
g = g + geom_bar(stat = "identity") +
  scale_fill_gradientn(colours = c("green", "black", "magenta", "darkred", "orange", "blue", "yellow")) +
  coord_flip() +
  theme_classic()
#g
pdf(file = paste(OutDirName, Cancer, "_entropy_cluster.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()
g = ggplot(Data_entropy, aes(x=Disease, y=samples, fill=cluster))
g = g + geom_bar(stat = "identity", position = "fill") +
  scale_fill_gradientn(colours = c("green", "black", "magenta", "darkred", "orange", "blue", "yellow")) +
  coord_flip() +
  theme_classic()
#g
pdf(file = paste(OutDirName, Cancer, "_entropy_cluster_2.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

gene_to_analyze = PATH$frequent_genes
gene_to_analyze = gene_to_analyze[gene_to_analyze %in% Gene_list]
Data_mutation_tmp = Data_mutation
Data_mutation_tmp$diagnosis = Data_survival$diagnosis
Summary_Gene_alteration = data.frame(matrix(0,
                                  nrow=length(Diseases),
                                  ncol=length(gene_to_analyze)))
colnames(Summary_Gene_alteration) = gene_to_analyze
rownames(Summary_Gene_alteration) = Diseases

for(i in Diseases){
  Data_tmp = Data_mutation_tmp %>% dplyr::filter(diagnosis == i)
  patient_no = length((Data_tmp)$TMB)
  for(j in gene_to_analyze){
    Summary_Gene_alteration[i,j] =
      sum(Data_tmp[,j]) / patient_no * 100
  }
}

Data_tmp = Summary_Gene_alteration
Data_tmp$Disease = rownames(Data_tmp)
Data_tmp = Data_tmp %>% gather(value = "freq", key = Gene, -Disease)
Data_tmp <- transform(Data_tmp, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))
g = ggplot(Data_tmp, aes(as.factor(Gene), as.factor(Disease))) +
     geom_tile(aes(fill = freq), color = "black",
#            lwd = 1,
            linetype = 1) + 
     geom_text(aes(label = round(freq, 0))) +
     scale_fill_gradient(low = "white", high = "red") +
     theme_classic() +
     labs(title = "frequently mutated genes and diagnoses") +
     theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank())
pdf(file = paste(OutDirName, Cancer, "_Oncogenic_mutation_heatmap_freq_genes.pdf", sep=""), width = 15, height = 10)
g
dev.off()

gene_to_analyze = PATH[target_pathway[[1]]][[1]]
gene_to_analyze = gene_to_analyze[gene_to_analyze %in% Gene_list]
Data_mutation_tmp = Data_mutation
Data_mutation_tmp$diagnosis = Data_survival$diagnosis
Summary_Gene_alteration = data.frame(matrix(0,
                                  nrow=length(Diseases),
                                  ncol=length(gene_to_analyze)))
colnames(Summary_Gene_alteration) = gene_to_analyze
rownames(Summary_Gene_alteration) = Diseases

for(i in Diseases){
  Data_tmp = Data_mutation_tmp %>% dplyr::filter(diagnosis == i)
  patient_no = length((Data_tmp)$TMB)
  for(j in gene_to_analyze){
    Summary_Gene_alteration[i,j] =
      sum(Data_tmp[,j]) / patient_no * 100
  }
}
Data_tmp = Summary_Gene_alteration
Data_tmp$Disease = rownames(Data_tmp)
Data_tmp = Data_tmp %>% gather(value = "freq", key = Gene, -Disease)
Data_tmp <- transform(Data_tmp, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))
g = ggplot(Data_tmp, aes(as.factor(Gene), as.factor(Disease))) +
     geom_tile(aes(fill = freq), color = "black",
#            lwd = 1,
            linetype = 1) + 
     geom_text(aes(label = round(freq, 0))) +
     scale_fill_gradient(low = "white", high = "red") +
     theme_classic() +
     labs(title = paste("Genes in ",  target_pathway[[1]], " and diagnoses", sep="")) +
     theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank())
pdf(file = paste(OutDirName, Cancer, "_Oncogenic_mutation_heatmap_target_pathway.pdf", sep=""), width = 15, height = 10)
g
dev.off()

gene_to_analyze = c("TP53", "CDK4", "MDM2", "WT1", "TERT", "FOXO1", "ATRX", "RB1", "SMARCB1", "KIT", "CDKN2A", "CDKN2B", "MYC", "NF1", "EED", "SUZ12", "PTEN", "PIK3CA", "TSC2", "BCOR", "CIC", "JUN", "GLI1", "CCNE1", "CCND3", "VEGFA", "IDH1", "FGFR4", "DICER1", "NRAS", "MYCN", "MED12", "KDR", "PDGFRA", "TP53BP1", "ERG", "STAG2")
gene_to_analyze = gene_to_analyze[gene_to_analyze %in% Gene_list]
Data_mutation_tmp = Data_mutation
Data_mutation_tmp$diagnosis = Data_survival$diagnosis
Data_mutation_tmp$YoungOld = Data_survival$YoungOld
Summary_Gene_alteration = data.frame(matrix(0,
                                  nrow=length(Diseases),
                                  ncol=length(gene_to_analyze)))
colnames(Summary_Gene_alteration) = gene_to_analyze
rownames(Summary_Gene_alteration) = Diseases

for(i in Diseases){
  Data_tmp = Data_mutation_tmp %>% dplyr::filter(diagnosis == i)
  patient_no = length((Data_tmp)$TMB)
  for(j in gene_to_analyze){
    Summary_Gene_alteration[i,j] =
      sum(Data_tmp[,j]) / patient_no * 100
  }
}
Data_mutation_tmp$idx = 1:length(Data_mutation[,1])
Data_mutation_tmp$cluster = dbscan_res_changed$cluster
tmp1 = Data_mutation_tmp
tmp1$C.CAT調査結果.基本項目.ハッシュID = rownames(tmp1)
tmp2 = Data_survival_EriTraPaz %>% dplyr::select(C.CAT調査結果.基本項目.ハッシュID, RECIST)
tmp3 = left_join(tmp1, tmp2, by = "C.CAT調査結果.基本項目.ハッシュID")

Data_report$C.CAT調査結果.変異情報TMB.TMB = as.numeric(str_sub(Data_report$C.CAT調査結果.変異情報TMB.TMB, start=1, end=6))
Data_report_TMB = Data_report %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, C.CAT調査結果.変異情報TMB.TMB) %>%
  dplyr::distinct() %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_survival$C.CAT調査結果.基本項目.ハッシュID) %>%
  dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
Data_survival = Data_survival %>%
  dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
Data_survival$TMB = Data_report_TMB$C.CAT調査結果.変異情報TMB.TMB

if(Cancer %in% c("Esophagus_Stomach", "Stomach", "Esophagus")){
  tmp4 = Data_survival %>% dplyr::select(
    C.CAT調査結果.基本項目.ハッシュID,
    症例.基本情報.性別.名称.,
    症例.基本情報.年齢,
    症例.背景情報.喫煙歴有無.名称.,
    症例.背景情報.喫煙年数,
    症例.背景情報.喫煙１日本数,
    症例.背景情報.アルコール多飲有無.名称.,
    症例.背景情報.ECOG.PS.名称.,
    症例.検体情報.パネル.名称.,
    症例.検体情報.腫瘍細胞含有割合,
    症例.検体情報.検体採取部位.名称.,
    症例.背景情報.病理診断名,
    症例.基本情報.がん種.OncoTree..名称.,
    症例.がん種情報.登録時転移有無.名称.,
    症例.がん種情報.食道.胃.腸.KRAS.名称.,
    症例.がん種情報.食道.胃.腸.KRAS.type.名称.,
    症例.がん種情報.食道.胃.腸.KRAS.検査方法.名称.,
    症例.がん種情報.食道.胃.腸.NRAS.名称.,
    症例.がん種情報.食道.胃.腸.NRAS.type.名称.,
    症例.がん種情報.食道.胃.腸.NRAS.検査方法.名称.,
    症例.がん種情報.食道.胃.腸.HER2.名称.,
    症例.がん種情報.食道.胃.腸.EGFR.IHC..名称.,
    症例.がん種情報.食道.胃.腸.BRAF.V600..名称.,
    directory)
} else{
  tmp4 = Data_survival %>% dplyr::select(
    C.CAT調査結果.基本項目.ハッシュID,
    症例.基本情報.性別.名称.,
    症例.基本情報.年齢,
    症例.背景情報.喫煙歴有無.名称.,
    症例.背景情報.喫煙年数,
    症例.背景情報.喫煙１日本数,
    症例.背景情報.アルコール多飲有無.名称.,
    症例.背景情報.ECOG.PS.名称.,
    症例.検体情報.パネル.名称.,
    症例.検体情報.腫瘍細胞含有割合,
    症例.背景情報.病理診断名,
    症例.管理情報.登録日,
    症例.EP後レジメン情報.エキスパートパネル開催日,
    症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日,
    症例.転帰情報.転帰.名称.,
    directory,
    TMB)
}
tmp3 = left_join(tmp3, tmp4, by = "C.CAT調査結果.基本項目.ハッシュID")
rownames(tmp3) = tmp3$C.CAT調査結果.基本項目.ハッシュID
tmp3 = tmp3[, colnames(tmp3) != "C.CAT調査結果.基本項目.ハッシュID"]
WriteXLS(Summary_Table, ExcelFileName = paste(OutDirName, Cancer, "_Summary.xls", sep=""))
WriteXLS(Summary_Gene_alteration, row.names = TRUE,
         ExcelFileName = paste(OutDirName, Cancer, "_Gene_alteration.xls", sep=""))
write.xlsx(tmp3, rowNames = TRUE, overwrite = TRUE,
         file = paste(OutDirName, Cancer, "_Gene_alteration_patient.xlsx", sep=""))
WriteXLS(Treatment_table, ExcelFileName = paste(OutDirName, Cancer, "_Treatment.xls", sep=""))

Data_tmp = Summary_Gene_alteration
Data_tmp$Disease = rownames(Data_tmp)
Data_tmp = Data_tmp %>% gather(value = "freq", key = Gene, -Disease)
Data_tmp <- transform(Data_tmp, Disease= factor(Disease, levels = sort(unique(Data_tmp$Disease), decreasing = TRUE)))
g = ggplot(Data_tmp, aes(as.factor(Gene), as.factor(Disease))) +
     geom_tile(aes(fill = freq), color = "black",
#            lwd = 1,
            linetype = 1) + 
     geom_text(aes(label = round(freq, 0))) +
     scale_fill_gradient(low = "white", high = "red") +
     theme_classic() +
     labs(title = paste("Selected genes and diagnoses", sep="")) +
     theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank())
pdf(file = paste(OutDirName, Cancer, "_Oncogenic_mutation_heatmap.pdf", sep=""), width = 15, height = 10)
g
dev.off()

Data_tmp_1 =data.frame(Summary_Table$Diseases)
Data_tmp_1$primary = Summary_Table$sample_is_primary_tumor
Data_tmp_1$Primary = "Yes"
Data_tmp_2 =data.frame(Summary_Table$Diseases)
Data_tmp_2$primary = Summary_Table$sample_is_metastatic_tumor
Data_tmp_2$Primary = "No"
Data_tmp_3 =data.frame(Summary_Table$Diseases)
Data_tmp_3$primary = Summary_Table$sample_is_unknown_tumor
Data_tmp_3$Primary = "Unknown"
Data_tmp = rbind(Data_tmp_1, Data_tmp_2, Data_tmp_3)
cancer_freq_order = c("Unknown", "No", "Yes")
Data_tmp <- transform(Data_tmp, Primary= factor(Primary, levels = cancer_freq_order))
Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_survival$diagnosis), decreasing = TRUE)))

chart_1 = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = primary, fill=Primary)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(legend.position = "none",
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())
chart_1_ = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = primary, fill=Primary)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(legend.position = "none",
      axis.ticks.y =  element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())
chart_1__ = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = primary, fill=Primary)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())


Data_tmp_1 =data.frame(Summary_Table$Diseases)
Data_tmp_1$purity = Summary_Table$tumor_purity_0_25
Data_tmp_1$Purity = "0-25%"
Data_tmp_2 =data.frame(Summary_Table$Diseases)
Data_tmp_2$purity = Summary_Table$tumor_purity_25_50
Data_tmp_2$Purity = "25-49%"
Data_tmp_3 =data.frame(Summary_Table$Diseases)
Data_tmp_3$purity = Summary_Table$tumor_purity_50_75
Data_tmp_3$Purity = "50-74%"
Data_tmp_4 =data.frame(Summary_Table$Diseases)
Data_tmp_4$purity = Summary_Table$tumor_purity_75_100
Data_tmp_4$Purity = "75-100%"
Data_tmp_5 =data.frame(Summary_Table$Diseases)
Data_tmp_5$purity = Summary_Table$tumor_purity_NA
Data_tmp_5$Purity = "NA"
Data_tmp = rbind(Data_tmp_1, Data_tmp_2, Data_tmp_3, Data_tmp_4, Data_tmp_5)
cancer_freq_order = c("NA", "75-100%", "50-74%", "25-49%", "0-25%")
Data_tmp <- transform(Data_tmp, Purity= factor(Purity, levels = cancer_freq_order))
Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_survival$diagnosis), decreasing = TRUE)))

chart_2 = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = purity, fill=Purity)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(legend.position = "none",
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())
chart_2_ = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = purity, fill=Purity)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())

Data_tmp = Data_survival
Data_tmp$Age = Data_survival$症例.基本情報.年齢
Data_tmp <- transform(Data_tmp, diagnosis= factor(diagnosis, levels = sort(unique(Data_survival$diagnosis), decreasing = TRUE)))

chart_3 = ggplot(data = Data_tmp, aes(x = diagnosis, y = Age)) +
  geom_boxplot(outlier.colour = NA) +
  coord_flip() +
  theme_classic() + 
  theme(legend.position = "none",
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())
chart_3_ = ggplot(data = Data_tmp, aes(x = diagnosis, y = Age)) +
  geom_boxplot(outlier.colour = NA) +
  coord_flip() +
  theme_classic() + 
  theme(
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())

Data_tmp_1 =data.frame(Summary_Table$Diseases)
Data_tmp_1$sex = Summary_Table$male
Data_tmp_1$Sex = "Male"
Data_tmp_2 =data.frame(Summary_Table$Diseases)
Data_tmp_2$sex = Summary_Table$female
Data_tmp_2$Sex = "Female"
Data_tmp_3 =data.frame(Summary_Table$Diseases)
Data_tmp_3$sex = Summary_Table$sex_unknown
Data_tmp_3$Sex = "Unknown"
Data_tmp = rbind(Data_tmp_1, Data_tmp_2, Data_tmp_3)
cancer_freq_order = c("Unknown", "Female", "Male")
Data_tmp <- transform(Data_tmp, Sex= factor(Sex, levels = cancer_freq_order))
Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_survival$diagnosis), decreasing = TRUE)))

chart_4 = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = sex, fill=Sex)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(legend.position = "none",
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())
chart_4_ = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = sex, fill=Sex)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())

Data_tmp_1 =data.frame(Summary_Table$Diseases)
Data_tmp_1$driver = Summary_Table$oncogenic_driver_plus
Data_tmp_1$Driver = "Yes"
Data_tmp_2 =data.frame(Summary_Table$Diseases)
Data_tmp_2$driver = Summary_Table$oncogenic_driver_minus
Data_tmp_2$Driver = "No"
Data_tmp = rbind(Data_tmp_1, Data_tmp_2)
cancer_freq_order = c("Unknown", "No", "Yes")
Data_tmp <- transform(Data_tmp, Driver= factor(Driver, levels = cancer_freq_order))
Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_survival$diagnosis), decreasing = TRUE)))

chart_5 = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = driver, fill=Driver)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(legend.position = "none",
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())
chart_5_ = ggplot(data = Data_tmp, aes(x = Summary_Table.Diseases, y = driver, fill=Driver)) +
  geom_col(position="fill") +
  coord_flip() +
  theme_classic() + 
  theme(
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank())

Data_tmp = Data_survival
Data_tmp <- transform(Data_tmp, diagnosis= factor(diagnosis, levels = sort(unique(Data_survival$diagnosis), decreasing = TRUE)))
chart_6 = ggplot(data = Data_tmp, aes(x = diagnosis, y = TMB)) +
  geom_boxplot(outlier.colour = NA) +
  coord_flip() +
  theme_classic() + 
  theme(legend.position = "none",
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank()) +
  scale_y_continuous(limits = c(0, 15))
chart_6_ = ggplot(data = Data_tmp, aes(x = diagnosis, y = TMB)) +
  geom_boxplot(outlier.colour = NA) +
  coord_flip() +
  theme_classic() + 
  theme(
      axis.ticks.y =  element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank()) +
  scale_y_continuous(limits = c(0, 15))

#grid.arrange(chart_1, chart_2, chart_3, chart_4, chart_5, chart_6, ncol = 6)
pdf(file = paste(OutDirName, Cancer, "_basic_data.pdf", sep=""), width = 10, height = 10)
grid.arrange(chart_1, chart_2, chart_3, chart_4, chart_5, chart_6, ncol = 6)
dev.off()
#grid.arrange(chart_1_, chart_1__, chart_2_, chart_3_, chart_4_, chart_5_, chart_6_, ncol = 7)
pdf(file = paste(OutDirName, Cancer, "_basic_data_legend.pdf", sep=""), width = 15, height = 10)
grid.arrange(chart_1_, chart_1__, chart_2_, chart_3_, chart_4_, chart_5_, chart_6_, ncol = 7)
dev.off()

Data_survival_save = Data_survival
Data_survival = Data_survival_EriTraPaz

Data_mutation_drug = Data_mutation[Data_survival$C.CAT調査結果.基本項目.ハッシュID,]
set.seed(1)
Data_mutation_umap = umap(Data_mutation_drug, random_state=1)
Data_mutation_cord = Data_mutation_umap$layout %>% as.data.frame()
p =  ggplot(Data_mutation_cord, aes(V1,V2,color = Data_survival$diagnosis)) +
geom_point() + 
  labs(x = "UMAP 1", y = "UMAP 2",
       title = "Unsupervised clustering of oncogenic alterations, pts with drug only") +
  theme_classic()
#p
pdf(file = paste(OutDirName, Cancer, "_clustering_with_drug_diagnosis.pdf", sep=""), width = 10, height = 10)
print(p, newpage = FALSE)
dev.off()

p =  ggplot(Data_mutation_cord, aes(V1,V2,color = Data_survival$YoungOld)) +
geom_point() + 
  labs(x = "UMAP 1", y = "UMAP 2",
  title = paste("Unsupervised clustering of oncogenic alterations, pts with drug only, age threshold:", Median_age)) +
  theme_classic()
pdf(file = paste(OutDirName, Cancer, "_clustering_with_drug_age.pdf", sep=""), width = 10, height = 10)
print(p, newpage = FALSE)
dev.off()

p =  ggplot(Data_mutation_cord, aes(V1,V2,color = Data_survival$RECIST)) +
geom_point() + 
  labs(x = "UMAP 1", y = "UMAP 2",
       title = paste("Unsupervised clustering of oncogenic alterations, pts with drug only, drug effect of", Regimen_name)) +
  theme_classic()
pdf(file = paste(OutDirName, Cancer, "_clustering_with_drug_drug_effect.pdf", sep=""), width = 10, height = 10)
print(p, newpage = FALSE)
dev.off()

#km_res <- kmeans(Data_mutation_cord, 10, nstart = 25)
#plot(Data_mutation_cord, col=km_res$cluster+1, main="K-means")
dbscan_res_changed <- dbscan(Data_mutation_cord, eps = 1.0, minPts = 5)

if(min(dbscan_res_changed$cluster) == 0){
  dbscan_res_changed$cluster = (dbscan_res_changed$cluster + 1)
  # plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
  #      pch = (dbscan_res_changed$cluster %% 26), main="DBSCAN clustering, cluster-1 is NOT-ASSIGNED, pts with drug only")
  # labels <- 1:(max(dbscan_res_changed$cluster))
  # legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  pdf(file = paste(OutDirName, Cancer, "_clustering_UMAP_with_drug.pdf", sep=""), width = 10, height = 10)
  plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
       pch = (dbscan_res_changed$cluster %% 26), main="DBSCAN clustering, cluster-1 is NOT-ASSIGNED, pts with drug only")
  labels <- 1:(max(dbscan_res_changed$cluster))
  legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  dev.off()
} else{
  # plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
  #      pch = (dbscan_res_changed$cluster %% 26), main="DBSCAN clustering, pts with drug only")
  # labels <- 1:(max(dbscan_res_changed$cluster))
  # legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  pdf(file = paste(OutDirName, Cancer, "_clustering_UMAP_with_drug.pdf", sep=""), width = 10, height = 10)
  plot(Data_mutation_cord, col=(dbscan_res_changed$cluster %% 29 + 1),
       pch = (dbscan_res_changed$cluster %% 26), main="DBSCAN clustering, pts with drug only")
  labels <- 1:(max(dbscan_res_changed$cluster))
  legend("topleft", legend = labels, pch=(labels %% 26), col= (labels %% 29 + 1))
  dev.off()
}

Summary_Table = data.frame(Diseases)
Summary_Table$entropy = 0

Data_survival$cluster_drug = dbscan_res_changed$cluster
Disease_cluster = data.frame(matrix(rep(0, length(Diseases) * max(Data_survival$cluster)),
                                    ncol = max(Data_survival$cluster)))
colnames(Disease_cluster) = seq(1,max(Data_survival$cluster))
rownames(Disease_cluster) = Diseases
for(i in 1:length(Diseases)){
  tmp_cluster = as.data.frame(table((Data_survival %>%
             dplyr::filter(diagnosis == Diseases[i]))$cluster))
  if(length(tmp_cluster[,1]) > 0){
    Summary_Table$entropy[i] =
      shannon.entropy(as.numeric(as.character(tmp_cluster[,2])), max(Data_survival$cluster))
    for(j in 1:length(tmp_cluster[,1])){
      Disease_cluster[i,as.numeric(as.character(tmp_cluster[j,1]))] =
        as.numeric(as.character(tmp_cluster[j,2]))
    }
  }
}

Disease_cluster$Disease = rownames(Disease_cluster)
Disease_cluster <- transform(Disease_cluster, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = TRUE)))
colnames(Disease_cluster) = c(seq(1,max(Data_survival$cluster)), "Disease")
WriteXLS(Disease_cluster, row.names = TRUE,
         ExcelFileName = paste(OutDirName, Cancer, "_disease_cluster_table_drug_only.xls", sep=""))

Age_group = unique(Data_survival$YoungOld)
Age_cluster = data.frame(matrix(rep(0, length(Age_group) * max(Data_survival$cluster)),
                                    ncol = max(Data_survival$cluster)))
colnames(Age_cluster) = seq(1,max(Data_survival$cluster))
rownames(Age_cluster) = Age_group
for(i in 1:length(Age_group)){
  tmp_cluster = as.data.frame(table((Data_survival %>%
             dplyr::filter(YoungOld == Age_group[i]))$cluster))
  if(length(tmp_cluster[,1]) > 0){
    for(j in 1:length(tmp_cluster[,1])){
      Age_cluster[i,as.numeric(as.character(tmp_cluster[j,1]))] =
        as.numeric(as.character(tmp_cluster[j,2]))
    }
  }
}

Age_cluster$Age = rownames(Age_cluster)
Age_cluster <- transform(Age_cluster, Age= factor(Age, levels = sort(unique(Age_cluster$Age), decreasing = TRUE)))
colnames(Age_cluster) = c(seq(1,max(Data_survival$cluster)), "Age")
WriteXLS(Age_cluster, row.names = TRUE,
         ExcelFileName = paste(OutDirName, Cancer, "_Age_cluster_table_drug_only.xls", sep=""))

Data_entropy = gather(Disease_cluster, key = cluster, value = samples, -Disease)
Data_entropy$cluster = as.numeric(Data_entropy$cluster)

Summary_Table = transform(Summary_Table, Diseases= factor(Diseases, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))
Data_entropy = transform(Data_entropy, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))

g = ggplot(Summary_Table, aes(x=Diseases, y=entropy, fill="black"))
g = g + geom_point(stat = "identity") +
  coord_flip() +
  theme_classic() +
  ggtitle("Patients with drug only")
#g
pdf(file = paste(OutDirName, Cancer, "_entropy_point_drug.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggplot(Data_entropy, aes(x=Disease, y=samples, fill=cluster))
g = g + geom_bar(stat = "identity") +
  scale_fill_gradientn(colours = c("green", "black", "magenta", "darkred", "orange", "blue", "yellow")) +
  coord_flip() +
  theme_classic() +
  ggtitle("Patients with drug only")
#g
pdf(file = paste(OutDirName, Cancer, "_entropy_cluster_drug.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()
g = ggplot(Data_entropy, aes(x=Disease, y=samples, fill=cluster))
g = g + geom_bar(stat = "identity", position = "fill") +
  scale_fill_gradientn(colours = c("green", "black", "magenta", "darkred", "orange", "blue", "yellow")) +
  coord_flip() +
  theme_classic() +
  ggtitle("Patients with drug only")
#g
pdf(file = paste(OutDirName, Cancer, "_entropy_cluster_drug_2.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()


trans <- data.frame(
  ID = rownames(Data_mutation_tmp),
  cluster = Data_mutation_tmp$cluster
)
Data_survival$cluster = lapply(list(Data_survival$C.CAT調査結果.基本項目.ハッシュID), function(x) {
  as.vector(trans$cluster[match(x, trans$ID)])
})[[1]]

Report_lolliplot_Patho = Data_report %>%
  dplyr::filter(C.CAT調査結果.変異情報.マーカー == lolliplot_gene &
                C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                C.CAT調査結果.変異情報.マーカー,
                C.CAT調査結果.変異情報.変異内容,
                C.CAT調査結果.変異情報.リファレンス塩基,
                C.CAT調査結果.変異情報.塩基変化,
                C.CAT調査結果.変異情報.物理位置) %>%
  dplyr::distinct()
Report_lolliplot_Patho$annotation = "Pathogenic"
Report_lolliplot_VUS = Data_report %>%
  dplyr::filter(C.CAT調査結果.変異情報.マーカー == lolliplot_gene) %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                C.CAT調査結果.変異情報.マーカー,
                C.CAT調査結果.変異情報.変異内容,
                C.CAT調査結果.変異情報.リファレンス塩基,
                C.CAT調査結果.変異情報.塩基変化,
                C.CAT調査結果.変異情報.物理位置) %>%
  dplyr::distinct()
Report_lolliplot_VUS$annotation = "VUS"
Report_lolliplot = rbind(Report_lolliplot_Patho, Report_lolliplot_VUS) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.変異情報.リファレンス塩基,
                  C.CAT調査結果.変異情報.塩基変化, .keep_all = TRUE) %>%
  dplyr::arrange(desc(C.CAT調査結果.変異情報.物理位置)) %>%
  dplyr::select(-C.CAT調査結果.変異情報.物理位置)
if(SD_is_effective == 2){
  Report_lolliplot = Report_lolliplot %>%
  dplyr::mutate(RECIST = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("CR", "PR", "LongSD")))$C.CAT調査結果.基本項目.ハッシュID ~ "CR/PR/LongSD",
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("PD")))$C.CAT調査結果.基本項目.ハッシュID ~ "PD",
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("NE", "SD")))$C.CAT調査結果.基本項目.ハッシュID ~ "NE/SD",
    TRUE ~ "unused"
    ))
} else if(SD_is_effective == 1){
  Report_lolliplot = Report_lolliplot %>%
  dplyr::mutate(RECIST = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("CR", "PR", "LongSD", "SD")))$C.CAT調査結果.基本項目.ハッシュID ~ "CR/PR/SD",
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("PD")))$C.CAT調査結果.基本項目.ハッシュID ~ "PD",
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("NE")))$C.CAT調査結果.基本項目.ハッシュID ~ "NE",
    TRUE ~ "unused"
    ))
} else{
  Report_lolliplot = Report_lolliplot %>%
  dplyr::mutate(RECIST = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("CR", "PR")))$C.CAT調査結果.基本項目.ハッシュID ~ "CR/PR",
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("PD")))$C.CAT調査結果.基本項目.ハッシュID ~ "PD",
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_survival %>% dplyr::filter(
      RECIST %in% c("NE", "SD")))$C.CAT調査結果.基本項目.ハッシュID ~ "NE/SD",
    TRUE ~ "unused"
    ))
}
write.csv(Report_lolliplot, file = paste(OutDirName, Cancer, "_lolliplot_", Regimen_name, "_effect.csv", sep=""))
if(Figure_lolliplot == 1){
  db = useMart("ensembl")
  hd = useDataset("hsapiens_gene_ensembl", mart = db)
  transcript = getBM(attributes = c("transcript_is_canonical", "ensembl_transcript_id"),
        filters = "hgnc_symbol",
        values = lolliplot_gene,
        mart = hd)
  transcript = (transcript %>% dplyr::filter(transcript_is_canonical == 1))$ensembl_transcript_id
  data_lolliplot = Report_lolliplot %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  annotation,
                  RECIST)
  data_lolliplot = cbind(data_lolliplot, transcript)
  colnames(data_lolliplot) = c("ID", "gene", "amino_acid_change", "annotation",
                               "RECIST", "transcript_name")
  data_lolliplot = data_lolliplot %>%
    dplyr::filter(!amino_acid_change %in% c("rearrangements",
                                            "truncation",
                                            "amplification",
                                            "fusion",
                                            "deletion",
                                            "loss")) %>%
    dplyr::filter(!str_detect(amino_acid_change, "c."))
  data_lolliplot$amino_acid_change = paste("p.", data_lolliplot$amino_acid_change, sep="")
  pdf(file = paste(OutDirName, Cancer, "_lolliplot.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  lolliplot((data_lolliplot %>% dplyr::filter(RECIST != "unused")),
            fillCol = "RECIST")
  dev.off()
}

testResult = NULL
detail_text = rep("", max(Data_mutation_tmp$cluster))
for(i in 1:max(Data_mutation_tmp$cluster)){
  ID_path = unique((Data_survival %>% dplyr::filter(
                    cluster == i))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival_tmp = Data_survival %>%
    dplyr::mutate(path_mut = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  Data_survival_tmp_RECIST = Data_survival_tmp
  if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "PD", "SD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else{
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  }
  EffectPathway = matrix(c(length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 0))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 0))$path_mut)),
                         nrow = 2,
                    dimnames = list(Effect = c("Drug_Effect +", "Drug_Effect -"),
                                  Pathway = c(paste("cluster", i), "Other clusters")))
  testResult = c(testResult, list(EffectPathway))
  testResult = c(testResult, list(fisher.test(EffectPathway)))
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(path_mut == 1)
  if(length(Data_survival_tmp[,1]) > 0){
    detail_text[[i]] = paste(
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
      ", ",
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
      sep="")
  }
}

testSummary = data.frame(matrix(rep(NA, length(testResult)*5.5), nrow=length(testResult)/2))
if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/LongSD)", "Cluster(+)Effect(-,PD/ShortSD/NE)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/SD)", "Cluster(+)Effect(-,PD/NE)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR)", "Cluster(+)Effect(-,PD/SD/NE)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/LongSD)", "Cluster(+)Effect(-,PD/ShortSD)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/SD)", "Cluster(+)Effect(-,PD)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else{
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR)", "Cluster(+)Effect(-,PD/SD)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
}
for(i in 1:(length(testResult)/2)){
  testSummary[i,1] = paste("Cluster", i)
  testSummary[i,2] = Regimen_name
  testSummary[i,3] = testResult[[i*2-1]][[1]]
  testSummary[i,4] = testResult[[i*2-1]][[2]]
  testSummary[i,5] = testResult[[i*2-1]][[3]]
  testSummary[i,6] = testResult[[i*2-1]][[4]]
  testSummary[i,7] = (testResult[[i*2]]$estimate)[[1]]
  testSummary[i,8] = (testResult[[i*2]]$conf.int)[[1]]
  testSummary[i,9] = (testResult[[i*2]]$conf.int)[[2]]
  testSummary[i,10] = (testResult[[i*2]]$p.value)[[1]]
  testSummary[i,11] = detail_text[[i]]
}
write.csv(testSummary, file = paste(OutDirName, Cancer, "_", Regimen_name, "_cluster_Summary.csv", sep=""))

testSummary = data.frame(matrix(rep("", max(Data_mutation_tmp$cluster)*4),
                                nrow=max(Data_mutation_tmp$cluster)))
colnames(testSummary) = c("Cluster", "Patients", "Positively_selected", "Negatively_selected")
OddsRatio_summary = NULL
Pvalue_summary = NULL
if(length(unique(Data_mutation_tmp$cluster)) > 1){
  for(i in 1:max(Data_mutation_tmp$cluster)){
    testSummary[i,1] = paste("Cluster", i)
    Data_cluster_plus = Data_mutation_tmp %>%
      dplyr::filter(cluster == i) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld) %>%
      colSums()
    Data_cluster_minus = Data_mutation_tmp %>%
      dplyr::filter(cluster != i) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld) %>%
      colSums()
    No_plus_neg = length((Data_mutation_tmp %>% dplyr::filter(cluster == i))[,1]) -
      Data_cluster_plus
    No_minus_neg = length((Data_mutation_tmp %>% dplyr::filter(cluster != i))[,1]) -
      Data_cluster_minus
    testSummary[i,2] = length((Data_mutation_tmp %>% dplyr::filter(cluster == i))[,1])
    OddsRatio = round(Data_cluster_plus / No_plus_neg /
                Data_cluster_minus * No_minus_neg, digits = 2)
    OddsRatio = replace(OddsRatio, which(is.na(OddsRatio)), 0)
    Pvalue = OddsRatio
    Pvalue[1:length(Pvalue)] = 1
    for(j in 1:length(Data_cluster_plus)){
      if(OddsRatio[j] > 2){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(paste("cluster", i), "Other clusters")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,3] = paste(testSummary[i,3],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
      if(OddsRatio[j] < 0.5){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(paste("cluster", i), "Other clusters")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,4] = paste(testSummary[i,4],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
    }
    OddsRatio_summary = rbind(OddsRatio_summary, OddsRatio)
    Pvalue_summary = rbind(Pvalue_summary, Pvalue)
  }
  write.csv(testSummary, file = paste(OutDirName, Cancer, "_cluster_characteristics.csv", sep=""))
}

Data_mutation_tmp_save = Data_mutation_tmp
Data_mutation_tmp$sampling_site = Data_survival_save$sampling_site
Sampling_site_group = unique(Data_mutation_tmp$sampling_site)
testSummary = data.frame(matrix(rep("", length(Sampling_site_group)*4),
                                nrow=length(Sampling_site_group)))
colnames(testSummary) = c("Sampling site", "Patients", "Positively_selected", "Negatively_selected")
OddsRatio_summary = NULL
Pvalue_summary = NULL
if(length(Sampling_site_group) > 1){
  for(i in 1:length(Sampling_site_group)){
    testSummary[i,1] = Sampling_site_group[i]
    Data_cluster_plus = Data_mutation_tmp %>%
      dplyr::filter(sampling_site == Sampling_site_group[i]) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld, -sampling_site) %>%
      colSums()
    Data_cluster_minus = Data_mutation_tmp %>%
      dplyr::filter(sampling_site != Sampling_site_group[i]) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld, -sampling_site) %>%
      colSums()
    No_plus_neg = length((Data_mutation_tmp %>% dplyr::filter(sampling_site == Sampling_site_group[i]))[,1]) -
      Data_cluster_plus
    No_minus_neg = length((Data_mutation_tmp %>% dplyr::filter(sampling_site != Sampling_site_group[i]))[,1]) -
      Data_cluster_minus
    testSummary[i,2] = length((Data_mutation_tmp %>% dplyr::filter(sampling_site == Sampling_site_group[i]))[,1])
    OddsRatio = round(Data_cluster_plus / No_plus_neg /
                Data_cluster_minus * No_minus_neg, digits = 2)
    OddsRatio = replace(OddsRatio, which(is.na(OddsRatio)), 0)
    Pvalue = OddsRatio
    Pvalue[1:length(Pvalue)] = 1
    for(j in 1:length(Data_cluster_plus)){
      if(OddsRatio[j] > 2){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(Sampling_site_group[i], "Others")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,3] = paste(testSummary[i,3],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
      if(OddsRatio[j] < 0.5){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(Sampling_site_group[i], "Others")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,4] = paste(testSummary[i,4],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
    }
    OddsRatio_summary = rbind(OddsRatio_summary, OddsRatio)
    Pvalue_summary = rbind(Pvalue_summary, Pvalue)
  }
  write.csv(testSummary, file = paste(OutDirName, Cancer, "_sampling_site_characteristics.csv", sep=""))
}

Data_mutation_tmp = Data_mutation_tmp_save
diagnosis_group = unique(Data_mutation_tmp$diagnosis)
testSummary = data.frame(matrix(rep("", length(diagnosis_group)*4),
                                nrow=length(diagnosis_group)))
colnames(testSummary) = c("Diagnosis", "Patients", "Positively_selected", "Negatively_selected")
OddsRatio_summary = NULL
Pvalue_summary = NULL
if(length(diagnosis_group) > 1){
  for(i in 1:length(diagnosis_group)){
    testSummary[i,1] = diagnosis_group[i]
    Data_cluster_plus = Data_mutation_tmp %>%
      dplyr::filter(diagnosis == diagnosis_group[i]) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld) %>%
      colSums()
    Data_cluster_minus = Data_mutation_tmp %>%
      dplyr::filter(diagnosis != diagnosis_group[i]) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld) %>%
      colSums()
    No_plus_neg = length((Data_mutation_tmp %>% dplyr::filter(diagnosis == diagnosis_group[i]))[,1]) -
      Data_cluster_plus
    No_minus_neg = length((Data_mutation_tmp %>% dplyr::filter(diagnosis != diagnosis_group[i]))[,1]) -
      Data_cluster_minus
    testSummary[i,2] = length((Data_mutation_tmp %>% dplyr::filter(diagnosis == diagnosis_group[i]))[,1])
    OddsRatio = round(Data_cluster_plus / No_plus_neg /
                Data_cluster_minus * No_minus_neg, digits = 2)
    OddsRatio = replace(OddsRatio, which(is.na(OddsRatio)), 0)
    Pvalue = OddsRatio
    Pvalue[1:length(Pvalue)] = 1
    for(j in 1:length(Data_cluster_plus)){
      if(OddsRatio[j] > 2){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(diagnosis_group[i], "Others")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,3] = paste(testSummary[i,3],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
      if(OddsRatio[j] < 0.5){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(diagnosis_group[i], "Others")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,4] = paste(testSummary[i,4],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
    }
    OddsRatio_summary = rbind(OddsRatio_summary, OddsRatio)
    Pvalue_summary = rbind(Pvalue_summary, Pvalue)
  }
  write.csv(testSummary, file = paste(OutDirName, Cancer, "_diagnosis_characteristics.csv", sep=""))
}

Data_mutation_tmp = Data_mutation_tmp_save
testSummary = data.frame(matrix(rep("", length(Grouping)*4),
                                nrow=length(Grouping)))
colnames(testSummary) = c("Age", "Patients", "Positively_selected", "Negatively_selected")
OddsRatio_summary = NULL
Pvalue_summary = NULL
if(length(Grouping) > 1){
  for(i in 1:length(Grouping)){
    testSummary[i,1] = Grouping[i]
    Data_cluster_plus = Data_mutation_tmp %>%
      dplyr::filter(YoungOld == Grouping[i]) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld) %>%
      colSums()
    Data_cluster_minus = Data_mutation_tmp %>%
      dplyr::filter(YoungOld != Grouping[i]) %>%
      dplyr::select(-diagnosis, -idx, -cluster, -YoungOld) %>%
      colSums()
    No_plus_neg = length((Data_mutation_tmp %>% dplyr::filter(YoungOld == Grouping[i]))[,1]) -
      Data_cluster_plus
    No_minus_neg = length((Data_mutation_tmp %>% dplyr::filter(YoungOld != Grouping[i]))[,1]) -
      Data_cluster_minus
    testSummary[i,2] = length((Data_mutation_tmp %>% dplyr::filter(YoungOld == Grouping[i]))[,1])
    OddsRatio = round(Data_cluster_plus / No_plus_neg /
                Data_cluster_minus * No_minus_neg, digits = 2)
    OddsRatio = replace(OddsRatio, which(is.na(OddsRatio)), 0)
    Pvalue = OddsRatio
    Pvalue[1:length(Pvalue)] = 1
    for(j in 1:length(Data_cluster_plus)){
      if(OddsRatio[j] > 2){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(Grouping[i], "Others")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,3] = paste(testSummary[i,3],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
      if(OddsRatio[j] < 0.5){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(Grouping[i], "Others")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,4] = paste(testSummary[i,4],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
    }
    OddsRatio_summary = rbind(OddsRatio_summary, OddsRatio)
    Pvalue_summary = rbind(Pvalue_summary, Pvalue)
  }
  write.csv(testSummary, file = paste(OutDirName, Cancer, "_Age_characteristics.csv", sep=""))
}

Data_mutation_tmp = Data_mutation_tmp_save

if(length(unique(Data_mutation_tmp$diagnosis)) > 1){
  testSummary = data.frame(matrix(rep("", max(Data_mutation_tmp$cluster)*4),
                                  nrow=max(Data_mutation_tmp$cluster)))
  colnames(testSummary) = c("Cluster", "Patients", "Positively_selected", "Negatively_selected")
  OddsRatio_summary = NULL
  Pvalue_summary = NULL
  Data_cluster = Data_mutation_tmp
  for(i in 1:max(Data_cluster$cluster)){
    testSummary[i,1] = paste("Cluster", i)
    pos_num = data.frame(Diseases)
    pos_num$n = 0
    colnames(pos_num) = c("diagnosis", "n")
    Data_cluster_plus = Data_cluster %>%
      dplyr::filter(cluster == i) %>%
        dplyr::count(diagnosis) %>%
        dplyr::arrange(-n)
    Data_cluster_plus = dplyr::full_join(Data_cluster_plus, pos_num, by = join_by(diagnosis))
    Data_cluster_plus = Data_cluster_plus[,1:2]
    colnames(Data_cluster_plus) = c("diagnosis", "n")
    Data_cluster_plus$n = replace_na(Data_cluster_plus$n, 0)
    Data_cluster_plus = Data_cluster_plus %>% dplyr::arrange(diagnosis)
    Data_cluster_plus = Data_cluster_plus$n
    pos_num = data.frame(Diseases)
    pos_num$n = 0
    colnames(pos_num) = c("diagnosis", "n")
    Data_cluster_minus = Data_cluster %>%
      dplyr::filter(cluster != i) %>%
        dplyr::count(diagnosis) %>%
        dplyr::arrange(-n)
    Data_cluster_minus = dplyr::full_join(Data_cluster_minus, pos_num, by = join_by(diagnosis))
    Data_cluster_minus = Data_cluster_minus[,1:2]
    colnames(Data_cluster_minus) = c("diagnosis", "n")
    Data_cluster_minus$n = replace_na(Data_cluster_minus$n, 0)
    Data_cluster_minus = Data_cluster_minus %>% dplyr::arrange(diagnosis)
    Data_cluster_minus = Data_cluster_minus$n
    No_plus_neg = length((Data_cluster %>% dplyr::filter(cluster == i))[,1]) -
      Data_cluster_plus
    No_minus_neg = length((Data_cluster %>% dplyr::filter(cluster != i))[,1]) -
      Data_cluster_minus
    testSummary[i,2] = length((Data_cluster %>% dplyr::filter(cluster == i))[,1])
    OddsRatio = round(Data_cluster_plus / No_plus_neg /
                Data_cluster_minus * No_minus_neg, digits = 2)
    OddsRatio = replace(OddsRatio, which(is.na(OddsRatio)), 0)
    names(OddsRatio) = sort(Diseases)
    Pvalue = OddsRatio
    Pvalue[1:length(Pvalue)] = 1
    for(j in 1:length(Data_cluster_plus)){
      if(OddsRatio[j] > 5){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Cancer = c("CancerType", "OtherCancerType"),
                                    Cluster = c(paste("cluster", i), "Other clusters")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.01){
          testSummary[i,3] = paste(testSummary[i,3],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
      if(OddsRatio[j] < 0.2){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Cancer = c("CancerType", "OtherCancerType"),
                                    Cluster = c(paste("cluster", i), "Other clusters")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.01){
          testSummary[i,4] = paste(testSummary[i,4],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
    }
    OddsRatio_summary = rbind(OddsRatio_summary, OddsRatio)
    Pvalue_summary = rbind(Pvalue_summary, Pvalue)
  }
  
  write.csv(testSummary, file = paste(OutDirName, Cancer, "_cluster_characteristics_disease.csv", sep=""))
}

Data_mutation_tmp = Data_mutation_tmp[Data_survival$C.CAT調査結果.基本項目.ハッシュID,]
Data_mutation_tmp_save = Data_mutation_tmp
Data_mutation_tmp$RECIST = Data_survival$RECIST

testSummary = data.frame(matrix(rep("", 8), nrow=2))
colnames(testSummary) = c("Drug_effect", "Patients", "Positively_selected", "Negatively_selected")
OddsRatio_summary = NULL
Pvalue_summary = NULL

if(Evaluate_NE_as_ineffective == 0){
  Data_mutation_tmp = Data_mutation_tmp %>%
    dplyr::filter(RECIST != "NE")
}
if(SD_is_effective == 2){
  Data_mutation_tmp = Data_mutation_tmp %>%
    dplyr::mutate(RECIST_binary = case_when(
      RECIST %in% c("CR", "PR", "LongSD") ~ "Effective",
      TRUE ~ "Ineffective"
    ))
} else if(SD_is_effective == 1){
  Data_mutation_tmp = Data_mutation_tmp %>%
    dplyr::mutate(RECIST_binary = case_when(
      RECIST %in% c("CR", "PR", "LongSD", "SD") ~ "Effective",
      TRUE ~ "Ineffective"
    ))
} else{
  Data_mutation_tmp = Data_mutation_tmp %>%
    dplyr::mutate(RECIST_binary = case_when(
      RECIST %in% c("CR", "PR") ~ "Effective",
      TRUE ~ "Ineffective"
    ))
}

Effect_binary = unique(Data_mutation_tmp$RECIST_binary)
Data_mutation_tmp = Data_mutation_tmp[, colnames(Data_mutation_tmp) != "RECIST"]

for(i in 1:2){
  testSummary[i,1] = Effect_binary[i]
  Data_cluster_plus = Data_mutation_tmp %>%
    dplyr::filter(RECIST_binary == Effect_binary[i]) %>%
    dplyr::select(-diagnosis, -idx, -cluster, -YoungOld, -RECIST_binary) %>%
    colSums()
  Data_cluster_minus = Data_mutation_tmp %>%
    dplyr::filter(RECIST_binary != Effect_binary[i]) %>%
    dplyr::select(-diagnosis, -idx, -cluster, -YoungOld, -RECIST_binary) %>%
    colSums()
  No_plus_neg = length((Data_mutation_tmp %>% dplyr::filter(RECIST_binary == Effect_binary[i]))[,1]) -
    Data_cluster_plus
  No_minus_neg = length((Data_mutation_tmp %>% dplyr::filter(RECIST_binary != Effect_binary[i]))[,1]) -
    Data_cluster_minus
  testSummary[i,2] = length((Data_mutation_tmp %>% dplyr::filter(RECIST_binary == Effect_binary[i]))[,1])
  OddsRatio = round(Data_cluster_plus / No_plus_neg /
              Data_cluster_minus * No_minus_neg, digits = 2)
  OddsRatio = replace(OddsRatio, which(is.na(OddsRatio)), 0)

  Pvalue = OddsRatio
  Pvalue[1:length(Pvalue)] = 1
  for(j in 1:length(Data_cluster_plus)){
    if(OddsRatio[j] > 2){
      mutation_matrix = matrix(c(Data_cluster_plus[j],
                             Data_cluster_minus[j],
                             No_plus_neg[j],
                             No_minus_neg[j]),
                    nrow = 2,
                    dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                  Cluster = c(Effect_binary[i], "Others")))
      Pvalue[j] = 
        round(fisher.test(mutation_matrix)$p.value, digits = 3)
      if(Pvalue[j] < 0.05){
        testSummary[i,3] = paste(testSummary[i,3],
                                 names(OddsRatio[j]), " OR=",
                                 OddsRatio[j], "(p=",
                                 Pvalue[j],") ",sep="")
      }
    }
    if(OddsRatio[j] < 0.5){
      mutation_matrix = matrix(c(Data_cluster_plus[j],
                             Data_cluster_minus[j],
                             No_plus_neg[j],
                             No_minus_neg[j]),
                    nrow = 2,
                    dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                  Cluster = c(Effect_binary[i], "Others")))
      Pvalue[j] = 
        round(fisher.test(mutation_matrix)$p.value, digits = 3)
      if(Pvalue[j] < 0.05){
        testSummary[i,4] = paste(testSummary[i,4],
                                 names(OddsRatio[j]), " OR=",
                                 OddsRatio[j], "(p=",
                                 Pvalue[j],") ",sep="")
      }
    }
  }
  OddsRatio_summary = rbind(OddsRatio_summary, OddsRatio)
  Pvalue_summary = rbind(Pvalue_summary, Pvalue)
}
write.csv(testSummary, file = paste(OutDirName, Cancer, "_Drug_effect_characteristics.csv", sep=""))

Data_mutation_tmp = Data_mutation_tmp_save
Data_mutation_tmp$cluster_drug = Data_survival$cluster_drug
testResult = NULL
detail_text = rep("", max(Data_mutation_tmp$cluster_drug))
for(i in 1:max(Data_mutation_tmp$cluster_drug)){
  ID_path = unique((Data_survival %>% dplyr::filter(
                    cluster_drug == i))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival_tmp = Data_survival %>%
    dplyr::mutate(path_mut = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  Data_survival_tmp_RECIST = Data_survival_tmp
  if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "PD", "SD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else{
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  }
  EffectPathway = matrix(c(length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 0))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 0))$path_mut)),
                         nrow = 2,
                    dimnames = list(Effect = c("Drug_Effect +", "Drug_Effect -"),
                                  Pathway = c(paste("cluster", i), "Other clusters")))
  testResult = c(testResult, list(EffectPathway))
  testResult = c(testResult, list(fisher.test(EffectPathway)))
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(path_mut == 1)
  if(length(Data_survival_tmp[,1]) > 0){
    detail_text[[i]] = paste(
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
      ", ",
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
      sep="")
  }
}

testSummary = data.frame(matrix(rep(NA, length(testResult)*5.5), nrow=length(testResult)/2))
if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/LongSD)", "Cluster(+)Effect(-,PD/ShortSD/NE)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/SD)", "Cluster(+)Effect(-,PD/NE)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR)", "Cluster(+)Effect(-,PD/SD/NE)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/LongSD)", "Cluster(+)Effect(-,PD/ShortSD)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR/SD)", "Cluster(+)Effect(-,PD)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else{
  colnames(testSummary) = c("Cluster", "Drug", "Cluster(+)Effect(+,CR/PR)", "Cluster(+)Effect(-,PD/SD)", "OtherCluster,Effect(+)", "OtherCluster,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
}
for(i in 1:(length(testResult)/2)){
  testSummary[i,1] = paste("Cluster", i)
  testSummary[i,2] = Regimen_name
  testSummary[i,3] = testResult[[i*2-1]][[1]]
  testSummary[i,4] = testResult[[i*2-1]][[2]]
  testSummary[i,5] = testResult[[i*2-1]][[3]]
  testSummary[i,6] = testResult[[i*2-1]][[4]]
  testSummary[i,7] = (testResult[[i*2]]$estimate)[[1]]
  testSummary[i,8] = (testResult[[i*2]]$conf.int)[[1]]
  testSummary[i,9] = (testResult[[i*2]]$conf.int)[[2]]
  testSummary[i,10] = (testResult[[i*2]]$p.value)[[1]]
  testSummary[i,11] = detail_text[[i]]
}
write.csv(testSummary, file = paste(OutDirName, Cancer, "_", Regimen_name, "_cluster_drug_Summary.csv", sep=""))

testSummary = data.frame(matrix(rep("", max(Data_mutation_tmp$cluster_drug)*4),
                                nrow=max(Data_mutation_tmp$cluster_drug)))
Data_mutation_tmp = Data_mutation_tmp[, colnames(Data_mutation_tmp) != "cluster"]
colnames(testSummary) = c("Cluster", "Patients", "Positively_selected", "Negatively_selected")
OddsRatio_summary = NULL
Pvalue_summary = NULL
if(length(unique(Data_mutation_tmp$cluster_drug)) > 1){
  for(i in 1:max(Data_mutation_tmp$cluster_drug)){
    testSummary[i,1] = paste("Cluster", i)
    Data_cluster_plus = Data_mutation_tmp %>%
      dplyr::filter(cluster_drug == i) %>%
      dplyr::select(-diagnosis, -idx, -cluster_drug, -YoungOld) %>%
      colSums()
    Data_cluster_minus = Data_mutation_tmp %>%
      dplyr::filter(cluster_drug != i) %>%
      dplyr::select(-diagnosis, -idx, -cluster_drug, -YoungOld) %>%
      colSums()
    No_plus_neg = length((Data_mutation_tmp %>% dplyr::filter(cluster_drug == i))[,1]) -
      Data_cluster_plus
    No_minus_neg = length((Data_mutation_tmp %>% dplyr::filter(cluster_drug != i))[,1]) -
      Data_cluster_minus
    testSummary[i,2] = length((Data_mutation_tmp %>% dplyr::filter(cluster_drug == i))[,1])
    OddsRatio = round(Data_cluster_plus / No_plus_neg /
                Data_cluster_minus * No_minus_neg, digits = 2)
    OddsRatio = replace(OddsRatio, which(is.na(OddsRatio)), 0)
    Pvalue = OddsRatio
    Pvalue[1:length(Pvalue)] = 1
    for(j in 1:length(Data_cluster_plus)){
      if(OddsRatio[j] > 2){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(paste("cluster", i), "Other clusters")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,3] = paste(testSummary[i,3],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
      if(OddsRatio[j] < 0.5){
        mutation_matrix = matrix(c(Data_cluster_plus[j],
                               Data_cluster_minus[j],
                               No_plus_neg[j],
                               No_minus_neg[j]),
                      nrow = 2,
                      dimnames = list(Mutation = c("Mutation +", "Mutation -"),
                                    Cluster = c(paste("cluster", i), "Other clusters")))
        Pvalue[j] = 
          round(fisher.test(mutation_matrix)$p.value, digits = 3)
        if(Pvalue[j] < 0.05){
          testSummary[i,4] = paste(testSummary[i,4],
                                   names(OddsRatio[j]), " OR=",
                                   OddsRatio[j], "(p=",
                                   Pvalue[j],") ",sep="")
        }
      }
    }
    OddsRatio_summary = rbind(OddsRatio_summary, OddsRatio)
    Pvalue_summary = rbind(Pvalue_summary, Pvalue)
  }
  
  write.csv(testSummary, file = paste(OutDirName, Cancer, "_cluster_drug_characteristics.csv", sep=""))
}

Data_mutation_tmp = Data_mutation_tmp_save
testResult = NULL
detail_text = rep("", length(diagnosis_group))
for(i in 1:length(diagnosis_group)){
  ID_path = unique((Data_survival %>% dplyr::filter(
                    diagnosis == diagnosis_group[i]))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival_tmp = Data_survival %>%
    dplyr::mutate(path_mut = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  Data_survival_tmp_RECIST = Data_survival_tmp
  if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "PD", "SD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD", "NE")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD") ~ 1,
        TRUE ~ 0
      ))
  } else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR", "LongSD", "SD") ~ 1,
        TRUE ~ 0
      ))
  } else{
    Data_survival_tmp = Data_survival_tmp %>%
      dplyr::filter(RECIST %in% c("CR", "PR", "LongSD", "SD", "PD")) %>%
      dplyr::mutate(Effect = case_when(
        RECIST %in% c("CR", "PR") ~ 1,
        TRUE ~ 0
      ))
  }
  EffectPathway = matrix(c(length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 1 & Effect == 0))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 1))$path_mut),
                           length((Data_survival_tmp %>% dplyr::filter(path_mut == 0 & Effect == 0))$path_mut)),
                         nrow = 2,
                    dimnames = list(Effect = c("Drug_Effect +", "Drug_Effect -"),
                                  Pathway = c(diagnosis_group[i], "Other clusters")))
  testResult = c(testResult, list(EffectPathway))
  testResult = c(testResult, list(fisher.test(EffectPathway)))
  Data_survival_tmp = Data_survival_tmp_RECIST %>%
    dplyr::filter(path_mut == 1)
  if(length(Data_survival_tmp[,1]) > 0){
    detail_text[[i]] = paste(
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,1]),collapse = "/"),
      ", ",
      paste0(as.character((as.data.frame(table(Data_survival_tmp$RECIST)))[,2]),collapse = "/"),
      sep="")
  }
}

testSummary = data.frame(matrix(rep(NA, length(testResult)*5.5), nrow=length(testResult)/2))
if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Diagnosis", "Drug", "Diagnosis(+)Effect(+,CR/PR/LongSD)", "Diagnosis(+)Effect(-,PD/ShortSD/NE)", "OtherDiagnosis,Effect(+)", "OtherDiagnosis,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Diagnosis", "Drug", "Diagnosis(+)Effect(+,CR/PR/SD)", "Diagnosis(+)Effect(-,PD/NE)", "OtherDiagnosis,Effect(+)", "OtherDiagnosis,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 0 & Evaluate_NE_as_ineffective == 1){
  colnames(testSummary) = c("Diagnosis", "Drug", "Diagnosis(+)Effect(+,CR/PR)", "Diagnosis(+)Effect(-,PD/SD/NE)", "OtherDiagnosis,Effect(+)", "OtherDiagnosis,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 2 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Diagnosis", "Drug", "Diagnosis(+)Effect(+,CR/PR/LongSD)", "Diagnosis(+)Effect(-,PD/ShortSD)", "OtherDiagnosis,Effect(+)", "OtherDiagnosis,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else if(SD_is_effective == 1 & Evaluate_NE_as_ineffective == 0){
  colnames(testSummary) = c("Diagnosis", "Drug", "Diagnosis(+)Effect(+,CR/PR/SD)", "Diagnosis(+)Effect(-,PD)", "OtherDiagnosis,Effect(+)", "OtherDiagnosis,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
} else{
  colnames(testSummary) = c("Diagnosis", "Drug", "Diagnosis(+)Effect(+,CR/PR)", "Diagnosis(+)Effect(-,PD/SD)", "OtherDiagnosis,Effect(+)", "OtherDiagnosis,Effect(-)", "OddsRatio_estimate", "OddsRatio_95CI_low", "OddsRatio_95CI_up", "p-value", "detail")
}
for(i in 1:(length(testResult)/2)){
  testSummary[i,1] = diagnosis_group[i]
  testSummary[i,2] = Regimen_name
  testSummary[i,3] = testResult[[i*2-1]][[1]]
  testSummary[i,4] = testResult[[i*2-1]][[2]]
  testSummary[i,5] = testResult[[i*2-1]][[3]]
  testSummary[i,6] = testResult[[i*2-1]][[4]]
  testSummary[i,7] = (testResult[[i*2]]$estimate)[[1]]
  testSummary[i,8] = (testResult[[i*2]]$conf.int)[[1]]
  testSummary[i,9] = (testResult[[i*2]]$conf.int)[[2]]
  testSummary[i,10] = (testResult[[i*2]]$p.value)[[1]]
  testSummary[i,11] = detail_text[[i]]
}
write.csv(testSummary, file = paste(OutDirName, Cancer, "_", Regimen_name, "_diagnosis_drug_Summary.csv", sep=""))

Data_survival_save = Data_survival
if(Evaluate_NE_as_ineffective == 0){
  Data_survival = Data_survival %>%
    dplyr::filter(RECIST != "NE")
}
if(SD_is_effective == 2){
  Data_survival = Data_survival %>%
    dplyr::mutate(RECIST_binary = case_when(
      RECIST %in% c("CR", "PR", "LongSD") ~ "Effective",
      TRUE ~ "Ineffective"
    ))
} else if(SD_is_effective == 1){
  Data_survival = Data_survival %>%
    dplyr::mutate(RECIST_binary = case_when(
      RECIST %in% c("CR", "PR", "LongSD", "SD") ~ "Effective",
      TRUE ~ "Ineffective"
    ))
} else{
  Data_survival = Data_survival %>%
    dplyr::mutate(RECIST_binary = case_when(
      RECIST %in% c("CR", "PR") ~ "Effective",
      TRUE ~ "Ineffective"
    ))
}

Data_survival = Data_survival %>% dplyr::arrange(RECIST_binary)

Data_oncoprint = Data_survival %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis, RECIST, RECIST_binary) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)

ID_target_drug = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)
mat_oncoprint = matrix("",
            nrow = length(Data_oncoprint$diagnosis),
            ncol = length(PATH$frequent_genes))
rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
colnames(mat_oncoprint) = PATH$frequent_genes
Data_major_gene_tmp = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_drug) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH$frequent_genes) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                  .keep_all=FALSE)

for(i in 1:length(Data_major_gene_tmp$C.CAT調査結果.基本項目.ハッシュID)){
  if(Data_major_gene_tmp[i,2] == "rearrangement"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            Data_major_gene_tmp[i,4],
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "copy_number_alteration"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            Data_major_gene_tmp[i,4],
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "small_scale_variant"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            "small_scale_variant",
            sep=";")
  } else if(Data_major_gene_tmp[i,2] == "other_biomarker"){
    mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
      paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
            "TMB_high",
            sep=";")
  }
}

rownames(mat_oncoprint) = Data_oncoprint$diagnosis
mat_oncoprint = t(mat_oncoprint)
col = c("amplification" = "red",
        "loss" = "blue",
        "TMB_high" = "orange",
        "small_scale_variant" = "green",
        "rearrangement" = "purple",
        "fusion" = "gray",
        "truncation" = "black",
        "rearrangements" = "yellow")
alter_fun = list(
    background = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = "#CCCCCC", col = NA))
    },
    amplification = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
            gp = gpar(fill = col["amplification"], col = NA))
    },
    loss = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
            gp = gpar(fill = col["loss"], col = NA))
    },
    TMB_high = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
            gp = gpar(fill = col["TMB_high"], col = NA))
    },
    small_scale_variant = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
            gp = gpar(fill = col["small_scale_variant"], col = NA))
    },
    rearrangement = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
            gp = gpar(fill = col["rearrangement"], col = NA))
    },
    fusion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
            gp = gpar(fill = col["fusion"], col = NA))
    },
    truncation = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
            gp = gpar(fill = col["truncation"], col = NA))
    },
    rearrangements = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
            gp = gpar(fill = col["rearrangements"], col = NA))
    }
)

column_title = paste("OncoPrint for selected diseases, frequently mutated genes,", Regimen_name)

tmp = mat_oncoprint
tmp[tmp != ""] = 1
tmp[tmp == ""] = 0
tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
tmp = data.frame(tmp)
colnames(tmp) = 1:ncol(tmp)
sample_order = 1:ncol(tmp)
colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))

anno_diag = colnames(mat_oncoprint)
anno_RECIST = Data_oncoprint$RECIST
diag_list = unique(anno_diag)
diag_col = colPal3(length(diag_list))
RECIST_list = unique(Data_oncoprint$RECIST)
RECIST_col = colPal3(length(RECIST_list))
names(diag_col) = diag_list
names(RECIST_col) = RECIST_list

for(l in tmp2){
  if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
    sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
    tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
  }
}
sample_order_final_1 = NULL
for(l in diag_list){
  sample_order_final_1 = c(sample_order_final_1,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
}
sample_order_final_2 = NULL
for(l in RECIST_list){
  sample_order_final_2 = c(sample_order_final_2,
                         sample_order_final_1[sample_order_final_1 %in% (1:length(sample_order_final_1))[(anno_RECIST == l)]])
}
sample_order_final_3 = NULL
for(l in RECIST_list){
  sample_order_final_3 = c(sample_order_final_3,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_RECIST == l)]])
}

heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
        labels = c("Oncogenic mut"))
ha = HeatmapAnnotation(Diagnosis = anno_diag,
  Drug_effect = anno_RECIST,
	col = list(Diagnosis = diag_col, Drug_effect = RECIST_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis"),
                                 Drug_effect = list(title = "RECIST")))
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final_1, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, pct_gp = gpar(fontsize = 5),
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_drug_effect_1.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final_2, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_drug_effect_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final_3, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_drug_effect_3.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()


column_title = paste("OncoPrint for selected diseases, frequently mutated genes,", Regimen_name)
heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
        labels = c("Oncogenic mut"))
tmp = mat_oncoprint
tmp[tmp != ""] = 1
tmp[tmp == ""] = 0
tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
tmp = data.frame(tmp)
colnames(tmp) = 1:ncol(tmp)
sample_order = 1:ncol(tmp)
colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
anno_diag = colnames(mat_oncoprint)
anno_RECIST = Data_oncoprint$RECIST_binary
diag_list = unique(anno_diag)
diag_col = colPal3(length(diag_list))
RECIST_list = unique(Data_oncoprint$RECIST_binary)
RECIST_col = colPal3(length(RECIST_list))
names(diag_col) = diag_list
names(RECIST_col) = RECIST_list

for(l in tmp2){
  if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
    sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
    tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
  }
}
sample_order_final_1 = NULL
for(l in diag_list){
  sample_order_final_1 = c(sample_order_final_1,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
}
sample_order_final_2 = NULL
for(l in RECIST_list){
  sample_order_final_2 = c(sample_order_final_2,
                         sample_order_final_1[sample_order_final_1 %in% (1:length(sample_order_final_1))[(anno_RECIST == l)]])
}
sample_order_final_3 = NULL
for(l in RECIST_list){
  sample_order_final_3 = c(sample_order_final_3,
                         sample_order[sample_order %in% (1:length(sample_order))[(anno_RECIST == l)]])
}
ha = HeatmapAnnotation(Diagnosis = anno_diag,
  Drug_effect = anno_RECIST,
	col = list(Diagnosis = diag_col, Drug_effect = RECIST_col),
	annotation_height = unit(c(5, 5, 15), "mm"),
  annotation_legend_param = list(Diagnosis = list(title = "Diagnosis"),
                                 Drug_effect = list(title = "RECIST"))
)
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final_1, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, pct_gp = gpar(fontsize = 5),
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_drug_effect_4.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final_2, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_drug_effect_5.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final_3, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_drug_effect_6.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()
ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
pdf(file = paste(OutDirName, Cancer, "_Oncoprint_frequent_genes_drug_effect_7.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
draw(ht)
dev.off()

if(length(target_pathway) > 0){
  Data_survival = Data_survival %>% dplyr::arrange(RECIST_binary)
  Data_oncoprint = Data_survival %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, diagnosis, RECIST, RECIST_binary) %>%
    dplyr::distinct() %>%
    dplyr::arrange(diagnosis)
  ID_target_drug = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)
  mat_oncoprint = matrix("",
              nrow = length(Data_oncoprint$diagnosis),
              ncol = length(PATH[target_pathway[1]][[1]]))
  rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
  colnames(mat_oncoprint) = PATH[target_pathway[1]][[1]]
  Data_major_gene_tmp = Data_report %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_drug) %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
    dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[1]][[1]]) %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.変異種類,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=FALSE)
  
  for(i in 1:length(Data_major_gene_tmp$C.CAT調査結果.基本項目.ハッシュID)){
    if(Data_major_gene_tmp[i,2] == "rearrangement"){
      mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
        paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
              Data_major_gene_tmp[i,4],
              sep=";")
    } else if(Data_major_gene_tmp[i,2] == "copy_number_alteration"){
      mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
        paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
              Data_major_gene_tmp[i,4],
              sep=";")
    } else if(Data_major_gene_tmp[i,2] == "small_scale_variant"){
      mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
        paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
              "small_scale_variant",
              sep=";")
    } else if(Data_major_gene_tmp[i,2] == "other_biomarker"){
      mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]] =
        paste(mat_oncoprint[Data_major_gene_tmp[i,1], Data_major_gene_tmp[i,3]],
              "TMB_high",
              sep=";")
    }
  }
  rownames(mat_oncoprint) = Data_oncoprint$diagnosis
  mat_oncoprint = t(mat_oncoprint)
  col = c("amplification" = "red",
          "loss" = "blue",
          "TMB_high" = "orange",
          "small_scale_variant" = "green",
          "rearrangement" = "purple",
          "fusion" = "gray",
          "truncation" = "black",
          "rearrangements" = "yellow")
  alter_fun = list(
      background = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = "#CCCCCC", col = NA))
      },
      amplification = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h-unit(2, "pt"), 
              gp = gpar(fill = col["amplification"], col = NA))
      },
      loss = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*1.0, 
              gp = gpar(fill = col["loss"], col = NA))
      },
      TMB_high = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.8, 
              gp = gpar(fill = col["TMB_high"], col = NA))
      },
      small_scale_variant = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.6, 
              gp = gpar(fill = col["small_scale_variant"], col = NA))
      },
      rearrangement = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.4, 
              gp = gpar(fill = col["rearrangement"], col = NA))
      },
      fusion = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.3, 
              gp = gpar(fill = col["fusion"], col = NA))
      },
      truncation = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.2, 
              gp = gpar(fill = col["truncation"], col = NA))
      },
      rearrangements = function(x, y, w, h) {
          grid.rect(x, y, w-unit(0, "pt"), h*0.1, 
              gp = gpar(fill = col["rearrangements"], col = NA))
      }
  )
    
  column_title = paste("OncoPrint for selected diseases,", target_pathway[1], "pathway,", Regimen_name)
  heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
          labels = c("Oncogenic mut"))

  tmp = mat_oncoprint
  tmp[tmp != ""] = 1
  tmp[tmp == ""] = 0
  tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
  tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
  tmp = data.frame(tmp)
  colnames(tmp) = 1:ncol(tmp)
  sample_order = 1:ncol(tmp)
  colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
  anno_diag = colnames(mat_oncoprint)
  anno_RECIST = Data_oncoprint$RECIST
  diag_list = unique(anno_diag)
  diag_col = colPal3(length(diag_list))
  RECIST_list = unique(Data_oncoprint$RECIST)
  RECIST_col = colPal3(length(RECIST_list))
  names(diag_col) = diag_list
  names(RECIST_col) = RECIST_list

  for(l in tmp2){
    if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
      sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
      tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
    }
  }
  sample_order_final = NULL
  for(l in diag_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
  }
  sample_order = sample_order_final
  sample_order_final = NULL
  for(l in RECIST_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_RECIST == l)]])
  }
  ha = HeatmapAnnotation(Diagnosis = anno_diag,
    Drug_effect = anno_RECIST,
  	col = list(Diagnosis = diag_col, Drug_effect = RECIST_col),
  	annotation_height = unit(c(5, 5, 15), "mm"),
    annotation_legend_param = list(Diagnosis = list(title = "Diagnosis"),
                                   Drug_effect = list(title = "RECIST")))
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
        column_order = sample_order, row_order = rev(tmp2),
        alter_fun = alter_fun, col = col, pct_gp = gpar(fontsize = 5),
        column_title = column_title,
        remove_empty_columns = FALSE, remove_empty_rows = FALSE,
        bottom_annotation = ha,
        use_raster = FALSE,
        alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_target_pathway_drug_effect_1.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
      column_order = sample_order_final, row_order = rev(tmp2),
      alter_fun = alter_fun, col = col, 
      column_title = column_title,
      remove_empty_columns = FALSE, remove_empty_rows = FALSE,
      bottom_annotation = ha,
      use_raster = FALSE,
      alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_target_pathway_drug_effect_2.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()


  column_title = paste("OncoPrint for selected diseases,", target_pathway[1], "pathway,", Regimen_name)
  heatmap_legend_param = list(title = "Alternations", at = c("Oncogenic mut"), 
          labels = c("Oncogenic mut"))

  tmp = mat_oncoprint
  tmp[tmp != ""] = 1
  tmp[tmp == ""] = 0
  tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
  tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
  tmp = data.frame(tmp)
  colnames(tmp) = 1:ncol(tmp)
  sample_order = 1:ncol(tmp)
  colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
  anno_diag = colnames(mat_oncoprint)
  anno_RECIST = Data_oncoprint$RECIST_binary
  diag_list = unique(anno_diag)
  diag_col = colPal3(length(diag_list))
  RECIST_list = unique(Data_oncoprint$RECIST_binary)
  RECIST_col = colPal3(length(RECIST_list))
  names(diag_col) = diag_list
  names(RECIST_col) = RECIST_list

  for(l in tmp2){
    if(!all(tmp[l,] == 1) & !all(tmp[l,] == 0)){
      sample_order = c(sample_order[tmp[l,] == 1], sample_order[tmp[l,] == 0])
      tmp = cbind(tmp[,tmp[l,] == 1], tmp[,tmp[l,] == 0])
    }
  }
  sample_order_final = NULL
  for(l in diag_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
  }
  sample_order = sample_order_final
  sample_order_final = NULL
  for(l in RECIST_list){
    sample_order_final = c(sample_order_final,
                           sample_order[sample_order %in% (1:length(sample_order))[(anno_RECIST == l)]])
  }
  ha = HeatmapAnnotation(Diagnosis = anno_diag,
        Drug_effect = anno_RECIST,
      	col = list(Diagnosis = diag_col, Drug_effect = RECIST_col),
      	annotation_height = unit(c(5, 5, 15), "mm"),
        annotation_legend_param = list(Diagnosis = list(title = "Diagnosis"),
                                       Drug_effect = list(title = "RECIST"))
  )
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
        column_order = sample_order, row_order = rev(tmp2),
        alter_fun = alter_fun, col = col, pct_gp = gpar(fontsize = 5),
        column_title = column_title,
        remove_empty_columns = FALSE, remove_empty_rows = FALSE,
        bottom_annotation = ha,
        use_raster = FALSE,
        alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_target_pathway_drug_effect_3.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
  ht = oncoPrint(mat_oncoprint, get_type = function(x) gsub(":.*$", "", strsplit(x, ";")[[1]]),
        column_order = sample_order_final, row_order = rev(tmp2),
        alter_fun = alter_fun, col = col, 
        column_title = column_title,
        remove_empty_columns = FALSE, remove_empty_rows = FALSE,
        bottom_annotation = ha,
        use_raster = FALSE,
        alter_fun_is_vectorized = TRUE)
  pdf(file = paste(OutDirName, Cancer, "_Oncoprint_target_pathway_drug_effect_4.pdf", sep=""), width = 20, height = 20, onefile = TRUE)
  draw(ht)
  dev.off()
}

pdf(file = paste(OutDirName, Cancer, "_diagnosis_pie_chart.pdf", sep=""), width = 10, height = 10, onefile = TRUE)

Data_survival = Data_survival_save
Data_survival$diagnosis_short = str_sub(Data_survival$diagnosis, 1, 5)

cancer_freq_order = names(sort(table(Data_survival$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival <- transform(Data_survival, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival <- transform(Data_survival, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))

Data_survival_pie = Data_survival
ggplot(data = Data_survival_pie, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title=paste("  All patients received", Regimen_name))

Data_survival_pie = Data_survival %>%
  dplyr::filter(RECIST != "NE")
cancer_freq_order = names(sort(table(Data_survival_pie$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival_pie$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival_pie <- transform(Data_survival_pie, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival_pie <- transform(Data_survival_pie, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))

ggplot(data = Data_survival_pie, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title=paste("  All patients received", Regimen_name, "and effect is not NE"))

Data_survival_pie = Data_survival %>%
  dplyr::filter(RECIST == "CR")
cancer_freq_order = names(sort(table(Data_survival_pie$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival_pie$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival_pie <- transform(Data_survival_pie, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival_pie <- transform(Data_survival_pie, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))
ggplot(data = Data_survival_pie, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title=paste("  All patients received", Regimen_name, "and effect is CR"))

Data_survival_pie = Data_survival %>%
  dplyr::filter(RECIST == "PR")
cancer_freq_order = names(sort(table(Data_survival_pie$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival_pie$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival_pie <- transform(Data_survival_pie, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival_pie <- transform(Data_survival_pie, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))
ggplot(data = Data_survival_pie, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title=paste("  All patients received", Regimen_name, "and effect is PR"))

Data_survival_pie = Data_survival %>%
  dplyr::filter(RECIST == "LongSD")
cancer_freq_order = names(sort(table(Data_survival_pie$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival_pie$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival_pie <- transform(Data_survival_pie, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival_pie <- transform(Data_survival_pie, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))
ggplot(data = Data_survival_pie, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title=paste("  All patients received", Regimen_name, "and effect is long_SD"))

Data_survival_pie = Data_survival %>%
  dplyr::filter(RECIST == "SD")
cancer_freq_order = names(sort(table(Data_survival_pie$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival_pie$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival_pie <- transform(Data_survival_pie, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival_pie <- transform(Data_survival_pie, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))
ggplot(data = Data_survival_pie, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title=paste("  All patients received", Regimen_name, "and effect is SD"))

Data_survival_pie = Data_survival %>%
  dplyr::filter(RECIST == "PD")
cancer_freq_order = names(sort(table(Data_survival_pie$diagnosis), decreasing = TRUE))
cancer_freq_order = c(cancer_freq_order[cancer_freq_order != "Other_sarcoma"], "Other_sarcoma")
cancer_freq_order_short = names(sort(table(Data_survival_pie$diagnosis_short), decreasing = TRUE))
cancer_freq_order_short = c(cancer_freq_order_short[cancer_freq_order_short != "Other"], "Other")
Data_survival_pie <- transform(Data_survival_pie, diagnosis= factor(diagnosis, levels = cancer_freq_order))
Data_survival_pie <- transform(Data_survival_pie, diagnosis_short= factor(diagnosis_short, levels = cancer_freq_order_short))
ggplot(data = Data_survival_pie, aes(x = 1, fill = factor(diagnosis), color = diagnosis_short)) +
  geom_bar(stat = "count", color = "black", position = position_stack(reverse = TRUE)) +
  coord_polar(theta = "y") + 
  theme_void() +
#  theme(legend.position = "none") +
 geom_text(stat = "count",
            aes(x = 1.1, y = ..count..,
                label = ..count..), color = "black",
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  geom_text(stat = "count",
            aes(x = 1.55, y = ..count..,
                label = str_c(..color.., "")),
            position = position_stack(vjust = 0.5, reverse = TRUE)) +
  labs(title=paste("  All patients received", Regimen_name, "and effect is PD"))
dev.off()

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.検体情報.パネル.名称.,
  RECIST,
  cluster,
  diagnosis) 
Data_summary = Data_summary %>%
  dplyr::mutate(Effect = case_when(
    RECIST == "PD" ~ "Ineffective",
    RECIST == "NE" & Evaluate_NE_as_ineffective == 1 ~ "Ineffective",
    RECIST == "NE" & Evaluate_NE_as_ineffective == 0 ~ "NE",
    RECIST == "SD" & SD_is_effective == 2 ~ "Ineffective",
    RECIST == "SD" & SD_is_effective == 1 ~ "Effective",
    RECIST == "SD" & SD_is_effective == 0 ~ "Ineffective",
    RECIST == "LongSD" & SD_is_effective == 2 ~ "Effective",
    RECIST == "LongSD" & SD_is_effective == 1 ~ "Effective",
    RECIST == "LongSD" & SD_is_effective == 0 ~ "Ineffective",
    RECIST == "PR" ~ "Effective",
    RECIST == "CR" ~ "Effective"
  ))
Data_summary$cluster = as.character(Data_summary$cluster)
table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. Patients with selected disease, treated with selected drug, Characteristics**") %>%
  bold_labels()
table_tmp

table_tmp = Data_summary %>%
  tbl_summary(
    by = RECIST,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment option exsisted**") %>%
  modify_caption("**Table. Drug effect by RECIST**") %>%
  bold_labels()
table_tmp

Data_summary = Data_summary %>%
  dplyr::filter(Effect != "NE")
table_tmp = Data_summary %>%
  tbl_summary(
    by = Effect,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table. Drug effect binary**") %>%
  bold_labels()
table_tmp

write.csv(Data_survival, file = paste(OutDirName, Cancer, "_Drug_used_patients_all_data.csv", sep=""))
```

```{r survival_from_entry, eval=(Analyze_enroll == 1), echo = (Source_code == 1 & Analyze_enroll == 1), warning=FALSE}
Data_survival = Data_case_target %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ "")) %>%
  dplyr::arrange(final_observe) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))
Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日))
Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$target_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_target_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:length(target_gene)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  } else{
    for(k in 1:length(target_gene)){
      if(target_gene[[k]][2] == "WT"){
        Data_survival$target_gene_count[i] = bitwOr(
          Data_survival$target_gene_count[i], 2^(k-1))
      }
    }
  }
}
if(sum(Data_survival$target_gene_count) == 0){
  target_gene = list(c(oncogenic_genes[1,1], "F"))
  gene_pattern = list(c(1))
  Data_target_gene = NULL
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.変異情報.変異内容 == target_gene[[i]][2]),
      Data_target_gene)
  }
  Data_target_gene = Data_target_gene %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=TRUE)
  
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[i]][2]) %>%
      dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                      C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT,
                    .keep_all=TRUE),
      Data_target_gene)
  }

  Data_survival$target_gene_count = 0
  ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
  for(i in 1:length(ID_target_gene)){
    tmp = Data_target_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
    if(dplyr::count(tmp)[[1]] > 0){
      for(j in 1:dplyr::count(tmp)[[1]]){
        for(k in 1:length(target_gene)){
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
        }
      }
    } else{
      for(k in 1:length(target_gene)){
        if(target_gene[[k]][2] == "WT"){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

Data_survival$oncogenic_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_major_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:min(length(oncogenic_genes$all_patients), gene_number)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == oncogenic_genes$gene_mutation[[k]]){
          Data_survival$oncogenic_gene_count[i] = bitwOr(
            Data_survival$oncogenic_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

Data_tmp = Data_case_target %>% dplyr::select(
  C.CAT調査結果.基本項目.ハッシュID,
  症例.EP後レジメン情報.治療ライン,
  症例.EP後レジメン情報.化学療法レジメン名称,
  症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
Data_tmp = tidyr::replace_na(Data_tmp,
                  replace = list(症例.EP後レジメン情報.治療ライン = 0,
                                 症例.EP後レジメン情報.化学療法レジメン名称 = "",
                                 症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = ""))
Data_tmp = Data_tmp %>%
  dplyr::filter(症例.EP後レジメン情報.治療ライン != 0 &(
                症例.EP後レジメン情報.化学療法レジメン名称 != "" |
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN. != ""))
ID_tmp = unique(Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
Data_survival$afterCGPtreat = Data_survival$EP_treat
for(i in ID_tmp){
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID == i, "afterCGPtreat"] = 1
}
Data_survival = Data_survival %>%
  dplyr::mutate(treat_group = case_when(
    afterCGPtreat == 1 & EP_treat == 1 ~ "Option(+)RecomTreat(+)",
    afterCGPtreat == 1 & EP_option == 1 ~ "Option(+)UnrecomTreat(+)",
    afterCGPtreat == 1 & EP_option == 0 ~ "Option(-)UnrecomTreat(+)",
    afterCGPtreat == 0 & EP_option == 1 ~ "Option(+)Treat(-)",
    afterCGPtreat == 0 & EP_option == 0 ~ "Option(-)Treat(-)"
  ))
Data_survival = Data_survival %>%
  dplyr::mutate(treat_group_2 = case_when(
    afterCGPtreat == 1 & EP_treat == 1 ~ "RecomTreat(+)",
    afterCGPtreat == 1 & EP_option == 1 ~ "UnrecomTreat(+)",
    afterCGPtreat == 1 & EP_option == 0 ~ "UnrecomTreat(+)",
    afterCGPtreat == 0 & EP_option == 1 ~ "Treat(-)",
    afterCGPtreat == 0 & EP_option == 0 ~ "Treat(-)"
  ))
Data_survival = Data_survival %>%
  dplyr::mutate(treat_group_3 = case_when(
    afterCGPtreat == 1 & EP_treat == 1 ~ "Treat(+)",
    afterCGPtreat == 1 & EP_option == 1 ~ "Treat(+)",
    afterCGPtreat == 1 & EP_option == 0 ~ "Treat(+)",
    afterCGPtreat == 0 & EP_option == 1 ~ "Treat(-)",
    afterCGPtreat == 0 & EP_option == 0 ~ "Treat(-)"
  ))
Data_survival = Data_survival %>%
  dplyr::mutate(treat_group_4 = case_when(
    afterCGPtreat == 1 & EP_treat == 1 ~ "RecomTreat(+)",
    afterCGPtreat == 1 & EP_option == 1 ~ "Option(+)RecomTreat(-)",
    afterCGPtreat == 1 & EP_option == 0 ~ "Option(-)",
    afterCGPtreat == 0 & EP_option == 1 ~ "Option(+)RecomTreat(-)",
    afterCGPtreat == 0 & EP_option == 0 ~ "Option(-)"
  ))
Data_survival = Data_survival %>%
  dplyr::mutate(treat_group_5 = case_when(
    afterCGPtreat == 1 & EP_treat == 1 ~ "RecomTreat(+)",
    afterCGPtreat == 1 & EP_option == 1 ~ "RecomTreat(-)",
    afterCGPtreat == 1 & EP_option == 0 ~ "RecomTreat(-)",
    afterCGPtreat == 0 & EP_option == 1 ~ "RecomTreat(-)",
    afterCGPtreat == 0 & EP_option == 0 ~ "RecomTreat(-)"
  ))

Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}
if(EGFR_positive_LUAD == 2){
  WriteXLS(Data_survival, ExcelFileName="EGFR_mut_with_survival_data_diagnosis_to_death.xls")
}
g = ggplot(Data_survival, aes(x = time_enroll_final))
g = g + geom_histogram(binwidth = 50, colour = "gray10", fill = "royalblue")
g = g + theme_classic()
g = g + ggtitle(paste(Cancer, "Distribution of time from enrollment to final observation (days)"))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_histogram_enrollment_to_final.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

Data_tmp = Data_survival %>% dplyr::filter(censor == 1)
g = ggplot(Data_tmp, aes(x = time_enroll_final))
g = g + geom_histogram(binwidth = 50, colour = "gray10", fill = "royalblue")
g = g + theme_classic()
g = g + ggtitle(paste(Cancer, "Distribution of time from enrollment to death (days)"))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_histogram_enrollment_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()


# Survival analysis for all patients
survival_simple = survfit(Surv(time_enroll_final, censor)~1,
                          data=Data_survival)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("All enrolled patients")
) + 
  labs(title = paste(Cancer, survival_simple[[1]], "patients"), 
       subtitle = paste("Median OS",
                summary(survival_simple)[[17]][[7]], "days"))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

# Survival analysis for enrollment timing
survival_simple = survfit(Surv(time_enroll_final, censor)~enrollment,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]] &
   length(unique(Data_survival$enrollment)) == 5){
diff_0 = survdiff(Surv(time_enroll_final, censor)~enrollment, data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~enrollment, data=Data_survival, rho=1)
#print(survival_simple)
#summary(survival_simple)
legend_enroll = paste("Entry timing:",
                      c("2019-06-01 to 2020-06-30",
                        "2020-07-01 to 2021-06-30",
                        "2021-07-01 to 2021-12-31",
                        "2022-01-01 to 2022-06-30",
                        "2022-07-01 to 2023-02-20")
                        , sep="")
legend_enroll = legend_enroll[1:(Entry_period + 1)]
subtitle_enroll = paste("Median OS",
             summary(survival_simple)[[18]][[6 * (Entry_period + 1) + min(1, (Entry_period + 1))]], "(1st)/",
             summary(survival_simple)[[18]][[6 * (Entry_period + 1) + min(2, (Entry_period + 1))]], "(2nd)/",
             summary(survival_simple)[[18]][[6 * (Entry_period + 1) + min(3, (Entry_period + 1))]], "(3rd)/",
             summary(survival_simple)[[18]][[6 * (Entry_period + 1) + min(4, (Entry_period + 1))]], "(4th)/",
             summary(survival_simple)[[18]][[6 * (Entry_period + 1) + min(5, (Entry_period + 1))]], "(latest) days,")
subtitle_enroll = subtitle_enroll[1:(Entry_period + 1)]
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = legend_enroll
) + 
  labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((survival_simple)[[1]]), "patients"), 
       subtitle = paste(subtitle_enroll,
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_entry.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
} else if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~enrollment, data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~enrollment, data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
) + 
  labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((survival_simple)[[1]]), "patients"), 
       subtitle = paste(
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_entry.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}


# Survival analysis for EP recommendation
survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_option, data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_option, data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = paste("Expert panel recommended option existed:",
                         0:1, sep="")
) + 
  labs(title = paste(Cancer, "EP option existed for", 
                     (survival_simple)[[1]][[2]], "patients"), 
       subtitle = paste("Median OS",
             summary(survival_simple)[[18]][[13]], "(0)/",
             summary(survival_simple)[[18]][[14]], "(1) days,",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_option.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# propensity score matching considering
# sex, age, performance status,
# enrollment timing, and multi cancer
if(length(unique(Data_survival$diagnosis)) > 1){
  if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_option ~
                          症例.基本情報.性別 + 症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + multi_cancer + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    } else{
      propensityScore = glm(EP_option ~
                          症例.基本情報.性別 + 症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    }
  } else{
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_option ~
                          症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + multi_cancer + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    } else{
      propensityScore = glm(EP_option ~
                          症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    }
  }
} else{
  if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_option ~
                        症例.基本情報.性別 + 症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment + multi_cancer,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    } else{
      propensityScore = glm(EP_option ~
                        症例.基本情報.性別 + 症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    }
  } else{
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_option ~
                        症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment + multi_cancer,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    } else{
      propensityScore = glm(EP_option ~
                        症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    }
  }
}
#summary(propensityScore)
Data_survival$PScore = propensityScore$fitted.values
matching = Match(Y = Data_survival$time_enroll_final,
                 Tr = Data_survival$EP_option,
                 X = Data_survival$PScore,
                 caliper=0.25,
                 ties=F,
                 replace=T)
if(!is.na(matching)[[1]]){
  #summary(matching)
  # ggplot(Data_survival, aes(x = PScore, y = after_stat(density))) +
  #   geom_histogram(color = "black") +
  #   facet_grid(rows = vars(EP_option), scales = "free_y") +
  #   labs(x = "Propensity score", y = "Probability density") +
  #   ggtitle("Expert panel treatment option")
  
  matched_data = Data_survival[c(matching$index.treated, matching$index.control),]
  matched_fit = survfit(Surv(event = censor,
                                 time = time_enroll_final) ~ EP_option, 
                                 data = matched_data)
  if(sum(matched_fit$n.event) > 0){
  if(summary(matched_fit)[[18]][7] > 2 & summary(matched_fit)[[18]][8] > 2){
    diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_option, data=matched_data, rho=0)
    diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_option, data=matched_data, rho=1)
    g = ggsurvplot(
        fit = matched_fit,
        data = matched_data,
        xlab = "Time (Days) from the enrollment",
        ylab = "Survival Probability",
        censor = TRUE,
        conf.int = FALSE,
        pval = FALSE,
        risk.table = TRUE,
        risk.table.y.text = FALSE,
        tables.theme = clean_theme(), 
        legend = c(0.8,0.8),
        xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
        break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
        legend.labs = c(paste("Expert panel treatment option", "(-), propensity score matched"),
                        paste("Expert panel treatment option", "(+), propensity score matched"))
        ) + 
        labs(title = "Expert panel treatment option existed or not, PS by age/sex/PS/entry date/multi-cancer/diagnosis", 
           subtitle = paste("Median OS",
               summary(matched_fit)[[18]][[13]], "(0)/",
               summary(matched_fit)[[18]][[14]], "(1) days,",
               "log-rank test, p-value:", format(
           1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
           "Gehan-Wilcoxon test, p-value:", format(
           1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)))
    #print(g)
    pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_option_PS_match.pdf", sep=""), width = 10, height = 10)
    print(g, newpage = FALSE)
    dev.off()
  }
  }
}

# Survival analysis for EP recommended treatment
survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                  data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = paste("Recommended treatment done:",
                         0:1, sep="")
) + 
  labs(title = paste(Cancer, "Recommended treatment done for", 
                     (survival_simple)[[1]][[2]], "patients"), 
       subtitle = paste("Median OS",
             summary(survival_simple)[[18]][[13]], "(0)/",
             summary(survival_simple)[[18]][[14]], "(1) days,",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# propensity score matching considering
# enrollment timing, and multi cancer
if(length(unique(Data_survival$diagnosis)) > 1){
  if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                          症例.基本情報.性別 + 症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + multi_cancer + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    } else{
      propensityScore = glm(EP_treat ~
                          症例.基本情報.性別 + 症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    }
  } else{
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                          症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + multi_cancer + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    } else{
      propensityScore = glm(EP_treat ~
                          症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival)
    }
  }
} else{
  if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                        症例.基本情報.性別 + 症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment + multi_cancer,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    } else{
      propensityScore = glm(EP_treat ~
                        症例.基本情報.性別 + 症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    }
  } else{
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                        症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment + multi_cancer,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    } else{
      propensityScore = glm(EP_treat ~
                        症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment,
                      family  = binomial(link = "logit"),
                      data    = Data_survival)
    }
  }
}

#summary(propensityScore)
Data_survival$PScore = propensityScore$fitted.values
matching = Match(Y = Data_survival$time_enroll_final,
                 Tr = Data_survival$EP_treat,
                 X = Data_survival$PScore,
                 caliper=0.25,
                 ties=F,
                 replace=T)
if(!is.na(matching)[[1]]){
  #summary(matching)
  # ggplot(Data_survival, aes(x = PScore, y = after_stat(density))) +
  #   geom_histogram(color = "black") +
  #   facet_grid(rows = vars(EP_treat), scales = "free_y") +
  #   labs(x = "Propensity score", y = "Probability density") +
  #   ggtitle("Expert panel treatment done")
  matched_data = Data_survival[c(matching$index.treated, matching$index.control),]
  matched_fit = survfit(Surv(event = censor,
                                 time = time_enroll_final) ~ EP_treat, 
                                 data = matched_data)
  if(sum(matched_fit$n.event) > 0){
  if(summary(matched_fit)[[18]][7] > 2 & summary(matched_fit)[[18]][8] > 2){
  diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                    data=matched_data, rho=0)
  diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                    data=matched_data, rho=1)
  g = ggsurvplot(
      fit = matched_fit,
      data = matched_data,
      xlab = "Time (Days) from the enrollment",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
      legend.labs = c(paste("Expert panel treatment done", "(-), propensity score matched"),
                      paste("Expert panel treatment done", "(+), propensity score matched"))
      ) + 
      labs(title = "Expert panel treatment done, PS by age/sex/PS/entry date/multi-cancer/diagnosis", 
         subtitle = paste("Median OS",
             summary(matched_fit)[[18]][[13]], "(0)/",
             summary(matched_fit)[[18]][[14]], "(1) days,",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         "Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3)))
    #print(g)
    pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment_PS_match.pdf", sep=""), width = 10, height = 10)
    print(g, newpage = FALSE)
    dev.off()
  }
  }
}

# Survival analysis for EP recommended treatment only patients with treatment option
Data_survival_option = Data_survival %>%
  dplyr::filter(EP_option == 1)
survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival_option)
if(length(Data_survival_option[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                  data=Data_survival_option, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                  data=Data_survival_option, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival_option,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival_option$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival_option$time_enroll_final) / 500) * 100,
  legend.labs = c("EP option exists but recommended treatment not done",
                  "EP option exists and recommended treatment done")
) + 
  labs(title = paste(Cancer, "Recommended treatment done for", 
                     (survival_simple)[[1]][[2]], "patients with treatment option"), 
       subtitle = paste("Median OS",
             summary(survival_simple)[[18]][[13]], "(0)/",
             summary(survival_simple)[[18]][[14]], "(1) days,",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_onlyPtWithOption_by_treatment_PS_match.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# propensity score matching considering
# enrollment timing, and multi cancer
if(length(unique(Data_survival$diagnosis)) > 1){
  if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                          症例.基本情報.性別 + 症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + multi_cancer + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival_option)
    } else{
      propensityScore = glm(EP_treat ~
                          症例.基本情報.性別 + 症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival_option)
    }
  } else{
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                          症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + multi_cancer + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival_option)
    } else{
      propensityScore = glm(EP_treat ~
                          症例.基本情報.年齢 +
                          症例.背景情報.ECOG.PS + 
                          enrollment + diagnosis,
                        family  = binomial(link = "logit"),
                        data    = Data_survival_option)
    }
  }
} else{
  if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                        症例.基本情報.性別 + 症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment + multi_cancer,
                      family  = binomial(link = "logit"),
                      data    = Data_survival_option)
    } else{
      propensityScore = glm(EP_treat ~
                        症例.基本情報.性別 + 症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment,
                      family  = binomial(link = "logit"),
                      data    = Data_survival_option)
    }
  } else{
    if(max(Data_survival$multi_cancer) == 1){
      propensityScore = glm(EP_treat ~
                        症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment + multi_cancer,
                      family  = binomial(link = "logit"),
                      data    = Data_survival_option)
    } else{
      propensityScore = glm(EP_treat ~
                        症例.基本情報.年齢 +
                        症例.背景情報.ECOG.PS + 
                        enrollment,
                      family  = binomial(link = "logit"),
                      data    = Data_survival_option)
    }
  }
}

#summary(propensityScore)
Data_survival_option$PScore = propensityScore$fitted.values
matching = Match(Y = Data_survival_option$time_enroll_final,
                 Tr = Data_survival_option$EP_treat,
                 X = Data_survival_option$PScore,
                 caliper=0.25,
                 ties=F,
                 replace=T)
if(!is.na(matching)[[1]]){
  #summary(matching)
  # ggplot(Data_survival_option, aes(x = PScore, y = after_stat(density))) +
  #   geom_histogram(color = "black") +
  #   facet_grid(rows = vars(EP_treat), scales = "free_y") +
  #   labs(x = "Propensity score", y = "Probability density") +
  #   ggtitle("Expert panel treatment done, only patients with treatment option")
  
  matched_data = Data_survival_option[c(matching$index.treated, matching$index.control),]
  matched_fit = survfit(Surv(event = censor,
                                 time = time_enroll_final) ~ EP_treat, 
                                 data = matched_data)
  if(sum(matched_fit$n.event) > 0){
  if(summary(matched_fit)[[18]][7] > 2 & summary(matched_fit)[[18]][8] > 2){
  diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                    data=matched_data, rho=0)
  diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                    data=matched_data, rho=1)
  g = ggsurvplot(
      fit = matched_fit,
      data = matched_data,
      xlab = "Time (Days) from the enrollment",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
    legend.labs = c("EP option exists but recommended treatment not done, PS matched",
                    "EP option exists and recommended treatment done, PS matched")
      ) + 
      labs(title = "Expert panel treatment done, only patients with treatment option, PS by age/sex/PS/entry date/multi-cancer/diagnosis", 
         subtitle = paste("Median OS",
             summary(matched_fit)[[18]][[13]], "(0)/",
             summary(matched_fit)[[18]][[14]], "(1) days,",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         "Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3)))
    #print(g)
    pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_onlyPtWithOption_by_treatment_PS_match.pdf", sep=""), width = 10, height = 10)
    print(g, newpage = FALSE)
    dev.off()
  }
  }
}

# Survival analysis for EP recommended treatment
survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group,
                  data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100
) + 
  labs(title = paste(Cancer,"Median OS",
             summary(survival_simple)[[18]][[31]], "/",
             summary(survival_simple)[[18]][[32]], "/",
             summary(survival_simple)[[18]][[33]], "/",
             summary(survival_simple)[[18]][[34]], "/",
             summary(survival_simple)[[18]][[35]], " days,",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "\nGehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment_pattern.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for EP recommended treatment
survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_2,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                  data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100
) + 
  labs(title = paste(Cancer,"Median OS",
             summary(survival_simple)[[18]][[19]], "(",
             summary(survival_simple)[[18]][[22]], "-",
             summary(survival_simple)[[18]][[25]], ")/",
             summary(survival_simple)[[18]][[20]], "(",
             summary(survival_simple)[[18]][[23]], "-",
             summary(survival_simple)[[18]][[26]], ").",
             summary(survival_simple)[[18]][[21]], "(",
             summary(survival_simple)[[18]][[24]], "-",
             summary(survival_simple)[[18]][[27]], ") days\n",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       ", Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment_pattern_2.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for EP recommended treatment
Data_survival_treated = Data_survival %>%
  dplyr::filter(treat_group_2 != "Treat(-)")
survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_2,
                          data=Data_survival_treated)
if(length(Data_survival_treated[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                  data=Data_survival_treated, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                  data=Data_survival_treated, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival_treated,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival_treated$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival_treated$time_enroll_final) / 500) * 100
) + 
  labs(title = paste(Cancer,"Median OS",
             summary(survival_simple)[[18]][[13]], "(",
             summary(survival_simple)[[18]][[15]], "-",
             summary(survival_simple)[[18]][[17]], ")/",
             summary(survival_simple)[[18]][[14]], "(",
             summary(survival_simple)[[18]][[16]], "-",
             summary(survival_simple)[[18]][[18]], ") days\n",
               paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "\nGehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment_pattern_2_2.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for EP recommended treatment
survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_3,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_3,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_3,
                  data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100
) + 
  labs(title = paste(Cancer,"Median OS",
             summary(survival_simple)[[18]][[13]], "(",
             summary(survival_simple)[[18]][[15]], "-",
             summary(survival_simple)[[18]][[17]], ")/",
             summary(survival_simple)[[18]][[14]], "(",
             summary(survival_simple)[[18]][[16]], "-",
             summary(survival_simple)[[18]][[18]], ") days\n",
               paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "\nGehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment_pattern_3.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for EP recommended treatment
survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_4,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_4,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_4,
                  data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100
) + 
  labs(title = paste(Cancer,"Median OS",
             summary(survival_simple)[[18]][[19]], "(",
             summary(survival_simple)[[18]][[22]], "-",
             summary(survival_simple)[[18]][[25]], ")/",
             summary(survival_simple)[[18]][[20]], "(",
             summary(survival_simple)[[18]][[23]], "-",
             summary(survival_simple)[[18]][[26]], ").",
             summary(survival_simple)[[18]][[21]], "(",
             summary(survival_simple)[[18]][[24]], "-",
             summary(survival_simple)[[18]][[27]], ") days\n",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "\nGehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment_pattern_4.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for EP recommended treatment
survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_5,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_5,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_5,
                  data=Data_survival, rho=1)
#summary(survival_simple)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.9),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100
) + 
  labs(title = paste(Cancer,"Median OS",
             summary(survival_simple)[[18]][[13]], "(",
             summary(survival_simple)[[18]][[15]], "-",
             summary(survival_simple)[[18]][[17]], ")/",
             summary(survival_simple)[[18]][[14]], "(",
             summary(survival_simple)[[18]][[16]], "-",
             summary(survival_simple)[[18]][[18]], ") days\n",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "\nGehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3))))
    #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_treatment_pattern_5.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for multiple cancer
if(max(Data_survival$multi_cancer) == 1){
  survival_simple = survfit(Surv(time_enroll_final, censor)~multi_cancer,
                            data=Data_survival)
  if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
  diff_0 = survdiff(Surv(time_enroll_final, censor)~multi_cancer,
                    data=Data_survival, rho=0)
  diff_1 = survdiff(Surv(time_enroll_final, censor)~multi_cancer,
                    data=Data_survival, rho=1)
  #summary(survival_simple)
  g = ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Data_survival,
    xlab = "Time (Days) from enrollment",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.7,0.9),
    xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
    legend.labs = paste("Multiple cancer existed:",
                           0:1, sep="")
  ) + 
    labs(title = paste(Cancer, "Multiple cancer existed for", 
                       (survival_simple)[[1]][[2]], "patients"), 
         subtitle = paste("Median OS",
               summary(survival_simple)[[18]][[13]], "(0)/",
               summary(survival_simple)[[18]][[14]], "(1) days,",
                  paste("log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         "Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3))))
    #print(g)
    pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_multiple_cancer.pdf", sep=""), width = 10, height = 10)
    print(g, newpage = FALSE)
    dev.off()
  }
}

# Survival analysis for ECOG performance status
survival_simple = survfit(Surv(time_enroll_final, censor)~症例.背景情報.ECOG.PS,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~症例.背景情報.ECOG.PS,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~症例.背景情報.ECOG.PS,
                  data=Data_survival, rho=1)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.8),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = paste("Performance status:",
                      sort(unique(Data_survival$症例.背景情報.ECOG.PS)),
                      sep="")
) + 
  labs(title = paste(Cancer, "ECOG PS (9=unknown)"), 
       subtitle = paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3)))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_performance_status.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for ECOG performance status, PS 0 vs 1 vs 2/3/4
Data_survival_PS = Data_survival %>%
  dplyr::filter(症例.背景情報.ECOG.PS != 9) %>%
  dplyr::mutate(PS_modified = case_when(
    症例.背景情報.ECOG.PS == 0 ~ "PS 0",
    症例.背景情報.ECOG.PS == 1 ~ "PS 1",
    TRUE ~ "PS 2-4"
  ))
survival_simple = survfit(Surv(time_enroll_final, censor)~PS_modified,
                          data=Data_survival_PS)
if(length(Data_survival_PS[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~PS_modified,
                  data=Data_survival_PS, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~PS_modified,
                  data=Data_survival_PS, rho=1)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival_PS,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.8),
  xlim = c(0, max(Data_survival_PS$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival_PS$time_enroll_final) / 500) * 100,
  legend.labs = paste("Performance status:",
                      sort(unique(Data_survival_PS$PS_modified)),
                      sep="")
) + 
  labs(title = paste(Cancer, "ECOG PS (PS unknown patients excluded)"), 
       subtitle = paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3)))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_performance_status_simple.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for ECOG performance status, PS 0 vs 1 vs 2/3/4
Data_survival_PS = Data_survival %>%
  dplyr::filter(症例.背景情報.ECOG.PS != 9) %>%
  dplyr::mutate(PS_modified = case_when(
    症例.背景情報.ECOG.PS == 0 ~ "PS 0",
    TRUE ~ "PS 1-4"
  ))
survival_simple = survfit(Surv(time_enroll_final, censor)~PS_modified,
                          data=Data_survival_PS)
if(length(Data_survival_PS[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~PS_modified,
                  data=Data_survival_PS, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~PS_modified,
                  data=Data_survival_PS, rho=1)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival_PS,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.8),
  xlim = c(0, max(Data_survival_PS$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival_PS$time_enroll_final) / 500) * 100,
  legend.labs = paste("Performance status:",
                      sort(unique(Data_survival_PS$PS_modified)),
                      sep="")
) + 
  labs(title = paste(Cancer, "ECOG PS (PS unknown patients excluded)"), 
       subtitle = paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3)))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_performance_status_simple_2.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# Survival analysis for diagnosis
survival_simple = survfit(Surv(time_enroll_final, censor)~diagnosis,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
diff_0 = survdiff(Surv(time_enroll_final, censor)~diagnosis,
                  data=Data_survival, rho=0)
diff_1 = survdiff(Surv(time_enroll_final, censor)~diagnosis,
                  data=Data_survival, rho=1)
g = ggsurvplot(
  fit = survival_simple,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.8),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = sort(unique(Data_survival$diagnosis))
) + 
  labs(title = paste(Cancer, "diagnosis"), 
       subtitle = paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
       digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
       digits=3)))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_giagnosis.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

g = NULL
for(i in 1:length(gene_pattern)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    for(k in 1:length(gene_pattern[[i]])){
      Data_survival$gene_select[[j]] = any(Data_survival$gene_select[[j]],
          (bitwAnd(gene_pattern[[i]][[k]],
                   Data_survival$target_gene_count[j]) ==
             gene_pattern[[i]][[k]]))
    }
  }
  survival_simple = survfit(Surv(time_enroll_final, censor)~gene_select,
                          data=Data_survival)
  mut_gene = ""
  for(l in 1:length(gene_pattern[[i]])){
    for(m in 1:length(target_gene)){
      if(bitwAnd(2^(m-1), gene_pattern[[i]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  if(i == 1){
    mut_gene_analyze = mut_gene
    ID_gene = unique((Data_survival %>%
            dplyr::filter(gene_select == 1))$C.CAT調査結果.基本項目.ハッシュID)
  }
  # print(mut_gene)
  # print(survival_simple)
  if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
    diff_0 = survdiff(Surv(time_enroll_final, censor)~gene_select,
                      data=Data_survival, rho=0)
    diff_1 = survdiff(Surv(time_enroll_final, censor)~gene_select,
                      data=Data_survival, rho=1)
    g=c(g, list(ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Data_survival,
    xlab = "Time (Days) from enrollment",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
    legend.labs = c("Other", mut_gene)
    ) + 
    labs(title = paste(Cancer, mut_gene), 
         subtitle = paste("Median OS",
             summary(survival_simple)[[18]][[13]], "(Negative)/",
             summary(survival_simple)[[18]][[14]], "(Positive) days,",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))))
    # propensity score matching considering
    # sex, age, performance status,
    # enrollment timing, and multi cancer
    if(length(unique(Data_survival$diagnosis)) > 1){
      if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        }
      } else{
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        }
      }
    } else{
      if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                            症例.基本情報.性別 + 症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment + multi_cancer,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                            症例.基本情報.性別 + 症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        }
      } else{
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                            症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment + multi_cancer,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                            症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        }
      }
    }
    # summary(propensityScore)
    Data_survival$PScore = propensityScore$fitted.values
    matching = Match(Y = Data_survival$time_enroll_final,
                     Tr = Data_survival$gene_select,
                     X = Data_survival$PScore,
                     caliper=0.25,
                     ties=F,
                     replace=T)
    if(!is.na(matching)[[1]]){
      # summary(matching)
      # g=c(g, list(ggplot(Data_survival, aes(x = PScore, y = after_stat(density))) +
      #   geom_histogram(color = "black") +
      #   facet_grid(rows = vars(gene_select), scales = "free_y") +
      #   labs(x = "Propensity score", y = "Probability density") +
      #   ggtitle(paste(mut_gene, "mutated or not"))))
      if (!is.na(matching[[1]])) {
      matched_data = Data_survival[c(matching$index.treated, matching$index.control),]
      matched_fit = survfit(Surv(event = censor,
                                     time = time_enroll_final) ~ gene_select, 
                                     data = matched_data)
      if(sum(matched_fit$n.event) > 0){
      if(summary(matched_fit)[[18]][7] > 2 & summary(matched_fit)[[18]][8] > 2){
      diff_0 = survdiff(Surv(time_enroll_final, censor)~gene_select, data=matched_data, rho=0)
      diff_1 = survdiff(Surv(time_enroll_final, censor)~gene_select, data=matched_data, rho=1)
      g=c(g, list(ggsurvplot(
          fit = matched_fit,
          data = matched_data,
          xlab = "Time (Days) from the enrollment",
          ylab = "Survival Probability",
          censor = TRUE,
          conf.int = FALSE,
          pval = TRUE,
          risk.table = TRUE,
          risk.table.y.text = FALSE,
          tables.theme = clean_theme(), 
          legend = c(0.8,0.8),
          xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
          break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
          legend.labs = c(paste(mut_gene, "(-), propensity score matched"),
                          paste(mut_gene, "(+), propensity score matched"))
          ) + 
          labs(title = paste(Cancer, mut_gene, "PS by age/sex/PS/entry date/multi-cancer/diagnosis"), 
         subtitle = paste("Median OS",
             summary(matched_fit)[[18]][[13]], "(0)/",
             summary(matched_fit)[[18]][[14]], "(1) days,",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
         "Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)))))
      }
      }
      }
    }
  }
}
#print(g)
pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_mutation.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

g = NULL
for(i in 1:length(drug_pattern)){
  Data_survival$drug_select = FALSE
  ID_list = NULL
  for(j in 1:length(drug_pattern[[i]])){
    ID_list_tmp = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)
    for(k in 1:length(selected_drug)){
      if(bitwAnd(drug_pattern[[i]][[j]],2^(k-1)) > 0){
        ID_list_tmp = intersect(ID_list_tmp, unique((Data_case_target %>% dplyr::filter(
            str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                       pattern=selected_drug[[k]]))
            )$C.CAT調査結果.基本項目.ハッシュID))
      }
    }
    ID_list = union(ID_list, ID_list_tmp)
  }
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_list,]$drug_select = TRUE 
  survival_simple = survfit(Surv(time_enroll_final, censor)~drug_select,
                          data=Data_survival)
  used_drug = ""
  for(l in 1:length(drug_pattern[[i]])){
    for(m in 1:length(selected_drug)){
      if(bitwAnd(2^(m-1), drug_pattern[[i]][[l]]) > 0){
        used_drug = paste(used_drug, selected_drug[[m]],
                     " and ", sep="")
      }
    }
    used_drug = substr(used_drug, 1, nchar(used_drug)-5)
    used_drug = paste(used_drug, " or ", sep="")
  }
  used_drug = substr(used_drug, 1, nchar(used_drug)-4)
  #print(used_drug)
  #print(survival_simple)
  if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
    diff_0 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=Data_survival, rho=0)
    diff_1 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=Data_survival, rho=1)
    g=c(g, list(ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Data_survival,
    xlab = "Time (Days) from enrollment",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
    legend.labs = c("Other", paste(used_drug, "was used before EP"))
    ) + 
    labs(title = paste(Cancer, used_drug, "was used before EP"), 
         subtitle = paste("Median OS",
             summary(survival_simple)[[18]][[13]], "(Negative)/",
             summary(survival_simple)[[18]][[14]], "(Positive) days,",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))))
    # propensity score matching considering
    # sex, age, performance status,
    # enrollment timing, and multi cancer
    if(length(unique(Data_survival$diagnosis)) > 1){
      if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(drug_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        } else{
          propensityScore = glm(drug_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        }
      } else{
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(drug_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        } else{
          propensityScore = glm(drug_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        }
      }
    } else{
      if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(drug_select ~
                            症例.基本情報.性別 + 症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment + multi_cancer,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        } else{
          propensityScore = glm(drug_select ~
                            症例.基本情報.性別 + 症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        }
      } else{
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(drug_select ~
                            症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment + multi_cancer,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        } else{
          propensityScore = glm(drug_select ~
                            症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        }
      }
    }
    #summary(propensityScore)
    Data_survival$PScore = propensityScore$fitted.values
    matching = Match(Y = Data_survival$time_enroll_final,
                     Tr = Data_survival$drug_select,
                     X = Data_survival$PScore,
                     caliper=0.25,
                     ties=F,
                     replace=T)
    if(!is.na(matching)[[1]]){
      #summary(matching)
      # g = c(g, list( ggplot(Data_survival, aes(x = PScore, y = after_stat(density))) +
      #   geom_histogram(color = "black") +
      #   facet_grid(rows = vars(drug_select), scales = "free_y") +
      #   labs(x = "Propensity score", y = "Probability density") +
      #   ggtitle(paste(target_drug[i], "was used before EP"))))
      if (!is.na(matching[[1]])) {
      matched_data = Data_survival[c(matching$index.treated, matching$index.control),]
      matched_fit = survfit(Surv(event = censor,
                                     time = time_enroll_final) ~ drug_select, 
                                     data = matched_data)
      if(sum(matched_fit$n.event) > 0){
      if(summary(matched_fit)[[18]][7] > 2 & summary(matched_fit)[[18]][8] > 2){
      diff_0 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=matched_data, rho=0)
      diff_1 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=matched_data, rho=1)
      g = c(g, list( ggsurvplot(
          fit = matched_fit,
          data = matched_data,
          xlab = "Time (Days) from the enrollment",
          ylab = "Survival Probability",
          censor = TRUE,
          conf.int = FALSE,
          pval = TRUE,
          risk.table = TRUE,
          risk.table.y.text = FALSE,
          tables.theme = clean_theme(), 
          legend = c(0.8,0.8),
          xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
          break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
          legend.labs = c(paste(used_drug, "(-), propensity score matched"),
                          paste(used_drug, "(+), propensity score matched"))
          ) + 
          labs(title = paste(Cancer, paste(used_drug, "was used before EP", "PS by age/sex/PS/entry date/multi-cancer/diagnosis")), 
         subtitle = paste("Median OS",
             summary(matched_fit)[[18]][[13]], "(0)/",
             summary(matched_fit)[[18]][[14]], "(1) days,",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
         "Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)))))
      }
      }
      }
    }
  }
}
#print(g)
pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_drug_before_CGP.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

g = NULL
for(i in 1:length(target_drug)){
  Data_survival$drug_select = FALSE
  ID_list = unique((Data_case_target %>% dplyr::filter(
            str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                       pattern=target_drug[i]))
            )$C.CAT調査結果.基本項目.ハッシュID)
  if(length(Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_list,]$drug_select) > 0){
    Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_list,]$drug_select = TRUE 
    survival_simple = survfit(Surv(time_enroll_final, censor)~drug_select,
                            data=Data_survival)
    # print(target_drug[i])
    # print(survival_simple)
    if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
      diff_0 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=Data_survival, rho=0)
      diff_1 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=Data_survival, rho=1)
      g=c(g, list(ggsurvplot(
      fit = survival_simple,
      combine = TRUE,
      data = Data_survival,
      xlab = "Time (Days) from enrollment",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
      legend.labs = c("Other", paste(target_drug[i], "was used before EP"))
      ) + 
      labs(title = paste(Cancer, target_drug[i], "was used before EP"), 
           subtitle = paste("Median OS",
               summary(survival_simple)[[18]][[13]], "(Negative)/",
               summary(survival_simple)[[18]][[14]], "(Positive) days,",
                  paste("log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
         "Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))))
      # propensity score matching considering
      # sex, age, performance status,
      # enrollment timing, and multi cancer
      if(length(unique(Data_survival$diagnosis)) > 1){
        if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
          if(max(Data_survival$multi_cancer) == 1){
            propensityScore = glm(drug_select ~
                                症例.基本情報.性別 + 症例.基本情報.年齢 +
                                症例.背景情報.ECOG.PS + 
                                enrollment + multi_cancer + diagnosis,
                              family  = binomial(link = "logit"),
                              data    = Data_survival)
          } else{
            propensityScore = glm(drug_select ~
                                症例.基本情報.性別 + 症例.基本情報.年齢 +
                                症例.背景情報.ECOG.PS + 
                                enrollment + diagnosis,
                              family  = binomial(link = "logit"),
                              data    = Data_survival)
          }
        } else{
          if(max(Data_survival$multi_cancer) == 1){
            propensityScore = glm(drug_select ~
                                症例.基本情報.年齢 +
                                症例.背景情報.ECOG.PS + 
                                enrollment + multi_cancer + diagnosis,
                              family  = binomial(link = "logit"),
                              data    = Data_survival)
          } else{
            propensityScore = glm(drug_select ~
                                症例.基本情報.年齢 +
                                症例.背景情報.ECOG.PS + 
                                enrollment + diagnosis,
                              family  = binomial(link = "logit"),
                              data    = Data_survival)
          }
        }
      } else{
        if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
          if(max(Data_survival$multi_cancer) == 1){
            propensityScore = glm(drug_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
          } else{
            propensityScore = glm(drug_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
          }
        } else{
          if(max(Data_survival$multi_cancer) == 1){
            propensityScore = glm(drug_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
          } else{
            propensityScore = glm(drug_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
          }
        }
      }
      # summary(propensityScore)
      Data_survival$PScore = propensityScore$fitted.values
      matching = Match(Y = Data_survival$time_enroll_final,
                       Tr = Data_survival$drug_select,
                       X = Data_survival$PScore,
                       caliper=0.25,
                       ties=F,
                       replace=T)
      if(!is.na(matching)[[1]]){
        # summary(matching)
        # g = c(g, list( ggplot(Data_survival, aes(x = PScore, y = after_stat(density))) +
        #   geom_histogram(color = "black") +
        #   facet_grid(rows = vars(drug_select), scales = "free_y") +
        #   labs(x = "Propensity score", y = "Probability density") +
        #   ggtitle(paste(target_drug[i], "was used before EP"))))
        if (!is.na(matching[[1]])) {
        matched_data = Data_survival[c(matching$index.treated, matching$index.control),]
        matched_fit = survfit(Surv(event = censor,
                                       time = time_enroll_final) ~ drug_select, 
                                       data = matched_data)
        if(sum(matched_fit$n.event) > 0){
        if(summary(matched_fit)[[18]][7] > 2 & summary(matched_fit)[[18]][8] > 2){
        diff_0 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=matched_data, rho=0)
        diff_1 = survdiff(Surv(time_enroll_final, censor)~drug_select, data=matched_data, rho=1)
        g = c(g, list( ggsurvplot(
            fit = matched_fit,
            data = matched_data,
            xlab = "Time (Days) from the enrollment",
            ylab = "Survival Probability",
            censor = TRUE,
            conf.int = FALSE,
            pval = TRUE,
            risk.table = TRUE,
            risk.table.y.text = FALSE,
            tables.theme = clean_theme(), 
            legend = c(0.8,0.8),
            xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
            break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
            legend.labs = c(paste(target_drug[i], "(-), propensity score matched"),
                            paste(target_drug[i], "(+), propensity score matched"))
            ) + 
            labs(title = paste(Cancer, paste(target_drug[i], "was used before EP, PS by age/sex/PS/entry date/multi-cancer/diagnosis")), 
           subtitle = paste("Median OS",
             summary(matched_fit)[[18]][[13]], "(0)/",
             summary(matched_fit)[[18]][[14]], "(1) days,",
             "log-rank test, p-value:", format(
           1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
           "Gehan-Wilcoxon test, p-value:", format(
           1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)))))
        }
        }
        }
      }
    }
  }
}
#print(g)
pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_freauent_drug.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

g = NULL
for(i in 1:min(length(oncogenic_genes$all_patients), gene_number)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    Data_survival$gene_select[[j]] = bitwAnd(
        Data_survival$oncogenic_gene_count[j], 2^(i-1))
  }
  Data_survival$gene_select = as.logical(Data_survival$gene_select)
  survival_simple = survfit(Surv(time_enroll_final, censor)~gene_select,
                          data=Data_survival)
  mut_gene = paste(oncogenic_genes$gene_mutation[i],
                   ", top ", i, " gene with oncogenic mutations", sep="")
  # print(mut_gene)
  # print(survival_simple)
  if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
    diff_0 = survdiff(Surv(time_enroll_final, censor)~gene_select, data=Data_survival, rho=0)
    diff_1 = survdiff(Surv(time_enroll_final, censor)~gene_select, data=Data_survival, rho=1)
    g=c(g, list(ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Data_survival,
    xlab = "Time (Days) from enrollment",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
    legend.labs = c("Other", mut_gene)
    ) + 
    labs(title = paste(Cancer, "top", gene_number,"genes with oncogenic mutations"), 
         subtitle = paste("Median OS",
             summary(survival_simple)[[18]][[13]], "(Negative)/",
             summary(survival_simple)[[18]][[14]], "(Positive) days,",
                paste("log-rank test, p-value:", format(
       1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
       "Gehan-Wilcoxon test, p-value:", format(
       1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3))))))
    # propensity score matching considering
    # sex, age, performance status,
    # enrollment timing, and multi cancer
    if(length(unique(Data_survival$diagnosis)) > 1){
      if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                              症例.基本情報.性別 + 症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        }
      } else{
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + multi_cancer + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                              症例.基本情報.年齢 +
                              症例.背景情報.ECOG.PS + 
                              enrollment + diagnosis,
                            family  = binomial(link = "logit"),
                            data    = Data_survival)
        }
      }
    } else{
      if(length(unique(Data_survival$症例.基本情報.性別)) > 1){
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                            症例.基本情報.性別 + 症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment + multi_cancer,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                            症例.基本情報.性別 + 症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        }
      } else{
        if(max(Data_survival$multi_cancer) == 1){
          propensityScore = glm(gene_select ~
                            症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment + multi_cancer,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        } else{
          propensityScore = glm(gene_select ~
                            症例.基本情報.年齢 +
                            症例.背景情報.ECOG.PS + 
                            enrollment,
                          family  = binomial(link = "logit"),
                          data    = Data_survival)
        }
      }
    }

    # summary(propensityScore)
    Data_survival$PScore = propensityScore$fitted.values
    matching = Match(Y = Data_survival$time_enroll_final,
                     Tr = Data_survival$gene_select,
                     X = Data_survival$PScore,
                     caliper=0.25,
                     ties=F,
                     replace=T)
    if(!is.na(matching)[[1]]){
      # summary(matching)
      # g=c(g, list(ggplot(Data_survival, aes(x = PScore, y = after_stat(density))) +
      #   geom_histogram(color = "black") +
      #   facet_grid(rows = vars(gene_select), scales = "free_y") +
      #   labs(x = "Propensity score", y = "Probability density") +
      #   ggtitle(paste(mut_gene, "mutated or not"))))
      if (!is.na(matching[[1]])) {
      matched_data = Data_survival[c(matching$index.treated, matching$index.control),]
      matched_fit = survfit(Surv(event = censor,
                                     time = time_enroll_final) ~ gene_select, 
                                     data = matched_data)
      if(sum(matched_fit$n.event) > 0){
      if(summary(matched_fit)[[18]][7] > 2 & summary(matched_fit)[[18]][8] > 2){
      diff_0 = survdiff(Surv(time_enroll_final, censor)~gene_select, data=matched_data, rho=0)
      diff_1 = survdiff(Surv(time_enroll_final, censor)~gene_select, data=matched_data, rho=1)
      g=c(g, list(ggsurvplot(
          fit = matched_fit,
          data = matched_data,
          xlab = "Time (Days) from the enrollment",
          ylab = "Survival Probability",
          censor = TRUE,
          conf.int = FALSE,
          pval = TRUE,
          risk.table = TRUE,
          risk.table.y.text = FALSE,
          tables.theme = clean_theme(), 
          legend = c(0.8,0.8),
          xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
          break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
          legend.labs = c(paste(mut_gene, "(-), propensity score matched"),
                          paste(mut_gene, "(+), propensity score matched"))
          ) + 
          labs(title = paste(Cancer, mut_gene, "PS by age/sex/PS/entry date/multi-cancer/diagnosis"), 
         subtitle = paste("Median OS",
             summary(matched_fit)[[18]][[13]], "(0)/",
             summary(matched_fit)[[18]][[14]], "(1) days,",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
         "Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)))))
      }
      }
      }
    }
  }
}
#print(g)
pdf(file = paste(OutDirName, Cancer, "_KMcurve_enrollment_to_final_by_freauent_mutation.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()


Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  treat_group_3,
  enrollment,
  time_enroll_final,
  diagnosis) 

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_option,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment option exsisted**") %>%
  modify_caption("**Table 5. Enrollment to final observation - Expert panel option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 6. Enrollment to final observation - Recommended treatment done**") %>%
  bold_labels()
}

Data_summary %>%
  tbl_summary(
    by = treat_group_3,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 6'. Enrollment to final observation - Any treatment done or not**") %>%
  bold_labels()

Data_summary_treat = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_enroll_final,
  diagnosis) %>%
  dplyr::filter(EP_option == 1)

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
survival_simple_ = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(survival_simple_[[1]][[1]] != survival_simple[[1]][[1]]){
Data_summary_treat %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 7. Enrollment to final observation - Recommended treatment done, only patients with treatment option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~diagnosis,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = diagnosis,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_caption("**Table 8. Enrollment to final observation - diagnosis**") %>%
  bold_labels()
}
```

```{r stan_program, echo = (Source_code == 1)}
Stan_code_loglog_weibull =
"
// log-Logistic survival model

functions {
  // Defines the log hazard
  vector log_h (vector t, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_H;
    for (i in 1:num_elements(t)) {
      log_H[i] = log(shape_tmp)-log(scale_tmp[i])+(shape_tmp-1)*(log(t[i])-log(scale_tmp[i]))-log(1+pow((t[i]/scale_tmp[i]),shape_tmp));
    }
    return log_H;
  }
  
  // Defines the log survival
  vector log_S (vector t, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_s;
    for (i in 1:num_elements(t)) {
      log_s[i] = -log(1+pow((t[i]/scale_tmp[i]),shape_tmp));
    }
    return log_s;
  }
  
  // Defines the sampling distribution
  real surv_loglogistic_lpdf (vector t, vector d, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_lik;
    real prob;
    log_lik = d .* log_h(t,shape_tmp,scale_tmp) + log_S(t,shape_tmp,scale_tmp);
    prob = sum(log_lik);
    return prob;
  }
}

data {
  int<lower=0> Median;
  int<lower=0> Nobs;
  int<lower=0> Ncen;
  int<lower=0> Ntot;
  int<lower=0> Nexp;
  int<lower=0> M_bg;
  vector[Nexp] ybef_exp;
  vector[Nobs] yobs;
  vector[Ncen] ycen;
  matrix[Nobs, M_bg] Xobs_bg;
  matrix[Ncen, M_bg] Xcen_bg;
}

transformed data {
  vector[Nobs] one_obs;
  vector[Ncen] zero_cen;

  matrix[Nobs, M_bg] Xobs_bg_early;
  matrix[Nobs, M_bg] Xobs_bg_late;
  matrix[Ncen, M_bg] Xcen_bg_early;
  matrix[Ncen, M_bg] Xcen_bg_late;

  for (i in 1:Nobs) {
    Xobs_bg_early[i,1] = 0;
    Xobs_bg_late[i,1] = 1;
    one_obs[i] = 1;
  }
  for (i in 1:Ncen) {
    Xcen_bg_early[i,1] = 0;
    Xcen_bg_late[i,1] = 1;
    zero_cen[i] = 0;
  }
}

parameters {
  vector[M_bg] alpha;
  real<lower=0> beta;
  real<lower=0> mu;

  real shape_exp;
  real<lower=0> scale_exp;
}

transformed parameters {
  vector[Nobs] linpredobs;
  vector[Ncen] linpredcen;
  vector[Nobs] muobs;
  vector[Ncen] mucen;
  linpredobs = Xobs_bg*alpha;
  for (i in 1:Nobs) {
    muobs[i] = exp(mu + linpredobs[i]);
  }
  linpredcen = Xcen_bg*alpha;
  for (i in 1:Ncen) {
    mucen[i] = exp(mu + linpredcen[i]);
  }
}

model {
  ybef_exp ~ weibull(exp(shape_exp), scale_exp);
  yobs ~ surv_loglogistic(one_obs, beta, muobs);
  ycen ~ surv_loglogistic(zero_cen, beta, mucen);
  mu ~ normal(6,3);
  alpha ~ normal(0,5);
  beta ~ normal(0,5);
  shape_exp ~ normal(0.0,5.0);
  scale_exp ~ normal(300.0,300.0);
}

generated quantities {
  real yhat_exp_uncens[Nobs + Ncen];
  real yhat_exp_total[Nobs + Ncen];
  real y_exptmp[Nobs + Ncen];
  real yhat_early[Nobs + Ncen];
  real yhat_late[Nobs + Ncen];
  for (i in 1:Nobs) {
    y_exptmp[i] = weibull_rng(exp(shape_exp), scale_exp);
    yhat_early[i] = exp(logistic_rng(((mu + Xobs_bg_early[i,] * alpha)), 1.0/beta));
    yhat_late[i] = exp(logistic_rng(((mu + Xobs_bg_late[i,] * alpha)), 1.0/beta));
    if (y_exptmp[i] <= Median)
      yhat_exp_uncens[i] = yhat_early[i];
    else
      yhat_exp_uncens[i] = yhat_late[i];
    yhat_exp_total[i] = yhat_exp_uncens[i] + y_exptmp[i];
  }
  for (i in 1:Ncen) {
    y_exptmp[Nobs + i] = weibull_rng(exp(shape_exp), scale_exp);
    yhat_early[Nobs + i] = exp(logistic_rng(((mu + Xcen_bg_early[i,] * alpha)), 1.0/beta));
    yhat_late[Nobs + i] = exp(logistic_rng(((mu + Xcen_bg_late[i,] * alpha)), 1.0/beta));
    if (y_exptmp[Nobs + i] <= Median)
      yhat_exp_uncens[Nobs + i] = yhat_early[Nobs + i];
    else
      yhat_exp_uncens[Nobs + i] = yhat_late[Nobs + i];
    yhat_exp_total[Nobs + i] = yhat_exp_uncens[Nobs + i] + y_exptmp[Nobs + i];
  }
}

"

Stan_code_loglog_weibull_factor <- 
"
// log-Logistic survival model

functions {
  // Defines the log hazard
  vector log_h (vector t, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_H;
    for (i in 1:num_elements(t)) {
      log_H[i] = log(shape_tmp)-log(scale_tmp[i])+(shape_tmp-1)*(log(t[i])-log(scale_tmp[i]))-log(1+pow((t[i]/scale_tmp[i]),shape_tmp));
    }
    return log_H;
  }
  
  // Defines the log survival
  vector log_S (vector t, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_s;
    for (i in 1:num_elements(t)) {
      log_s[i] = -log(1+pow((t[i]/scale_tmp[i]),shape_tmp));
    }
    return log_s;
  }
  
  // Defines the sampling distribution
  real surv_loglogistic_lpdf (vector t, vector d, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_lik;
    real prob;
    log_lik = d .* log_h(t,shape_tmp,scale_tmp) + log_S(t,shape_tmp,scale_tmp);
    prob = sum(log_lik);
    return prob;
  }
}

data {
  int<lower=0> Median_pos;
  int<lower=0> Median_neg;
  int<lower=0> Nobs;
  int<lower=0> Ncen;
  int<lower=0> Ntot;
  int<lower=0> Nexp;
  int<lower=0> M_bg;
  vector[Nexp] ybef_exp;
  vector[Nexp] xfactor;
  vector[Nobs] yobs;
  vector[Ncen] ycen;
  matrix[Nobs, M_bg] Xobs_bg;
  matrix[Ncen, M_bg] Xcen_bg;
}

transformed data {
  vector[Nobs] one_obs;
  vector[Ncen] zero_cen;

  matrix[Nobs, M_bg] Xobs_bg_early_pos = Xobs_bg;
  matrix[Nobs, M_bg] Xobs_bg_late_pos = Xobs_bg;
  matrix[Ncen, M_bg] Xcen_bg_early_pos = Xcen_bg;
  matrix[Ncen, M_bg] Xcen_bg_late_pos = Xcen_bg;
  matrix[Nobs, M_bg] Xobs_bg_early_neg = Xobs_bg;
  matrix[Nobs, M_bg] Xobs_bg_late_neg = Xobs_bg;
  matrix[Ncen, M_bg] Xcen_bg_early_neg = Xcen_bg;
  matrix[Ncen, M_bg] Xcen_bg_late_neg = Xcen_bg;

  for (i in 1:Nobs) {
    Xobs_bg_early_pos[i,1] = 0;
    Xobs_bg_late_pos[i,1] = 1;
    Xobs_bg_early_neg[i,1] = 0;
    Xobs_bg_late_neg[i,1] = 1;
    Xobs_bg_early_pos[i,2] = 1;
    Xobs_bg_late_pos[i,2] = 1;
    Xobs_bg_early_neg[i,2] = 0;
    Xobs_bg_late_neg[i,2] = 0;
    one_obs[i] = 1;
  }
  for (i in 1:Ncen) {
    Xcen_bg_early_pos[i,1] = 0;
    Xcen_bg_late_pos[i,1] = 1;
    Xcen_bg_early_neg[i,1] = 0;
    Xcen_bg_late_neg[i,1] = 1;
    Xcen_bg_early_pos[i,2] = 1;
    Xcen_bg_late_pos[i,2] = 1;
    Xcen_bg_early_neg[i,2] = 0;
    Xcen_bg_late_neg[i,2] = 0;
    zero_cen[i] = 0;
  }
}

parameters {
  vector[M_bg] alpha;
  real<lower=0> beta;
  real factor;
  real<lower=0> mu;
  real shape_exp;
  real<lower=0> scale_exp;
}

transformed parameters {
  vector[Nobs] linpredobs;
  vector[Ncen] linpredcen;
  vector[Nobs] muobs;
  vector[Ncen] mucen;
  linpredobs = Xobs_bg*alpha;
  for (i in 1:Nobs) {
    muobs[i] = exp(mu + linpredobs[i]);
  }
  linpredcen = Xcen_bg*alpha;
  for (i in 1:Ncen) {
    mucen[i] = exp(mu + linpredcen[i]);
  }
}

model {
  ybef_exp ~ weibull(exp(shape_exp), scale_exp + xfactor * factor);
  yobs ~ surv_loglogistic(one_obs, beta, muobs);
  ycen ~ surv_loglogistic(zero_cen, beta, mucen);

  mu ~ normal(6,3);
  alpha ~ normal(0,5);
  beta ~ normal(0,5);
  shape_exp ~ normal(0.0,5.0);
  scale_exp ~ normal(300.0,300.0);
  factor ~ normal(0.0,100);
}

generated quantities {
  real yhat_total[Nobs + Ncen];
  real yhat_factor_total[Nobs + Ncen];
  real y_bef;
  real y_aft;
  for (i in 1:Nobs) {
    y_bef = weibull_rng(exp(shape_exp), scale_exp);
    if (y_bef <= Median_neg) {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_early_neg[i,] * alpha)), 1.0/beta));
      yhat_total[i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_late_neg[i,] * alpha)), 1.0/beta));
      yhat_total[i] = y_aft + y_bef;
    }
    y_bef =  weibull_rng(exp(shape_exp), scale_exp + factor);
    if (y_bef <= Median_pos) {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_early_pos[i,] * alpha)), 1.0/beta));
      yhat_factor_total[i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_late_pos[i,] * alpha)), 1.0/beta));
      yhat_factor_total[i] = y_aft + y_bef;
    }
  }
  for (i in 1:Ncen) {
    y_bef = weibull_rng(exp(shape_exp), scale_exp);
    if (y_bef <= Median_neg) {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_early_neg[i,] * alpha)), 1.0/beta));
      yhat_total[Nobs + i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_late_neg[i,] * alpha)), 1.0/beta));
      yhat_total[Nobs + i] = y_aft + y_bef;
    }
    y_bef =  weibull_rng(exp(shape_exp), scale_exp + factor);
    if (y_bef <= Median_pos) {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_early_pos[i,] * alpha)), 1.0/beta));
      yhat_factor_total[Nobs + i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_late_pos[i,] * alpha)), 1.0/beta));
      yhat_factor_total[Nobs + i] = y_aft + y_bef;
    }
  }
}
"

Stan_code_loglog_weibull_3_factor <- 
"
// log-Logistic survival model

functions {
  // Defines the log hazard
  vector log_h (vector t, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_H;
    for (i in 1:num_elements(t)) {
      log_H[i] = log(shape_tmp)-log(scale_tmp[i])+(shape_tmp-1)*(log(t[i])-log(scale_tmp[i]))-log(1+pow((t[i]/scale_tmp[i]),shape_tmp));
    }
    return log_H;
  }
  
  // Defines the log survival
  vector log_S (vector t, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_s;
    for (i in 1:num_elements(t)) {
      log_s[i] = -log(1+pow((t[i]/scale_tmp[i]),shape_tmp));
    }
    return log_s;
  }
  
  // Defines the sampling distribution
  real surv_loglogistic_lpdf (vector t, vector d, real shape_tmp, vector scale_tmp) {
    vector[num_elements(t)] log_lik;
    real prob;
    log_lik = d .* log_h(t,shape_tmp,scale_tmp) + log_S(t,shape_tmp,scale_tmp);
    prob = sum(log_lik);
    return prob;
  }
}

data {
  int<lower=0> Median_2;
  int<lower=0> Median_1;
  int<lower=0> Median_0;
  int<lower=0> Nobs;
  int<lower=0> Ncen;
  int<lower=0> Ntot;
  int<lower=0> Nexp;
  int<lower=0> M_bg;
  vector[Nexp] ybef_exp;
  vector[Nexp] xfactor;
  vector[Nexp] xfactor2;
  vector[Nobs] yobs;
  vector[Ncen] ycen;
  matrix[Nobs, M_bg] Xobs_bg;
  matrix[Ncen, M_bg] Xcen_bg;
}

transformed data {
  vector[Nobs] one_obs;
  vector[Ncen] zero_cen;

  matrix[Nobs, M_bg] Xobs_bg_early_2;
  matrix[Nobs, M_bg] Xobs_bg_late_2;
  matrix[Ncen, M_bg] Xcen_bg_early_2;
  matrix[Ncen, M_bg] Xcen_bg_late_2;
  matrix[Nobs, M_bg] Xobs_bg_early_1;
  matrix[Nobs, M_bg] Xobs_bg_late_1;
  matrix[Ncen, M_bg] Xcen_bg_early_1;
  matrix[Ncen, M_bg] Xcen_bg_late_1;
  matrix[Nobs, M_bg] Xobs_bg_early_0;
  matrix[Nobs, M_bg] Xobs_bg_late_0;
  matrix[Ncen, M_bg] Xcen_bg_early_0;
  matrix[Ncen, M_bg] Xcen_bg_late_0;

  for (i in 1:Nobs) {
    Xobs_bg_early_2[i,1] = 0;
    Xobs_bg_late_2[i,1] = 1;
    Xobs_bg_early_1[i,1] = 0;
    Xobs_bg_late_1[i,1] = 1;
    Xobs_bg_early_0[i,1] = 0;
    Xobs_bg_late_0[i,1] = 1;
    Xobs_bg_early_2[i,2] = 0;
    Xobs_bg_late_2[i,2] = 0;
    Xobs_bg_early_1[i,2] = 1;
    Xobs_bg_late_1[i,2] = 1;
    Xobs_bg_early_0[i,2] = 0;
    Xobs_bg_late_0[i,2] = 0;
    Xobs_bg_early_2[i,3] = 1;
    Xobs_bg_late_2[i,3] = 1;
    Xobs_bg_early_1[i,3] = 0;
    Xobs_bg_late_1[i,3] = 0;
    Xobs_bg_early_0[i,3] = 0;
    Xobs_bg_late_0[i,3] = 0;
    one_obs[i] = 1;
  }
  for (i in 1:Ncen) {
    Xcen_bg_early_2[i,1] = 0;
    Xcen_bg_late_2[i,1] = 1;
    Xcen_bg_early_1[i,1] = 0;
    Xcen_bg_late_1[i,1] = 1;
    Xcen_bg_early_0[i,1] = 0;
    Xcen_bg_late_0[i,1] = 1;
    Xcen_bg_early_2[i,2] = 0;
    Xcen_bg_late_2[i,2] = 0;
    Xcen_bg_early_1[i,2] = 1;
    Xcen_bg_late_1[i,2] = 1;
    Xcen_bg_early_0[i,2] = 0;
    Xcen_bg_late_0[i,2] = 0;
    Xcen_bg_early_2[i,3] = 1;
    Xcen_bg_late_2[i,3] = 1;
    Xcen_bg_early_1[i,3] = 0;
    Xcen_bg_late_1[i,3] = 0;
    Xcen_bg_early_0[i,3] = 0;
    Xcen_bg_late_0[i,3] = 0;
    zero_cen[i] = 0;
  }
}

parameters {
  vector[M_bg] alpha;
  real<lower=0> beta;
  real factor;
  real factor2;
  real<lower=0> mu;
  real shape_exp;
  real<lower=0> scale_exp;
}

transformed parameters {
  vector[Nobs] linpredobs;
  vector[Ncen] linpredcen;
  vector[Nobs] muobs;
  vector[Ncen] mucen;
  linpredobs = Xobs_bg*alpha;
  for (i in 1:Nobs) {
    muobs[i] = exp(mu + linpredobs[i]);
  }
  linpredcen = Xcen_bg*alpha;
  for (i in 1:Ncen) {
    mucen[i] = exp(mu + linpredcen[i]);
  }
}

model {
  ybef_exp ~ weibull(exp(shape_exp), scale_exp + xfactor * factor + xfactor2 * factor2);
  yobs ~ surv_loglogistic(one_obs, beta, muobs);
  ycen ~ surv_loglogistic(zero_cen, beta, mucen);

  mu ~ normal(6,3);
  alpha ~ normal(0,5);
  beta ~ normal(0,5);
  shape_exp ~ normal(0.0,5.0);
  scale_exp ~ normal(300.0,300.0);
  factor ~ normal(0.0,100);
  factor2 ~ normal(0.0,100);
}

generated quantities {
  real yhat_total[Nobs + Ncen];
  real yhat_factor_total[Nobs + Ncen];
  real yhat_factor2_total[Nobs + Ncen];
  real y_bef;
  real y_aft;
  for (i in 1:Nobs) {
    y_bef = weibull_rng(exp(shape_exp), scale_exp);
    if (y_bef <= Median_0) {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_early_0[i,] * alpha)), 1.0/beta));
      yhat_total[i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_late_0[i,] * alpha)), 1.0/beta));
      yhat_total[i] = y_aft + y_bef;
    }
    y_bef =  weibull_rng(exp(shape_exp), scale_exp + factor);
    if (y_bef <= Median_1) {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_early_1[i,] * alpha)), 1.0/beta));
      yhat_factor_total[i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_late_1[i,] * alpha)), 1.0/beta));
      yhat_factor_total[i] = y_aft + y_bef;
    }
    y_bef =  weibull_rng(exp(shape_exp), scale_exp + factor2);
    if (y_bef <= Median_2) {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_early_2[i,] * alpha)), 1.0/beta));
      yhat_factor2_total[i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xobs_bg_late_2[i,] * alpha)), 1.0/beta));
      yhat_factor2_total[i] = y_aft + y_bef;
    }
  }
  for (i in 1:Ncen) {
    y_bef = weibull_rng(exp(shape_exp), scale_exp);
    if (y_bef <= Median_0) {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_early_0[i,] * alpha)), 1.0/beta));
      yhat_total[Nobs + i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_late_0[i,] * alpha)), 1.0/beta));
      yhat_total[Nobs + i] = y_aft + y_bef;
    }
    y_bef =  weibull_rng(exp(shape_exp), scale_exp + factor);
    if (y_bef <= Median_1) {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_early_1[i,] * alpha)), 1.0/beta));
      yhat_factor_total[Nobs + i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_late_1[i,] * alpha)), 1.0/beta));
      yhat_factor_total[Nobs + i] = y_aft + y_bef;
    }
    y_bef =  weibull_rng(exp(shape_exp), scale_exp + factor2);
    if (y_bef <= Median_2) {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_early_2[i,] * alpha)), 1.0/beta));
      yhat_factor2_total[Nobs + i] = y_aft + y_bef;
    } else {
      y_aft = exp(logistic_rng(((mu + Xcen_bg_late_2[i,] * alpha)), 1.0/beta));
      yhat_factor2_total[Nobs + i] = y_aft + y_bef;
    }
  }
}

"

if(!file.exists("source/Stan_code_loglog_weibull.stan")){
  write(Stan_code_loglog_weibull, file = "source/Stan_code_loglog_weibull.stan")
}
if(!file.exists("source/Stan_code_loglog_weibull_factor.stan")){
  write(Stan_code_loglog_weibull_factor, file = "source/Stan_code_loglog_weibull_factor.stan")
}
if(!file.exists("source/Stan_code_loglog_weibull_3_factor.stan")){
  write(Stan_code_loglog_weibull_3_factor, file = "source/Stan_code_loglog_weibull_3_factor.stan")
}
```

```{r survival_from_chemo, eval = (Analyze_chemo >= 1), echo = (Source_code == 1 & Analyze_chemo >= 1), warning=FALSE}
## Survival from the first palliative chemotherapy
## Survival analysis considering left truncation

Data_survival = Data_case_target %>%
  dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))

Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日))
Data_survival = Data_survival %>%
  dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))
Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
  Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日))
Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$target_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_target_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:length(target_gene)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  } else{
    for(k in 1:length(target_gene)){
      if(target_gene[[k]][2] == "WT"){
        Data_survival$target_gene_count[i] = bitwOr(
          Data_survival$target_gene_count[i], 2^(k-1))
      }
    }
  }
}

if(sum(Data_survival$target_gene_count) == 0){
  target_gene = list(c(oncogenic_genes[1,1], "F"))
  gene_pattern = list(c(1))
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.変異情報.変異内容 == target_gene[[i]][2]),
      Data_target_gene)
  }
  Data_target_gene = Data_target_gene %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=TRUE)
  
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[i]][2]) %>%
      dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                      C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT,
                    .keep_all=TRUE),
      Data_target_gene)
  }
  Data_survival$target_gene_count = 0
  ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
  for(i in 1:length(ID_target_gene)){
    tmp = Data_target_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
    if(dplyr::count(tmp)[[1]] > 0){
      for(j in 1:dplyr::count(tmp)[[1]]){
        for(k in 1:length(target_gene)){
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
        }
      }
    } else{
      for(k in 1:length(target_gene)){
        if(target_gene[[k]][2] == "WT"){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "SPT to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "SPT to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "SPT to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "SPT to entry 3-4 year",
    TRUE ~ "SPT to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
    TRUE ~ "SPT to entry later"))

## Target gene selection
Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.エビデンス.エビデンスレベル,
                  .keep_all=TRUE) %>%
  dplyr::mutate(gene_mut = paste(C.CAT調査結果.変異情報.マーカー,
                                 C.CAT調査結果.変異情報.変異内容,
                                 sep=" _ ")) %>%
  dplyr::mutate(gene_patho = C.CAT調査結果.変異情報.マーカー)

oncogenic_mutations = Data_major_gene %>%
                dplyr::count(gene_mut) %>%
                dplyr::arrange(-n)

c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_mutations
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_mut)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_mut)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_mutations = cbind(oncogenic_mutations, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_mutations) =c_name_gene

DT::datatable(oncogenic_mutations, filter = 'top', extensions = 'Scroller',
             caption = "Oncogenic mutations, SPT to death analysis")

oncogenic_genes = Data_major_gene %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations, SPT to death analysis (counts)")

oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
  dplyr::distinct() %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
      dplyr::distinct() %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations, SPT to death analysis (patients)")
Data_survival$oncogenic_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_major_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:min(length(oncogenic_genes$all_patients), gene_number)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == oncogenic_genes$gene_mutation[[k]]){
          Data_survival$oncogenic_gene_count[i] = bitwOr(
            Data_survival$oncogenic_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

```

```{r loglog_stan, eval = (Analyze_chemo == 1), echo = (Source_code == 1 & Analyze_chemo == 1), warning=FALSE}
time_tmp = quantile(Data_survival$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
time_10 = as.integer(time_tmp[[1]])
time_90 = as.integer(time_tmp[[2]])
time_100 = as.integer(time_tmp[[3]])
Data_survival_tmp = Data_survival %>% dplyr::filter(
  time_palliative_enroll >= time_10 &
    time_palliative_enroll <= time_90)
Data_survival_tmp$time_palliative_enroll =
  Data_survival_tmp$time_palliative_enroll - (time_10 - 1)

Data_survival_tmp2 = Data_survival %>% dplyr::filter(
  time_palliative_enroll > time_90 &
    time_palliative_enroll <= time_100)
Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
idx <- 1:nrow(Data_survival_tmp2)
Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
Median_entry = median(Data_survival_tmp$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
    TRUE ~ "SPT to entry later"))

stan_weibull_survival_model_data <-
    list(
        Median = as.integer(Median_entry),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 1,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1], ncol = 1),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0], ncol = 1)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1)

#mcmc_rhat(rhat(stan_weibull_survival_model_fit))

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_bef_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
    tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_aft_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
    tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_ear <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
    tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_lat <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
    tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
    dplyr::select(-key) 


median_os_list_weibull_total_exp = NULL
median_os_list_weibull_bef_exp = NULL
median_os_list_weibull_aft_exp = NULL
median_os_list_weibull_ear = NULL
median_os_list_weibull_lat = NULL

for(i in 1:8000){
  idx = seq(i, length(stan_weibull_survival_model_draws_total_exp$yhat_total_exp), by=8000)
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_bef_exp = c(median_os_list_weibull_bef_exp,
      median(stan_weibull_survival_model_draws_bef_exp[idx,]$yhat_bef_exp, na.rm = T))
    median_os_list_weibull_aft_exp = c(median_os_list_weibull_aft_exp,
      median(stan_weibull_survival_model_draws_aft_exp[idx,]$yhat_aft_exp, na.rm = T))
    median_os_list_weibull_ear = c(median_os_list_weibull_ear,
      median(stan_weibull_survival_model_draws_ear[idx,]$yhat_ear, na.rm = T))
    median_os_list_weibull_lat = c(median_os_list_weibull_lat,
      median(stan_weibull_survival_model_draws_lat[idx,]$yhat_lat, na.rm = T))
}

median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bef_exp = quantile(median_os_list_weibull_bef_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_aft_exp = quantile(median_os_list_weibull_aft_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_ear = quantile(median_os_list_weibull_ear, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_lat = quantile(median_os_list_weibull_lat, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_bef_exp = sort(stan_weibull_survival_model_draws_bef_exp$yhat_bef_exp)
yhat_weibull_sort_aft_exp = sort(stan_weibull_survival_model_draws_aft_exp$yhat_aft_exp)
yhat_weibull_sort_ear = sort(stan_weibull_survival_model_draws_ear$yhat_ear)
yhat_weibull_sort_lat = sort(stan_weibull_survival_model_draws_lat$yhat_lat)
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_bef_exp = yhat_weibull_sort_bef_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_aft_exp = yhat_weibull_sort_aft_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_ear = yhat_weibull_sort_ear[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_lat = yhat_weibull_sort_lat[1:length(Data_survival$censor) * 8000]

Data_survival$left_adj = yhat_weibull_sort_average_total_exp
Data_survival$weibull_bef_exp = yhat_weibull_sort_average_bef_exp
Data_survival$weibull_ear = yhat_weibull_sort_average_ear
Data_survival$weibull_lat = yhat_weibull_sort_average_lat

traditional_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival)
traditional_fit_2 = survfit(Surv(event = rep(1,length(Data_survival_tmp$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival_tmp)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_bef_exp)),
                               time = weibull_bef_exp) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit_1 = traditional_fit_1,
             traditional_fit_2 = traditional_fit_2,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  #data = Data_survival,
  xlab = "Time from SPT to CGP (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
  legend.labs = c("raw data",
                  "middle data",
                  "Weibull curve")
) + labs(title = paste(Cancer, " ", traditional_fit_1[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit_1)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_1)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_1)[[17]][[9]]),") (raw)/",
             as.integer(summary(traditional_fit_2)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_2)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_2)[[17]][[9]]),") (middle)\n",
             as.integer(median_os_summary_weibull_bef_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_bef_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_bef_exp[[3]]),
             ") (Weibull) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_weibull_SPT_to_CGP.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$weibull_ear)),
                               time = weibull_ear) ~ 1, 
                               data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_lat)),
                               time = weibull_lat) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_1 = simulation_fit_1,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data",
                  "CGP enrollment at early timing, Weibull curve",
                  "CGP enrollment at late timing, Weibull curve")
) + labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (raw, early)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (raw, late)/\n",
             as.integer(median_os_summary_weibull_ear[[2]]), 
             " (", as.integer(median_os_summary_weibull_ear[[1]]), "-",
             as.integer(median_os_summary_weibull_ear[[3]]), ") (Weibull, early)/",
             as.integer(median_os_summary_weibull_lat[[2]]), " (",
             as.integer(median_os_summary_weibull_lat[[1]]), "-",
             as.integer(median_os_summary_weibull_lat[[3]]),
             ") (Weibull, late) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_CGP_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggsurvplot(
  fit = traditional_fit,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  pval = T,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data")
)
pdf(file = paste(OutDirName, Cancer, "_CGP_to_death_raw.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

# check Kendall's tau
# p-value < 0.05 -> quasi-independence of truncation and death is rejected 
independence_check = with(Data_survival, trSurvfit(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  tFun = "linear",
  plots = FALSE))
# independence_check
pdf(file = paste(OutDirName, Cancer, "_tranSurv.pdf", sep=""), width = 10, height = 10)
plot(independence_check)
text.default(x=0, y=0, adj = 0, paste("cKendall tau =",
                      round(independence_check$iniKendall, 3)))
dev.off()

if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to positive dependent-left truncation.")
  print("Patients with CGP test at early timing lived short survival.")
} else if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to negative dependent-left truncation.")
  print("Patients with CGP test at early timing lived long survival.")
} else{
  print("This survival data was subjected to independent-left truncation.")
}

independence_check = with(Data_survival, cKendall(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  trans = "linear"))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ 1, 
                               data = Data_survival)
delayed_entry_fit = survfit(Surv(event = censor,
                                 time = time_palliative_enroll,
                                 time2 = time_palliative_final) ~ 1, 
                                 data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$left_adj)),
                               time = left_adj) ~ 1, 
                               data = Data_survival)
tmp_num = 18
if(length(summary(delayed_entry_fit)[[18]]) == 1){
  tmp_num = 17
}
g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             delayed_entry_fit = delayed_entry_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from SPT to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Number at risk adjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + 
  labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[7]]), " (",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[8]]),"-",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[9]]),") (risk adj.)/\n",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(delayed_entry_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (risk adj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_loglog_SPT_to_death_3curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from SPT to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + 

  labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS\n",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_loglog_SPT_to_death_2curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

# Survival analysis for enrollment timing
traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ enrollment, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$enrollment)) == 5){
  print(traditional_fit)
  legend_enroll =  c("2019-06-01 to 2020-06-30",
                     "2020-07-01 to 2021-06-30",
                     "2021-07-01 to 2021-12-31",
                     "2022-01-01 to 2022-06-30",
                     "2022-07-01 to 2023-02-20")
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.90),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = legend_enroll
  ) + 
    labs(title = paste(Cancer, "Enrollment timing for", 
                       sum((traditional_fit)[[1]]), "patients, w/o adjustment"))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_entry_period.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  print(traditional_fit)
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.85),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100
  ) + labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((traditional_fit)[[1]]), "patients"))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_entry_period.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# comparing between diagnoses
if(length(unique(Data_survival$diagnosis)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ diagnosis, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = sort(unique(Data_survival$diagnosis))[[1]]
    name_2 = sort(unique(Data_survival$diagnosis))[[2]]
    name_3 = sort(unique(Data_survival$diagnosis))[[3]]
    name_1_short = str_sub(name_1,1,5)
    name_2_short = str_sub(name_2,1,5)
    name_3_short = str_sub(name_3,1,5)

    Data_survival$gene_select =
      as.numeric(Data_survival$diagnosis == name_2) +
      2 * as.numeric(Data_survival$diagnosis == name_3)
    
    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    time_tmp = quantile(Data_survival_tmp_2$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    time_tmp = quantile(Data_survival_tmp_1$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    time_tmp = quantile(Data_survival_tmp_0$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp6 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp6$time_palliative_enroll = Data_survival_tmp6$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp6)
    Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx %% 10 != 1],])
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_0_2 = median_os_list_weibull_total - median_os_list_weibull_total_exp2
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_0_2 = quantile(median_os_list_weibull_bias_0_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Bonferroni-like correction, 98.3% CI)\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_1_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_2[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (98.3% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_3_diagnosis.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# comparing between diagnoses
if(length(unique(Data_survival$diagnosis)) == 2){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ diagnosis, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = sort(unique(Data_survival$diagnosis))[[1]]
    name_2 = sort(unique(Data_survival$diagnosis))[[2]]
    name_1_short = str_sub(name_1,1,5)
    name_2_short = str_sub(name_2,1,5)
    
    Data_survival$gene_select = Data_survival$diagnosis == name_2
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
  
  
  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (",
               name_1_short,", unadj.)/\n",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (",
               name_2_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/\n",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (", name_2_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_2_diagnosis.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# comparing among designated prognostic scores
Data_survival$gene_select = 0
for(i in 1:length(target_gene)){
  ID_tmp = unique((Data_report %>% dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == target_gene[[i]][2] & C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1]))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] = 
    Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] + 1
}
Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    gene_select > 2 ~ 2,
    TRUE ~ gene_select
  ))
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = "No muts in selected genes"
    name_2 = "Single mut in selected genes"
    name_3 = "Multiple muts in selected genes"
    name_1_short = "No"
    name_2_short = "Single"
    name_3_short = "Multiple"

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    time_tmp = quantile(Data_survival_tmp_2$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    time_tmp = quantile(Data_survival_tmp_1$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    time_tmp = quantile(Data_survival_tmp_0$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp6 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp6$time_palliative_enroll = Data_survival_tmp6$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp6)
    Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx %% 10 != 1],])
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS\n",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Dunnett-like correction, ",
                       paste(unlist(target_gene)[seq(1, length(target_gene) * 2 - 1, 2)], collapse = ","),
                       ")\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (97.5% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_prognostic_scoring.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# Double-hit is a prognostic factor?
Data_survival$gene_select = 0
  ID_tmp = (Data_report %>% dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" & C.CAT調査結果.変異情報.マーカー == double_hit_gene))$C.CAT調査結果.基本項目.ハッシュID
  table_tmp = data.frame(table(ID_tmp))
  ID_multihit = as.character(table_tmp[table_tmp$Freq>1,]$ID_tmp)
  ID_singlehit = as.character(table_tmp[table_tmp$Freq==1,]$ID_tmp)
  ID_nohit = as.character(table_tmp[table_tmp$Freq==0,]$ID_tmp)
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_multihit, "gene_select"] = 2
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_singlehit, "gene_select"] = 1
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_nohit, "gene_select"] = 0

Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    gene_select > 2 ~ 2,
    TRUE ~ gene_select
  ))
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = paste("No muts in", double_hit_gene)
    name_2 = paste("Single mut in", double_hit_gene)
    name_3 = paste("Multiple muts in", double_hit_gene)
    name_1_short = "No"
    name_2_short = "Single"
    name_3_short = "Multiple"

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    time_tmp = quantile(Data_survival_tmp_2$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    time_tmp = quantile(Data_survival_tmp_1$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    time_tmp = quantile(Data_survival_tmp_0$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp6 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp6$time_palliative_enroll = Data_survival_tmp6$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp6)
    Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx %% 10 != 1],])
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS\n",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Dunnett-like correction, ",
                       double_hit_gene,
                       ")\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (97.5% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_multiple_hit.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# analysis for common oncogenic mutations
g = NULL
gene_table = data.frame(oncogenic_genes$gene_mutation[1:min(length(oncogenic_genes$all_patients), gene_number)])
colnames(gene_table) = c("Gene")
gene_table$cancer_type = Cancer
gene_table$positive_patients = oncogenic_genes$all_patients[1:min(length(oncogenic_genes$all_patients), gene_number)]
gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$delay)) / 10
gene_table$positive_median = 0
gene_table$positive_LL = 0
gene_table$positive_UL = 0
gene_table$negative_median = 0
gene_table$negative_LL = 0
gene_table$negative_UL = 0
gene_table$diff_median = 0
gene_table$diff_LL = 0
gene_table$diff_UL = 0
gene_table$positive_mortal_event = 0
gene_table$negative_mortal_event = 0
  
for(i in 1:min(length(oncogenic_genes$all_patients), gene_number)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    Data_survival$gene_select[[j]] = bitwAnd(
        Data_survival$oncogenic_gene_count[j], 2^(i-1))
  }
  Data_survival$gene_select = as.logical(Data_survival$gene_select)
  mut_gene = paste(oncogenic_genes$gene_mutation[i],
                   ", top ", i, " gene with oncogenic mutation", sep="")
  file.create(paste(OutDirName, "top_", i, "_gene.tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    gene_table$negative_mortal_event[i] = summary(traditional_fit)[[18]][7]
    gene_table$positive_mortal_event[i] = summary(traditional_fit)[[18]][8]
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
      gene_table$positive_median[i] = as.integer(median_os_summary_weibull_total_exp[[2]])
      gene_table$positive_LL[i] = as.integer(median_os_summary_weibull_total_exp[[1]])
      gene_table$positive_UL[i] = as.integer(median_os_summary_weibull_total_exp[[3]])
      gene_table$negative_median[i] = as.integer(median_os_summary_weibull_total[[2]])
      gene_table$negative_LL[i] = as.integer(median_os_summary_weibull_total[[1]])
      gene_table$negative_UL[i] = as.integer(median_os_summary_weibull_total[[3]])
      gene_table$diff_median[i] = -as.integer(median_os_summary_weibull_bias[[2]])
      gene_table$diff_LL[i] = -as.integer(median_os_summary_weibull_bias[[3]])
      gene_table$diff_UL[i] = -as.integer(median_os_summary_weibull_bias[[1]])

    }
  }
  file.remove(paste(OutDirName, "top_", i, "_gene.tmp", sep=""))
}

#g
pdf(file = paste(OutDirName, Cancer, "_freq_gene_SPT_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

write.csv(gene_table, file = paste(OutDirName, Cancer, "_FreqGene_loglog_",
          length(Data_survival$症例.基本情報.年齢), "_patients.csv", sep=""))

Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))
p <- 
  gene_table |>
  ggplot(aes(y = fct_rev(Gene))) + 
  theme_classic()
p <- p +
  geom_point(aes(x=diff_median), shape=15, size=3) +
  geom_linerange(aes(xmin=diff_LL, xmax=diff_UL)) 
p <- p +
  geom_vline(xintercept = 0, linetype="dashed") +
  labs(x="Survival difference by mutation", y="Genes with oncogenic alteration")
p <- p +
  coord_cartesian(ylim=c(1,length(Gene_arrange) + 1), xlim=c(min(gene_table$diff_LL) - 15, max(gene_table$diff_UL) + 15))
p <- p +
  annotate("text", x = min(gene_table$diff_LL)/2 - 15, y = length(Gene_arrange) + 1, label = "mut=short survival") +
  annotate("text", x = max(gene_table$diff_UL)/2 + 15, y = length(Gene_arrange) + 1, label = "mut=long survival")
p_mid <- p + 
  theme(axis.line.y = element_blank(),
        axis.ticks.y= element_blank(),
        axis.text.y= element_blank(),
        axis.title.y= element_blank())

gene_table <- gene_table %>%
  dplyr::mutate(
  estimate_lab = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
  dplyr::mutate(
    patients = paste0(positive_patients, " (", positive_freq, "%)"))
gene_table = 
  dplyr::bind_rows(
    data.frame(
      Gene = "Gene",
      estimate_lab = "Survival difference (95% CI)",
      patients = "Patients"
    ), gene_table  )
Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))

p_left <-
  gene_table  %>%
  ggplot(aes(y = fct_rev(Gene)))
p_left <- 
  p_left +
  geom_text(aes(x = 0, label = Gene), hjust = 0, fontface = "bold")
p_left <- 
  p_left +
  geom_text(
    aes(x = 1, label = estimate_lab),
    hjust = 0,
    fontface = ifelse(gene_table$estimate_lab == "Survival difference (95% CI)", "bold", "plain")
  )
p_left <-
  p_left +
  theme_void() +
  coord_cartesian(xlim = c(0, 4))

# right side of plot - pvalues
p_right <-
  gene_table  |>
  ggplot() +
  geom_text(
    aes(x = 0, y = fct_rev(Gene), label = patients),
    hjust = 0,
    fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")
  ) +
  theme_void() 

layout <- c(
  area(t = 0, l = 0, b = 30, r = 3), # left plot, starts at the top of the page (0) and goes 30 units down and 3 units to the right
  area(t = 1, l = 4, b = 30, r = 9), # middle plot starts a little lower (t=1) because there's no title. starts 1 unit right of the left plot (l=4, whereas left plot is r=3), goes to the bottom of the page (30 units), and 6 units further over from the left plot (r=9 whereas left plot is r=3)
  area(t = 0, l = 9, b = 30, r = 11) # right most plot starts at top of page, begins where middle plot ends (l=9, and middle plot is r=9), goes to bottom of page (b=30), and extends two units wide (r=11)
)
# final plot arrangement
pdf(file = paste(OutDirName, Cancer, "_Gene_forest-plot.pdf", sep=""), width = 10, height = 10)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()

poor_prognostic_factor = 
  gene_table %>% dplyr::arrange(diff_UL)
poor_prognostic_factor = poor_prognostic_factor %>%
  dplyr::filter(diff_LL != diff_UL)
poor_prognostic_factor = 
  as.character(poor_prognostic_factor[1:6, "Gene"])

if(all(c("CDKN2A", "CDKN2B") %in% poor_prognostic_factor) == TRUE){
  poor_prognostic_factor =
    poor_prognostic_factor[-which(poor_prognostic_factor %in% "CDKN2B")] 
}
poor_prognostic_factor = poor_prognostic_factor[1:4]


g = NULL
for(i in 1:length(gene_pattern)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    for(k in 1:length(gene_pattern[[i]])){
      Data_survival$gene_select[[j]] = any(Data_survival$gene_select[[j]],
          (bitwAnd(gene_pattern[[i]][[k]],
                   Data_survival$target_gene_count[j]) ==
             gene_pattern[[i]][[k]]))
    }
  }
  mut_gene = ""
  for(l in 1:length(gene_pattern[[i]])){
    for(m in 1:length(target_gene)){
      if(bitwAnd(2^(m-1), gene_pattern[[i]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  file.create(paste(OutDirName, mut_gene, ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(OutDirName, mut_gene, ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_gene_SPT_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()


g = NULL
for(i in 1:length(target_pathway)){
  Data_report_tmp = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[[i]]][[1]])
  ID_path = unique(Data_report_tmp$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival = Data_survival %>%
    dplyr::mutate(gene_select = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  mut_gene = paste(target_pathway[i], " mutation", sep="")
  file.create(paste(OutDirName, target_pathway[i], ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(OutDirName, target_pathway[i], ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_pathway_SPT_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

# comparing among prognostic scores; auto gene selection
Data_survival$gene_select = 0
for(i in 1:length(poor_prognostic_factor)){
  ID_tmp = unique((Data_report %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー == poor_prognostic_factor[[i]] & C.CAT調査結果.エビデンス.エビデンスレベル == "F"))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] = 
    Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] + 1
}
Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    gene_select > 2 ~ 2,
    TRUE ~ gene_select
  ))
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = "No muts in selected genes"
    name_2 = "Single mut in selected genes"
    name_3 = "Multiple muts in selected genes"
    name_1_short = "No"
    name_2_short = "Single"
    name_3_short = "Multiple"

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    time_tmp = quantile(Data_survival_tmp_2$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    time_tmp = quantile(Data_survival_tmp_1$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    time_tmp = quantile(Data_survival_tmp_0$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp6 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp6$time_palliative_enroll = Data_survival_tmp6$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp6)
    Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx %% 10 != 1],])
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS\n",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Dunnett-like correction, ",
                       paste(unlist(poor_prognostic_factor), collapse = ","),
                       " oncogenic muts)\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (97.5% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_prognostic_scoring_auto.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. Patients with selected disease, SPT-death, Characteristics**") %>%
  bold_labels()
table_tmp

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_option,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment option exsisted**") %>%
  modify_caption("**Table 9. First chemotherapy to final observation - Expert panel option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 10. Enrollment to final observation - Recommended treatment done**") %>%
  bold_labels()
}

Data_summary_treat = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) %>%
  dplyr::filter(EP_option == 1)

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
survival_simple_ = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(survival_simple_[[1]][[1]] != survival_simple[[1]][[1]] & length(survival_simple_) == length(survival_simple)){
Data_summary_treat %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 2''. First chemotherapy to final observation - Recommended treatment done, only patients with treatment option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~diagnosis,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = diagnosis,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 3,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 3)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_caption("**Table 11. First chemotherapy to final observation - diagnosis**") %>%
  bold_labels()
}
```

```{r loglog_stan_new, eval = (Analyze_chemo == 2), echo = (Source_code == 1 & Analyze_chemo == 2), warning=FALSE}

Data_survival_tmp = Data_survival %>% dplyr::filter(
  time_palliative_enroll >= time_10)
Data_survival_tmp$time_palliative_enroll =
  Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
Median_entry = median(Data_survival_tmp$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
    TRUE ~ "SPT to entry later"))

stan_weibull_survival_model_data <-
    list(
        Median = as.integer(Median_entry),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 1,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1], ncol = 1),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0], ncol = 1)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1)

#mcmc_rhat(rhat(stan_weibull_survival_model_fit))

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_bef_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
    tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_aft_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
    tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_ear <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
    tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_lat <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
    tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
    dplyr::select(-key) 


median_os_list_weibull_total_exp = NULL
median_os_list_weibull_bef_exp = NULL
median_os_list_weibull_aft_exp = NULL
median_os_list_weibull_ear = NULL
median_os_list_weibull_lat = NULL

for(i in 1:8000){
  idx = seq(i, length(stan_weibull_survival_model_draws_total_exp$yhat_total_exp), by=8000)
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_bef_exp = c(median_os_list_weibull_bef_exp,
      median(stan_weibull_survival_model_draws_bef_exp[idx,]$yhat_bef_exp, na.rm = T))
    median_os_list_weibull_aft_exp = c(median_os_list_weibull_aft_exp,
      median(stan_weibull_survival_model_draws_aft_exp[idx,]$yhat_aft_exp, na.rm = T))
    median_os_list_weibull_ear = c(median_os_list_weibull_ear,
      median(stan_weibull_survival_model_draws_ear[idx,]$yhat_ear, na.rm = T))
    median_os_list_weibull_lat = c(median_os_list_weibull_lat,
      median(stan_weibull_survival_model_draws_lat[idx,]$yhat_lat, na.rm = T))
}

median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bef_exp = quantile(median_os_list_weibull_bef_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_aft_exp = quantile(median_os_list_weibull_aft_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_ear = quantile(median_os_list_weibull_ear, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_lat = quantile(median_os_list_weibull_lat, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_bef_exp = sort(stan_weibull_survival_model_draws_bef_exp$yhat_bef_exp)
yhat_weibull_sort_aft_exp = sort(stan_weibull_survival_model_draws_aft_exp$yhat_aft_exp)
yhat_weibull_sort_ear = sort(stan_weibull_survival_model_draws_ear$yhat_ear)
yhat_weibull_sort_lat = sort(stan_weibull_survival_model_draws_lat$yhat_lat)
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_bef_exp = yhat_weibull_sort_bef_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_aft_exp = yhat_weibull_sort_aft_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_ear = yhat_weibull_sort_ear[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_lat = yhat_weibull_sort_lat[1:length(Data_survival$censor) * 8000]

Data_survival$left_adj = yhat_weibull_sort_average_total_exp
Data_survival$weibull_bef_exp = yhat_weibull_sort_average_bef_exp
Data_survival$weibull_ear = yhat_weibull_sort_average_ear
Data_survival$weibull_lat = yhat_weibull_sort_average_lat

traditional_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_bef_exp)),
                               time = weibull_bef_exp) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit_1 = traditional_fit_1,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  #data = Data_survival,
  xlab = "Time from SPT to CGP (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
  legend.labs = c("raw data",
                  "Weibull curve")
) + labs(title = paste(Cancer, " ", traditional_fit_1[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit_1)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_1)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_1)[[17]][[9]]),") (raw)/",
             as.integer(median_os_summary_weibull_bef_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_bef_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_bef_exp[[3]]),
             ") (Weibull) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_weibull_SPT_to_CGP.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$weibull_ear)),
                               time = weibull_ear) ~ 1, 
                               data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_lat)),
                               time = weibull_lat) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_1 = simulation_fit_1,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data",
                  "CGP enrollment at early timing, Weibull curve",
                  "CGP enrollment at late timing, Weibull curve")
) + labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (raw, early)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (raw, late)/\n",
             as.integer(median_os_summary_weibull_ear[[2]]), 
             " (", as.integer(median_os_summary_weibull_ear[[1]]), "-",
             as.integer(median_os_summary_weibull_ear[[3]]), ") (Weibull, early)/",
             as.integer(median_os_summary_weibull_lat[[2]]), " (",
             as.integer(median_os_summary_weibull_lat[[1]]), "-",
             as.integer(median_os_summary_weibull_lat[[3]]),
             ") (Weibull, late) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_CGP_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggsurvplot(
  fit = traditional_fit,
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  pval = T,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data")
)
pdf(file = paste(OutDirName, Cancer, "_CGP_to_death_raw.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

# check Kendall's tau
# p-value < 0.05 -> quasi-independence of truncation and death is rejected 
independence_check = with(Data_survival, trSurvfit(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  tFun = "linear",
  plots = FALSE))
# independence_check
pdf(file = paste(OutDirName, Cancer, "_tranSurv.pdf", sep=""), width = 10, height = 10)
plot(independence_check)
text.default(x=0, y=0, adj = 0, paste("cKendall tau =",
                      round(independence_check$iniKendall, 3)))
dev.off()

if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to positive dependent-left truncation.")
  print("Patients with CGP test at early timing lived short survival.")
} else if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to negative dependent-left truncation.")
  print("Patients with CGP test at early timing lived long survival.")
} else{
  print("This survival data was subjected to independent-left truncation.")
}

independence_check = with(Data_survival, cKendall(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  trans = "linear"))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ 1, 
                               data = Data_survival)
delayed_entry_fit = survfit(Surv(event = censor,
                                 time = time_palliative_enroll,
                                 time2 = time_palliative_final) ~ 1, 
                                 data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$left_adj)),
                               time = left_adj) ~ 1, 
                               data = Data_survival)
tmp_num = 18
if(length(summary(delayed_entry_fit)[[18]]) == 1){
  tmp_num = 17
}
g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             delayed_entry_fit = delayed_entry_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from SPT to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Number at risk adjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + 
  labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[7]]), " (",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[8]]),"-",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[9]]),") (risk adj.)/\n",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(delayed_entry_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (risk adj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_loglog_SPT_to_death_3curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from SPT to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + 

  labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS\n",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_loglog_SPT_to_death_2curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

# Survival analysis for enrollment timing
traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ enrollment, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$enrollment)) == 5){
  print(traditional_fit)
  legend_enroll =  c("2019-06-01 to 2020-06-30",
                     "2020-07-01 to 2021-06-30",
                     "2021-07-01 to 2021-12-31",
                     "2022-01-01 to 2022-06-30",
                     "2022-07-01 to 2023-02-20")
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.90),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = legend_enroll
  ) + 
    labs(title = paste(Cancer, "Enrollment timing for", 
                       sum((traditional_fit)[[1]]), "patients, w/o adjustment"))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_entry_period.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  print(traditional_fit)
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.85),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100
  ) + labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((traditional_fit)[[1]]), "patients"))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_entry_period.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

# comparing between diagnoses
if(length(unique(Data_survival$diagnosis)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ diagnosis, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = sort(unique(Data_survival$diagnosis))[[1]]
    name_2 = sort(unique(Data_survival$diagnosis))[[2]]
    name_3 = sort(unique(Data_survival$diagnosis))[[3]]
    name_1_short = str_sub(name_1,1,5)
    name_2_short = str_sub(name_2,1,5)
    name_3_short = str_sub(name_3,1,5)

    Data_survival$gene_select =
      as.numeric(Data_survival$diagnosis == name_2) +
      2 * as.numeric(Data_survival$diagnosis == name_3)
    
    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_0_2 = median_os_list_weibull_total - median_os_list_weibull_total_exp2
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_0_2 = quantile(median_os_list_weibull_bias_0_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Bonferroni-like correction, 98.3% CI)\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_1_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_2[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (98.3% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_3_diagnosis.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# comparing between diagnoses
if(length(unique(Data_survival$diagnosis)) == 2){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ diagnosis, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = sort(unique(Data_survival$diagnosis))[[1]]
    name_2 = sort(unique(Data_survival$diagnosis))[[2]]
    name_1_short = str_sub(name_1,1,5)
    name_2_short = str_sub(name_2,1,5)
    
    Data_survival$gene_select = Data_survival$diagnosis == name_2
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
  
  
  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (",
               name_1_short,", unadj.)/\n",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (",
               name_2_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/\n",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (", name_2_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_2_diagnosis.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# comparing among designated prognostic scores
Data_survival$gene_select = 0
for(i in 1:length(target_gene)){
  ID_tmp = unique((Data_report %>% dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == target_gene[[i]][2] & C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1]))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] = 
    Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] + 1
}
Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    gene_select > 2 ~ 2,
    TRUE ~ gene_select
  ))
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = "No muts in selected genes"
    name_2 = "Single mut in selected genes"
    name_3 = "Multiple muts in selected genes"
    name_1_short = "No"
    name_2_short = "Single"
    name_3_short = "Multiple"

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS\n",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Dunnett-like correction, ",
                       paste(unlist(target_gene)[seq(1, length(target_gene) * 2 - 1, 2)], collapse = ","),
                       ")\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (97.5% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_prognostic_scoring.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# Double-hit is a prognostic factor?
Data_survival$gene_select = 0
  ID_tmp = (Data_report %>% dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" & C.CAT調査結果.変異情報.マーカー == double_hit_gene))$C.CAT調査結果.基本項目.ハッシュID
  table_tmp = data.frame(table(ID_tmp))
  ID_multihit = as.character(table_tmp[table_tmp$Freq>1,]$ID_tmp)
  ID_singlehit = as.character(table_tmp[table_tmp$Freq==1,]$ID_tmp)
  ID_nohit = as.character(table_tmp[table_tmp$Freq==0,]$ID_tmp)
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_multihit, "gene_select"] = 2
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_singlehit, "gene_select"] = 1
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_nohit, "gene_select"] = 0

Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    gene_select > 2 ~ 2,
    TRUE ~ gene_select
  ))
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = paste("No muts in", double_hit_gene)
    name_2 = paste("Single mut in", double_hit_gene)
    name_3 = paste("Multiple muts in", double_hit_gene)
    name_1_short = "No"
    name_2_short = "Single"
    name_3_short = "Multiple"

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS\n",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Dunnett-like correction, ",
                       double_hit_gene,
                       ")\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (97.5% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_multiple_hit.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

# analysis for common oncogenic mutations
g = NULL
gene_table = data.frame(oncogenic_genes$gene_mutation[1:min(length(oncogenic_genes$all_patients), gene_number)])
colnames(gene_table) = c("Gene")
gene_table$cancer_type = Cancer
gene_table$positive_patients = oncogenic_genes$all_patients[1:min(length(oncogenic_genes$all_patients), gene_number)]
gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$delay)) / 10
gene_table$positive_median = 0
gene_table$positive_LL = 0
gene_table$positive_UL = 0
gene_table$negative_median = 0
gene_table$negative_LL = 0
gene_table$negative_UL = 0
gene_table$diff_median = 0
gene_table$diff_LL = 0
gene_table$diff_UL = 0
gene_table$positive_mortal_event = 0
gene_table$negative_mortal_event = 0
  
for(i in 1:min(length(oncogenic_genes$all_patients), gene_number)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    Data_survival$gene_select[[j]] = bitwAnd(
        Data_survival$oncogenic_gene_count[j], 2^(i-1))
  }
  Data_survival$gene_select = as.logical(Data_survival$gene_select)
  mut_gene = paste(oncogenic_genes$gene_mutation[i],
                   ", top ", i, " gene with oncogenic mutation", sep="")
  file.create(paste(OutDirName, "top_", i, "_gene.tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    gene_table$negative_mortal_event[i] = summary(traditional_fit)[[18]][7]
    gene_table$positive_mortal_event[i] = summary(traditional_fit)[[18]][8]
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
      gene_table$positive_median[i] = as.integer(median_os_summary_weibull_total_exp[[2]])
      gene_table$positive_LL[i] = as.integer(median_os_summary_weibull_total_exp[[1]])
      gene_table$positive_UL[i] = as.integer(median_os_summary_weibull_total_exp[[3]])
      gene_table$negative_median[i] = as.integer(median_os_summary_weibull_total[[2]])
      gene_table$negative_LL[i] = as.integer(median_os_summary_weibull_total[[1]])
      gene_table$negative_UL[i] = as.integer(median_os_summary_weibull_total[[3]])
      gene_table$diff_median[i] = -as.integer(median_os_summary_weibull_bias[[2]])
      gene_table$diff_LL[i] = -as.integer(median_os_summary_weibull_bias[[3]])
      gene_table$diff_UL[i] = -as.integer(median_os_summary_weibull_bias[[1]])

    }
  }
  file.remove(paste(OutDirName, "top_", i, "_gene.tmp", sep=""))
}

#g
pdf(file = paste(OutDirName, Cancer, "_freq_gene_SPT_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

write.csv(gene_table, file = paste(OutDirName, Cancer, "_FreqGene_loglog_",
          length(Data_survival$症例.基本情報.年齢), "_patients.csv", sep=""))

Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))
p <- 
  gene_table |>
  ggplot(aes(y = fct_rev(Gene))) + 
  theme_classic()
p <- p +
  geom_point(aes(x=diff_median), shape=15, size=3) +
  geom_linerange(aes(xmin=diff_LL, xmax=diff_UL)) 
p <- p +
  geom_vline(xintercept = 0, linetype="dashed") +
  labs(x="Survival difference by mutation", y="Genes with oncogenic alteration")
p <- p +
  coord_cartesian(ylim=c(1,length(Gene_arrange) + 1), xlim=c(min(gene_table$diff_LL) - 15, max(gene_table$diff_UL) + 15))
p <- p +
  annotate("text", x = min(gene_table$diff_LL)/2 - 15, y = length(Gene_arrange) + 1, label = "mut=short survival") +
  annotate("text", x = max(gene_table$diff_UL)/2 + 15, y = length(Gene_arrange) + 1, label = "mut=long survival")
p_mid <- p + 
  theme(axis.line.y = element_blank(),
        axis.ticks.y= element_blank(),
        axis.text.y= element_blank(),
        axis.title.y= element_blank())

gene_table <- gene_table %>%
  dplyr::mutate(
  estimate_lab = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
  dplyr::mutate(
    patients = paste0(positive_patients, " (", positive_freq, "%)"))
gene_table = 
  dplyr::bind_rows(
    data.frame(
      Gene = "Gene",
      estimate_lab = "Survival difference (95% CI)",
      patients = "Patients"
    ), gene_table  )
Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))

p_left <-
  gene_table  %>%
  ggplot(aes(y = fct_rev(Gene)))
p_left <- 
  p_left +
  geom_text(aes(x = 0, label = Gene), hjust = 0, fontface = "bold")
p_left <- 
  p_left +
  geom_text(
    aes(x = 1, label = estimate_lab),
    hjust = 0,
    fontface = ifelse(gene_table$estimate_lab == "Survival difference (95% CI)", "bold", "plain")
  )
p_left <-
  p_left +
  theme_void() +
  coord_cartesian(xlim = c(0, 4))

# right side of plot - pvalues
p_right <-
  gene_table  |>
  ggplot() +
  geom_text(
    aes(x = 0, y = fct_rev(Gene), label = patients),
    hjust = 0,
    fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")
  ) +
  theme_void() 

layout <- c(
  area(t = 0, l = 0, b = 30, r = 3), # left plot, starts at the top of the page (0) and goes 30 units down and 3 units to the right
  area(t = 1, l = 4, b = 30, r = 9), # middle plot starts a little lower (t=1) because there's no title. starts 1 unit right of the left plot (l=4, whereas left plot is r=3), goes to the bottom of the page (30 units), and 6 units further over from the left plot (r=9 whereas left plot is r=3)
  area(t = 0, l = 9, b = 30, r = 11) # right most plot starts at top of page, begins where middle plot ends (l=9, and middle plot is r=9), goes to bottom of page (b=30), and extends two units wide (r=11)
)
# final plot arrangement
pdf(file = paste(OutDirName, Cancer, "_Gene_forest-plot.pdf", sep=""), width = 12, height = 10)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()
pdf(file = paste(OutDirName, Cancer, "_Gene_forest-plot_middle_size.pdf", sep=""), width = 12, height = 6)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()
pdf(file = paste(OutDirName, Cancer, "_Gene_forest-plot_small_size.pdf", sep=""), width = 10, height = 6)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()
pdf(file = paste(OutDirName, Cancer, "_Gene_forest-plot_mini_size.pdf", sep=""), width = 8, height = 6)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()
pdf(file = paste(OutDirName, Cancer, "_Gene_forest-plot_petit_size.pdf", sep=""), width = 6, height = 4)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()

poor_prognostic_factor = 
  gene_table %>% dplyr::arrange(diff_UL)
poor_prognostic_factor = poor_prognostic_factor %>%
  dplyr::filter(diff_LL != diff_UL)
poor_prognostic_factor = 
  as.character(poor_prognostic_factor[1:6, "Gene"])

if(all(c("CDKN2A", "CDKN2B") %in% poor_prognostic_factor) == TRUE){
  poor_prognostic_factor =
    poor_prognostic_factor[-which(poor_prognostic_factor %in% "CDKN2B")] 
}
poor_prognostic_factor = poor_prognostic_factor[1:4]


g = NULL
for(i in 1:length(gene_pattern)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    for(k in 1:length(gene_pattern[[i]])){
      Data_survival$gene_select[[j]] = any(Data_survival$gene_select[[j]],
          (bitwAnd(gene_pattern[[i]][[k]],
                   Data_survival$target_gene_count[j]) ==
             gene_pattern[[i]][[k]]))
    }
  }
  mut_gene = ""
  for(l in 1:length(gene_pattern[[i]])){
    for(m in 1:length(target_gene)){
      if(bitwAnd(2^(m-1), gene_pattern[[i]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  file.create(paste(OutDirName, mut_gene, ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(OutDirName, mut_gene, ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_gene_SPT_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()


g = NULL
for(i in 1:length(target_pathway)){
  Data_report_tmp = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[[i]]][[1]])
  ID_path = unique(Data_report_tmp$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival = Data_survival %>%
    dplyr::mutate(gene_select = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  mut_gene = paste(target_pathway[i], " mutation", sep="")
  file.create(paste(OutDirName, target_pathway[i], ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(OutDirName, target_pathway[i], ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_pathway_SPT_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

# comparing among prognostic scores; auto gene selection
Data_survival$gene_select = 0
for(i in 1:length(poor_prognostic_factor)){
  ID_tmp = unique((Data_report %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー == poor_prognostic_factor[[i]] & C.CAT調査結果.エビデンス.エビデンスレベル == "F"))$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] = 
    Data_survival[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp, "gene_select"] + 1
}
Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    gene_select > 2 ~ 2,
    TRUE ~ gene_select
  ))
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = "No muts in selected genes"
    name_2 = "Single mut in selected genes"
    name_3 = "Multiple muts in selected genes"
    name_1_short = "No"
    name_2_short = "Single"
    name_3_short = "Multiple"

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/2, 0.5, 1 - 0.025/2))
  median_os_summary_weibull_bias_0_1_ = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025, 0.5, 1 - 0.025))
  median_os_summary_weibull_bias_1_2_ = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025, 0.5, 1 - 0.025))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS\n",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Dunnett-like correction, ",
                       paste(unlist(poor_prognostic_factor), collapse = ","),
                       " oncogenic muts)\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (97.5% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_prognostic_scoring_auto.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(name_1, ", Unadjusted"),
                      paste(name_2, ", Unadjusted"),
                      paste(name_3, ", Unadjusted"),
                      paste(name_1, ", Adjusted for Delayed Entry"),
                      paste(name_2, ", Adjusted for Delayed Entry"),
                      paste(name_3, ", Adjusted for Delayed Entry"))
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS\n",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference, ",
                       paste(unlist(poor_prognostic_factor), collapse = ","),
                       " oncogenic muts)\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1_[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1_[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1_[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2_[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2_[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2_[[3]]),
            ") days, median (95% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_prognostic_scoring_auto_2.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. Patients with selected disease, SPT-death, Characteristics**") %>%
  bold_labels()
table_tmp

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_option,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment option exsisted**") %>%
  modify_caption("**Table 9. First chemotherapy to final observation - Expert panel option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 10. Enrollment to final observation - Recommended treatment done**") %>%
  bold_labels()
}

Data_summary_treat = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) %>%
  dplyr::filter(EP_option == 1)

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
survival_simple_ = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(survival_simple_[[1]][[1]] != survival_simple[[1]][[1]] & length(survival_simple_) == length(survival_simple)){
Data_summary_treat %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 2''. First chemotherapy to final observation - Recommended treatment done, only patients with treatment option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~diagnosis,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = diagnosis,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 3,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 3)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_caption("**Table 11. First chemotherapy to final observation - diagnosis**") %>%
  bold_labels()
}
```

```{r BREAST, eval = (Cancer == "Breast" & Subanalysis == 1), echo = (Source_code == 1 & Cancer == "Breast" & Subanalysis == 1), warning=FALSE}

# analysis for common oncogenic mutations
g = NULL
i=1
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    for(k in 1:length(gene_pattern[[i]])){
      Data_survival$gene_select[[j]] = any(Data_survival$gene_select[[j]],
          (bitwAnd(gene_pattern[[i]][[k]],
                   Data_survival$target_gene_count[j]) ==
             gene_pattern[[i]][[k]]))
    }
  }
  mut_gene = ""
  for(l in 1:length(gene_pattern[[i]])){
    for(m in 1:length(target_gene)){
      if(bitwAnd(2^(m-1), gene_pattern[[i]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  file.create(paste("top_", i, "_gene.tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
 
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
    
    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  #rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  #mcmc_rhat(rhat(stan_weibull_survival_model_fit))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))
  print(g)
  pdf(file = paste(Cancer, "_target_mutation.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
    g=ggsurvplot(
      fit = list(simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    )
  print(g)
  #pdf(file = paste(Cancer, "_target_mutation_simple.pdf", sep=""), width = 10, height = 10)
  #print(g, newpage = FALSE)
  #dev.off()
  }
}
```

```{r pancreas, eval = (Cancer == "Pancreas" & Subanalysis == 1), echo = (Source_code == 1 & Cancer == "Pancreas" & Subanalysis == 1), warning=FALSE}
## Survival analysis considering left truncation
Drug_GP = c("Gemcitabine", "nab-paclitaxel")
Drug_FOLFIRINOX = c("Oxaliplatin", "Fluorouracil", "Irinotecan")

Data_survival = Data_case_target %>%
  dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))

Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日))
Data_survival = Data_survival %>%
  dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))
Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
  Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

  
Data_tmp_pre = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.投与開始日,
                症例.EP前レジメン情報.投与終了日,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_tmp_post = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.投与開始日,
                症例.EP後レジメン情報.投与終了日,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "")
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::mutate(drug_group = case_when(
    drug_name ==  "Gemcitabine" ~ "Gem",
    drug_name ==  "nab-paclitaxel" ~ "Npt",
    drug_name ==  "Oxaliplatin" ~ "Oxa",
    drug_name ==  "Fluorouracil" ~ "Flu",
    drug_name ==  "Irinotecan" ~ "Iri",
    TRUE ~ "Oth"))

Data_survival = Data_survival %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日))
Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$First_SPT = ""
drug_seq = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
ID_exclude = NULL
for(i in 1:length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)){
  Data_tmp = Data_drug_sequence[Data_drug_sequence$C.CAT調査結果.基本項目.ハッシュID == Data_survival$C.CAT調査結果.基本項目.ハッシュID[i],]
  drug_seq[i] = paste(Data_tmp$drug_group, collapse = "")
  if(length(Data_tmp[[1]]) > 2){
    if(
     Data_tmp$drug_start_date[[1]] == Data_tmp$drug_start_date[[2]] &
     str_sub(drug_seq[i],1,6) %in% c("GemNpt", "NptGem")){
      Data_survival$First_SPT[i] = "GEM+nab-Paclitaxel"
    }
  }
  if(length(Data_tmp[[1]]) >= 3){
    if(Data_tmp$drug_start_date[[1]] == Data_tmp$drug_start_date[[2]] &
     Data_tmp$drug_start_date[[1]] == Data_tmp$drug_start_date[[3]] &
     str_sub(drug_seq[i],1,9) %in% c("OxaFluIri", "OxaIriFlu", "FluOxaIri", "FluIriOxa", "IriOxaFlu", "IriFluOxa")){
      Data_survival$First_SPT[i] = "FOLFIRINOX"
    }
  }
  if(length(Data_tmp[[1]]) >= 4){
    if(Data_tmp$drug_start_date[[1]] == Data_tmp$drug_start_date[[2]] &
     Data_tmp$drug_start_date[[1]] == Data_tmp$drug_start_date[[3]] &
     Data_tmp$drug_start_date[[1]] == Data_tmp$drug_start_date[[3]] &
     str_replace_all(str_sub(drug_seq[i],1,12), pattern="Oth", replacement = "") %in% c("OxaFluIri", "OxaIriFlu", "FluOxaIri", "FluIriOxa", "IriOxaFlu", "IriFluOxa")){
      Data_survival$First_SPT[i] = "FOLFIRINOX"
    }
  }
}

Data_survival = Data_survival %>%
  dplyr::filter(First_SPT != "")

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "SPT to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "SPT to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "SPT to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "SPT to entry 3-4 year",
    TRUE ~ "SPT to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
    TRUE ~ "SPT to entry later"))

# GEM+nab-Paclitaxel vs FOLFIRINOX
traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ First_SPT, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  Data_survival_tmp_pos = Data_survival %>%
    dplyr::filter(First_SPT == "FOLFIRINOX")
  Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp$time_palliative_enroll =
    Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
  
  Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
  
  Data_survival_tmp_neg = Data_survival %>%
    dplyr::filter(First_SPT != "FOLFIRINOX")
  Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp3$time_palliative_enroll =
    Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
  
  Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
  Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
 
    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          First_SPT == "FOLFIRINOX" ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          First_SPT == "FOLFIRINOX" ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          First_SPT != "FOLFIRINOX" ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          First_SPT != "FOLFIRINOX" ~ "SPT to entry later"))

  if(sum((Data_survival %>% dplyr::filter(First_SPT == "FOLFIRINOX" & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(First_SPT != "FOLFIRINOX" & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(First_SPT == "FOLFIRINOX" & delay_1 != "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(First_SPT != "FOLFIRINOX" & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$First_SPT != "FOLFIRINOX"),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$First_SPT != "FOLFIRINOX")[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$First_SPT != "FOLFIRINOX")[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

    #rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
    #mcmc_rhat(rhat(stan_weibull_survival_model_fit))

    stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
    stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
   stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
    median_os_list_weibull_total = NULL
    median_os_list_weibull_total_exp = NULL
    for(j in 1:8000){
    idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
      median_os_list_weibull_total = c(median_os_list_weibull_total,
        median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
      median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
        median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    }
    median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
  
    median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
    median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
    median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
    yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
    yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
    yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
    yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  
    Data_survival$factor_neg = yhat_weibull_sort_average_total
    Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
    Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
    Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

    traditional_fit = survfit(Surv(event = censor,
                                   time = time_palliative_final) ~ First_SPT,
                                   data = Data_survival)
    simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                   time = factor_neg) ~ 1, 
                                   data = Data_survival)
    simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                   time = factor_pos) ~ 1, 
                                   data = Data_survival)
    
    g=ggsurvplot(
        fit = list(traditional_fit = traditional_fit,
                   simulation_fit_neg = simulation_fit_neg,
                   simulation_fit_pos = simulation_fit_pos),
        combine = TRUE,
        data = Data_survival,
        xlab = "Time from SPT to final observation (days)",
        ylab = "Survival Probability",
        censor = TRUE,
        conf.int = FALSE,
        pval = FALSE,
        risk.table = TRUE,
        risk.table.y.text = FALSE,
        tables.theme = clean_theme(), 
        legend = c(0.8,0.8),
        xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
        break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
        legend.labs = c("FOLFIRINOX first, Unadjusted",
                        "GEM+nab-Paclitaxel first, Unadjusted",
                        "FOLFIRINOX first, Adjusted for Delayed Entry",
                        "GEM+nab-Paclitaxel first, Adjusted for Delayed Entry")
      ) +   
      labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                  "Median OS ",
                 as.integer(summary(traditional_fit)[[18]][[13]])," (",
                 as.integer(summary(traditional_fit)[[18]][[15]]),"-",
                 as.integer(summary(traditional_fit)[[18]][[17]]),") (FOLFIRINOX, unadj.)/",
                 as.integer(summary(traditional_fit)[[18]][[14]]), " (",
                 as.integer(summary(traditional_fit)[[18]][[16]]),"-",
                 as.integer(summary(traditional_fit)[[18]][[18]]),") (GEM+nab-Pac, unadj.)/\n",
                 as.integer(median_os_summary_weibull_total[[2]]), 
                 " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
                 as.integer(median_os_summary_weibull_total[[3]]), ") (FOLFIRINOX, adj.)/",
                 as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
                 as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
                 as.integer(median_os_summary_weibull_total_exp[[3]]),
                 ") (GEM+nab-Pac, adj.) days",
                 sep=""),
        subtitle = paste("Survival difference: ",
              as.integer(median_os_summary_weibull_bias[[2]]), " (",
              as.integer(median_os_summary_weibull_bias[[1]]), "-", 
              as.integer(median_os_summary_weibull_bias[[3]]), 
              ") days, median (95% CI)",
              sep=""))
    #print(g)
    pdf(file = paste(OutDirName, Cancer, "_SPT_to_death_GEMNAB_FOLFIRINOX.pdf", sep=""), width = 10, height = 10)
    print(g, newpage = FALSE)
    dev.off()
}
}
```

```{r prostate, eval = (Cancer == "Prostate" & Subanalysis == 1), echo = (Source_code == 1 & Cancer == "Prostate" & Subanalysis == 1), warning=FALSE}
Drug_vintage = c("Bicalutamide", "Flutamide")
Drug_taxane = c("Docetaxel", "Cabazitaxel")
Drug_ARAT = c("Abiraterone", "Enzalutamide", "Apalutamide", "Darolutamide")
Gene_DDR = c("ATM", "ATR", "BLM", "BRCA1", "BRCA2", "EME1", "EXO1",
             "FANCF", "FANCM", "FANCN", "GEN1", "MRE11", "NBS1",
             "RAD50", "RAD51", "RAD52", "RAD54", "RACQ4", "RPA",
             "WRN", "XRCC2", "XRCC3")

Data_survival = Data_case_target %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))

Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))

Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_tmp_pre = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.投与開始日,
                症例.EP前レジメン情報.投与終了日,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_tmp_post = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.投与開始日,
                症例.EP後レジメン情報.投与終了日,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "")
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::mutate(drug_group = case_when(
    drug_name %in% Drug_vintage ~ "Vin",
    drug_name %in% Drug_ARAT ~ "Ara",
    drug_name %in% Drug_taxane ~ "Tax",
    TRUE ~ "Oth"))
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::mutate(drug_group_tax = case_when(
    drug_name %in% Drug_vintage ~ "Vin",
    drug_name %in% Drug_ARAT ~ "Ara",
    drug_name == "Docetaxel" ~ "Doc",
    drug_name == "Cabazitaxel"  ~ "Cab",
    TRUE ~ "Oth"))
ID_vintage = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "Vin"))$C.CAT調査結果.基本項目.ハッシュID)
ID_ARAT = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "Ara"))$C.CAT調査結果.基本項目.ハッシュID)
ID_taxane = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "Tax"))$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_survival %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_vintage) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_taxane) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ARAT) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日))
Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$target_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_target_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:length(target_gene)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.変異情報.変異内容  == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  } else{
    for(k in 1:length(target_gene)){
      if(target_gene[[k]][2] == "WT"){
        Data_survival$target_gene_count[i] = bitwOr(
          Data_survival$target_gene_count[i], 2^(k-1))
      }
    }
  }
}

if(sum(Data_survival$target_gene_count) == 0){
  target_gene = list(c(oncogenic_genes[1,1], "F"))
  gene_pattern = list(c(1))
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.変異情報.変異内容  == target_gene[[i]][2]),
      Data_target_gene)
  }
  Data_target_gene = Data_target_gene %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容 ,
                    .keep_all=TRUE)
  
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[i]][2]) %>%
      dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                      C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容 ,
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT,
                    .keep_all=TRUE),
      Data_target_gene)
  }
  Data_survival$target_gene_count = 0
  ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
  for(i in 1:length(ID_target_gene)){
    tmp = Data_target_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
    if(dplyr::count(tmp)[[1]] > 0){
      for(j in 1:dplyr::count(tmp)[[1]]){
        for(k in 1:length(target_gene)){
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.変異情報.変異内容  == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
        }
      }
    } else{
      for(k in 1:length(target_gene)){
        if(target_gene[[k]][2] == "WT"){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

Data_survival$oncogenic_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_major_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:min(length(oncogenic_genes$all_patients), gene_number)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == oncogenic_genes$gene_mutation[[k]]){
          Data_survival$oncogenic_gene_count[i] = bitwOr(
            Data_survival$oncogenic_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

drug_seq = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
drug_seq_tax = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
final_vin_date = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
ID_exclude = NULL
for(i in 1:length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)){
  Data_tmp = Data_drug_sequence[Data_drug_sequence$C.CAT調査結果.基本項目.ハッシュID == Data_survival$C.CAT調査結果.基本項目.ハッシュID[i],]
  drug_seq[i] = paste(Data_tmp$drug_group, collapse = "")
  drug_seq_tax[i] = paste(Data_tmp$drug_group_tax, collapse = "")
  if(rev(Data_tmp[Data_tmp$drug_group == "Vin", "drug_end_date"])[[1]] == "" |
     is.na(rev(Data_tmp[Data_tmp$drug_group == "Vin", "drug_end_date"])[[1]])){
    ID_exclude = c(ID_exclude, Data_survival$C.CAT調査結果.基本項目.ハッシュID[i])
    final_vin_date[i] = "no data"
  } else{
    final_vin_date[i] = rev(Data_tmp[Data_tmp$drug_group == "Vin", "drug_end_date"])[[1]]
  }
}

drug_seq = str_replace_all(drug_seq,
          pattern="Oth",
          replacement = "")
drug_seq = str_replace_all(drug_seq,
          pattern="VinVin",
          replacement = "Vin")
drug_seq = str_replace_all(drug_seq,
          pattern="VinVin",
          replacement = "Vin")
drug_seq = str_replace_all(drug_seq,
          pattern="VinVin",
          replacement = "Vin")
drug_seq = str_replace_all(drug_seq,
          pattern="AraAra",
          replacement = "Ara")
drug_seq = str_replace_all(drug_seq,
          pattern="AraAra",
          replacement = "Ara")
drug_seq = str_replace_all(drug_seq,
          pattern="AraAra",
          replacement = "Ara")
drug_seq = str_replace_all(drug_seq,
          pattern="TaxTax",
          replacement = "Tax")
drug_seq = str_replace_all(drug_seq,
          pattern="TaxTax",
          replacement = "Tax")
drug_seq = str_replace_all(drug_seq,
          pattern="TaxTax",
          replacement = "Tax")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="Oth",
          replacement = "")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="VinVin",
          replacement = "Vin")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="VinVin",
          replacement = "Vin")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="VinVin",
          replacement = "Vin")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="AraAra",
          replacement = "Ara")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="AraAra",
          replacement = "Ara")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="AraAra",
          replacement = "Ara")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="DocDoc",
          replacement = "Doc")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="DocDoc",
          replacement = "Doc")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="DocDoc",
          replacement = "Doc")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="CabCab",
          replacement = "Cab")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="CabCab",
          replacement = "Cab")
drug_seq_tax = str_replace_all(drug_seq_tax,
          pattern="CabCab",
          replacement = "Cab")
Data_survival$drug_seq = drug_seq
Data_survival$drug_seq_tax = drug_seq_tax
Data_survival$final_vin_date = final_vin_date
if(length(ID_exclude > 0)){
  Data_survival = Data_survival %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_exclude)
}

Data_survival = Data_survival %>%
  dplyr::filter(str_sub(drug_seq, 1, 3) == "Vin")

Data_survival = Data_survival %>%
  dplyr::mutate(VAT_VTA = case_when(
    str_sub(drug_seq, 1, 9) == "VinAraTax" ~ "Vintage_ARAT_Taxane",
    str_sub(drug_seq, 1, 9) == "VinTaxAra" ~ "Vintage_Taxane_ARAT",
    TRUE ~ "other"))

Data_survival = Data_survival %>%
  dplyr::filter(VAT_VTA != "other")

Data_survival = Data_survival %>%
  dplyr::mutate(final_vin_date = as.Date(Data_survival$final_vin_date))
Data_survival = Data_survival %>%
  dplyr::mutate(time_vintage_enroll =
                  as.integer(difftime(Data_survival$症例.管理情報.登録日,
                                      Data_survival$final_vin_date,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_vintage_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$final_vin_date,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_vintage_enroll/365)

Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final = Data_survival$time_vintage_final)
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll = Data_survival$time_vintage_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "Vintage chemotherapy to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "Vintage chemotherapy to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "Vintage chemotherapy to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "Vintage chemotherapy to entry 3-4 year",
    TRUE ~ "Vintage chemotherapy to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "Vintage chemotherapy to entry early",
    TRUE ~ "Vintage chemotherapy to entry later"))

# Survival analysis for delayed entry
traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$delay)) == 5){
print(traditional_fit)
ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = paste("Enrollment timing:",
                      c("Vintage chemotherapy to entry <=1 year",
                        "Vintage chemotherapy to entry 1-2 year",
                        "Vintage chemotherapy to entry 2-3 year",
                        "Vintage chemotherapy to entry 3-4 year",
                        "Vintage chemotherapy to entry >4 year")
                        , sep="")
) + 
  labs(title = paste(Cancer, "Delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
print(traditional_fit)
ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
) + 
  labs(title = paste(Cancer, "Delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
}

# Survival analysis for delayed entry, < median or not
traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$delay_1)) == 2){
print(traditional_fit)
ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = paste("Enrollment timing:",
                      c("Vintage chemotherapy to entry early",
                        "Vintage chemotherapy to entry later")
                        , sep="")
) + 
  labs(title = paste(Cancer, "Delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
}

# survival curve: from the last vintage chemotherapy to enrollment
Data_survival_tmp = Data_survival %>% dplyr::filter(
  time_palliative_enroll >= time_10)
Data_survival_tmp$time_palliative_enroll =
  Data_survival_tmp$time_palliative_enroll - (time_10 - 1)

Median_entry = median(Data_survival_tmp$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "Vintage chemotherapy to entry early",
    TRUE ~ "Vintage chemotherapy to entry later"))

stan_weibull_survival_model_data <-
    list(
        Median = as.integer(Median_entry),
        Nobs = sum(Data_survival$censor == 1),
        Ncen = sum(Data_survival$censor == 0),
        Ntot = length(Data_survival$censor),
        M_bg = 1,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        Xobs_bg = matrix(as.numeric(Data_survival$delay_1 == "Vintage chemotherapy to entry later")[Data_survival$censor == 1]),
        Xcen_bg = matrix(as.numeric(Data_survival$delay_1 == "Vintage chemotherapy to entry later")[Data_survival$censor == 0])
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1)

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_bef_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
    tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_aft_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
    tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_ear <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
    tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_lat <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
    tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
    dplyr::select(-key) 


median_os_list_weibull_total_exp = NULL
median_os_list_weibull_bef_exp = NULL
median_os_list_weibull_aft_exp = NULL
median_os_list_weibull_ear = NULL
median_os_list_weibull_lat = NULL

for(i in 1:8000){
  idx = seq(i, length(stan_weibull_survival_model_draws_total_exp$yhat_total_exp), by=8000)
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_bef_exp = c(median_os_list_weibull_bef_exp,
      median(stan_weibull_survival_model_draws_bef_exp[idx,]$yhat_bef_exp, na.rm = T))
    median_os_list_weibull_aft_exp = c(median_os_list_weibull_aft_exp,
      median(stan_weibull_survival_model_draws_aft_exp[idx,]$yhat_aft_exp, na.rm = T))
    median_os_list_weibull_ear = c(median_os_list_weibull_ear,
      median(stan_weibull_survival_model_draws_ear[idx,]$yhat_ear, na.rm = T))
    median_os_list_weibull_lat = c(median_os_list_weibull_lat,
      median(stan_weibull_survival_model_draws_lat[idx,]$yhat_lat, na.rm = T))
}

median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bef_exp = quantile(median_os_list_weibull_bef_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_aft_exp = quantile(median_os_list_weibull_aft_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_ear = quantile(median_os_list_weibull_ear, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_lat = quantile(median_os_list_weibull_lat, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_bef_exp = sort(stan_weibull_survival_model_draws_bef_exp$yhat_bef_exp)
yhat_weibull_sort_aft_exp = sort(stan_weibull_survival_model_draws_aft_exp$yhat_aft_exp)
yhat_weibull_sort_ear = sort(stan_weibull_survival_model_draws_ear$yhat_ear)
yhat_weibull_sort_lat = sort(stan_weibull_survival_model_draws_lat$yhat_lat)
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_bef_exp = yhat_weibull_sort_bef_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_aft_exp = yhat_weibull_sort_aft_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_ear = yhat_weibull_sort_ear[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_lat = yhat_weibull_sort_lat[1:length(Data_survival$censor) * 8000]

Data_survival$left_adj = yhat_weibull_sort_average_total_exp
Data_survival$weibull_bef_exp = yhat_weibull_sort_average_bef_exp
Data_survival$weibull_ear = yhat_weibull_sort_average_ear
Data_survival$weibull_lat = yhat_weibull_sort_average_lat

traditional_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival)
traditional_fit_2 = survfit(Surv(event = rep(1,length(Data_survival_tmp$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival_tmp)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_bef_exp)),
                               time = weibull_bef_exp) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit_1 = traditional_fit_1,
             traditional_fit_2 = traditional_fit_2,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  #data = Data_survival,
  xlab = "Time from Vintage chemotherapy to CGP (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
  legend.labs = c("raw data",
                  "middle data",
                  "Weibull curve")
) + labs(title = paste(Cancer, " ", traditional_fit_1[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit_1)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_1)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_1)[[17]][[9]]),") (raw)/",
             as.integer(summary(traditional_fit_2)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_2)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_2)[[17]][[9]]),") (middle)\n",
             as.integer(median_os_summary_weibull_bef_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_bef_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_bef_exp[[3]]),
             ") (Weibull) days",
             sep=""))
print(g)
# pdf(file = paste(Cancer, "_weibull_SPT_to_CGP.pdf", sep=""), width = 10, height = 10)
# print(g, newpage = FALSE)
# dev.off()

traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$weibull_ear)),
                               time = weibull_ear) ~ 1, 
                               data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_lat)),
                               time = weibull_lat) ~ 1, 
                               data = Data_survival)


g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_1 = simulation_fit_1,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data",
                  "CGP enrollment at early timing, Weibull curve",
                  "CGP enrollment at late timing, Weibull curve")
) + labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (raw, early)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (raw, late)/\n",
             as.integer(median_os_summary_weibull_ear[[2]]), 
             " (", as.integer(median_os_summary_weibull_ear[[1]]), "-",
             as.integer(median_os_summary_weibull_ear[[3]]), ") (Weibull, early)/",
             as.integer(median_os_summary_weibull_lat[[2]]), " (",
             as.integer(median_os_summary_weibull_lat[[1]]), "-",
             as.integer(median_os_summary_weibull_lat[[3]]),
             ") (Weibull, late) days",
             sep=""))
print(g)

independence_check = with(Data_survival, cKendall(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  trans = "linear"))


traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ 1, 
                               data = Data_survival)
delayed_entry_fit = survfit(Surv(event = censor,
                                 time = time_palliative_enroll,
                                 time2 = time_palliative_final) ~ 1, 
                                 data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$left_adj)),
                               time = left_adj) ~ 1, 
                               data = Data_survival)
tmp_num = 18
if(length(summary(delayed_entry_fit)[[18]]) == 1){
  tmp_num = 17
}

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             delayed_entry_fit = delayed_entry_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from Vintage chemotherapy to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Number at risk adjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + 

  labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[7]]), " (",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[8]]),"-",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[9]]),") (risk adj.)/\n",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(delayed_entry_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (risk adj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
print(g)


# Survival analysis for enrollment timing
traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ enrollment, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$enrollment)) == 5){
  print(traditional_fit)
  legend_enroll =  c("2019-06-01 to 2020-06-30",
                     "2020-07-01 to 2021-06-30",
                     "2021-07-01 to 2021-12-31",
                     "2022-01-01 to 2022-06-30",
                     "2022-07-01 to 2023-02-20")
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from Vintage chemotherapy to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.90),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = legend_enroll
  ) + 
    labs(title = paste(Cancer, "Enrollment timing for", 
                       sum((traditional_fit)[[1]]), "patients"))
  print(g)
  # pdf(file = paste(Cancer, "_entry_period.pdf", sep=""), width = 10, height = 10)
  # print(g, newpage = FALSE)
  # dev.off()
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  print(traditional_fit)
  ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from Vintage chemotherapy to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.85),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100
  ) + labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((traditional_fit)[[1]]), "patients"))
}

Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    VAT_VTA == "Vintage_ARAT_Taxane" ~ 0,
    VAT_VTA == "Vintage_Taxane_ARAT" ~ 1))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  Data_survival_tmp_pos = Data_survival %>%
    dplyr::filter(gene_select == TRUE)
  Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp$time_palliative_enroll =
    Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
  
  Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

  Data_survival_tmp_neg = Data_survival %>%
    dplyr::filter(gene_select == FALSE)
  Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp3$time_palliative_enroll =
    Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
  
  Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
  Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

  Data_survival = Data_survival %>%
    dplyr::mutate(delay_1 = case_when(
      Data_survival$time_palliative_enroll <= Median_entry_pos &
        gene_select == TRUE ~ "Vintage chemotherapy to entry early",
      Data_survival$time_palliative_enroll > Median_entry_pos &
        gene_select == TRUE ~ "Vintage chemotherapy to entry later",
      Data_survival$time_palliative_enroll <= Median_entry_neg &
        gene_select == FALSE ~ "Vintage chemotherapy to entry early",
      Data_survival$time_palliative_enroll > Median_entry_neg &
        gene_select == FALSE ~ "Vintage chemotherapy to entry later"))

  if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "Vintage chemotherapy to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "Vintage chemotherapy to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "Vintage chemotherapy to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "Vintage chemotherapy to entry later"))$censor) >= 1){

    stan_weibull_survival_model_data <-
      list(
        Median_pos = as.integer(Median_entry_pos),
        Median_neg = as.integer(Median_entry_neg),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 2,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        xfactor = as.numeric(Data_survival_tmp$gene_select),
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "Vintage chemotherapy to entry later")[Data_survival$censor == 1],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "Vintage chemotherapy to entry later")[Data_survival$censor == 0],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull_factor.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                         "yhat_total", "yhat_factor_total"))

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
    dplyr::select(-key) 

median_os_list_weibull_total = NULL
median_os_list_weibull_total_exp = NULL
for(j in 1:8000){
idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
  median_os_list_weibull_total = c(median_os_list_weibull_total,
    median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
  median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
    median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
}
median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

Data_survival$factor_neg = yhat_weibull_sort_average_total
Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select,
                               data = Data_survival)
simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                               time = factor_neg) ~ 1, 
                               data = Data_survival)
simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                               time = factor_pos) ~ 1, 
                               data = Data_survival)

ggsurvplot(
    fit = list(traditional_fit = traditional_fit,
               simulation_fit_neg = simulation_fit_neg,
               simulation_fit_pos = simulation_fit_pos),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from vintage chemotherapy to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = c("ARAT->Taxane, Unadjusted",
                    "Taxane->ARAT, Unadjusted",
                    "ARAT->Taxane, Adjusted for Delayed Entry",
                    "Taxane->ARAT, Adjusted for Delayed Entry")
  ) +   
  labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (ARAT->Taxane, unadj.)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (Taxane->ARAT, unadj.)/\n",
             as.integer(median_os_summary_weibull_total[[2]]), 
             " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
             as.integer(median_os_summary_weibull_total[[3]]), ") (ARAT->Taxane, adj.)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (Taxane->ARAT, adj.) days",
             sep=""),
    subtitle = paste("Survival difference with treatment sequence: ",
          as.integer(median_os_summary_weibull_bias[[2]]), " (",
          as.integer(median_os_summary_weibull_bias[[1]]), "-", 
          as.integer(median_os_summary_weibull_bias[[3]]), 
          ") days, median (95% CI)",
          sep=""))
  }
}


Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_option,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment sequence**") %>%
  modify_caption("**Table X. Vintage chemotherapy to final observation - treatment sequence**") %>%
  bold_labels()
}
```

```{r LUAD_ICI, eval = (Cancer %in% c("Lung", "LUAD", "LUSC") & Subanalysis == 1), echo = (Source_code == 1 & Cancer %in% c("Lung", "LUAD", "LUSC") & Subanalysis == 1), warning=FALSE}
## Survival analysis considering left truncation
ICI_drugs = c("Atezolizumab", "Nivolumab", "Pembrolizumab", "Ipilimumab", "Durvalumab")
Platinum_drugs = c("Carboplatin", "Cisplatin")
Gene_target_drugs = c("Osimertinib", "Afatinib", "Erlotinib", "Crizotinib", "Gefitinib", "Lorlatinib", 
                      "Alectinib", "Ceritinib", "Tepotinib", "Capmatinib", "Entrectinib", "Larotrectinib", "Dabrafenib",
                      "Trametinib", "Trastuzumab Deruxtecan", "Dacomitinib", "Brigatinib", "Sotorasib")
EGFR_ALK = c("EGFR", "ALK")
#Driver_genes = c("KRAS", "MET", "FGFR1", "FGFR2", "FGFR3", "FGFR4", "RET", "BRAF", "ERBB2", "ROS1",
#                 "NTRK1", "NTRK2", "NTRK3", "PIK3CA", "PTEN", "AKT1", "AKT2", "AKT3", "MAPK1", "MAP2K1",
#                 "VEGFA", "KIT", "NRAS")
Driver_genes = c("MET", "BRAF", "ERBB2", "RET", "ROS1", "NTRK1", "NTRK2", "NTRK3")
Driver_genes_1 =c("MET")
Driver_genes_2 =c("BRAF", "ERBB2")
Driver_genes_3 =c("RET", "ROS1", "NTRK1", "NTRK2", "NTRK3")

Data_survival = Data_case_target %>%
  dplyr::filter(!症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日))

ID_ICI_used_periop = NULL
for(i in ICI_drugs){
  ID_ICI_used_periop = c(ID_ICI_used_periop,
    (unique((Data_survival %>% dplyr::filter(
         str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                    pattern=i))
    )$C.CAT調査結果.基本項目.ハッシュID)))
}

ID_ICI_used_periop = unique(ID_ICI_used_periop)

Data_survival = Data_case_target %>%
  dplyr::filter(!症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日))

ID_target_used_periop = NULL
for(i in Gene_target_drugs){
  ID_target_used_periop = c(ID_target_used_periop,
    (unique((Data_survival %>% dplyr::filter(
         str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                    pattern=i))
    )$C.CAT調査結果.基本項目.ハッシュID)))
}

ID_target_used_periop = unique(ID_target_used_periop)

Data_survival = Data_case_target %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI_used_periop) %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_target_used_periop) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))

Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          " & final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))

Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_tmp_pre = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.投与開始日,
                症例.EP前レジメン情報.投与終了日,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_tmp_post = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.投与開始日,
                症例.EP後レジメン情報.投与終了日,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "")

for(i in ICI_drugs){
  ICI_drugs = c(ICI_drugs,
    (unique((Data_survival %>% dplyr::filter(
         str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                    pattern=i))
    )$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.)))
  ICI_drugs = c(ICI_drugs,
    (unique((Data_survival %>% dplyr::filter(
         str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                    pattern=i))
    )$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.)))
}

ICI_drugs = unique(ICI_drugs)

for(i in Gene_target_drugs){
  Gene_target_drugs = c(Gene_target_drugs,
    (unique((Data_survival %>% dplyr::filter(
         str_detect(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                    pattern=i))
    )$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.)))
  Gene_target_drugs = c(Gene_target_drugs,
    (unique((Data_survival %>% dplyr::filter(
         str_detect(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                    pattern=i))
    )$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.)))
}

Gene_target_drugs = unique(Gene_target_drugs)

Data_drug_sequence = Data_drug_sequence %>%
  dplyr::mutate(drug_group = case_when(
    drug_name %in% ICI_drugs ~ "Ici",
    drug_name %in% Gene_target_drugs ~ "Tar",
    TRUE ~ "Oth"))
ID_ICI = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "Ici"))$C.CAT調査結果.基本項目.ハッシュID)
ID_TAR = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "Tar"))$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_survival %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日))
Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "SPT to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "SPT to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "SPT to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "SPT to entry 3-4 year",
    TRUE ~ "SPT to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)
Median_entry_raw = Median_entry

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
    TRUE ~ "SPT to entry later"))

Data_survival$First_SPT = "Oth"
drug_seq = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
for(i in 1:length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)){
  Data_tmp = Data_drug_sequence[Data_drug_sequence$
                C.CAT調査結果.基本項目.ハッシュID == Data_survival$
                  C.CAT調査結果.基本項目.ハッシュID[i],]
  if(length(Data_tmp$C.CAT調査結果.基本項目.ハッシュID > 0)){
    Data_tmp = Data_tmp %>% dplyr::filter(
      drug_start_date == Data_tmp$drug_start_date[[1]]
    )
    drug_seq[i] = paste(Data_tmp$drug_group, collapse = "")
    if(str_detect(drug_seq[i], "Ici") & !str_detect(drug_seq[i], "Tar")){
      Data_survival$First_SPT[i] = "ICI"
    }
    if(str_detect(drug_seq[i], "Tar") & !str_detect(drug_seq[i], "Ici")){
      Data_survival$First_SPT[i] = "TAR"
    }
  }
}

drug_seq = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
drug_seq_tax = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
for(i in 1:length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)){
  Data_tmp = Data_drug_sequence[Data_drug_sequence$C.CAT調査結果.基本項目.ハッシュID == Data_survival$C.CAT調査結果.基本項目.ハッシュID[i],]
  drug_seq[i] = paste(Data_tmp$drug_group, collapse = "")
  drug_seq_tax[i] = paste(Data_tmp$drug_group_tax, collapse = "")
}

drug_seq = str_replace_all(drug_seq,
          pattern="Oth",
          replacement = "")
drug_seq = str_replace_all(drug_seq,
          pattern="IciIci",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="IciIci",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="IciIci",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="IciIci",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")

Data_survival$drug_seq = drug_seq

Data_survival = Data_survival %>%
  dplyr::mutate(Treatment_seq = case_when(
    str_sub(drug_seq, 1, 3) == "Ici" ~ "ICI_first",
    str_sub(drug_seq, 1, 3) == "Tar" ~ "Target_therapy_first",
    TRUE ~ "other"))

Data_survival = Data_survival %>%
  dplyr::mutate(Target_done = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% ID_TAR ~ "done",
    TRUE ~ ""))
Data_survival = Data_survival %>%
  dplyr::mutate(ICI_done = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI ~ "done",
    TRUE ~ ""))

Data_survival_save = Data_survival

ID = Data_survival$C.CAT調査結果.基本項目.ハッシュID

Data_case_tmp = Data_case_target %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID)

tmp_report = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.基本項目.ハッシュID %in% ID)
tmp_report_drug = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル %in% c( "E", "C", "D", "B", "A") &
                  C.CAT調査結果.基本項目.ハッシュID %in% ID) %>%
    dplyr::mutate(tmp2 = paste(tmp, C.CAT調査結果.エビデンス.薬剤, C.CAT調査結果.エビデンス.エビデンスレベル, sep="_"))
tmp_case_pre = Data_case_tmp %>% dplyr::filter(
                  !is.na(症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) &
                  !is.na(症例.EP前レジメン情報.化学療法レジメン名称)) %>%
              dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
              dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                            症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                            症例.EP前レジメン情報.薬剤名.YJ一般名JP.,
                            症例.EP前レジメン情報.薬剤名.入力商品名.,
                            症例.EP前レジメン情報.実施目的.名称.,
                            症例.EP前レジメン情報.化学療法レジメン名称,
                            症例.EP前レジメン情報.投与開始日,
                            症例.EP前レジメン情報.治療ライン.名称.
                            ) %>%
              dplyr::distinct()
tmp_case_post = Data_case_tmp %>% dplyr::filter(
                  !is.na(症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) &
                  !is.na(症例.EP後レジメン情報.化学療法レジメン名称)) %>%
              dplyr::arrange(症例.EP後レジメン情報.投与開始日) %>%
              dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                            症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                            症例.EP後レジメン情報.薬剤名.YJ一般名JP.,
                            症例.EP後レジメン情報.薬剤名.入力商品名.,
                            症例.EP後レジメン情報.治療方針.名称.,
                            症例.EP後レジメン情報.化学療法レジメン名称,
                            症例.EP後レジメン情報.投与開始日,
                            症例.EP後レジメン情報.治療ライン.名称.
                            ) %>%
              dplyr::distinct()

Treatment_table = data.frame(ID)
Treatment_table$histology = ""
Treatment_table$BTS_STS = ""
Treatment_table$metastasis_site = ""
Treatment_table$Target_done = ""
Treatment_table$ICI_done = ""
Treatment_table$First_SPT = ""
Treatment_table$Treatment_seq = ""
Treatment_table$Treated_drug_Eng_pre = ""
Treatment_table$Treated_drug_Jap_pre = ""
Treatment_table$Treated_drug_Japanese_general_pre = ""
Treatment_table$Treated_drug_Japanese_product_pre = ""
Treatment_table$Treated_regimen_pre = ""
Treatment_table$Treatment_method_pre = ""
Treatment_table$Treatment_line_pre = ""
Treatment_table$Treatment_date_pre = ""
Treatment_table$mutation = ""
Treatment_table$matched_drug = ""
Treatment_table$Treatment_recommendation = ""
Treatment_table$Treated_with_matched_drug = ""
Treatment_table$Reason_matched_drug_unused = ""
Treatment_table$Treated_drug_Eng_post = ""
Treatment_table$Treated_drug_Jap_post = ""
Treatment_table$Treated_drug_Japanese_general_post = ""
Treatment_table$Treated_drug_Japanese_product_post = ""
Treatment_table$Treated_regimen_post = ""
Treatment_table$Treatment_method_post = ""
Treatment_table$Treatment_line_post = ""
Treatment_table$Treatment_date_post = ""

for(i in 1:length(ID)){
  Treatment_table$histology[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$diagnosis
  Treatment_table$BTS_STS[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$directory
  Treatment_table$Target_done[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$Target_done
  Treatment_table$ICI_done[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$ICI_done
  Treatment_table$First_SPT[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$First_SPT
  Treatment_table$Treatment_seq[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$Treatment_seq
  if(length((tmp_report %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))[,1]) > 0){
    Treatment_table$mutation[i] = paste0(as.character((tmp_report %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$tmp), collapse = " -> ")
  }
  if(length((tmp_report_drug %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))[,1]) > 0){
    Treatment_table$matched_drug[i] = paste0(as.character((tmp_report_drug %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$tmp2), collapse = " -> ")
  }
  Treatment_table$Treated_with_matched_drug[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.提示された治療薬を投与した.名称.
  Treatment_table$Treatment_recommendation[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称.
  if(length(Data_case_target %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i])) > 0){
    Treatment_table$Treated_drug_Eng_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.)), collapse = " -> ")
    Treatment_table$Treated_drug_Jap_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.YJ一般名JP.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_general_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.入力一般名.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_product_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.薬剤名.入力商品名.)), collapse = " -> ")
    Treatment_table$Treated_regimen_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.化学療法レジメン名称)), collapse = " -> ")
  Treatment_table$Treatment_method_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.実施目的.名称.)), collapse = " -> ")
  Treatment_table$Treatment_line_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.治療ライン.名称.)), collapse = " -> ")
  Treatment_table$Treatment_date_pre[i] = paste0(as.character(((tmp_case_pre %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP前レジメン情報.投与開始日)), collapse = " -> ")
    Treatment_table$Treated_drug_Eng_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.YJ一般名.EN.)), collapse = " -> ")
    Treatment_table$Treated_drug_Jap_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.YJ一般名JP.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_genera_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.入力一般名.)), collapse = " -> ")
    Treatment_table$Treated_drug_Japanese_product_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.薬剤名.入力商品名.)), collapse = " -> ")
    Treatment_table$Treated_regimen_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.化学療法レジメン名称)), collapse = " -> ")
  Treatment_table$Treatment_method_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.治療方針.名称.)), collapse = " -> ")
  Treatment_table$Treatment_line_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.治療ライン.名称.)), collapse = " -> ")
  Treatment_table$Treatment_date_post[i] = paste0(as.character(((tmp_case_post %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.投与開始日)), collapse = " -> ")
    Treatment_table$metastasis_site[i] = paste0(as.character(unique((Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.がん種情報.登録時転移部位.名称.)), collapse = " -> ")
}
  Treatment_table$Reason_matched_drug_unused[i] = (Data_survival %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID[i]))$症例.EP後レジメン情報.提示された治療薬を投与しなかった理由.名称.
}
WriteXLS(Treatment_table, ExcelFileName = paste(OutDirName, Cancer, "_Treatment_ICI_target.xls", sep=""))

# ICI->Target therapy, or Target therapy->ICI?
Data_survival = Data_survival_save %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_TAR)

Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    Treatment_seq == "ICI_first" ~ 0,
    Treatment_seq == "Target_therapy_first" ~ 1))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  Data_survival_tmp_pos = Data_survival %>%
    dplyr::filter(gene_select == TRUE)
#  time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
   #as.integer(time_tmp[[1]])
  #time_90 = as.integer(time_tmp[[2]])
  #time_100 = as.integer(time_tmp[[3]])
  Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
    time_palliative_enroll >= time_10)# &
      #time_palliative_enroll <= time_90)
  Data_survival_tmp$time_palliative_enroll =
    Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
  
  # Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
  #   time_palliative_enroll > time_90 &
  #     time_palliative_enroll <= time_100)
  # Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
  # idx <- 1:nrow(Data_survival_tmp2)
  # Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
  Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

  Data_survival_tmp_neg = Data_survival %>%
    dplyr::filter(gene_select == FALSE)
#  time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
   #as.integer(time_tmp[[1]])
  #time_90 = as.integer(time_tmp[[2]])
  #time_100 = as.integer(time_tmp[[3]])
  Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
    time_palliative_enroll >= time_10)# &
      #time_palliative_enroll <= time_90)
  Data_survival_tmp3$time_palliative_enroll =
    Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
  
  # Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
  #   time_palliative_enroll > time_90 &
  #     time_palliative_enroll <= time_100)
  # Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
  # idx <- 1:nrow(Data_survival_tmp4)
  # Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
  Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
  Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

  Data_survival = Data_survival %>%
    dplyr::mutate(delay_1 = case_when(
      Data_survival$time_palliative_enroll <= Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry later",
      Data_survival$time_palliative_enroll <= Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry later"))

  if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

    stan_weibull_survival_model_data <-
      list(
        Median_pos = as.integer(Median_entry_pos),
        Median_neg = as.integer(Median_entry_neg),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 2,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        xfactor = as.numeric(Data_survival_tmp$gene_select),
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull_factor.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1,
                pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                         "yhat_total", "yhat_factor_total"), )

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
    dplyr::select(-key) 

median_os_list_weibull_total = NULL
median_os_list_weibull_total_exp = NULL
for(j in 1:8000){
idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
  median_os_list_weibull_total = c(median_os_list_weibull_total,
    median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
  median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
    median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
}
median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

Data_survival$factor_neg = yhat_weibull_sort_average_total
Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select,
                               data = Data_survival)
simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                               time = factor_neg) ~ 1, 
                               data = Data_survival)
simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                               time = factor_pos) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit,
               simulation_fit_neg = simulation_fit_neg,
               simulation_fit_pos = simulation_fit_pos),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = c("ICI->Target therapy, Unadjusted",
                    "Target therapy->ICI, Unadjusted",
                    "ICI->Target therapy, Adjusted for Delayed Entry",
                    "Target therapy->ICI, Adjusted for Delayed Entry"
                   )
  ) +   
  labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (ICI, unadj)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (Target, unadj.)/\n",
             as.integer(median_os_summary_weibull_total[[2]]), 
             " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
             as.integer(median_os_summary_weibull_total[[3]]), ") (ICI, adj)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (Target, adj) days",
             sep=""),
    subtitle = paste("Survival difference with treatment sequence: ",
          as.integer(median_os_summary_weibull_bias[[2]]), " (",
          as.integer(median_os_summary_weibull_bias[[1]]), "-", 
          as.integer(median_os_summary_weibull_bias[[3]]), 
          ") days, median (95% CI)",
          sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_ICI_first_target_first.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. Target->ICI or ICI->Target, Characteristics**") %>%
  bold_labels()
table_tmp


# # ICI vs Target therapy in EGFR/ALK positive patients
# Data_survival = Data_survival_save %>%
#   dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI) %>%
#   dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_TAR)
# ID_EGFR_ALK = unique((Data_report %>%
#       dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% EGFR_ALK &
#                     C.CAT調査結果.エビデンス.エビデンスレベル == "F")
#              )$C.CAT調査結果.基本項目.ハッシュID)
# Data_survival = Data_survival %>%
#       dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_EGFR_ALK)
# Data_survival = Data_survival %>%
#   dplyr::mutate(gene_select = case_when(
#     Treatment_seq == "ICI_first" ~ 0,
#     Treatment_seq == "Target_therapy_first" ~ 1))
# 
# traditional_fit = survfit(Surv(event = censor,
#                                time = time_palliative_final) ~ gene_select, 
#                                data = Data_survival)
# if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
#   Data_survival_tmp_pos = Data_survival %>%
#     dplyr::filter(gene_select == TRUE)
#   #time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
#   #as.integer(time_tmp[[1]])
#   #time_90 = as.integer(time_tmp[[2]])
#   #time_100 = as.integer(time_tmp[[3]])
#   Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
#     time_palliative_enroll >= time_10)# &
#       #time_palliative_enroll <= time_90)
#   Data_survival_tmp$time_palliative_enroll =
#     Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
#   
#   # Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
#   #   time_palliative_enroll > time_90 &
#   #     time_palliative_enroll <= time_100)
#   # Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
#   # idx <- 1:nrow(Data_survival_tmp2)
#   # Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
#   Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
# 
#   Data_survival_tmp_neg = Data_survival %>%
#     dplyr::filter(gene_select == FALSE)
#   #time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
#   #as.integer(time_tmp[[1]])
#   #time_90 = as.integer(time_tmp[[2]])
#   #time_100 = as.integer(time_tmp[[3]])
#   Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
#     time_palliative_enroll >= time_10)# &
#       #time_palliative_enroll <= time_90)
#   Data_survival_tmp3$time_palliative_enroll =
#     Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
#   
#   # Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
#   #   time_palliative_enroll > time_90 &
#   #     time_palliative_enroll <= time_100)
#   # Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
#   # idx <- 1:nrow(Data_survival_tmp4)
#   # Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
#   Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
#   Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
# 
#   Data_survival = Data_survival %>%
#     dplyr::mutate(delay_1 = case_when(
#       Data_survival$time_palliative_enroll <= Median_entry_pos &
#         gene_select == TRUE ~ "SPT to entry early",
#       Data_survival$time_palliative_enroll > Median_entry_pos &
#         gene_select == TRUE ~ "SPT to entry later",
#       Data_survival$time_palliative_enroll <= Median_entry_neg &
#         gene_select == FALSE ~ "SPT to entry early",
#       Data_survival$time_palliative_enroll > Median_entry_neg &
#         gene_select == FALSE ~ "SPT to entry later"))
# 
#   if(sum((Data_survival_tmp %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
#      sum((Data_survival_tmp %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
#      sum((Data_survival_tmp %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
#      sum((Data_survival_tmp %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
# 
#     stan_weibull_survival_model_data <-
#       list(
#         Median_pos = as.integer(Median_entry_pos),
#         Median_neg = as.integer(Median_entry_neg),
#         ## Number of event individuals
#         Nobs = sum(Data_survival$censor == 1),
#         ## Number of censored individuals
#         Ncen = sum(Data_survival$censor == 0),
#         ## Number of total individuals
#         Ntot = length(Data_survival$censor),
#         ## Number of covariates
#         M_bg = 2,
#         Nexp = length(Data_survival_tmp$time_palliative_enroll),
#         ## Times for CGP testing
#         ybef_exp = Data_survival_tmp$time_palliative_enroll,
#         xfactor = as.numeric(Data_survival_tmp$gene_select),
#         ## Times for event individuals
#         yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
#         ## Times for censored individuals
#         ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
#         ## Covariates for event individuals as a matrix
#         Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
#                          as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
#         ## Covariates for censored individuals as a matrix
#         Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
#                          as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
#         )
# 
# stan_weibull_survival_model_fit <-
#     rstan::stan(file =
#                   "source/Stan_code_loglog_weibull_factor.stan",
#                 data = stan_weibull_survival_model_data,
#                 control=list(adapt_delta=0.99, max_treedepth = 10),
#                 seed=12, chains=4, iter=3000, warmup=1000, thin=1,
#                 pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
#                          "yhat_total", "yhat_factor_total"))
# 
# stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
# stan_weibull_survival_model_draws_total <-
#     stan_weibull_survival_model_draws %>%
#     dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
#     tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
#     dplyr::select(-key) 
# stan_weibull_survival_model_draws_total_exp <-
#     stan_weibull_survival_model_draws %>%
#     dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
#     tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
#     dplyr::select(-key) 
# 
# median_os_list_weibull_total = NULL
# median_os_list_weibull_total_exp = NULL
# for(j in 1:8000){
# idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
#   median_os_list_weibull_total = c(median_os_list_weibull_total,
#     median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
#   median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
#     median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
# }
# median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
# 
# 
# median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
# median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
# median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
# yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
# yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
# yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
# yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
# 
# Data_survival$factor_neg = yhat_weibull_sort_average_total
# Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
# 
# traditional_fit = survfit(Surv(event = censor,
#                                time = time_palliative_final) ~ gene_select,
#                                data = Data_survival)
# simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
#                                time = factor_neg) ~ 1, 
#                                data = Data_survival)
# simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
#                                time = factor_pos) ~ 1, 
#                                data = Data_survival)
# 
# g = ggsurvplot(
#     fit = list(traditional_fit = traditional_fit,
#                simulation_fit_neg = simulation_fit_neg,
#                simulation_fit_pos = simulation_fit_pos),
#     combine = TRUE,
#     data = Data_survival,
#     xlab = "Time from SPT to final observation (days)",
#     ylab = "Survival Probability",
#     censor = TRUE,
#     conf.int = FALSE,
#     pval = FALSE,
#     risk.table = TRUE,
#     risk.table.y.text = FALSE,
#     tables.theme = clean_theme(), 
#     legend = c(0.8,0.8),
#     xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
#     break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
#     legend.labs = c("ICI->Target therapy, Unadjusted",
#                     "Target therapy->ICI, Unadjusted",
#                     "ICI->Target therapy, Adjusted for Delayed Entry",
#                     "Target therapy->ICI, Adjusted for Delayed Entry"
#                    )
#   ) +   
#   labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
#               "Median OS ",
#              as.integer(summary(traditional_fit)[[18]][[13]])," (",
#              as.integer(summary(traditional_fit)[[18]][[15]]),"-",
#              as.integer(summary(traditional_fit)[[18]][[17]]),") (ICI, unadj)/",
#              as.integer(summary(traditional_fit)[[18]][[14]]), " (",
#              as.integer(summary(traditional_fit)[[18]][[16]]),"-",
#              as.integer(summary(traditional_fit)[[18]][[18]]),") (Target, unadj.)/\n",
#              as.integer(median_os_summary_weibull_total[[2]]), 
#              " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
#              as.integer(median_os_summary_weibull_total[[3]]), ") (ICI, adj)/",
#              as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
#              as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
#              as.integer(median_os_summary_weibull_total_exp[[3]]),
#              ") (Target, adj) days",
#              sep=""),
#     subtitle = paste("Survival difference with treatment sequence: ",
#           as.integer(median_os_summary_weibull_bias[[2]]), " (",
#           as.integer(median_os_summary_weibull_bias[[1]]), "-", 
#           as.integer(median_os_summary_weibull_bias[[3]]), 
#           ") days, median (95% CI), EGFR/ALK positive",
#           sep=""))
# #print(g)
# pdf(file = paste(OutDirName, Cancer, "_target_or_ICI_EGFRorALK.pdf", sep=""), width = 10, height = 10)
# print(g, newpage = FALSE)
# dev.off()
#   }
# }
# 
# 
# # ICI vs Target therapy for driver genes other than EGFR/ALK positive patients
# Data_survival = Data_survival_save %>%
#   dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI) %>%
#   dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_TAR)
# ID_Driver = unique((Data_report %>%
#       dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% Driver_genes &
#                     C.CAT調査結果.エビデンス.エビデンスレベル == "F")
#              )$C.CAT調査結果.基本項目.ハッシュID)
# Data_survival = Data_survival %>%
#       dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_EGFR_ALK) %>%
#       dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_Driver)
# Data_survival = Data_survival %>%
#   dplyr::mutate(gene_select = case_when(
#     Treatment_seq == "ICI_first" ~ 0,
#     Treatment_seq == "Target_therapy_first" ~ 1))
# 
# traditional_fit = survfit(Surv(event = censor,
#                                time = time_palliative_final) ~ gene_select, 
#                                data = Data_survival)
# if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
#   Data_survival_tmp_pos = Data_survival %>%
#     dplyr::filter(gene_select == TRUE)
#   #time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
#   #as.integer(time_tmp[[1]])
#   #time_90 = as.integer(time_tmp[[2]])
#   #time_100 = as.integer(time_tmp[[3]])
#   Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
#     time_palliative_enroll >= time_10)# &
#       #time_palliative_enroll <= time_90)
#   Data_survival_tmp$time_palliative_enroll =
#     Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
#   
#   # Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
#   #   time_palliative_enroll > time_90 &
#   #     time_palliative_enroll <= time_100)
#   # Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
#   # idx <- 1:nrow(Data_survival_tmp2)
#   # Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
#   Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
# 
#   Data_survival_tmp_neg = Data_survival %>%
#     dplyr::filter(gene_select == FALSE)
#   #time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
#   #as.integer(time_tmp[[1]])
#   #time_90 = as.integer(time_tmp[[2]])
#   #time_100 = as.integer(time_tmp[[3]])
#   Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
#     time_palliative_enroll >= time_10)# &
#       #time_palliative_enroll <= time_90)
#   Data_survival_tmp3$time_palliative_enroll =
#     Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
#   
#   # Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
#   #   time_palliative_enroll > time_90 &
#   #     time_palliative_enroll <= time_100)
#   # Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
#   # idx <- 1:nrow(Data_survival_tmp4)
#   # Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
#   Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
#   Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
# 
#   Data_survival = Data_survival %>%
#     dplyr::mutate(delay_1 = case_when(
#       Data_survival$time_palliative_enroll <= Median_entry_pos &
#         gene_select == TRUE ~ "SPT to entry early",
#       Data_survival$time_palliative_enroll > Median_entry_pos &
#         gene_select == TRUE ~ "SPT to entry later",
#       Data_survival$time_palliative_enroll <= Median_entry_neg &
#         gene_select == FALSE ~ "SPT to entry early",
#       Data_survival$time_palliative_enroll > Median_entry_neg &
#         gene_select == FALSE ~ "SPT to entry later"))
# 
#   if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
#      sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
#      sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
#      sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
# 
#     stan_weibull_survival_model_data <-
#       list(
#         Median_pos = as.integer(Median_entry_pos),
#         Median_neg = as.integer(Median_entry_neg),
#         ## Number of event individuals
#         Nobs = sum(Data_survival$censor == 1),
#         ## Number of censored individuals
#         Ncen = sum(Data_survival$censor == 0),
#         ## Number of total individuals
#         Ntot = length(Data_survival$censor),
#         ## Number of covariates
#         M_bg = 2,
#         Nexp = length(Data_survival_tmp$time_palliative_enroll),
#         ## Times for CGP testing
#         ybef_exp = Data_survival_tmp$time_palliative_enroll,
#         xfactor = as.numeric(Data_survival_tmp$gene_select),
#         ## Times for event individuals
#         yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
#         ## Times for censored individuals
#         ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
#         ## Covariates for event individuals as a matrix
#         Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
#                          as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
#         ## Covariates for censored individuals as a matrix
#         Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
#                          as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
#         )
# 
# stan_weibull_survival_model_fit <-
#     rstan::stan(file =
#                   "source/Stan_code_loglog_weibull_factor.stan",
#                 data = stan_weibull_survival_model_data,
#                 control=list(adapt_delta=0.99, max_treedepth = 10),
#                 seed=12, chains=4, iter=3000, warmup=1000, thin=1,
#                 pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
#                          "yhat_total", "yhat_factor_total"))
# 
# stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
# stan_weibull_survival_model_draws_total <-
#     stan_weibull_survival_model_draws %>%
#     dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
#     tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
#     dplyr::select(-key) 
# stan_weibull_survival_model_draws_total_exp <-
#     stan_weibull_survival_model_draws %>%
#     dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
#     tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
#     dplyr::select(-key) 
# 
# median_os_list_weibull_total = NULL
# median_os_list_weibull_total_exp = NULL
# for(j in 1:8000){
# idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
#   median_os_list_weibull_total = c(median_os_list_weibull_total,
#     median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
#   median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
#     median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
# }
# median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
# 
# 
# median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
# median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
# median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
# yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
# yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
# yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
# yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
# 
# Data_survival$factor_neg = yhat_weibull_sort_average_total
# Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
# 
# traditional_fit = survfit(Surv(event = censor,
#                                time = time_palliative_final) ~ gene_select,
#                                data = Data_survival)
# simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
#                                time = factor_neg) ~ 1, 
#                                data = Data_survival)
# simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
#                                time = factor_pos) ~ 1, 
#                                data = Data_survival)
# 
# g = ggsurvplot(
#     fit = list(traditional_fit = traditional_fit,
#                simulation_fit_neg = simulation_fit_neg,
#                simulation_fit_pos = simulation_fit_pos),
#     combine = TRUE,
#     data = Data_survival,
#     xlab = "Time from SPT to final observation (days)",
#     ylab = "Survival Probability",
#     censor = TRUE,
#     conf.int = FALSE,
#     pval = FALSE,
#     risk.table = TRUE,
#     risk.table.y.text = FALSE,
#     tables.theme = clean_theme(), 
#     legend = c(0.8,0.8),
#     xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
#     break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
#     legend.labs = c("ICI->Target therapy, Unadjusted",
#                     "Target therapy->ICI, Unadjusted",
#                     "ICI->Target therapy, Adjusted for Delayed Entry",
#                     "Target therapy->ICI, Adjusted for Delayed Entry"
#                    )
#   ) +   
#   labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
#               "Median OS ",
#              as.integer(summary(traditional_fit)[[18]][[13]])," (",
#              as.integer(summary(traditional_fit)[[18]][[15]]),"-",
#              as.integer(summary(traditional_fit)[[18]][[17]]),") (ICI, unadj)/",
#              as.integer(summary(traditional_fit)[[18]][[14]]), " (",
#              as.integer(summary(traditional_fit)[[18]][[16]]),"-",
#              as.integer(summary(traditional_fit)[[18]][[18]]),") (Target, unadj.)/\n",
#              as.integer(median_os_summary_weibull_total[[2]]), 
#              " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
#              as.integer(median_os_summary_weibull_total[[3]]), ") (ICI, adj)/",
#              as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
#              as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
#              as.integer(median_os_summary_weibull_total_exp[[3]]),
#              ") (Target, adj) days",
#              sep=""),
#     subtitle = paste("Survival difference with treatment sequence: ",
#           as.integer(median_os_summary_weibull_bias[[2]]), " (",
#           as.integer(median_os_summary_weibull_bias[[1]]), "-", 
#           as.integer(median_os_summary_weibull_bias[[3]]), 
#           ") days, median (95% CI), driver genes other than EGFR/ALK positive",
#           sep=""))
# #print(g)
# pdf(file = paste(OutDirName, Cancer, "_ICI_or_target_notEGFR_notALK.pdf", sep=""), width = 10, height = 10)
# print(g, newpage = FALSE)
# dev.off()
#   }
# }

# ICI->Target therapy, or Target therapy->ICI?
Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    First_SPT == "ICI" ~ 0,
    First_SPT == "TAR" ~ 1,
    First_SPT == "Oth" ~ 2))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    name_1 = "1st-line:ICI"
    name_2 = "1st-line:Target therapy"
    name_3 = "1st-line:Other therapy"
    name_1_short = "ICI"
    name_2_short = "Target therapy"
    name_3_short = "Other therapy"

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)
  
    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_0_2 = median_os_list_weibull_total - median_os_list_weibull_total_exp2
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_0_2 = quantile(median_os_list_weibull_bias_0_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c("ICI first, Unadjusted",
                      "Target therapy first, Unadjusted",
                      "Other therapy first, Unadjusted",
                      "ICI first, Adjusted for Delayed Entry",
                      "Target therapy first, Adjusted for Delayed Entry",
                      "Other therapy first, Adjusted for Delayed Entry"
                   )
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Bonferroni-like correction, 98.3% CI)\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_1_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_2[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (98.3% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_3_first_CTx.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. First-line treatment, target or ICI or other, Characteristics**") %>%
  bold_labels()
table_tmp

# ICI->Target therapy, or Target therapy->ICI?
Data_survival = Data_survival_save %>%
  dplyr::filter(First_SPT != "Oth") %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_TAR)

Data_survival = Data_survival %>%
  dplyr::mutate(gene_select = case_when(
    Treatment_seq == "ICI_first" ~ 0,
    Treatment_seq == "Target_therapy_first" ~ 1))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  Data_survival_tmp_pos = Data_survival %>%
    dplyr::filter(gene_select == TRUE)
#  time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
   #as.integer(time_tmp[[1]])
  #time_90 = as.integer(time_tmp[[2]])
  #time_100 = as.integer(time_tmp[[3]])
  Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
    time_palliative_enroll >= time_10)# &
      #time_palliative_enroll <= time_90)
  Data_survival_tmp$time_palliative_enroll =
    Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
  
  # Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
  #   time_palliative_enroll > time_90 &
  #     time_palliative_enroll <= time_100)
  # Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
  # idx <- 1:nrow(Data_survival_tmp2)
  # Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
  Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

  Data_survival_tmp_neg = Data_survival %>%
    dplyr::filter(gene_select == FALSE)
#  time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
   #as.integer(time_tmp[[1]])
  #time_90 = as.integer(time_tmp[[2]])
  #time_100 = as.integer(time_tmp[[3]])
  Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
    time_palliative_enroll >= time_10)# &
      #time_palliative_enroll <= time_90)
  Data_survival_tmp3$time_palliative_enroll =
    Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
  
  # Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
  #   time_palliative_enroll > time_90 &
  #     time_palliative_enroll <= time_100)
  # Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
  # idx <- 1:nrow(Data_survival_tmp4)
  # Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
  Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
  Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

  Data_survival = Data_survival %>%
    dplyr::mutate(delay_1 = case_when(
      Data_survival$time_palliative_enroll <= Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry later",
      Data_survival$time_palliative_enroll <= Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry later"))

  if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

    stan_weibull_survival_model_data <-
      list(
        Median_pos = as.integer(Median_entry_pos),
        Median_neg = as.integer(Median_entry_neg),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 2,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        xfactor = as.numeric(Data_survival_tmp$gene_select),
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull_factor.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                         "yhat_total", "yhat_factor_total"))

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
    dplyr::select(-key) 

median_os_list_weibull_total = NULL
median_os_list_weibull_total_exp = NULL
for(j in 1:8000){
idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
  median_os_list_weibull_total = c(median_os_list_weibull_total,
    median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
  median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
    median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
}
median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

Data_survival$factor_neg = yhat_weibull_sort_average_total
Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select,
                               data = Data_survival)
simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                               time = factor_neg) ~ 1, 
                               data = Data_survival)
simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                               time = factor_pos) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit,
               simulation_fit_neg = simulation_fit_neg,
               simulation_fit_pos = simulation_fit_pos),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = c("1stLineICI->Target therapy, Unadjusted",
                    "1stLineTarget therapy->ICI, Unadjusted",
                    "1stLineICI->Target therapy, Adjusted for Delayed Entry",
                    "1stLineTarget therapy->ICI, Adjusted for Delayed Entry"
                   )
  ) +   
  labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (ICI, unadj)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (Target, unadj.)/\n",
             as.integer(median_os_summary_weibull_total[[2]]), 
             " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
             as.integer(median_os_summary_weibull_total[[3]]), ") (ICI, adj)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (Target, adj) days",
             sep=""),
    subtitle = paste("Survival difference with treatment sequence: ",
          as.integer(median_os_summary_weibull_bias[[2]]), " (",
          as.integer(median_os_summary_weibull_bias[[1]]), "-", 
          as.integer(median_os_summary_weibull_bias[[3]]), 
          ") days, median (95% CI)",
          sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_ICI_or_target_first_sequence.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. TargetAsFirstLine->ICI or ICIAsFirstLine->Target, Characteristics**") %>%
  bold_labels()
table_tmp


```

```{r STS_2ndline, eval = (Cancer == "Soft_tissue" & Subanalysis == 1), echo = (Source_code == 1 & Cancer == "Soft_tissue" & Subanalysis == 1), warning=FALSE}
## Survival analysis considering left truncation
Drugs_A = c("Eribulin")
Drugs_B = c("Pazopanib")
Drugs_C = c("Trabectedin")

Data_survival = Data_case_target %>%
  dplyr::filter(!症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% Drugs_A)
ID_A_used_periop = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_case_target %>%
  dplyr::filter(!症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% Drugs_A)
ID_B_used_periop = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_case_target %>%
  dplyr::filter(!症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% Drugs_C)
ID_C_used_periop = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_case_target %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_A_used_periop) %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_B_used_periop) %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_C_used_periop) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))

Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))

Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_tmp_pre = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.投与開始日,
                症例.EP前レジメン情報.投与終了日,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_tmp_post = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.投与開始日,
                症例.EP後レジメン情報.投与終了日,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "")

Data_drug_sequence = Data_drug_sequence %>%
  dplyr::mutate(drug_group = case_when(
    drug_name %in% Drugs_A ~ "A__",
    drug_name %in% Drugs_B ~ "B__",
    drug_name %in% Drugs_C ~ "C__",
    TRUE ~ "Oth"))
ID_A = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "A__"))$C.CAT調査結果.基本項目.ハッシュID)
ID_B = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "B__"))$C.CAT調査結果.基本項目.ハッシュID)
ID_C = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "C__"))$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_survival %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日))
Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "SPT to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "SPT to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "SPT to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "SPT to entry 3-4 year",
    TRUE ~ "SPT to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
    TRUE ~ "SPT to entry later"))

Data_survival$First_SPT = "Oth"
drug_seq = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
for(i in 1:length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)){
  Data_tmp = Data_drug_sequence[Data_drug_sequence$
                C.CAT調査結果.基本項目.ハッシュID == Data_survival$
                  C.CAT調査結果.基本項目.ハッシュID[i],]
  if(length(Data_tmp$C.CAT調査結果.基本項目.ハッシュID > 0)){
    Data_tmp = Data_tmp %>% dplyr::filter(
      drug_start_date == Data_tmp$drug_start_date[[1]]
    )
    drug_seq[i] = paste(Data_tmp$drug_group, collapse = "")
    if(str_detect(drug_seq[i], "A__") &
       !str_detect(drug_seq[i], "B__") &
       !str_detect(drug_seq[i], "C__")){
      Data_survival$First_SPT[i] = "A__"
    }
    if(!str_detect(drug_seq[i], "A__") &
       str_detect(drug_seq[i], "B__") &
       !str_detect(drug_seq[i], "C__")){
      Data_survival$First_SPT[i] = "B__"
    }
    if(!str_detect(drug_seq[i], "A__") &
       !str_detect(drug_seq[i], "B__") &
       str_detect(drug_seq[i], "C__")){
      Data_survival$First_SPT[i] = "C__"
    }
  }
}

drug_seq = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
drug_seq_tax = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
for(i in 1:length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)){
  Data_tmp = Data_drug_sequence[Data_drug_sequence$C.CAT調査結果.基本項目.ハッシュID == Data_survival$C.CAT調査結果.基本項目.ハッシュID[i],]
  drug_seq[i] = paste(Data_tmp$drug_group, collapse = "")
  drug_seq_tax[i] = paste(Data_tmp$drug_group_tax, collapse = "")
}

drug_seq = str_replace_all(drug_seq,
          pattern="Oth",
          replacement = "")
drug_seq = str_replace_all(drug_seq,
          pattern="A__A__",
          replacement = "A__")
drug_seq = str_replace_all(drug_seq,
          pattern="A__A__",
          replacement = "A__")
drug_seq = str_replace_all(drug_seq,
          pattern="A__A__",
          replacement = "A__")
drug_seq = str_replace_all(drug_seq,
          pattern="A__A__",
          replacement = "A__")
drug_seq = str_replace_all(drug_seq,
          pattern="B__B__",
          replacement = "B__")
drug_seq = str_replace_all(drug_seq,
          pattern="B__B__",
          replacement = "B__")
drug_seq = str_replace_all(drug_seq,
          pattern="B__B__",
          replacement = "B__")
drug_seq = str_replace_all(drug_seq,
          pattern="B__B__",
          replacement = "B__")
drug_seq = str_replace_all(drug_seq,
          pattern="C__C__",
          replacement = "C__")
drug_seq = str_replace_all(drug_seq,
          pattern="C__C__",
          replacement = "C__")
drug_seq = str_replace_all(drug_seq,
          pattern="C__C__",
          replacement = "C__")
drug_seq = str_replace_all(drug_seq,
          pattern="C__C__",
          replacement = "C__")

Data_survival$drug_seq = drug_seq

Data_survival = Data_survival %>%
  dplyr::mutate(Treatment_seq = case_when(
    str_sub(drug_seq, 1, 3) == "A__" ~ "A_first",
    str_sub(drug_seq, 1, 3) == "B__" ~ "B_first",
    str_sub(drug_seq, 1, 3) == "C__" ~ "C_first",
    TRUE ~ "other"))
Data_survival = Data_survival %>%
  dplyr::mutate(Treatment_seq_AB = case_when(
    str_sub(drug_seq, 1, 3) == "A__" ~ "A_first",
    str_sub(drug_seq, 1, 3) == "B__" ~ "B_first",
    str_sub(drug_seq, 1, 3) == "C__A__" ~ "A_first",
    str_sub(drug_seq, 1, 3) == "C__B__" ~ "B_first",
    TRUE ~ "other"))
Data_survival = Data_survival %>%
  dplyr::mutate(Treatment_seq_BC = case_when(
    str_sub(drug_seq, 1, 3) == "B__" ~ "B_first",
    str_sub(drug_seq, 1, 3) == "C__" ~ "C_first",
    str_sub(drug_seq, 1, 3) == "A__B__" ~ "B_first",
    str_sub(drug_seq, 1, 3) == "A__C__" ~ "C_first",
    TRUE ~ "other"))
Data_survival = Data_survival %>%
  dplyr::mutate(Treatment_seq_AC = case_when(
    str_sub(drug_seq, 1, 3) == "A__" ~ "A_first",
    str_sub(drug_seq, 1, 3) == "C__" ~ "C_first",
    str_sub(drug_seq, 1, 3) == "B__A__" ~ "A_first",
    str_sub(drug_seq, 1, 3) == "B__C__" ~ "C_first",
    TRUE ~ "other"))

Data_survival_save = Data_survival

# Drug sequence, A->B vs B->A
Data_survival = Data_survival_save %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_A) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_B)

Data_survival = Data_survival %>%
  dplyr::filter(Treatment_seq_AB != "other") %>%
  dplyr::mutate(gene_select = case_when(
    Treatment_seq_AB == "A_first" ~ 0,
    Treatment_seq_AB == "B_first" ~ 1))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(Data_survival[,1]) > 20){
  Data_survival_tmp_pos = Data_survival %>%
    dplyr::filter(gene_select == TRUE)
  
  Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp$time_palliative_enroll =
    Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
  Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

  Data_survival_tmp_neg = Data_survival %>%
    dplyr::filter(gene_select == FALSE)
  
  Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp3$time_palliative_enroll =
    Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
  Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
  Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

  Data_survival = Data_survival %>%
    dplyr::mutate(delay_1 = case_when(
      Data_survival$time_palliative_enroll <= Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry later",
      Data_survival$time_palliative_enroll <= Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry later"))

  if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

    stan_weibull_survival_model_data <-
      list(
        Median_pos = as.integer(Median_entry_pos),
        Median_neg = as.integer(Median_entry_neg),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 2,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        xfactor = as.numeric(Data_survival_tmp$gene_select),
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull_factor.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1,
                pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                         "yhat_total", "yhat_factor_total"), )

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
    dplyr::select(-key) 

median_os_list_weibull_total = NULL
median_os_list_weibull_total_exp = NULL
for(j in 1:8000){
idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
  median_os_list_weibull_total = c(median_os_list_weibull_total,
    median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
  median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
    median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
}
median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

Data_survival$factor_neg = yhat_weibull_sort_average_total
Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select,
                               data = Data_survival)
simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                               time = factor_neg) ~ 1, 
                               data = Data_survival)
simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                               time = factor_pos) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit,
               simulation_fit_neg = simulation_fit_neg,
               simulation_fit_pos = simulation_fit_pos),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = c(paste(paste(str_sub(Drugs_A,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_B,1,3),collapse = ","),
                          ", unadj", sep=""),
                    paste(paste(str_sub(Drugs_B,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_A,1,3),collapse = ","),
                          ", unadj", sep=""),
                    paste(paste(str_sub(Drugs_A,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_B,1,3),collapse = ","),
                          ", adj", sep=""),
                    paste(paste(str_sub(Drugs_B,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_A,1,3),collapse = ","),
                          ", adj", sep="")
                   )
  ) +
  labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (",
             paste(str_sub(Drugs_A,1,3),collapse = ","), " first, unadj)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (",
             paste(str_sub(Drugs_B,1,3),collapse = ","), " first, unadj.)/\n",
             as.integer(median_os_summary_weibull_total[[2]]), 
             " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
             as.integer(median_os_summary_weibull_total[[3]]), ") (",
             paste(str_sub(Drugs_A,1,3),collapse = ","), " first, adj)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),") (",
             paste(str_sub(Drugs_B,1,3),collapse = ","), " first, adj) days",
             sep=""),
    subtitle = paste("Survival difference: ",
          as.integer(median_os_summary_weibull_bias[[2]]), " (",
          as.integer(median_os_summary_weibull_bias[[1]]), "-", 
          as.integer(median_os_summary_weibull_bias[[3]]), 
          ") days, median (95% CI)",
          sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_3drugs_sequence_A_B.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. 3 drugs, A or B, Characteristics**") %>%
  bold_labels()
table_tmp

# Drug sequence, B->C vs C->B
Data_survival = Data_survival_save %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_A) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_B)

Data_survival = Data_survival %>%
  dplyr::filter(Treatment_seq_BC != "other") %>%
  dplyr::mutate(gene_select = case_when(
    Treatment_seq_BC == "B_first" ~ 0,
    Treatment_seq_BC == "C_first" ~ 1))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(Data_survival[,1]) > 20){
  Data_survival_tmp_pos = Data_survival %>%
    dplyr::filter(gene_select == TRUE)
  
  Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp$time_palliative_enroll =
    Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
  Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

  Data_survival_tmp_neg = Data_survival %>%
    dplyr::filter(gene_select == FALSE)
  
  Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp3$time_palliative_enroll =
    Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
  Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
  Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

  Data_survival = Data_survival %>%
    dplyr::mutate(delay_1 = case_when(
      Data_survival$time_palliative_enroll <= Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry later",
      Data_survival$time_palliative_enroll <= Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry later"))

  if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

    stan_weibull_survival_model_data <-
      list(
        Median_pos = as.integer(Median_entry_pos),
        Median_neg = as.integer(Median_entry_neg),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 2,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        xfactor = as.numeric(Data_survival_tmp$gene_select),
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull_factor.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1,
                pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                         "yhat_total", "yhat_factor_total"), )

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
    dplyr::select(-key) 

median_os_list_weibull_total = NULL
median_os_list_weibull_total_exp = NULL
for(j in 1:8000){
idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
  median_os_list_weibull_total = c(median_os_list_weibull_total,
    median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
  median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
    median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
}
median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

Data_survival$factor_neg = yhat_weibull_sort_average_total
Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select,
                               data = Data_survival)
simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                               time = factor_neg) ~ 1, 
                               data = Data_survival)
simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                               time = factor_pos) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit,
               simulation_fit_neg = simulation_fit_neg,
               simulation_fit_pos = simulation_fit_pos),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = c(paste(paste(str_sub(Drugs_B,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_C,1,3),collapse = ","),
                          ", unadj", sep=""),
                    paste(paste(str_sub(Drugs_C,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_B,1,3),collapse = ","),
                          ", unadj", sep=""),
                    paste(paste(str_sub(Drugs_B,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_C,1,3),collapse = ","),
                          ", adj", sep=""),
                    paste(paste(str_sub(Drugs_C,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_B,1,3),collapse = ","),
                          ", adj", sep="")
                   )
  ) +
  labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (",
             paste(str_sub(Drugs_B,1,3),collapse = ","), " first, unadj)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (",
             paste(str_sub(Drugs_C,1,3),collapse = ","), " first, unadj.)/\n",
             as.integer(median_os_summary_weibull_total[[2]]), 
             " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
             as.integer(median_os_summary_weibull_total[[3]]), ") (",
             paste(str_sub(Drugs_B,1,3),collapse = ","), " first, adj)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),") (",
             paste(str_sub(Drugs_C,1,3),collapse = ","), " first, adj) days",
             sep=""),
    subtitle = paste("Survival difference: ",
          as.integer(median_os_summary_weibull_bias[[2]]), " (",
          as.integer(median_os_summary_weibull_bias[[1]]), "-", 
          as.integer(median_os_summary_weibull_bias[[3]]), 
          ") days, median (95% CI)",
          sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_3drugs_sequence_B_C.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. 3 drugs, B or C, Characteristics**") %>%
  bold_labels()
table_tmp

# Drug sequence, A->C vs C->A
Data_survival = Data_survival_save %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_A) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_B)

Data_survival = Data_survival %>%
  dplyr::filter(Treatment_seq_AC != "other") %>%
  dplyr::mutate(gene_select = case_when(
    Treatment_seq_AC == "A_first" ~ 0,
    Treatment_seq_AC == "C_first" ~ 1))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(Data_survival[,1]) > 20){
  Data_survival_tmp_pos = Data_survival %>%
    dplyr::filter(gene_select == TRUE)
  
  Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp$time_palliative_enroll =
    Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
  Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

  Data_survival_tmp_neg = Data_survival %>%
    dplyr::filter(gene_select == FALSE)
  
  Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
    time_palliative_enroll >= time_10)
  Data_survival_tmp3$time_palliative_enroll =
    Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
  Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
  Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

  Data_survival = Data_survival %>%
    dplyr::mutate(delay_1 = case_when(
      Data_survival$time_palliative_enroll <= Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_pos &
        gene_select == TRUE ~ "SPT to entry later",
      Data_survival$time_palliative_enroll <= Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry early",
      Data_survival$time_palliative_enroll > Median_entry_neg &
        gene_select == FALSE ~ "SPT to entry later"))

  if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
     sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

    stan_weibull_survival_model_data <-
      list(
        Median_pos = as.integer(Median_entry_pos),
        Median_neg = as.integer(Median_entry_neg),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 2,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        xfactor = as.numeric(Data_survival_tmp$gene_select),
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                         as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull_factor.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1,
                pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                         "yhat_total", "yhat_factor_total"), )

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
    dplyr::select(-key) 

median_os_list_weibull_total = NULL
median_os_list_weibull_total_exp = NULL
for(j in 1:8000){
idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
  median_os_list_weibull_total = c(median_os_list_weibull_total,
    median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
  median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
    median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
}
median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

Data_survival$factor_neg = yhat_weibull_sort_average_total
Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select,
                               data = Data_survival)
simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                               time = factor_neg) ~ 1, 
                               data = Data_survival)
simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                               time = factor_pos) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit,
               simulation_fit_neg = simulation_fit_neg,
               simulation_fit_pos = simulation_fit_pos),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from SPT to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = c(paste(paste(str_sub(Drugs_A,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_C,1,3),collapse = ","),
                          ", unadj", sep=""),
                    paste(paste(str_sub(Drugs_C,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_A,1,3),collapse = ","),
                          ", unadj", sep=""),
                    paste(paste(str_sub(Drugs_A,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_C,1,3),collapse = ","),
                          ", adj", sep=""),
                    paste(paste(str_sub(Drugs_C,1,3),collapse = ","),"->",
                          paste(str_sub(Drugs_A,1,3),collapse = ","),
                          ", adj", sep="")
                   )
  ) +
  labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (",
             paste(str_sub(Drugs_A,1,3),collapse = ","), " first, unadj)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (",
             paste(str_sub(Drugs_C,1,3),collapse = ","), " first, unadj.)/\n",
             as.integer(median_os_summary_weibull_total[[2]]), 
             " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
             as.integer(median_os_summary_weibull_total[[3]]), ") (",
             paste(str_sub(Drugs_A,1,3),collapse = ","), " first, adj)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),") (",
             paste(str_sub(Drugs_C,1,3),collapse = ","), " first, adj) days",
             sep=""),
    subtitle = paste("Survival difference: ",
          as.integer(median_os_summary_weibull_bias[[2]]), " (",
          as.integer(median_os_summary_weibull_bias[[1]]), "-", 
          as.integer(median_os_summary_weibull_bias[[3]]), 
          ") days, median (95% CI)",
          sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_3drugs_sequence_A_C.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    by = diagnosis,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. 3 drugs, A or C, Characteristics**") %>%
  bold_labels()
table_tmp

# ICI->Target therapy, or Target therapy->ICI?
Data_survival = Data_survival_save %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% c(ID_A, ID_B, ID_C))

Data_survival = Data_survival %>%
  dplyr::filter(Treatment_seq != "other") %>%
  dplyr::mutate(gene_select = case_when(
    Treatment_seq == "A_first" ~ 0,
    Treatment_seq == "B_first" ~ 1,
    Treatment_seq == "C_first" ~ 2))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ gene_select, 
                               data = Data_survival)
if(length(unique(Data_survival$gene_select)) == 3){
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
     length(Data_survival[,1]) > 20){
    name_1 = paste("1st-line:", paste(str_sub(Drugs_A,1,3),collapse = ","), sep="")
    name_2 = paste("1st-line:", paste(str_sub(Drugs_B,1,3),collapse = ","), sep="")
    name_3 = paste("1st-line:", paste(str_sub(Drugs_C,1,3),collapse = ","), sep="")
    name_1_short = paste(str_sub(Drugs_A,1,3),collapse = ",")
    name_2_short = paste(str_sub(Drugs_B,1,3),collapse = ",")
    name_3_short = paste(str_sub(Drugs_C,1,3),collapse = ",")

    Data_survival_tmp_2 = Data_survival %>%
      dplyr::filter(gene_select == 2)
    
    Data_survival_tmp = Data_survival_tmp_2 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)

    Median_entry_2 = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_1 = Data_survival %>%
      dplyr::filter(gene_select == 1)
    
    Data_survival_tmp3 = Data_survival_tmp_1 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)

    Median_entry_1 = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  
  
    Data_survival_tmp_0 = Data_survival %>%
      dplyr::filter(gene_select == 0)
    
    Data_survival_tmp5 = Data_survival_tmp_0 %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp5$time_palliative_enroll =
      Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)

    Median_entry_0 = median(Data_survival_tmp5$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp5)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_2 &
          gene_select == 2 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_1 &
          gene_select == 1 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_0 &
          gene_select == 0 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_2 &
          gene_select == 2 ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_1 &
          gene_select == 1 ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_0 &
          gene_select == 0 ~ "SPT to entry later"))
  
    if(sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
  
      stan_weibull_survival_model_data <-
        list(
          Median_2 = as.integer(Median_entry_2),
          Median_1 = as.integer(Median_entry_1),
          Median_0 = as.integer(Median_entry_0),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 3,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select == 1),
          xfactor2 = as.numeric(Data_survival_tmp$gene_select == 2),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 1],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 1]),
                           ncol = 3),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 1)[Data_survival$censor == 0],
                             as.numeric(Data_survival$gene_select == 2)[Data_survival$censor == 0]),
                           ncol = 3)
          )
  
  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
                           "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
  
  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))
  
  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp2 <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  median_os_list_weibull_total_exp2 = NULL
 for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_total_exp2 = c(median_os_list_weibull_total_exp2,
      median(stan_weibull_survival_model_draws_total_exp2[idx,]$yhat_total_exp2, na.rm = T))
  }
  median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
  median_os_list_weibull_bias_0_2 = median_os_list_weibull_total - median_os_list_weibull_total_exp2
  median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2

  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_0_2 = quantile(median_os_list_weibull_bias_0_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025/3, 0.5, 1 - 0.025/3))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * 8000]
  
  Data_survival$factor_0 = yhat_weibull_sort_average_total
  Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
  Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
  Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
  Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
  Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
  
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
                                 time = factor_0) ~ 1, 
                                 data = Data_survival)
  simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
                                 time = factor_1) ~ 1, 
                                 data = Data_survival)
  simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
                                 time = factor_2) ~ 1, 
                                 data = Data_survival)

  g = ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_0 = simulation_fit_0,
                 simulation_fit_1 = simulation_fit_1,
                 simulation_fit_2 = simulation_fit_2),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(
                    paste(paste(str_sub(Drugs_A,1,3),collapse = ","),
                          " first, unadj", sep=""),
                    paste(paste(str_sub(Drugs_B,1,3),collapse = ","),
                          " first, unadj", sep=""),
                    paste(paste(str_sub(Drugs_C,1,3),collapse = ","),
                          " first, unadj", sep=""),
                    paste(paste(str_sub(Drugs_A,1,3),collapse = ","),
                          " first, adj", sep=""),
                    paste(paste(str_sub(Drugs_B,1,3),collapse = ","),
                          " first, adj", sep=""),
                    paste(paste(str_sub(Drugs_C,1,3),collapse = ","),
                          " first, adj", sep="")
                   )
    ) +
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[19]])," (",
               as.integer(summary(traditional_fit)[[18]][[22]]),"-",
               as.integer(summary(traditional_fit)[[18]][[25]]),") (",
               name_1_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[20]]), " (",
               as.integer(summary(traditional_fit)[[18]][[23]]),"-",
               as.integer(summary(traditional_fit)[[18]][[26]]),") (",
               name_2_short,", unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[21]]), " (",
               as.integer(summary(traditional_fit)[[18]][[24]]),"-",
               as.integer(summary(traditional_fit)[[18]][[27]]),") (",
               name_3_short,", unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (",
               name_1_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), 
               " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
               name_2_short,", adj.)/",
               as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp2[[3]]),
               ") (", name_3_short,", adj.) days",
               sep=""),
      subtitle = paste("Survival difference (Bonferroni-like correction, 98.3% CI)\n",
            name_1_short, "-", name_2_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
            name_1_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_0_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_0_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_0_2[[3]]), "), ",
            name_2_short, "-", name_3_short, ": ",
            as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
            as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias_1_2[[3]]),
            ") days, median (98.3% CI)",
            sep=""))
      #print(g)
      pdf(file = paste(OutDirName, Cancer, "_3drugs_first_CTx.pdf", sep=""), width = 10, height = 10)
      print(g, newpage = FALSE)
      dev.off()
    }
  }
}

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    by = diagnosis,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. First-line treatment, Characteristics**") %>%
  bold_labels()
table_tmp


```

```{r Bladder, eval = (Cancer == "Bladder" & Subanalysis == 1), echo = (Source_code == 1 & Cancer == "Bladder" & Subanalysis == 1), warning=FALSE}
## Survival analysis considering left truncation
ICI_drugs = c("Pembrolizumab")
Gene_target_drugs = c("Gemcitabine")

# GCについて評価してみたい　奏効性に関与する遺伝子変異
Data_survival = Data_case_target %>%
  dplyr::filter(!症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日))
Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% ICI_drugs)
ID_ICI_used_periop = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_case_target %>%
  dplyr::filter(!症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日))
Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. %in% Gene_target_drugs)
ID_target_used_periop = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)

Data_survival = Data_case_target %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI_used_periop) %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_target_used_periop) %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日))

Data_survival = Data_survival %>%
  dplyr::filter(症例.EP前レジメン情報.投与開始日 != "") %>%
  dplyr::filter(!is.na(症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
  dplyr::filter(症例.転帰情報.転帰.名称. != "") %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(症例.転帰情報.最終生存確認日 != "" |
                症例.転帰情報.死亡日 != "") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))

Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_tmp_pre = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.投与開始日,
                症例.EP前レジメン情報.投与終了日,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_tmp_post = Data_survival %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.投与開始日,
                症例.EP後レジメン情報.投与終了日,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_start_date",
                           "drug_end_date",
                           "drug_name")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "")

Data_drug_sequence = Data_drug_sequence %>%
  dplyr::mutate(drug_group = case_when(
    drug_name %in% ICI_drugs ~ "Ici",
    drug_name %in% Gene_target_drugs ~ "Tar",
    TRUE ~ "Oth"))
ID_ICI = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "Ici"))$C.CAT調査結果.基本項目.ハッシュID)
ID_TAR = unique((Data_drug_sequence %>%
  dplyr::filter(drug_group == "Tar"))$C.CAT調査結果.基本項目.ハッシュID)
Data_survival = Data_survival %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_TAR)

Data_survival = Data_survival %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE)

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日))
Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

drug_seq = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
ICI_date = rep("", length(Data_survival$C.CAT調査結果.基本項目.ハッシュID))
ID_exclude = NULL
for(i in 1:length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)){
  Data_tmp = Data_drug_sequence[Data_drug_sequence$C.CAT調査結果.基本項目.ハッシュID == Data_survival$C.CAT調査結果.基本項目.ハッシュID[i],]
  drug_seq[i] = paste(Data_tmp$drug_group, collapse = "")
  if((Data_tmp[Data_tmp$drug_group == "Ici", "drug_start_date"])[[1]] == "" |
     is.na((Data_tmp[Data_tmp$drug_group == "Ici", "drug_start_date"])[[1]])){
    ID_exclude = c(ID_exclude, Data_survival$C.CAT調査結果.基本項目.ハッシュID[i])
    ICI_date[i] = "no data"
  } else{
    ICI_date[i] = (Data_tmp[Data_tmp$drug_group == "Ici", "drug_start_date"])[[1]]
  }
}

drug_seq = str_replace_all(drug_seq,
          pattern="Oth",
          replacement = "")
drug_seq = str_replace_all(drug_seq,
          pattern="Ici",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="IciIci",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="IciIci",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="IciIci",
          replacement = "Ici")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")
drug_seq = str_replace_all(drug_seq,
          pattern="TarTar",
          replacement = "Tar")

Data_survival$drug_seq = drug_seq

Data_survival = Data_survival %>%
  dplyr::mutate(Treatment_seq = case_when(
    str_sub(drug_seq, 1, 3) == "Ici" ~ "ICI_first",
    str_sub(drug_seq, 1, 3) == "Tar" ~ "Target_therapy_first",
    TRUE ~ "other"))


Data_survival$drug_seq = drug_seq
Data_survival$ICI_date = ICI_date
if(length(ID_exclude > 0)){
  Data_survival = Data_survival %>%
  dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_exclude)
}

Data_survival = Data_survival %>%
  dplyr::filter(str_sub(drug_seq, 1, 3) == "Tar")
Data_survival = Data_survival %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ICI) %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_TAR)
Data_survival = Data_survival %>%
    dplyr::filter(Treatment_seq == "Target_therapy_first")

Data_survival = Data_survival %>%
  dplyr::mutate(ICI_date = as.Date(Data_survival$ICI_date))
Data_survival = Data_survival %>%
  dplyr::mutate(time_ICI_enroll =
                  as.integer(difftime(Data_survival$症例.管理情報.登録日,
                                      Data_survival$ICI_date,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_ICI_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$ICI_date,
                                      units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.EP前レジメン情報.投与開始日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
                  Data_survival$time_palliative_final -
                  Data_survival$time_enroll_final)

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_ICI_enroll/365)

Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_ICI_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final = Data_survival$time_ICI_final)
Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll = Data_survival$time_ICI_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "ICI to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "ICI to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "ICI to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "ICI to entry 3-4 year",
    TRUE ~ "ICI to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "ICI to entry early",
    TRUE ~ "ICI to entry later"))

# Survival analysis for delayed entry
traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$delay)) == 5){
print(traditional_fit)
ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = paste("Enrollment timing:",
                      c("Vintage chemotherapy to entry <=1 year",
                        "Vintage chemotherapy to entry 1-2 year",
                        "Vintage chemotherapy to entry 2-3 year",
                        "Vintage chemotherapy to entry 3-4 year",
                        "Vintage chemotherapy to entry >4 year")
                        , sep="")
) + 
  labs(title = paste(Cancer, "Delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
print(traditional_fit)
ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
) + 
  labs(title = paste(Cancer, "Delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
}

# Survival analysis for delayed entry, < median or not
traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$delay_1)) == 2){
print(traditional_fit)
ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from enrollment",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = paste("Enrollment timing:",
                      c("Vintage chemotherapy to entry early",
                        "Vintage chemotherapy to entry later")
                        , sep="")
) + 
  labs(title = paste(Cancer, "Delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
}

# survival curve: from the last vintage chemotherapy to enrollment
Data_survival_tmp = Data_survival %>% dplyr::filter(
  time_palliative_enroll >= time_10)
Data_survival_tmp$time_palliative_enroll =
  Data_survival_tmp$time_palliative_enroll - (time_10 - 1)

Median_entry = median(Data_survival_tmp$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "ICI to entry early",
    TRUE ~ "ICI to entry later"))

stan_weibull_survival_model_data <-
    list(
        Median = as.integer(Median_entry),
        Nobs = sum(Data_survival$censor == 1),
        Ncen = sum(Data_survival$censor == 0),
        Ntot = length(Data_survival$censor),
        M_bg = 1,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        Xobs_bg = matrix(as.numeric(Data_survival$delay_1 == "ICI to entry later")[Data_survival$censor == 1]),
        Xcen_bg = matrix(as.numeric(Data_survival$delay_1 == "ICI to entry later")[Data_survival$censor == 0])
        )

stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1)

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_bef_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
    tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_aft_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
    tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_ear <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
    tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_lat <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
    tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
    dplyr::select(-key) 


median_os_list_weibull_total_exp = NULL
median_os_list_weibull_bef_exp = NULL
median_os_list_weibull_aft_exp = NULL
median_os_list_weibull_ear = NULL
median_os_list_weibull_lat = NULL

for(i in 1:8000){
  idx = seq(i, length(stan_weibull_survival_model_draws_total_exp$yhat_total_exp), by=8000)
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_bef_exp = c(median_os_list_weibull_bef_exp,
      median(stan_weibull_survival_model_draws_bef_exp[idx,]$yhat_bef_exp, na.rm = T))
    median_os_list_weibull_aft_exp = c(median_os_list_weibull_aft_exp,
      median(stan_weibull_survival_model_draws_aft_exp[idx,]$yhat_aft_exp, na.rm = T))
    median_os_list_weibull_ear = c(median_os_list_weibull_ear,
      median(stan_weibull_survival_model_draws_ear[idx,]$yhat_ear, na.rm = T))
    median_os_list_weibull_lat = c(median_os_list_weibull_lat,
      median(stan_weibull_survival_model_draws_lat[idx,]$yhat_lat, na.rm = T))
}

median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bef_exp = quantile(median_os_list_weibull_bef_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_aft_exp = quantile(median_os_list_weibull_aft_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_ear = quantile(median_os_list_weibull_ear, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_lat = quantile(median_os_list_weibull_lat, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_bef_exp = sort(stan_weibull_survival_model_draws_bef_exp$yhat_bef_exp)
yhat_weibull_sort_aft_exp = sort(stan_weibull_survival_model_draws_aft_exp$yhat_aft_exp)
yhat_weibull_sort_ear = sort(stan_weibull_survival_model_draws_ear$yhat_ear)
yhat_weibull_sort_lat = sort(stan_weibull_survival_model_draws_lat$yhat_lat)
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_bef_exp = yhat_weibull_sort_bef_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_aft_exp = yhat_weibull_sort_aft_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_ear = yhat_weibull_sort_ear[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_lat = yhat_weibull_sort_lat[1:length(Data_survival$censor) * 8000]

Data_survival$left_adj = yhat_weibull_sort_average_total_exp
Data_survival$weibull_bef_exp = yhat_weibull_sort_average_bef_exp
Data_survival$weibull_ear = yhat_weibull_sort_average_ear
Data_survival$weibull_lat = yhat_weibull_sort_average_lat

traditional_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival)
traditional_fit_2 = survfit(Surv(event = rep(1,length(Data_survival_tmp$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival_tmp)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_bef_exp)),
                               time = weibull_bef_exp) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit_1 = traditional_fit_1,
             traditional_fit_2 = traditional_fit_2,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  #data = Data_survival,
  xlab = "Time from ICI to CGP (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
  legend.labs = c("raw data",
                  "middle data",
                  "Weibull curve")
) + labs(title = paste(Cancer, " ", traditional_fit_1[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit_1)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_1)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_1)[[17]][[9]]),") (raw)/",
             as.integer(summary(traditional_fit_2)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_2)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_2)[[17]][[9]]),") (middle)\n",
             as.integer(median_os_summary_weibull_bef_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_bef_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_bef_exp[[3]]),
             ") (Weibull) days",
             sep=""))
print(g)
# pdf(file = paste(Cancer, "_weibull_SPT_to_CGP.pdf", sep=""), width = 10, height = 10)
# print(g, newpage = FALSE)
# dev.off()

traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$weibull_ear)),
                               time = weibull_ear) ~ 1, 
                               data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_lat)),
                               time = weibull_lat) ~ 1, 
                               data = Data_survival)


g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_1 = simulation_fit_1,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data",
                  "CGP enrollment at early timing, Weibull curve",
                  "CGP enrollment at late timing, Weibull curve")
) + labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (raw, early)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (raw, late)/\n",
             as.integer(median_os_summary_weibull_ear[[2]]), 
             " (", as.integer(median_os_summary_weibull_ear[[1]]), "-",
             as.integer(median_os_summary_weibull_ear[[3]]), ") (Weibull, early)/",
             as.integer(median_os_summary_weibull_lat[[2]]), " (",
             as.integer(median_os_summary_weibull_lat[[1]]), "-",
             as.integer(median_os_summary_weibull_lat[[3]]),
             ") (Weibull, late) days",
             sep=""))
print(g)

independence_check = with(Data_survival, cKendall(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  trans = "linear"))


traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ 1, 
                               data = Data_survival)
delayed_entry_fit = survfit(Surv(event = censor,
                                 time = time_palliative_enroll,
                                 time2 = time_palliative_final) ~ 1, 
                                 data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$left_adj)),
                               time = left_adj) ~ 1, 
                               data = Data_survival)
tmp_num = 18
if(length(summary(delayed_entry_fit)[[18]]) == 1){
  tmp_num = 17
}

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             delayed_entry_fit = delayed_entry_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from ICI to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Number at risk adjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + 

  labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[7]]), " (",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[8]]),"-",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[9]]),") (risk adj.)/\n",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(delayed_entry_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (risk adj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
print(g)


# Survival analysis for enrollment timing
traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ enrollment, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$enrollment)) == 5){
  print(traditional_fit)
  legend_enroll =  c("2019-06-01 to 2020-06-30",
                     "2020-07-01 to 2021-06-30",
                     "2021-07-01 to 2021-12-31",
                     "2022-01-01 to 2022-06-30",
                     "2022-07-01 to 2023-02-20")
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from ICI to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.90),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = legend_enroll
  ) + 
    labs(title = paste(Cancer, "Enrollment timing for", 
                       sum((traditional_fit)[[1]]), "patients"))
  print(g)
  # pdf(file = paste(Cancer, "_entry_period.pdf", sep=""), width = 10, height = 10)
  # print(g, newpage = FALSE)
  # dev.off()
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  print(traditional_fit)
  ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from ICI to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.85),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100
  ) + labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((traditional_fit)[[1]]), "patients"))
}

g = NULL
for(i in 1:length(target_pathway)){
  Data_report_tmp = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[[i]]][[1]])
  ID_path = unique(Data_report_tmp$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival = Data_survival %>%
    dplyr::mutate(gene_select = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  mut_gene = paste(target_pathway[i], " mutation", sep="")
  file.create(paste(target_pathway[i], ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "SPT to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "SPT to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "SPT to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from SPT to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(target_pathway[i], ".tmp", sep=""))
}
g

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_option,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment sequence**") %>%
  modify_caption("**Table X. ICI (after GEM) to final observation**") %>%
  bold_labels()
}
```

```{r STAD_ERBB2, eval = (Cancer == "Stomach" & Subanalysis == 1), echo = (Source_code == 1 & Cancer == "Stomach" & Subanalysis == 1), warning=FALSE}
ID_ERBB2 = unique((Data_report %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% c("ERBB2") & C.CAT調査結果.エビデンス.エビデンスレベル == "F"))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_wt = unique((Data_report %>% dplyr::filter(!C.CAT調査結果.変異情報.マーカー %in% c("ERBB2")))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_amp = unique((Data_report %>% dplyr::filter(tmp %in% c("ERBB2 _ amplification")))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_mut = unique((Data_report %>% dplyr::filter(!tmp %in% c("ERBB2 _ amplification")) %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% c("ERBB2") & C.CAT調査結果.エビデンス.エビデンスレベル == "F"))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_mut_only = ID_ERBB2[!ID_ERBB2 %in% ID_ERBB2_amp]

Data_ERBB2 = Data_case_target %>% dplyr::filter(症例.がん種情報.食道.胃.腸.HER2.名称. %in% c("陽性（3+）",   "境界域（2+）", "陰性（1+）", "陰性"))
tmp1 = (Data_ERBB2 %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp1$ERBB2_amplification = "+"
tmp1$ERBB2_mutation = "+"
tmp2 = (Data_ERBB2 %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp2$ERBB2_amplification = "-"
tmp2$ERBB2_mutation = "+"
tmp3 = (Data_ERBB2 %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp3$ERBB2_amplification = "+"
tmp3$ERBB2_mutation = "-"
tmp4 = (Data_ERBB2 %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp4$ERBB2_amplification = "-"
tmp4$ERBB2_mutation = "-"
Data_ERBB2_concordance = rbind(tmp1, tmp2, tmp3, tmp4)

Data_ERBB2_concordance = Data_ERBB2_concordance %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
Data_ERBB2_concordance = Data_ERBB2_concordance %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_ERBB2_concordance$症例.管理情報.登録日)) %>%
  dplyr::mutate(症例.転帰情報.死亡日 =
                    as.Date(Data_ERBB2_concordance$症例.転帰情報.死亡日)) %>%
  dplyr::mutate(症例.転帰情報.最終生存確認日 =
                    as.Date(Data_ERBB2_concordance$症例.転帰情報.最終生存確認日)) %>%
  dplyr::mutate(final_observe = as.Date(Data_ERBB2_concordance$final_observe)) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)

tmp1 = Data_ERBB2 %>%
  dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trastuzumab") %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                  症例.EP前レジメン情報.化学療法レジメン名称,
                  症例.EP前レジメン情報.投与開始日,
                  症例.EP前レジメン情報.投与終了日,
                  症例.EP前レジメン情報.最良総合効果.名称.) %>%
  dplyr::distinct()
tmp1 = tmp1 %>%
  dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                    as.Date(tmp1$症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::mutate(症例.EP前レジメン情報.投与終了日 =
                    as.Date(tmp1$症例.EP前レジメン情報.投与終了日))
tmp1 = tmp1 %>%
  dplyr::mutate(ERBB2_inhibitor_length = 
                  as.integer(difftime(tmp1$症例.EP前レジメン情報.投与終了日,
                                      tmp1$症例.EP前レジメン情報.投与開始日,
                                      units = "days"))) %>%
  dplyr::arrange(症例.EP前レジメン情報.投与終了日) %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.化学療法レジメン名称,
                症例.EP前レジメン情報.最良総合効果.名称.,
                ERBB2_inhibitor_length)

colnames(tmp1) = c("C.CAT調査結果.基本項目.ハッシュID",
                   "レジメン",
                   "最良総合効果",
                   "ERBB2_inhibitor_length")
tmp2 = Data_ERBB2 %>%
  dplyr::filter(症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trastuzumab") %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                  症例.EP後レジメン情報.化学療法レジメン名称,
                  症例.EP後レジメン情報.投与開始日,
                  症例.EP後レジメン情報.投与終了日,
                  症例.EP後レジメン情報.最良総合効果.名称.) %>%
  dplyr::distinct()
tmp2 = tmp2 %>%
  dplyr::mutate(症例.EP後レジメン情報.投与開始日 =
                    as.Date(tmp2$症例.EP後レジメン情報.投与開始日)) %>%
  dplyr::mutate(症例.EP後レジメン情報.投与終了日 =
                    as.Date(tmp2$症例.EP後レジメン情報.投与終了日))
tmp2 = tmp2 %>%
  dplyr::mutate(ERBB2_inhibitor_length = 
                  as.integer(difftime(tmp2$症例.EP後レジメン情報.投与終了日,
                                      tmp2$症例.EP後レジメン情報.投与開始日,
                                      units = "days"))) %>%
  dplyr::arrange(症例.EP後レジメン情報.投与開始日) %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.化学療法レジメン名称,
                症例.EP後レジメン情報.最良総合効果.名称.,
                ERBB2_inhibitor_length)
colnames(tmp2) = c("C.CAT調査結果.基本項目.ハッシュID",
                   "レジメン",
                   "最良総合効果",
                   "ERBB2_inhibitor_length")
Data_ERBB2_inhibitor = rbind(tmp1, tmp2)

Data_ERBB2_inhibitor = Data_ERBB2_inhibitor %>%
  tidyr::replace_na(replace = list(ERBB2_inhibitor_length = 0))
Data_ERBB2_inhibitor_unique = Data_ERBB2_inhibitor %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)

#for(i in 1:length(Data_ERBB2_inhibitor_unique$C.CAT調査結果.基本項目.ハッシュID)){
#  tmp1 = Data_ERBB2_inhibitor %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == Data_ERBB2_inhibitor_unique$C.CAT調査結果.基本項目.ハッシュID[[i]])
#  Data_ERBB2_inhibitor_unique$ERBB2_inhibitor_length[[i]] = sum(tmp1$ERBB2_inhibitor_length)
#}

Summary_ERBB2 = left_join(Data_ERBB2_concordance, Data_ERBB2_inhibitor_unique, by="C.CAT調査結果.基本項目.ハッシュID")

Summary_ERBB2 = Summary_ERBB2 %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.転帰.名称. == "死亡" ~ 1,
    TRUE ~ 0
  )) %>%
  dplyr::mutate(
    OS_after_CGP = as.integer(difftime(Summary_ERBB2$final_observe,
                                      Summary_ERBB2$症例.管理情報.登録日,
                                      units = "days"))
  ) %>%
  dplyr::mutate(ERBB2_IHC = case_when(
    症例.がん種情報.食道.胃.腸.HER2.名称. == "陽性（3+）" ~ "陽性",
    TRUE ~ "陰性"
  )) %>%
  dplyr::mutate(ERBB2_IHC_2 = case_when(
    症例.がん種情報.食道.胃.腸.HER2.名称. == "陽性（3+）" ~ "陽性",
    症例.がん種情報.食道.胃.腸.HER2.名称. == "境界域（2+）" ~ "陽性",
    TRUE ~ "陰性"
  ))
if(SD_is_effective == 2){
  Summary_ERBB2 = Summary_ERBB2 %>%
    dplyr::mutate(effect = case_when(
      最良総合効果 == "PR" ~ 1,
      最良総合効果 == "CR" ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length > LongSD_days ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length <= LongSD_days ~ 0,
      最良総合効果 == "NE" ~ 0,
      最良総合効果 == "PD" ~ 0,
      TRUE ~ -1
    ))
} else if(SD_is_effective == 1){
  Summary_ERBB2 = Summary_ERBB2 %>%
    dplyr::mutate(effect = case_when(
      最良総合効果 == "PR" ~ 1,
      最良総合効果 == "CR" ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length > LongSD_days ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length <= LongSD_days ~ 1,
      最良総合効果 == "NE" ~ 0,
      最良総合効果 == "PD" ~ 0,
      TRUE ~ -1
    ))
} else{
  Summary_ERBB2 = Summary_ERBB2 %>%
    dplyr::mutate(effect = case_when(
      最良総合効果 == "PR" ~ 1,
      最良総合効果 == "CR" ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length > LongSD_days ~ 0,
      最良総合効果 == "SD" & ERBB2_inhibitor_length <= LongSD_days ~ 0,
      最良総合効果 == "NE" ~ 0,
      最良総合効果 == "PD" ~ 0,
      TRUE ~ -1
    ))
}

WriteXLS(Summary_ERBB2, ExcelFileName="summary_erbb2.xls")
Summary_ERBB2$all_one = 1

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  tidyr::drop_na(ERBB2_inhibitor_length) %>%
  dplyr::filter(ERBB2_inhibitor_length > 0)

Data_histogram = Summary_ERBB2_inhibitor %>% 
  dplyr::select(ERBB2_inhibitor_length, ERBB2_amplification)

g = Summary_ERBB2_inhibitor %>%
  ggplot(aes(x = ERBB2_inhibitor_length, fill = ERBB2_amplification))
g <- g + geom_histogram(position = "dodge")
g <- g + scale_fill_npg()
plot(g)
ggplot( Summary_ERBB2_inhibitor, aes( x = ERBB2_inhibitor_length, y = ..density.., color = ERBB2_amplification, fill = ERBB2_amplification ) ) + 
  geom_density( alpha = 0.5 ) 

####
    diff_0 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_amplification, data=Summary_ERBB2_inhibitor, rho=0)
    diff_1 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_amplification, data=Summary_ERBB2_inhibitor, rho=1)
    survival_simple = survfit(Surv(event = all_one,
                                   time = ERBB2_inhibitor_length) ~ ERBB2_amplification, 
                                   data = Summary_ERBB2_inhibitor)
    ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Summary_ERBB2_inhibitor,
    xlab = paste("Time from trastuzumab start date (days)"),
    ylab = "Survival Probability",
    censor = TRUE,
    pval = FALSE,
    conf.int = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(),
    legend = c(0.7,0.9),
    xlim = c(0, max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) * 1.05),
    break.x.by = ceiling(max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) / 500) * 100
  ) + labs(title = paste(Cancer, " trastuzumab duration,"),
        subtitle = paste("Median OS(95%CI), ERBB2_amplification(-), ", 
                         summary(survival_simple)[[18]][[13]], "(",
                         summary(survival_simple)[[18]][[15]], "-",
                         summary(survival_simple)[[18]][[17]], "), ERBB2_amplification(+), ",
                         summary(survival_simple)[[18]][[14]], "(",
                         summary(survival_simple)[[18]][[16]], "-",
                         summary(survival_simple)[[18]][[18]], ")\n",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         ", Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3), sep=""))

    diff_0 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_IHC,  data=Summary_ERBB2_inhibitor, rho=0)
    diff_1 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_IHC, data=Summary_ERBB2_inhibitor, rho=1)
    survival_simple = survfit(Surv(event = all_one,
                                   time = ERBB2_inhibitor_length) ~ ERBB2_IHC, 
                                   data = Summary_ERBB2_inhibitor)
    ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Summary_ERBB2_inhibitor,
    xlab = paste("Time from trastuzumab start date (days)"),
    ylab = "Survival Probability",
    censor = TRUE,
    pval = FALSE,
    conf.int = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(),
    legend = c(0.7,0.9),
    xlim = c(0, max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) * 1.05),
    break.x.by = ceiling(max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) / 500) * 100
  ) + labs(title = paste(Cancer, " trastuzumab duration,"),
        subtitle = paste("Median OS(95%CI), ERBB2_IHC(-), ", 
                         summary(survival_simple)[[18]][[13]], "(",
                         summary(survival_simple)[[18]][[15]], "-",
                         summary(survival_simple)[[18]][[17]], "), ERBB2_IHC(+), ",
                         summary(survival_simple)[[18]][[14]], "(",
                         summary(survival_simple)[[18]][[16]], "-",
                         summary(survival_simple)[[18]][[18]], ")\n",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         ", Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3), sep=""))

if(SD_is_effective == 2){
  title_name = "CR_PR_long_SD"
} else if(SD_is_effective == 1){
  title_name = "CR_PR_SD"
} else if(SD_is_effective == 0){
  title_name = "CR_PR"
}

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1) %>%
  dplyr::filter(diagnosis != "Stomach Adenocarcinoma, NOS")
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification + diagnosis,
           data=Summary_ERBB2_inhibitor, family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "STAD_diff_and_undiff_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_diagnosis_IHC_amp.csv", sep=""))

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1)
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification,
           data=Summary_ERBB2_inhibitor,family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "All_STAD_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_IHC_amp.csv", sep=""))


Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1) %>%
  dplyr::filter(症例.がん種情報.食道.胃.腸.HER2.名称. %in%
                  c("陽性（3+）", "境界域（2+）"))
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification,
           data=Summary_ERBB2_inhibitor,family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "All_STAD_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_IHC_2_or_3_only_amp.csv", sep=""))

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1) %>%
  dplyr::filter(diagnosis == "Stomach Adenocarcinoma, differentiated")
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification,
           data=Summary_ERBB2_inhibitor,family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "STAD_differentiated_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_IHC_amp.csv", sep=""))

```

```{r Breast_ERBB2, eval = (Cancer == "Breast" & Subanalysis == 1), echo = (Source_code == 1 & Cancer == "Breast" & Subanalysis == 1), warning=FALSE}
ID_ERBB2 = unique((Data_report %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% c("ERBB2") & C.CAT調査結果.エビデンス.エビデンスレベル == "F"))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_wt = unique((Data_report %>% dplyr::filter(!C.CAT調査結果.変異情報.マーカー %in% c("ERBB2")))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_amp = unique((Data_report %>% dplyr::filter(tmp %in% c("ERBB2 _ amplification")))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_mut = unique((Data_report %>% dplyr::filter(!tmp %in% c("ERBB2 _ amplification")) %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー %in% c("ERBB2") & C.CAT調査結果.エビデンス.エビデンスレベル == "F"))$C.CAT調査結果.基本項目.ハッシュID)
ID_ERBB2_mut_only = ID_ERBB2[!ID_ERBB2 %in% ID_ERBB2_amp]

Data_ERBB2 = Data_case_target %>% dplyr::filter(症例.がん種情報.食道.胃.腸.HER2.名称. %in% c("陽性（3+）",   "境界域（2+）", "陰性（1+）", "陰性"))
tmp1 = (Data_ERBB2 %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp1$ERBB2_amplification = "+"
tmp1$ERBB2_mutation = "+"
tmp2 = (Data_ERBB2 %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp2$ERBB2_amplification = "-"
tmp2$ERBB2_mutation = "+"
tmp3 = (Data_ERBB2 %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp3$ERBB2_amplification = "+"
tmp3$ERBB2_mutation = "-"
tmp4 = (Data_ERBB2 %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_amp) %>%
    dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_ERBB2_mut) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.基本情報.性別.名称.,
                  症例.基本情報.年齢,
                  症例.背景情報.喫煙年数,
                  症例.背景情報.喫煙１日本数,
                  症例.背景情報.アルコール多飲有無.名称.,
                  症例.背景情報.ECOG.PS,
                  症例.管理情報.登録日,
                  症例.がん種情報.食道.胃.腸.HER2.名称.,
                  症例.転帰情報.転帰.名称.,
                  症例.転帰情報.最終生存確認日,
                  症例.転帰情報.死亡日,
                  diagnosis) %>%
    dplyr::distinct()
  )
tmp4$ERBB2_amplification = "-"
tmp4$ERBB2_mutation = "-"
Data_ERBB2_concordance = rbind(tmp1, tmp2, tmp3, tmp4)

Data_ERBB2_concordance = Data_ERBB2_concordance %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))
Data_ERBB2_concordance = Data_ERBB2_concordance %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_ERBB2_concordance$症例.管理情報.登録日)) %>%
  dplyr::mutate(症例.転帰情報.死亡日 =
                    as.Date(Data_ERBB2_concordance$症例.転帰情報.死亡日)) %>%
  dplyr::mutate(症例.転帰情報.最終生存確認日 =
                    as.Date(Data_ERBB2_concordance$症例.転帰情報.最終生存確認日)) %>%
  dplyr::mutate(final_observe = as.Date(Data_ERBB2_concordance$final_observe)) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)

tmp1 = Data_ERBB2 %>%
  dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "Trastuzumab") %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                  症例.EP前レジメン情報.化学療法レジメン名称,
                  症例.EP前レジメン情報.投与開始日,
                  症例.EP前レジメン情報.投与終了日,
                  症例.EP前レジメン情報.最良総合効果.名称.) %>%
  dplyr::distinct()
tmp1 = tmp1 %>%
  dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                    as.Date(tmp1$症例.EP前レジメン情報.投与開始日)) %>%
  dplyr::mutate(症例.EP前レジメン情報.投与終了日 =
                    as.Date(tmp1$症例.EP前レジメン情報.投与終了日))
tmp1 = tmp1 %>%
  dplyr::mutate(ERBB2_inhibitor_length = 
                  as.integer(difftime(tmp1$症例.EP前レジメン情報.投与終了日,
                                      tmp1$症例.EP前レジメン情報.投与開始日,
                                      units = "days"))) %>%
  dplyr::arrange(症例.EP前レジメン情報.投与終了日) %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.化学療法レジメン名称,
                症例.EP前レジメン情報.最良総合効果.名称.,
                ERBB2_inhibitor_length)

colnames(tmp1) = c("C.CAT調査結果.基本項目.ハッシュID",
                   "レジメン",
                   "最良総合効果",
                   "ERBB2_inhibitor_length")
tmp2 = Data_ERBB2 %>%
  dplyr::filter(症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == "Trastuzumab") %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                  症例.EP後レジメン情報.化学療法レジメン名称,
                  症例.EP後レジメン情報.投与開始日,
                  症例.EP後レジメン情報.投与終了日,
                  症例.EP後レジメン情報.最良総合効果.名称.) %>%
  dplyr::distinct()
tmp2 = tmp2 %>%
  dplyr::mutate(症例.EP後レジメン情報.投与開始日 =
                    as.Date(tmp2$症例.EP後レジメン情報.投与開始日)) %>%
  dplyr::mutate(症例.EP後レジメン情報.投与終了日 =
                    as.Date(tmp2$症例.EP後レジメン情報.投与終了日))
tmp2 = tmp2 %>%
  dplyr::mutate(ERBB2_inhibitor_length = 
                  as.integer(difftime(tmp2$症例.EP後レジメン情報.投与終了日,
                                      tmp2$症例.EP後レジメン情報.投与開始日,
                                      units = "days"))) %>%
  dplyr::arrange(症例.EP後レジメン情報.投与開始日) %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.化学療法レジメン名称,
                症例.EP後レジメン情報.最良総合効果.名称.,
                ERBB2_inhibitor_length)
colnames(tmp2) = c("C.CAT調査結果.基本項目.ハッシュID",
                   "レジメン",
                   "最良総合効果",
                   "ERBB2_inhibitor_length")
Data_ERBB2_inhibitor = rbind(tmp1, tmp2)

Data_ERBB2_inhibitor = Data_ERBB2_inhibitor %>%
  tidyr::replace_na(replace = list(ERBB2_inhibitor_length = 0))
Data_ERBB2_inhibitor_unique = Data_ERBB2_inhibitor %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)

#for(i in 1:length(Data_ERBB2_inhibitor_unique$C.CAT調査結果.基本項目.ハッシュID)){
#  tmp1 = Data_ERBB2_inhibitor %>% dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == Data_ERBB2_inhibitor_unique$C.CAT調査結果.基本項目.ハッシュID[[i]])
#  Data_ERBB2_inhibitor_unique$ERBB2_inhibitor_length[[i]] = sum(tmp1$ERBB2_inhibitor_length)
#}

Summary_ERBB2 = left_join(Data_ERBB2_concordance, Data_ERBB2_inhibitor_unique, by="C.CAT調査結果.基本項目.ハッシュID")

Summary_ERBB2 = Summary_ERBB2 %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.転帰.名称. == "死亡" ~ 1,
    TRUE ~ 0
  )) %>%
  dplyr::mutate(
    OS_after_CGP = as.integer(difftime(Summary_ERBB2$final_observe,
                                      Summary_ERBB2$症例.管理情報.登録日,
                                      units = "days"))
  ) %>%
  dplyr::mutate(ERBB2_IHC = case_when(
    症例.がん種情報.食道.胃.腸.HER2.名称. == "陽性（3+）" ~ "陽性",
    TRUE ~ "陰性"
  )) %>%
  dplyr::mutate(ERBB2_IHC_2 = case_when(
    症例.がん種情報.食道.胃.腸.HER2.名称. == "陽性（3+）" ~ "陽性",
    症例.がん種情報.食道.胃.腸.HER2.名称. == "境界域（2+）" ~ "陽性",
    TRUE ~ "陰性"
  ))
if(SD_is_effective == 2){
  Summary_ERBB2 = Summary_ERBB2 %>%
    dplyr::mutate(effect = case_when(
      最良総合効果 == "PR" ~ 1,
      最良総合効果 == "CR" ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length > LongSD_days ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length <= LongSD_days ~ 0,
      最良総合効果 == "NE" ~ 0,
      最良総合効果 == "PD" ~ 0,
      TRUE ~ -1
    ))
} else if(SD_is_effective == 1){
  Summary_ERBB2 = Summary_ERBB2 %>%
    dplyr::mutate(effect = case_when(
      最良総合効果 == "PR" ~ 1,
      最良総合効果 == "CR" ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length > LongSD_days ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length <= LongSD_days ~ 1,
      最良総合効果 == "NE" ~ 0,
      最良総合効果 == "PD" ~ 0,
      TRUE ~ -1
    ))
} else{
  Summary_ERBB2 = Summary_ERBB2 %>%
    dplyr::mutate(effect = case_when(
      最良総合効果 == "PR" ~ 1,
      最良総合効果 == "CR" ~ 1,
      最良総合効果 == "SD" & ERBB2_inhibitor_length > LongSD_days ~ 0,
      最良総合効果 == "SD" & ERBB2_inhibitor_length <= LongSD_days ~ 0,
      最良総合効果 == "NE" ~ 0,
      最良総合効果 == "PD" ~ 0,
      TRUE ~ -1
    ))
}

WriteXLS(Summary_ERBB2, ExcelFileName="summary_erbb2.xls")
Summary_ERBB2$all_one = 1

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  tidyr::drop_na(ERBB2_inhibitor_length) %>%
  dplyr::filter(ERBB2_inhibitor_length > 0)

Data_histogram = Summary_ERBB2_inhibitor %>% 
  dplyr::select(ERBB2_inhibitor_length, ERBB2_amplification)

g = Summary_ERBB2_inhibitor %>%
  ggplot(aes(x = ERBB2_inhibitor_length, fill = ERBB2_amplification))
g <- g + geom_histogram(position = "dodge")
g <- g + scale_fill_npg()
plot(g)
ggplot( Summary_ERBB2_inhibitor, aes( x = ERBB2_inhibitor_length, y = ..density.., color = ERBB2_amplification, fill = ERBB2_amplification ) ) + 
  geom_density( alpha = 0.5 ) 

####
    diff_0 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_amplification, data=Summary_ERBB2_inhibitor, rho=0)
    diff_1 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_amplification, data=Summary_ERBB2_inhibitor, rho=1)
    survival_simple = survfit(Surv(event = all_one,
                                   time = ERBB2_inhibitor_length) ~ ERBB2_amplification, 
                                   data = Summary_ERBB2_inhibitor)
    ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Summary_ERBB2_inhibitor,
    xlab = paste("Time from trastuzumab start date (days)"),
    ylab = "Survival Probability",
    censor = TRUE,
    pval = FALSE,
    conf.int = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(),
    legend = c(0.7,0.9),
    xlim = c(0, max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) * 1.05),
    break.x.by = ceiling(max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) / 500) * 100
  ) + labs(title = paste(Cancer, " trastuzumab time-to-treatment-failure,"),
        subtitle = paste("Median OS(95%CI), ERBB2_amplification(-), ", 
                         summary(survival_simple)[[18]][[13]], "(",
                         summary(survival_simple)[[18]][[15]], "-",
                         summary(survival_simple)[[18]][[17]], "), ERBB2_amplification(+), ",
                         summary(survival_simple)[[18]][[14]], "(",
                         summary(survival_simple)[[18]][[16]], "-",
                         summary(survival_simple)[[18]][[18]], ")\n",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         ", Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3), sep=""))

    diff_0 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_IHC,  data=Summary_ERBB2_inhibitor, rho=0)
    diff_1 = survdiff(Surv(ERBB2_inhibitor_length, all_one)~ERBB2_IHC, data=Summary_ERBB2_inhibitor, rho=1)
    survival_simple = survfit(Surv(event = all_one,
                                   time = ERBB2_inhibitor_length) ~ ERBB2_IHC, 
                                   data = Summary_ERBB2_inhibitor)
    ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Summary_ERBB2_inhibitor,
    xlab = paste("Time from trastuzumab start date (days)"),
    ylab = "Survival Probability",
    censor = TRUE,
    pval = FALSE,
    conf.int = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(),
    legend = c(0.7,0.9),
    xlim = c(0, max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) * 1.05),
    break.x.by = ceiling(max(Summary_ERBB2_inhibitor$ERBB2_inhibitor_length) / 500) * 100
  ) + labs(title = paste(Cancer, " trastuzumab time-to-treatment-failure,"),
        subtitle = paste("Median OS(95%CI), ERBB2_IHC(-), ", 
                         summary(survival_simple)[[18]][[13]], "(",
                         summary(survival_simple)[[18]][[15]], "-",
                         summary(survival_simple)[[18]][[17]], "), ERBB2_IHC(+), ",
                         summary(survival_simple)[[18]][[14]], "(",
                         summary(survival_simple)[[18]][[16]], "-",
                         summary(survival_simple)[[18]][[18]], ")\n",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         ", Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3), sep=""))

if(SD_is_effective == 2){
  title_name = "CR_PR_long_SD"
} else if(SD_is_effective == 1){
  title_name = "CR_PR_SD"
} else if(SD_is_effective == 0){
  title_name = "CR_PR"
}

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1) %>%
  dplyr::filter(diagnosis != "Stomach Adenocarcinoma, NOS")
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification + diagnosis,
           data=Summary_ERBB2_inhibitor, family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "STAD_diff_and_undiff_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_diagnosis_IHC_amp.csv", sep=""))

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1)
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification,
           data=Summary_ERBB2_inhibitor,family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "All_STAD_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_IHC_amp.csv", sep=""))


Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1) %>%
  dplyr::filter(症例.がん種情報.食道.胃.腸.HER2.名称. %in%
                  c("陽性（3+）", "境界域（2+）"))
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification,
           data=Summary_ERBB2_inhibitor,family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "All_STAD_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_IHC_2_or_3_only_amp.csv", sep=""))

Summary_ERBB2_inhibitor = Summary_ERBB2 %>%
  dplyr::filter(effect != -1) %>%
  dplyr::filter(diagnosis == "Stomach Adenocarcinoma, differentiated")
ans <- glm(Summary_ERBB2_inhibitor$effect~ ERBB2_IHC + ERBB2_amplification,
           data=Summary_ERBB2_inhibitor,family=binomial)
s.ans = summary(ans)
coe <- s.ans$coefficient
RR <- exp(coe[,1])
RRlow <- exp(coe[,1]-1.96*coe[,2])
RRup <- exp(coe[,1]+1.96*coe[,2])
N <- nrow(data)
aic <- AIC(ans)
result <- cbind(coe,RR,RRlow,RRup,aic,N)
colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
write.csv(result, file = paste(
  "STAD_differentiated_trastuzumab_effective_indicates_", title_name,
  "_logistic_regression_IHC_amp.csv", sep=""))

```

```{r diagnosis, eval = (Analyze_diagnosis == 1), echo = (Source_code == 1 & Analyze_diagnosis == 1), warning=FALSE}
## Survival analysis considering left truncation
Data_survival = Data_case_target %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))

Data_survival = Data_survival %>%
  dplyr::filter(症例.背景情報.診断日 != "") %>%
  dplyr::filter(!is.na(症例.背景情報.診断日))
Data_survival = Data_survival %>%
  dplyr::arrange(症例.背景情報.診断日) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))
Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
  Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日)) %>%
  dplyr::mutate(症例.背景情報.診断日 =
                    as.Date(Data_survival$症例.背景情報.診断日))

Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.背景情報.診断日,
                                units = "days")))

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
            as.integer(difftime(Data_survival$症例.管理情報.登録日,
                                Data_survival$症例.背景情報.診断日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "SPT to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "SPT to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "SPT to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "SPT to entry 3-4 year",
    TRUE ~ "SPT to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)
Median_entry_raw = Median_entry
Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "diagnosis to entry early",
    TRUE ~ "diagnosis to entry later"))

g = ggplot(Data_survival, aes(x = time_enroll_final))
g = g + geom_histogram(binwidth = 50, colour = "gray10", fill = "royalblue")
g = g + theme_classic()
g = g + ggtitle(paste(Cancer, "Distribution of time from enrollment to final observation (days)"))
#plot(g)
pdf(file = paste(OutDirName, Cancer, "_entry_to_final_observation.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

ID_target = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID) 

## Target gene selection
Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.エビデンス.エビデンスレベル,
                  .keep_all=TRUE) %>%
  dplyr::mutate(gene_mut = paste(C.CAT調査結果.変異情報.マーカー,
                                 C.CAT調査結果.変異情報.変異内容,
                                 sep=" _ ")) %>%
  dplyr::mutate(gene_patho = C.CAT調査結果.変異情報.マーカー)

oncogenic_mutations = Data_major_gene %>%
                dplyr::count(gene_mut) %>%
                dplyr::arrange(-n)

c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_mutations
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_mut)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_mut)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_mutations = cbind(oncogenic_mutations, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_mutations) =c_name_gene

DT::datatable(oncogenic_mutations, filter = 'top', extensions = 'Scroller',
             caption = "Oncogenic mutations, diagnosis to death analysis")

oncogenic_genes = Data_major_gene %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations, diagnosis to death analysis (counts)")

oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
  dplyr::distinct() %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
      dplyr::distinct() %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations, diagnosis to death analysis (patients)")

Data_tmp_pre = Data_case_target %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
                ) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_name")
Data_tmp_post = Data_case_target %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_name")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "") %>%
  dplyr::distinct() %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target)
for(i in 1:length(Data_drug_sequence[,1])){
  if(length(str_split(Data_drug_sequence[i,2], ",")[[1]]) > 1){
    for(j in 2:length(str_split(Data_drug_sequence[i,2], ",")[[1]])){
      data_drug_correction = Data_drug_sequence[i,]
      data_drug_correction[1,2] = str_split(Data_drug_sequence[i,2], ",")[[1]][[j]]
      Data_drug_sequence = rbind(Data_drug_sequence, data_drug_correction)
    }
    Data_drug_sequence[i,2] = str_split(Data_drug_sequence[i,2], ",")[[1]][[1]]
  }
}
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::distinct()

drug_list = unique(Data_drug_sequence$drug_name)
if(length(which(drug_list %in% c("", "Unknown"))) > 0){
  drug_list = drug_list[-which(drug_list %in% c("", "Unknown"))]
}
if (length(drug_list[-which(is.na(drug_list))] > 0)){
  drug_list = drug_list[-which(is.na(drug_list))]
}
drug_list = unique(drug_list)

drug_list = data.frame(drug_list)
drug_list$patients = 0

for(i in 1:length(drug_list[,1])){
  drug_list[i,2] = length((Data_drug_sequence %>% dplyr::filter(
    drug_name == drug_list[i,1]))$C.CAT調査結果.基本項目.ハッシュID)
}
drug_list = drug_list %>%
  dplyr::arrange(-patients)

DT::datatable(drug_list,
             caption = "Used drugs, diagnosis to death analysis (patients)", filter = 'top', extensions = 'Scroller')


# Performance status and delayed entry
traditional_fit = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 症例.背景情報.ECOG.PS,
                               data = Data_survival)
print(traditional_fit)
g = ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from diagnosis to entry",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
) + 
  labs(title = paste(Cancer, "PS and delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
pdf(file = paste(OutDirName, Cancer, "_diag_entry_performance_status.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()


# Survival analysis for delayed entry - early half vs late half
traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$delay_1)) == 2){
  print(traditional_fit)
  g=ggsurvplot(
    fit = traditional_fit,
    data = Data_survival,
    xlab = "Time from test entry (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = TRUE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.88,1.05),
    xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
    legend.labs = paste("",
                        c("diagnosis to entry early",
                          "diagnosis to entry later")
                          , sep="")
  ) + 
  labs(title = paste(Cancer, " ", 
                     sum((traditional_fit)[[1]]), " patients, median interval:", Median_entry_raw, "days", sep=""),
        subtitle = paste("Median OS (95%) ",
             summary(traditional_fit)[[18]][[13]],
             " (", summary(traditional_fit)$table[[15]],
             "-", summary(traditional_fit)$table[[17]], ") (early)/",
             summary(traditional_fit)[[18]][[14]],
             " (", summary(traditional_fit)$table[[16]],
             "-", summary(traditional_fit)$table[[18]], ") (later) days",
             sep=""))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_diag_entry_timing_2_group.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

time_tmp = quantile(Data_survival$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
time_10 = as.integer(time_tmp[[1]])
time_90 = as.integer(time_tmp[[2]])
time_100 = as.integer(time_tmp[[3]])
Data_survival_tmp = Data_survival %>% dplyr::filter(
  time_palliative_enroll >= time_10 &
    time_palliative_enroll <= time_90)
Data_survival_tmp$time_palliative_enroll =
  Data_survival_tmp$time_palliative_enroll - (time_10 - 1)

Data_survival_tmp2 = Data_survival %>% dplyr::filter(
  time_palliative_enroll > time_90 &
    time_palliative_enroll <= time_100)
Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
idx <- 1:nrow(Data_survival_tmp2)
Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
Median_entry = median(Data_survival_tmp$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "diagnosis to entry early",
    TRUE ~ "diagnosis to entry later"))

stan_weibull_survival_model_data <-
    list(
        Median = as.integer(Median_entry),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 1,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        ## Times for CGP testing
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1]),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0])
        )


stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1)

# pdf(file = paste(OutDirName, Cancer, "_diag_loglog_HMC.pdf", sep=""), width = 10, height = 10)
# rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape", "scale", "scale_exp", "shape_exp"))
# mcmc_rhat(rhat(stan_weibull_survival_model_fit))
# dev.off()

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_bef_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
    tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_aft_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
    tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_ear <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
    tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_lat <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
    tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
    dplyr::select(-key) 


median_os_list_weibull_total_exp = NULL
median_os_list_weibull_bef_exp = NULL
median_os_list_weibull_aft_exp = NULL
median_os_list_weibull_ear = NULL
median_os_list_weibull_lat = NULL

for(i in 1:8000){
  idx = seq(i, length(stan_weibull_survival_model_draws_lat$yhat_lat), by=8000)
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_bef_exp = c(median_os_list_weibull_bef_exp,
      median(stan_weibull_survival_model_draws_bef_exp[idx,]$yhat_bef_exp, na.rm = T))
    median_os_list_weibull_aft_exp = c(median_os_list_weibull_aft_exp,
      median(stan_weibull_survival_model_draws_aft_exp[idx,]$yhat_aft_exp, na.rm = T))
    median_os_list_weibull_ear = c(median_os_list_weibull_ear,
      median(stan_weibull_survival_model_draws_ear[idx,]$yhat_ear, na.rm = T))
    median_os_list_weibull_lat = c(median_os_list_weibull_lat,
      median(stan_weibull_survival_model_draws_lat[idx,]$yhat_lat, na.rm = T))
}

median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bef_exp = quantile(median_os_list_weibull_bef_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_aft_exp = quantile(median_os_list_weibull_aft_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_ear = quantile(median_os_list_weibull_ear, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_lat = quantile(median_os_list_weibull_lat, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_bef_exp = sort(stan_weibull_survival_model_draws_bef_exp$yhat_bef_exp)
yhat_weibull_sort_aft_exp = sort(stan_weibull_survival_model_draws_aft_exp$yhat_aft_exp)
yhat_weibull_sort_ear = sort(stan_weibull_survival_model_draws_ear$yhat_ear)
yhat_weibull_sort_lat = sort(stan_weibull_survival_model_draws_lat$yhat_lat)
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_bef_exp = yhat_weibull_sort_bef_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_aft_exp = yhat_weibull_sort_aft_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_ear = yhat_weibull_sort_ear[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_lat = yhat_weibull_sort_lat[1:length(Data_survival$censor) * 8000]

Data_survival$left_adj = yhat_weibull_sort_average_total_exp
Data_survival$weibull_bef_exp = yhat_weibull_sort_average_bef_exp
Data_survival$weibull_ear = yhat_weibull_sort_average_ear
Data_survival$weibull_lat = yhat_weibull_sort_average_lat

traditional_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival)
traditional_fit_2 = survfit(Surv(event = rep(1,length(Data_survival_tmp$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival_tmp)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_bef_exp)),
                               time = weibull_bef_exp) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit_1 = traditional_fit_1,
             traditional_fit_2 = traditional_fit_2,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  xlab = "Time from diagnosis to CGP (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
  legend.labs = c("raw data",
                  "middle data",
                  "Weibull curve")
) + labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit_1)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_1)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_1)[[17]][[9]]),") (raw)/",
             as.integer(summary(traditional_fit_2)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_2)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_2)[[17]][[9]]),") (middle)\n",
             as.integer(median_os_summary_weibull_bef_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_bef_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_bef_exp[[3]]),
             ") (Weibull) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_weibull_SPT_to_CGP.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$weibull_ear)),
                               time = weibull_ear) ~ 1, 
                               data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_lat)),
                               time = weibull_lat) ~ 1, 
                               data = Data_survival)


g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_1 = simulation_fit_1,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data",
                  "CGP enrollment at early timing, Weibull curve",
                  "CGP enrollment at late timing, Weibull curve")
) + labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (raw, early)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (raw, late)/\n",
             as.integer(median_os_summary_weibull_ear[[2]]), 
             " (", as.integer(median_os_summary_weibull_ear[[1]]), "-",
             as.integer(median_os_summary_weibull_ear[[3]]), ") (Weibull, early)/",
             as.integer(median_os_summary_weibull_lat[[2]]), " (",
             as.integer(median_os_summary_weibull_lat[[1]]), "-",
             as.integer(median_os_summary_weibull_lat[[3]]),
             ") (Weibull, late) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_loglog_CGP_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

#check Kendall's tau
#p-value < 0.05 -> quasi-independence of truncation and death is rejected
independence_check = with(Data_survival, trSurvfit(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  tFun = "linear",
  plots = FALSE))
#independence_check
pdf(file = paste(OutDirName, Cancer, "_diag_tranSurv.pdf", sep=""), width = 10, height = 10)
plot(independence_check)
dev.off()

if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to positive dependent-left truncation.")
  print("Patients with CGP test at early timing lived short survival.")
} else if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to negative dependent-left truncation.")
  print("Patients with CGP test at early timing lived long survival.")
} else{
  print("This survival data was subjected to independent-left truncation.")
}

independence_check = with(Data_survival, cKendall(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  trans = "linear"))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ 1, 
                               data = Data_survival)
delayed_entry_fit = survfit(Surv(event = censor,
                                 time = time_palliative_enroll,
                                 time2 = time_palliative_final) ~ 1, 
                                 data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$left_adj)),
                               time = left_adj) ~ 1, 
                               data = Data_survival)
tmp_num = 18
if(length(summary(delayed_entry_fit)[[18]]) == 1){
  tmp_num = 17
}

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             delayed_entry_fit = delayed_entry_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from diagnosis to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Number at risk adjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[7]]), " (",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[8]]),"-",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[9]]),") (risk adj.)/\n",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(delayed_entry_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (risk adj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_loglog_SPT_to_death_3curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from diagnosis to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS\n",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_loglog_SPT_to_death_2curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

# Survival analysis for enrollment timing
traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ enrollment, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$enrollment)) == 5){
  print(traditional_fit)
  legend_enroll =  c("2019-06-01 to 2020-06-30",
                     "2020-07-01 to 2021-06-30",
                     "2021-07-01 to 2021-12-31",
                     "2022-01-01 to 2022-06-30",
                     "2022-07-01 to 2023-02-20")
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from diagnosis to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.90),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = legend_enroll
  ) + 
    labs(title = paste(Cancer, "Enrollment timing for", 
                       sum((traditional_fit)[[1]]), "patients"))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_entry_period_diagnosis.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  #print(traditional_fit)
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from diagnosis to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.7,0.85),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100
  ) + labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((traditional_fit)[[1]]), "patients"))
  pdf(file = paste(OutDirName, Cancer, "_entry_period_diagnosis.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()

}


# analysis for common oncogenic mutations
ID_prog = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)
oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho)
oncogenic_genes = oncogenic_genes %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
  dplyr::distinct() %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_survival %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
      dplyr::distinct() %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

Data_survival$target_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_target_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:length(target_gene)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  } else{
    for(k in 1:length(target_gene)){
      if(target_gene[[k]][2] == "WT"){
        Data_survival$target_gene_count[i] = bitwOr(
          Data_survival$target_gene_count[i], 2^(k-1))
      }
    }
  }
}

if(sum(Data_survival$target_gene_count) == 0){
  target_gene = list(c(oncogenic_genes[1,1], "F"))
  gene_pattern = list(c(1))
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.変異情報.変異内容 == target_gene[[i]][2]),
      Data_target_gene)
  }
  Data_target_gene = Data_target_gene %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=TRUE)
  
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[i]][2]) %>%
      dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                      C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT,
                    .keep_all=TRUE),
      Data_target_gene)
  }
  Data_survival$target_gene_count = 0
  ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
  for(i in 1:length(ID_target_gene)){
    tmp = Data_target_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
    if(dplyr::count(tmp)[[1]] > 0){
      for(j in 1:dplyr::count(tmp)[[1]]){
        for(k in 1:length(target_gene)){
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
        }
      }
    } else{
      for(k in 1:length(target_gene)){
        if(target_gene[[k]][2] == "WT"){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

Data_survival$oncogenic_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_major_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:min(length(oncogenic_genes$all_patients), gene_number)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == oncogenic_genes$gene_mutation[[k]]){
          Data_survival$oncogenic_gene_count[i] = bitwOr(
            Data_survival$oncogenic_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

g = NULL
gene_table = data.frame(oncogenic_genes$gene_mutation[1:min(length(oncogenic_genes$all_patients), gene_number)])
colnames(gene_table) = c("Gene")
gene_table$cancer_type = Cancer
gene_table$positive_patients = oncogenic_genes$all_patients[1:min(length(oncogenic_genes$all_patients), gene_number)]
gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$delay)) / 10
gene_table$positive_median = 0
gene_table$positive_LL = 0
gene_table$positive_UL = 0
gene_table$negative_median = 0
gene_table$negative_LL = 0
gene_table$negative_UL = 0
gene_table$diff_median = 0
gene_table$diff_LL = 0
gene_table$diff_UL = 0
gene_table$positive_mortal_event = 0
gene_table$negative_mortal_event = 0
  
for(i in 1:min(length(oncogenic_genes$all_patients), gene_number)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    Data_survival$gene_select[[j]] = bitwAnd(
        Data_survival$oncogenic_gene_count[j], 2^(i-1))
  }
  Data_survival$gene_select = as.logical(Data_survival$gene_select)
  mut_gene = paste(oncogenic_genes$gene_mutation[i],
                   ", top ", i, " gene with oncogenic mutation", sep="")
  file.create(paste(OutDirName,"top_", i, "_gene.tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    gene_table$negative_mortal_event[i] = summary(traditional_fit)[[18]][7]
    gene_table$positive_mortal_event[i] = summary(traditional_fit)[[18]][8]
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
    
    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry later"))
    
    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "diagnosis to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  #g=c(g, list(
  #rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  #))
  #g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from diagnosis to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
      gene_table$positive_median[i] = as.integer(median_os_summary_weibull_total_exp[[2]])
      gene_table$positive_LL[i] = as.integer(median_os_summary_weibull_total_exp[[1]])
      gene_table$positive_UL[i] = as.integer(median_os_summary_weibull_total_exp[[3]])
      gene_table$negative_median[i] = as.integer(median_os_summary_weibull_total[[2]])
      gene_table$negative_LL[i] = as.integer(median_os_summary_weibull_total[[1]])
      gene_table$negative_UL[i] = as.integer(median_os_summary_weibull_total[[3]])
      gene_table$diff_median[i] = -as.integer(median_os_summary_weibull_bias[[2]])
      gene_table$diff_LL[i] = -as.integer(median_os_summary_weibull_bias[[3]])
      gene_table$diff_UL[i] = -as.integer(median_os_summary_weibull_bias[[1]])

    }
  }
  file.remove(paste(OutDirName,"top_", i, "_gene.tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_freq_gene_diagnosis_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

write.csv(gene_table, file = paste(OutDirName, Cancer, "FreqGene_diag_loglog_",  length(Data_survival$症例.基本情報.年齢), "_patients.csv", sep=""))

Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))
p <- 
  gene_table |>
  ggplot(aes(y = fct_rev(Gene))) + 
  theme_classic()
p <- p +
  geom_point(aes(x=diff_median), shape=15, size=3) +
  geom_linerange(aes(xmin=diff_LL, xmax=diff_UL)) 
p <- p +
  geom_vline(xintercept = 0, linetype="dashed") +
  labs(x="Survival difference by mutation", y="Genes with oncogenic alteration")
p <- p +
  coord_cartesian(ylim=c(1,length(Gene_arrange) + 1), xlim=c(min(gene_table$diff_LL) - 15, max(gene_table$diff_UL) + 15))
p <- p +
  annotate("text", x = min(gene_table$diff_LL)/2 - 15, y = length(Gene_arrange) + 1, label = "mut=short survival") +
  annotate("text", x = max(gene_table$diff_UL)/2 + 15, y = length(Gene_arrange) + 1, label = "mut=long survival")
p_mid <- p + 
  theme(axis.line.y = element_blank(),
        axis.ticks.y= element_blank(),
        axis.text.y= element_blank(),
        axis.title.y= element_blank())

gene_table <- gene_table %>%
  dplyr::mutate(
  estimate_lab = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
  dplyr::mutate(
    patients = paste0(positive_patients, " (", positive_freq, "%)"))
gene_table = 
  dplyr::bind_rows(
    data.frame(
      Gene = "Gene",
      estimate_lab = "Survival difference (95% CI)",
      patients = "Patients"
    ), gene_table  )
Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))

p_left <-
  gene_table  %>%
  ggplot(aes(y = fct_rev(Gene)))
p_left <- 
  p_left +
  geom_text(aes(x = 0, label = Gene), hjust = 0, fontface = "bold")
p_left <- 
  p_left +
  geom_text(
    aes(x = 1, label = estimate_lab),
    hjust = 0,
    fontface = ifelse(gene_table$estimate_lab == "Survival difference (95% CI)", "bold", "plain")
  )
p_left <-
  p_left +
  theme_void() +
  coord_cartesian(xlim = c(0, 4))

# right side of plot - pvalues
p_right <-
  gene_table  |>
  ggplot() +
  geom_text(
    aes(x = 0, y = fct_rev(Gene), label = patients),
    hjust = 0,
    fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")
  ) +
  theme_void() 

layout <- c(
  area(t = 0, l = 0, b = 30, r = 3), # left plot, starts at the top of the page (0) and goes 30 units down and 3 units to the right
  area(t = 1, l = 4, b = 30, r = 9), # middle plot starts a little lower (t=1) because there's no title. starts 1 unit right of the left plot (l=4, whereas left plot is r=3), goes to the bottom of the page (30 units), and 6 units further over from the left plot (r=9 whereas left plot is r=3)
  area(t = 0, l = 9, b = 30, r = 11) # right most plot starts at top of page, begins where middle plot ends (l=9, and middle plot is r=9), goes to bottom of page (b=30), and extends two units wide (r=11)
)
# final plot arrangement
pdf(file = paste(OutDirName, Cancer, "_diag_to_final_Gene_forest-plot.pdf", sep=""), width = 10, height = 10)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()

for(i in 1:length(gene_pattern)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    for(k in 1:length(gene_pattern[[i]])){
      Data_survival$gene_select[[j]] = any(Data_survival$gene_select[[j]],
          (bitwAnd(gene_pattern[[i]][[k]],
                   Data_survival$target_gene_count[j]) ==
             gene_pattern[[i]][[k]]))
    }
  }
  mut_gene = ""
  for(l in 1:length(gene_pattern[[i]])){
    for(m in 1:length(target_gene)){
      if(bitwAnd(2^(m-1), gene_pattern[[i]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  file.create(paste(mut_gene, ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "diagnosis to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from diagnosis to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(mut_gene, ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_gene_diagnosis_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

g = NULL
for(i in 1:length(target_pathway)){
  Data_report_tmp = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[[i]]][[1]])
  ID_path = unique(Data_report_tmp$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival = Data_survival %>%
    dplyr::mutate(gene_select = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  mut_gene = paste(target_pathway[i], " mutation", sep="")
  file.create(paste(target_pathway[i], ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    time_tmp = quantile(Data_survival_tmp_pos$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp2$time_palliative_enroll = Data_survival_tmp2$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp2)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx %% 10 != 1],])
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    time_tmp = quantile(Data_survival_tmp_neg$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
    time_10 = as.integer(time_tmp[[1]])
    time_90 = as.integer(time_tmp[[2]])
    time_100 = as.integer(time_tmp[[3]])
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10 &
        time_palliative_enroll <= time_90)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll > time_90 &
        time_palliative_enroll <= time_100)
    Data_survival_tmp4$time_palliative_enroll = Data_survival_tmp4$time_palliative_enroll - (time_10 - 1)
    idx <- 1:nrow(Data_survival_tmp4)
    Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx %% 10 != 1],])
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "diagnosis to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from diagnosis to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(target_pathway[i], ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_pathway_diagnosis_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. Patients with selected disease, diagnosis-death, Characteristics**") %>%
  bold_labels()
table_tmp

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_option,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment option exsisted**") %>%
  modify_caption("**Table 9. Diagnosis to final observation - Expert panel option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 10. Enrollment to final observation - Recommended treatment done**") %>%
  bold_labels()
}

Data_summary_treat = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) %>%
  dplyr::filter(EP_option == 1)

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
survival_simple_ = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(survival_simple_[[1]][[1]] != survival_simple[[1]][[1]] & length(survival_simple_) == length(survival_simple)){
Data_summary_treat %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 2''. Diagnosis to final observation - Recommended treatment done, only patients with treatment option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~diagnosis,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = diagnosis,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 3,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 3)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_caption("**Table 11. Diagnosis to final observation - diagnosis**") %>%
  bold_labels()
}

Report_lolliplot_diagnosis = Report_lolliplot %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_survival$C.CAT調査結果.基本項目.ハッシュID)
write.csv(Report_lolliplot_diagnosis, file = paste(OutDirName, Cancer, "_only_pts_with_diagnosis_to_death_info_", target_gene[[1]][[1]], ".csv", sep=""))

Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F")
oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, C.CAT調査結果.変異情報.マーカー) %>%
  dplyr::distinct() %>%
                dplyr::count(C.CAT調査結果.変異情報.マーカー) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "All patients", "Selected disease with target gene mut", "Selected disease without target gene mut", "Other diseases")
tmp_no = oncogenic_genes
Data_case_tmp = Data_case %>%
  dplyr::mutate(diagnosis = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% Data_case_target$C.CAT調査結果.基本項目.ハッシュID &
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_major_gene %>%
          dplyr::filter(C.CAT調査結果.変異情報.マーカー ==
            target_gene[[1]][[1]]))$C.CAT調査結果.基本項目.ハッシュID ~ "Selected disease with target gene mutation",
    C.CAT調査結果.基本項目.ハッシュID %in% Data_case_target$C.CAT調査結果.基本項目.ハッシュID ~ "Selected disease without target gene mutation",
    TRUE ~ "Other diseases"
  ))
disease_list = c("Selected disease with target gene mutation", "Selected disease without target gene mutation", "Other diseases")
if(length(unique(Data_case_tmp$diagnosis)) == 3){
  for(i in 1:3){
    tmp_no[,2] = 0
    ID_target_tmp = unique((Data_case_tmp %>% dplyr::filter(diagnosis == disease_list[[i]]))
                           $C.CAT調査結果.基本項目.ハッシュID) 
    tmp = Data_major_gene %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, C.CAT調査結果.変異情報.マーカー) %>%
        dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
        dplyr::distinct() %>%
                  dplyr::count(C.CAT調査結果.変異情報.マーカー)
    tmp_no$n = lapply(tmp_no[,1],
                                 function(k) {
      as.vector(tmp$n[match(k,
                tmp$C.CAT調査結果.変異情報.マーカー)])
    })
    tmp_no[is.na(tmp_no)] <- 0
    oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  }
  colnames(oncogenic_genes) =c_name_gene
  
  DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
               caption = "Genes with oncogenic mutations in entire cohorts (patients)")
}

```

```{r diagnosis_new, eval = (Analyze_diagnosis == 2), echo = (Source_code == 1 & Analyze_diagnosis == 2), warning=FALSE}
## Survival analysis considering left truncation
Data_survival = Data_case_target %>%
  dplyr::mutate(final_observe = case_when(
    症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
    症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
    症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
    TRUE ~ ""))

Data_survival = Data_survival %>%
  dplyr::filter(症例.背景情報.診断日 != "") %>%
  dplyr::filter(!is.na(症例.背景情報.診断日))
Data_survival = Data_survival %>%
  dplyr::arrange(症例.背景情報.診断日) %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
  dplyr::filter(症例.管理情報.登録日 != "") %>%
  dplyr::filter(final_observe != "9999-99-99"  & final_observe != "          "& final_observe != "          ") %>%
  dplyr::mutate(censor = case_when(
    症例.転帰情報.死亡日 != "" ~ 1,
    TRUE ~ 0))
Data_survival = Data_survival %>%
  dplyr::mutate(enrollment = case_when(
    Data_survival$症例.管理情報.登録日 <= "2020-06-30" ~ "2019-06-01 to 2020-06-30",
    Data_survival$症例.管理情報.登録日 >= "2020-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-06-30" ~ "2020-07-01 to 2021-06-30",
    Data_survival$症例.管理情報.登録日 >= "2021-07-01" &
    Data_survival$症例.管理情報.登録日 <= "2021-12-31" ~ "2021-07-01 to 2021-12-31",
    Data_survival$症例.管理情報.登録日 >= "2022-01-01" &
    Data_survival$症例.管理情報.登録日 <= "2022-06-30" ~ "2022-01-01 to 2022-06-30",
    TRUE ~ "2022-07-01 to 2023-02-20"))
  Data_survival = Data_survival %>%
  dplyr::mutate(multi_cancer = case_when(
    is.na(症例.背景情報.重複がん有無.異なる臓器..名称.) ~ 0,
    症例.背景情報.重複がん有無.異なる臓器..名称. == "なし" ~ 0,
    TRUE ~ 1))
if(Ex_inactive_multi == 1){
  Data_survival =  Data_survival %>% 
  dplyr::mutate(multi_cancer = case_when(
    multi_cancer == 0 ~ 0,
    multi_cancer == 1 & is.na(症例.背景情報.重複がん.活動性) ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 2 ~ 0,
    multi_cancer == 1 & 症例.背景情報.重複がん.活動性 == 9 ~ 0,
    TRUE ~ 1))
}

Data_survival = Data_survival %>%
  dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
  dplyr::mutate(症例.管理情報.登録日 =
                    as.Date(Data_survival$症例.管理情報.登録日)) %>%
  dplyr::mutate(症例.背景情報.診断日 =
                    as.Date(Data_survival$症例.背景情報.診断日))

Data_survival = Data_survival %>%
  dplyr::mutate(time_enroll_final =
                  as.integer(difftime(Data_survival$final_observe,
                                      Data_survival$症例.管理情報.登録日,
                                      units = "days")))

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_final =
            as.integer(difftime(Data_survival$final_observe,
                                Data_survival$症例.背景情報.診断日,
                                units = "days")))

Data_survival = Data_survival %>%
  dplyr::mutate(time_palliative_enroll =
            as.integer(difftime(Data_survival$症例.管理情報.登録日,
                                Data_survival$症例.背景情報.診断日,
                                units = "days")))
Data_survival = Data_survival %>%
  dplyr::mutate(EP_option = case_when(
    症例.EP後レジメン情報.EPの結果治療薬の選択肢が提示された.名称. ==
      "はい" ~ 1,
    TRUE ~ 0)) %>%
  dplyr::mutate(EP_treat = case_when(
    症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
      "投与した" ~ 1,
    TRUE ~ 0))

Data_survival$症例.基本情報.年齢_entry =
  Data_survival$症例.基本情報.年齢 - as.integer(Data_survival$time_palliative_enroll/365)
Data_survival = Data_survival %>%
  dplyr::filter(time_enroll_final > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll > 0)
Data_survival = Data_survival %>%
  dplyr::filter(time_palliative_enroll < 10000)

if(Entry_period < 4){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-07-01 to 2023-02-20")
}
if(Entry_period < 3){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2022-01-01 to 2022-06-30")
}
if(Entry_period < 2){
  Data_survival = Data_survival %>%
    dplyr::filter(enrollment != "2021-07-01 to 2021-12-31")
}

Data_survival = Data_survival %>%
  dplyr::mutate(delay = case_when(
    Data_survival$time_palliative_enroll <= 365 ~ "SPT to entry 0-1 year",
    Data_survival$time_palliative_enroll >= 366 &
    Data_survival$time_palliative_enroll <= 730 ~ "SPT to entry 1-2 year",
    Data_survival$time_palliative_enroll >= 731 &
    Data_survival$time_palliative_enroll <= 1095 ~ "SPT to entry 2-3 year",
    Data_survival$time_palliative_enroll >= 1096 &
    Data_survival$time_palliative_enroll <= 1460 ~ "SPT to entry 3-4 year",
    TRUE ~ "SPT to entry 4- year"))

Median_entry = median(Data_survival$time_palliative_enroll)
Median_entry_raw = Median_entry

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "diagnosis to entry early",
    TRUE ~ "diagnosis to entry later"))

g = ggplot(Data_survival, aes(x = time_enroll_final))
g = g + geom_histogram(binwidth = 50, colour = "gray10", fill = "royalblue")
g = g + theme_classic()
g = g + ggtitle(paste(Cancer, "Distribution of time from enrollment to final observation (days)"))
#plot(g)
pdf(file = paste(OutDirName, Cancer, "_entry_to_final_observation.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

ID_target = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID) 

## Target gene selection
Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target) %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F") %>%
  dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                  C.CAT調査結果.変異情報.マーカー,
                  C.CAT調査結果.変異情報.変異内容,
                  C.CAT調査結果.エビデンス.エビデンスレベル,
                  .keep_all=TRUE) %>%
  dplyr::mutate(gene_mut = paste(C.CAT調査結果.変異情報.マーカー,
                                 C.CAT調査結果.変異情報.変異内容,
                                 sep=" _ ")) %>%
  dplyr::mutate(gene_patho = C.CAT調査結果.変異情報.マーカー)

oncogenic_mutations = Data_major_gene %>%
                dplyr::count(gene_mut) %>%
                dplyr::arrange(-n)

c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_mutations
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_mut)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_mut)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_mutations = cbind(oncogenic_mutations, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_mutations) =c_name_gene

DT::datatable(oncogenic_mutations, filter = 'top', extensions = 'Scroller',
             caption = "Oncogenic mutations, diagnosis to death analysis")

oncogenic_genes = Data_major_gene %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations, diagnosis to death analysis (counts)")

oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
  dplyr::distinct() %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_case_target %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
      dplyr::distinct() %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
             caption = "Genes with oncogenic mutations, diagnosis to death analysis (patients)")

Data_tmp_pre = Data_case_target %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
                ) %>%
  dplyr::distinct()
colnames(Data_tmp_pre) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_name")
Data_tmp_post = Data_case_target %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                症例.EP後レジメン情報.薬剤名.YJ一般名.EN.) %>%
  dplyr::distinct()
colnames(Data_tmp_post) = c("C.CAT調査結果.基本項目.ハッシュID",
                           "drug_name")
Data_drug_sequence = rbind(Data_tmp_pre, Data_tmp_post)
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::filter(!is.na(drug_name)) %>%
  dplyr::filter(drug_name != "") %>%
  dplyr::distinct() %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target)
for(i in 1:length(Data_drug_sequence[,1])){
  if(length(str_split(Data_drug_sequence[i,2], ",")[[1]]) > 1){
    for(j in 2:length(str_split(Data_drug_sequence[i,2], ",")[[1]])){
      data_drug_correction = Data_drug_sequence[i,]
      data_drug_correction[1,2] = str_split(Data_drug_sequence[i,2], ",")[[1]][[j]]
      Data_drug_sequence = rbind(Data_drug_sequence, data_drug_correction)
    }
    Data_drug_sequence[i,2] = str_split(Data_drug_sequence[i,2], ",")[[1]][[1]]
  }
}
Data_drug_sequence = Data_drug_sequence %>%
  dplyr::distinct()

drug_list = unique(Data_drug_sequence$drug_name)
if(length(which(drug_list %in% c("", "Unknown"))) > 0){
  drug_list = drug_list[-which(drug_list %in% c("", "Unknown"))]
}
if (length(drug_list[-which(is.na(drug_list))] > 0)){
  drug_list = drug_list[-which(is.na(drug_list))]
}
drug_list = unique(drug_list)

drug_list = data.frame(drug_list)
drug_list$patients = 0

for(i in 1:length(drug_list[,1])){
  drug_list[i,2] = length((Data_drug_sequence %>% dplyr::filter(
    drug_name == drug_list[i,1]))$C.CAT調査結果.基本項目.ハッシュID)
}
drug_list = drug_list %>%
  dplyr::arrange(-patients)

DT::datatable(drug_list,
             caption = "Used drugs, diagnosis to death analysis (patients)", filter = 'top', extensions = 'Scroller')

# Performance status and delayed entry
traditional_fit = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 症例.背景情報.ECOG.PS,
                               data = Data_survival)
print(traditional_fit)
g = ggsurvplot(
  fit = traditional_fit,
  data = Data_survival,
  xlab = "Time (Days) from diagnosis to entry",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  pval = TRUE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.7,0.85),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
) + 
  labs(title = paste(Cancer, "PS and delayed entry for", 
                     sum((traditional_fit)[[1]]), "patients"))
pdf(file = paste(OutDirName, Cancer, "_diag_entry_performance_status.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()


# Survival analysis for delayed entry - early half vs late half
traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$delay_1)) == 2){
  print(traditional_fit)
  g=ggsurvplot(
    fit = traditional_fit,
    data = Data_survival,
    xlab = "Time from test entry (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = TRUE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.88,1.05),
    xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
    legend.labs = paste("",
                        c("diagnosis to entry early",
                          "diagnosis to entry later")
                          , sep="")
  ) + 
  labs(title = paste(Cancer, " ", 
                     sum((traditional_fit)[[1]]), " patients, median interval:", Median_entry_raw, "days", sep=""),
        subtitle = paste("Median OS (95%) ",
             summary(traditional_fit)[[18]][[13]],
             " (", summary(traditional_fit)$table[[15]],
             "-", summary(traditional_fit)$table[[17]], ") (early)/",
             summary(traditional_fit)[[18]][[14]],
             " (", summary(traditional_fit)$table[[16]],
             "-", summary(traditional_fit)$table[[18]], ") (later) days",
             sep=""))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_diag_entry_timing_2_group.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
}

time_tmp = quantile(Data_survival$time_palliative_enroll,  probs = c(0.10, 0.90, 1.00))
Data_survival_tmp = Data_survival %>% dplyr::filter(
  time_palliative_enroll >= time_10)
Data_survival_tmp$time_palliative_enroll =
  Data_survival_tmp$time_palliative_enroll - (time_10 - 1)

Median_entry = median(Data_survival_tmp$time_palliative_enroll)

Data_survival = Data_survival %>%
  dplyr::mutate(delay_1 = case_when(
    Data_survival$time_palliative_enroll <= Median_entry ~ "diagnosis to entry early",
    TRUE ~ "diagnosis to entry later"))

stan_weibull_survival_model_data <-
    list(
        Median = as.integer(Median_entry),
        ## Number of event individuals
        Nobs = sum(Data_survival$censor == 1),
        ## Number of censored individuals
        Ncen = sum(Data_survival$censor == 0),
        ## Number of total individuals
        Ntot = length(Data_survival$censor),
        ## Number of covariates
        M_bg = 1,
        Nexp = length(Data_survival_tmp$time_palliative_enroll),
        ## Times for CGP testing
        ybef_exp = Data_survival_tmp$time_palliative_enroll,
        ## Times for CGP testing
        ## Times for event individuals
        yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
        ## Times for censored individuals
        ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
        ## Covariates for event individuals as a matrix
        Xobs_bg = matrix(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1]),
        ## Covariates for censored individuals as a matrix
        Xcen_bg = matrix(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0])
        )


stan_weibull_survival_model_fit <-
    rstan::stan(file =
                  "source/Stan_code_loglog_weibull.stan",
                data = stan_weibull_survival_model_data,
                control=list(adapt_delta=0.99, max_treedepth = 10),
                seed=123, chains=4, iter=3000, warmup=1000, thin=1)

# pdf(file = paste(OutDirName, Cancer, "_diag_loglog_HMC.pdf", sep=""), width = 10, height = 10)
# rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape", "scale", "scale_exp", "shape_exp"))
# mcmc_rhat(rhat(stan_weibull_survival_model_fit))
# dev.off()

stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
stan_weibull_survival_model_draws_total_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_bef_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
    tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_aft_exp <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
    tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_ear <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
    tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
    dplyr::select(-key) 
stan_weibull_survival_model_draws_lat <-
    stan_weibull_survival_model_draws %>%
    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
    tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
    dplyr::select(-key) 


median_os_list_weibull_total_exp = NULL
median_os_list_weibull_bef_exp = NULL
median_os_list_weibull_aft_exp = NULL
median_os_list_weibull_ear = NULL
median_os_list_weibull_lat = NULL

for(i in 1:8000){
  idx = seq(i, length(stan_weibull_survival_model_draws_lat$yhat_lat), by=8000)
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
    median_os_list_weibull_bef_exp = c(median_os_list_weibull_bef_exp,
      median(stan_weibull_survival_model_draws_bef_exp[idx,]$yhat_bef_exp, na.rm = T))
    median_os_list_weibull_aft_exp = c(median_os_list_weibull_aft_exp,
      median(stan_weibull_survival_model_draws_aft_exp[idx,]$yhat_aft_exp, na.rm = T))
    median_os_list_weibull_ear = c(median_os_list_weibull_ear,
      median(stan_weibull_survival_model_draws_ear[idx,]$yhat_ear, na.rm = T))
    median_os_list_weibull_lat = c(median_os_list_weibull_lat,
      median(stan_weibull_survival_model_draws_lat[idx,]$yhat_lat, na.rm = T))
}

median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_bef_exp = quantile(median_os_list_weibull_bef_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_aft_exp = quantile(median_os_list_weibull_aft_exp, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_ear = quantile(median_os_list_weibull_ear, probs = c(0.025, 0.5, 0.975))
median_os_summary_weibull_lat = quantile(median_os_list_weibull_lat, probs = c(0.025, 0.5, 0.975))
yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
yhat_weibull_sort_bef_exp = sort(stan_weibull_survival_model_draws_bef_exp$yhat_bef_exp)
yhat_weibull_sort_aft_exp = sort(stan_weibull_survival_model_draws_aft_exp$yhat_aft_exp)
yhat_weibull_sort_ear = sort(stan_weibull_survival_model_draws_ear$yhat_ear)
yhat_weibull_sort_lat = sort(stan_weibull_survival_model_draws_lat$yhat_lat)
yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_bef_exp = yhat_weibull_sort_bef_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_aft_exp = yhat_weibull_sort_aft_exp[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_ear = yhat_weibull_sort_ear[1:length(Data_survival$censor) * 8000]
yhat_weibull_sort_average_lat = yhat_weibull_sort_lat[1:length(Data_survival$censor) * 8000]

Data_survival$left_adj = yhat_weibull_sort_average_total_exp
Data_survival$weibull_bef_exp = yhat_weibull_sort_average_bef_exp
Data_survival$weibull_ear = yhat_weibull_sort_average_ear
Data_survival$weibull_lat = yhat_weibull_sort_average_lat

traditional_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival)
traditional_fit_2 = survfit(Surv(event = rep(1,length(Data_survival_tmp$time_palliative_enroll)),
                               time = time_palliative_enroll) ~ 1, 
                               data = Data_survival_tmp)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_bef_exp)),
                               time = weibull_bef_exp) ~ 1, 
                               data = Data_survival)

g = ggsurvplot(
  fit = list(traditional_fit_1 = traditional_fit_1,
             traditional_fit_2 = traditional_fit_2,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  xlab = "Time from diagnosis to CGP (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_enroll) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_enroll) / 500) * 100,
  legend.labs = c("raw data",
                  "middle data",
                  "Weibull curve")
) + labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit_1)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_1)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_1)[[17]][[9]]),") (raw)/",
             as.integer(summary(traditional_fit_2)[[17]][[7]])," (",
             as.integer(summary(traditional_fit_2)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit_2)[[17]][[9]]),") (middle)\n",
             as.integer(median_os_summary_weibull_bef_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_bef_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_bef_exp[[3]]),
             ") (Weibull) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_weibull_SPT_to_CGP.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

traditional_fit = survfit(Surv(event = censor,
                               time = time_enroll_final) ~ delay_1, 
                               data = Data_survival)
simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$weibull_ear)),
                               time = weibull_ear) ~ 1, 
                               data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$weibull_lat)),
                               time = weibull_lat) ~ 1, 
                               data = Data_survival)


g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_1 = simulation_fit_1,
             simulation_fit_2 = simulation_fit_2),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from CGP testing to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_enroll_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_enroll_final) / 500) * 100,
  legend.labs = c("CGP enrollment at early timing, raw data",
                  "CGP enrollment at late timing, raw data",
                  "CGP enrollment at early timing, Weibull curve",
                  "CGP enrollment at late timing, Weibull curve")
) + labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[18]][[13]])," (",
             as.integer(summary(traditional_fit)[[18]][[15]]),"-",
             as.integer(summary(traditional_fit)[[18]][[17]]),") (raw, early)/",
             as.integer(summary(traditional_fit)[[18]][[14]]), " (",
             as.integer(summary(traditional_fit)[[18]][[16]]),"-",
             as.integer(summary(traditional_fit)[[18]][[18]]),") (raw, late)/\n",
             as.integer(median_os_summary_weibull_ear[[2]]), 
             " (", as.integer(median_os_summary_weibull_ear[[1]]), "-",
             as.integer(median_os_summary_weibull_ear[[3]]), ") (Weibull, early)/",
             as.integer(median_os_summary_weibull_lat[[2]]), " (",
             as.integer(median_os_summary_weibull_lat[[1]]), "-",
             as.integer(median_os_summary_weibull_lat[[3]]),
             ") (Weibull, late) days",
             sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_loglog_CGP_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

#check Kendall's tau
#p-value < 0.05 -> quasi-independence of truncation and death is rejected
independence_check = with(Data_survival, trSurvfit(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  tFun = "linear",
  plots = FALSE))
#independence_check
pdf(file = paste(OutDirName, Cancer, "_diag_tranSurv.pdf", sep=""), width = 10, height = 10)
plot(independence_check)
dev.off()

if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to positive dependent-left truncation.")
  print("Patients with CGP test at early timing lived short survival.")
} else if(independence_check$iniP < 0.05 & independence_check$iniKendall > 0){
  print("This survival data was subjected to negative dependent-left truncation.")
  print("Patients with CGP test at early timing lived long survival.")
} else{
  print("This survival data was subjected to independent-left truncation.")
}

independence_check = with(Data_survival, cKendall(
  trun = time_palliative_enroll,
  obs = time_palliative_final,
  delta = censor,
  trans = "linear"))

traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ 1, 
                               data = Data_survival)
delayed_entry_fit = survfit(Surv(event = censor,
                                 time = time_palliative_enroll,
                                 time2 = time_palliative_final) ~ 1, 
                                 data = Data_survival)
simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$left_adj)),
                               time = left_adj) ~ 1, 
                               data = Data_survival)
tmp_num = 18
if(length(summary(delayed_entry_fit)[[18]]) == 1){
  tmp_num = 17
}

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             delayed_entry_fit = delayed_entry_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from diagnosis to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Number at risk adjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS ",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[7]]), " (",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[8]]),"-",
             as.integer(summary(delayed_entry_fit)[[tmp_num]][[9]]),") (risk adj.)/\n",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(delayed_entry_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (risk adj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_loglog_SPT_to_death_3curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

g = ggsurvplot(
  fit = list(traditional_fit = traditional_fit,
             simulation_fit_2 = simulation_fit_2
             ),
  combine = TRUE,
  data = Data_survival,
  xlab = "Time from diagnosis to final observation (days)",
  ylab = "Survival Probability",
  censor = TRUE,
  conf.int = FALSE,
  risk.table = TRUE,
  risk.table.y.text = FALSE,
  tables.theme = clean_theme(), 
  legend = c(0.8,0.90),
  xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
  break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
  legend.labs = c("All patients, Unadjusted",
                  "All patients, Adj. right cens. & left trunc.")
) + labs(title = paste(Cancer, " ", traditional_fit[[1]], " patients ",
              "Median OS\n",
             as.integer(summary(traditional_fit)[[17]][[7]])," (",
             as.integer(summary(traditional_fit)[[17]][[8]]),"-",
             as.integer(summary(traditional_fit)[[17]][[9]]),") (unadj.)/",
             as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
             as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
             as.integer(median_os_summary_weibull_total_exp[[3]]),
             ") (adj. right & left) days",
             sep=""),
      subtitle = paste(k_year, "-year rate ",
             format(summary(traditional_fit, time = k_year * 365)$surv * 100,digits=3),
             "% (unadj.)/",
             format(summary(simulation_fit_2, time = k_year * 365)$surv * 100,digits=3),
            "% (adj. right & left)\n",
            "cKendall tau= ", format(independence_check$PE,digits=2),
            ", SE= ", format(independence_check$SE,digits=2),
            ", p= ", format(independence_check$p.value,digits=2), sep=""))
#print(g)
pdf(file = paste(OutDirName, Cancer, "_diag_loglog_SPT_to_death_2curves.pdf", sep=""), width = 10, height = 10)
print(g, newpage = FALSE)
dev.off()

# Survival analysis for enrollment timing
traditional_fit = survfit(Surv(event = censor,
                               time = time_palliative_final) ~ enrollment, 
                               data = Data_survival)
if(length(Data_survival[,1]) != traditional_fit[[1]][[1]] &
   length(unique(Data_survival$enrollment)) == 5){
  print(traditional_fit)
  legend_enroll =  c("2019-06-01 to 2020-06-30",
                     "2020-07-01 to 2021-06-30",
                     "2021-07-01 to 2021-12-31",
                     "2022-01-01 to 2022-06-30",
                     "2022-07-01 to 2023-02-20")
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from diagnosis to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.68,0.90),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
    legend.labs = legend_enroll
  ) + 
    labs(title = paste(Cancer, "Enrollment timing for", 
                       sum((traditional_fit)[[1]]), "patients"))
  #print(g)
  pdf(file = paste(OutDirName, Cancer, "_entry_period_diagnosis.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()
} else if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
  #print(traditional_fit)
  g = ggsurvplot(
    fit = list(traditional_fit = traditional_fit),
    combine = TRUE,
    data = Data_survival,
    xlab = "Time from diagnosis to final observation (days)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    pval = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.7,0.85),
    xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
    break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100
  ) + labs(title = paste(Cancer, "Enrollment timing for", 
                     sum((traditional_fit)[[1]]), "patients"))
  pdf(file = paste(OutDirName, Cancer, "_entry_period_diagnosis.pdf", sep=""), width = 10, height = 10)
  print(g, newpage = FALSE)
  dev.off()

}


# analysis for common oncogenic mutations
ID_prog = unique(Data_survival$C.CAT調査結果.基本項目.ハッシュID)
oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho)
oncogenic_genes = oncogenic_genes %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_prog) %>%
  dplyr::distinct() %>%
                dplyr::count(gene_patho) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "all_patients")
tmp_no = oncogenic_genes
for(i in 1:length(Diseases)){
  tmp_no[,2] = 0
  Data_case_tmp = Data_survival %>%
                     dplyr::filter(diagnosis %in% Diseases[[i]])
  ID_target_tmp = unique(Data_case_tmp$C.CAT調査結果.基本項目.ハッシュID) 
  tmp = Data_major_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID, gene_patho) %>%
      dplyr::distinct() %>%
                dplyr::count(gene_patho)
  tmp_no$n = lapply(tmp_no[,1],
                               function(k) {
    as.vector(tmp$n[match(k,
              tmp$gene_patho)])
  })
  tmp_no[is.na(tmp_no)] <- 0
  oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  c_name_gene = c(c_name_gene, Diseases[[i]])
}
colnames(oncogenic_genes) =c_name_gene

Data_survival$target_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_target_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:length(target_gene)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
           tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  } else{
    for(k in 1:length(target_gene)){
      if(target_gene[[k]][2] == "WT"){
        Data_survival$target_gene_count[i] = bitwOr(
          Data_survival$target_gene_count[i], 2^(k-1))
      }
    }
  }
}

if(sum(Data_survival$target_gene_count) == 0){
  target_gene = list(c(oncogenic_genes[1,1], "F"))
  gene_pattern = list(c(1))
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.変異情報.変異内容 == target_gene[[i]][2]),
      Data_target_gene)
  }
  Data_target_gene = Data_target_gene %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                    C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    .keep_all=TRUE)
  
  for(i in 1:length(target_gene)){
    Data_target_gene = rbind(
      Data_report %>%
      dplyr::filter(C.CAT調査結果.変異情報.マーカー == target_gene[[i]][1] &
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[i]][2]) %>%
      dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                      C.CAT調査結果.変異情報.マーカー,
                    C.CAT調査結果.変異情報.変異内容,
                    C.CAT調査結果.エビデンス.エビデンスレベル_WT,
                    .keep_all=TRUE),
      Data_target_gene)
  }
  Data_survival$target_gene_count = 0
  ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
  for(i in 1:length(ID_target_gene)){
    tmp = Data_target_gene %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
    if(dplyr::count(tmp)[[1]] > 0){
      for(j in 1:dplyr::count(tmp)[[1]]){
        for(k in 1:length(target_gene)){
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.変異情報.変異内容 == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
          if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == target_gene[[k]][1] &
             tmp[j,]$C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[k]][2]){
            Data_survival$target_gene_count[i] = bitwOr(
              Data_survival$target_gene_count[i], 2^(k-1))
          }
        }
      }
    } else{
      for(k in 1:length(target_gene)){
        if(target_gene[[k]][2] == "WT"){
          Data_survival$target_gene_count[i] = bitwOr(
            Data_survival$target_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

Data_survival$oncogenic_gene_count = 0
ID_target_gene = Data_survival$C.CAT調査結果.基本項目.ハッシュID
for(i in 1:length(ID_target_gene)){
  tmp = Data_major_gene %>%
    dplyr::filter(C.CAT調査結果.基本項目.ハッシュID == ID_target_gene[i])
  if(dplyr::count(tmp)[[1]] > 0){
    for(j in 1:dplyr::count(tmp)[[1]]){
      for(k in 1:min(length(oncogenic_genes$all_patients), gene_number)){
        if(tmp[j,]$C.CAT調査結果.変異情報.マーカー == oncogenic_genes$gene_mutation[[k]]){
          Data_survival$oncogenic_gene_count[i] = bitwOr(
            Data_survival$oncogenic_gene_count[i], 2^(k-1))
        }
      }
    }
  }
}

g = NULL
gene_table = data.frame(oncogenic_genes$gene_mutation[1:min(length(oncogenic_genes$all_patients), gene_number)])
colnames(gene_table) = c("Gene")
gene_table$cancer_type = Cancer
gene_table$positive_patients = oncogenic_genes$all_patients[1:min(length(oncogenic_genes$all_patients), gene_number)]
gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$delay)) / 10
gene_table$positive_median = 0
gene_table$positive_LL = 0
gene_table$positive_UL = 0
gene_table$negative_median = 0
gene_table$negative_LL = 0
gene_table$negative_UL = 0
gene_table$diff_median = 0
gene_table$diff_LL = 0
gene_table$diff_UL = 0
gene_table$positive_mortal_event = 0
gene_table$negative_mortal_event = 0
  
for(i in 1:min(length(oncogenic_genes$all_patients), gene_number)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    Data_survival$gene_select[[j]] = bitwAnd(
        Data_survival$oncogenic_gene_count[j], 2^(i-1))
  }
  Data_survival$gene_select = as.logical(Data_survival$gene_select)
  mut_gene = paste(oncogenic_genes$gene_mutation[i],
                   ", top ", i, " gene with oncogenic mutation", sep="")
  file.create(paste(OutDirName, "top_", i, "_gene.tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    gene_table$negative_mortal_event[i] = summary(traditional_fit)[[18]][7]
    gene_table$positive_mortal_event[i] = summary(traditional_fit)[[18]][8]
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)
    
    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry later"))
    
    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "diagnosis to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file = "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  #g=c(g, list(
  #rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  #))
  #g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from diagnosis to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
      gene_table$positive_median[i] = as.integer(median_os_summary_weibull_total_exp[[2]])
      gene_table$positive_LL[i] = as.integer(median_os_summary_weibull_total_exp[[1]])
      gene_table$positive_UL[i] = as.integer(median_os_summary_weibull_total_exp[[3]])
      gene_table$negative_median[i] = as.integer(median_os_summary_weibull_total[[2]])
      gene_table$negative_LL[i] = as.integer(median_os_summary_weibull_total[[1]])
      gene_table$negative_UL[i] = as.integer(median_os_summary_weibull_total[[3]])
      gene_table$diff_median[i] = -as.integer(median_os_summary_weibull_bias[[2]])
      gene_table$diff_LL[i] = -as.integer(median_os_summary_weibull_bias[[3]])
      gene_table$diff_UL[i] = -as.integer(median_os_summary_weibull_bias[[1]])

    }
  }
  file.remove(paste(OutDirName, "top_", i, "_gene.tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_freq_gene_diagnosis_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

write.csv(gene_table, file = paste(OutDirName, Cancer, "FreqGene_diag_loglog_",  length(Data_survival$症例.基本情報.年齢), "_patients.csv", sep=""))

Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))
p <- 
  gene_table |>
  ggplot(aes(y = fct_rev(Gene))) + 
  theme_classic()
p <- p +
  geom_point(aes(x=diff_median), shape=15, size=3) +
  geom_linerange(aes(xmin=diff_LL, xmax=diff_UL)) 
p <- p +
  geom_vline(xintercept = 0, linetype="dashed") +
  labs(x="Survival difference by mutation", y="Genes with oncogenic alteration")
p <- p +
  coord_cartesian(ylim=c(1,length(Gene_arrange) + 1), xlim=c(min(gene_table$diff_LL) - 15, max(gene_table$diff_UL) + 15))
p <- p +
  annotate("text", x = min(gene_table$diff_LL)/2 - 15, y = length(Gene_arrange) + 1, label = "mut=short survival") +
  annotate("text", x = max(gene_table$diff_UL)/2 + 15, y = length(Gene_arrange) + 1, label = "mut=long survival")
p_mid <- p + 
  theme(axis.line.y = element_blank(),
        axis.ticks.y= element_blank(),
        axis.text.y= element_blank(),
        axis.title.y= element_blank())

gene_table <- gene_table %>%
  dplyr::mutate(
  estimate_lab = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
  dplyr::mutate(
    patients = paste0(positive_patients, " (", positive_freq, "%)"))
gene_table = 
  dplyr::bind_rows(
    data.frame(
      Gene = "Gene",
      estimate_lab = "Survival difference (95% CI)",
      patients = "Patients"
    ), gene_table  )
Gene_arrange = gene_table$Gene
gene_table = gene_table %>% dplyr::mutate(
  Gene = factor(gene_table$Gene, levels = Gene_arrange))

p_left <-
  gene_table  %>%
  ggplot(aes(y = fct_rev(Gene)))
p_left <- 
  p_left +
  geom_text(aes(x = 0, label = Gene), hjust = 0, fontface = "bold")
p_left <- 
  p_left +
  geom_text(
    aes(x = 1, label = estimate_lab),
    hjust = 0,
    fontface = ifelse(gene_table$estimate_lab == "Survival difference (95% CI)", "bold", "plain")
  )
p_left <-
  p_left +
  theme_void() +
  coord_cartesian(xlim = c(0, 4))

# right side of plot - pvalues
p_right <-
  gene_table  |>
  ggplot() +
  geom_text(
    aes(x = 0, y = fct_rev(Gene), label = patients),
    hjust = 0,
    fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")
  ) +
  theme_void() 

layout <- c(
  area(t = 0, l = 0, b = 30, r = 3), # left plot, starts at the top of the page (0) and goes 30 units down and 3 units to the right
  area(t = 1, l = 4, b = 30, r = 9), # middle plot starts a little lower (t=1) because there's no title. starts 1 unit right of the left plot (l=4, whereas left plot is r=3), goes to the bottom of the page (30 units), and 6 units further over from the left plot (r=9 whereas left plot is r=3)
  area(t = 0, l = 9, b = 30, r = 11) # right most plot starts at top of page, begins where middle plot ends (l=9, and middle plot is r=9), goes to bottom of page (b=30), and extends two units wide (r=11)
)
# final plot arrangement
pdf(file = paste(OutDirName, Cancer, "_diag_to_final_Gene_forest-plot.pdf", sep=""), width = 10, height = 10)
p_left + p_mid + p_right + plot_layout(design = layout)
dev.off()

for(i in 1:length(gene_pattern)){
  Data_survival$gene_select = FALSE
  for(j in 1:length(Data_survival$gene_select)){
    for(k in 1:length(gene_pattern[[i]])){
      Data_survival$gene_select[[j]] = any(Data_survival$gene_select[[j]],
          (bitwAnd(gene_pattern[[i]][[k]],
                   Data_survival$target_gene_count[j]) ==
             gene_pattern[[i]][[k]]))
    }
  }
  mut_gene = ""
  for(l in 1:length(gene_pattern[[i]])){
    for(m in 1:length(target_gene)){
      if(bitwAnd(2^(m-1), gene_pattern[[i]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  file.create(paste(mut_gene, ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "diagnosis to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from diagnosis to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(mut_gene, ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_gene_diagnosis_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

g = NULL
for(i in 1:length(target_pathway)){
  Data_report_tmp = Data_report %>%
    dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F" &
                  C.CAT調査結果.変異情報.マーカー %in% PATH[target_pathway[[i]]][[1]])
  ID_path = unique(Data_report_tmp$C.CAT調査結果.基本項目.ハッシュID)
  Data_survival = Data_survival %>%
    dplyr::mutate(gene_select = case_when(
      C.CAT調査結果.基本項目.ハッシュID %in% ID_path ~ 1,
      TRUE ~ 0
    ))
  mut_gene = paste(target_pathway[i], " mutation", sep="")
  file.create(paste(target_pathway[i], ".tmp", sep=""))
  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select, 
                                 data = Data_survival)
  if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
    Data_survival_tmp_pos = Data_survival %>%
      dplyr::filter(gene_select == TRUE)
    Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp$time_palliative_enroll =
      Data_survival_tmp$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_pos = median(Data_survival_tmp$time_palliative_enroll)

    Data_survival_tmp_neg = Data_survival %>%
      dplyr::filter(gene_select == FALSE)
    Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      time_palliative_enroll >= time_10)
    Data_survival_tmp3$time_palliative_enroll =
      Data_survival_tmp3$time_palliative_enroll - (time_10 - 1)
    
    Median_entry_neg = median(Data_survival_tmp3$time_palliative_enroll)
    Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp3)  

    Data_survival = Data_survival %>%
      dplyr::mutate(delay_1 = case_when(
        Data_survival$time_palliative_enroll <= Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_pos &
          gene_select == TRUE ~ "diagnosis to entry later",
        Data_survival$time_palliative_enroll <= Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry early",
        Data_survival$time_palliative_enroll > Median_entry_neg &
          gene_select == FALSE ~ "diagnosis to entry later"))

    if(sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 == "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == TRUE & delay_1 != "diagnosis to entry later"))$censor) >= 1 &
       sum((Data_survival %>% dplyr::filter(gene_select == FALSE & delay_1 != "diagnosis to entry later"))$censor) >= 1){

      stan_weibull_survival_model_data <-
        list(
          Median_pos = as.integer(Median_entry_pos),
          Median_neg = as.integer(Median_entry_neg),
          ## Number of event individuals
          Nobs = sum(Data_survival$censor == 1),
          ## Number of censored individuals
          Ncen = sum(Data_survival$censor == 0),
          ## Number of total individuals
          Ntot = length(Data_survival$censor),
          ## Number of covariates
          M_bg = 2,
          Nexp = length(Data_survival_tmp$time_palliative_enroll),
          ## Times for CGP testing
          ybef_exp = Data_survival_tmp$time_palliative_enroll,
          xfactor = as.numeric(Data_survival_tmp$gene_select),
          ## Times for event individuals
          yobs = Data_survival$time_enroll_final[Data_survival$censor == 1],
          ## Times for censored individuals
          ycen = Data_survival$time_enroll_final[Data_survival$censor == 0],
          ## Covariates for event individuals as a matrix
          Xobs_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 1],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 1]),ncol = 2),
          ## Covariates for censored individuals as a matrix
          Xcen_bg = matrix(c(as.numeric(Data_survival$delay_1 == "diagnosis to entry later")[Data_survival$censor == 0],
                           as.numeric(Data_survival$gene_select)[Data_survival$censor == 0]),ncol = 2)
          )

  stan_weibull_survival_model_fit <-
      rstan::stan(file =
                    #"Stan_code_weibull_weibull_factor.stan",
                    "source/Stan_code_loglog_weibull_factor.stan",
                  data = stan_weibull_survival_model_data,
                  control=list(adapt_delta=0.99, max_treedepth = 10),
                  seed=12, chains=4, iter=3000, warmup=1000, thin=1,
                  pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
                           "yhat_total", "yhat_factor_total"))

  # g=c(g, list(
  # rstan::traceplot(stan_weibull_survival_model_fit, par = c("alpha","mu", "shape_exp", "scale_exp", "factor"))
  # ))
  # g=c(g, list(mcmc_rhat(rhat(stan_weibull_survival_model_fit))))

  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
  stan_weibull_survival_model_draws_total <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      dplyr::select(-key) 
  stan_weibull_survival_model_draws_total_exp <-
      stan_weibull_survival_model_draws %>%
      dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      dplyr::select(-key) 
  
  median_os_list_weibull_total = NULL
  median_os_list_weibull_total_exp = NULL
  for(j in 1:8000){
  idx = seq(j, length(stan_weibull_survival_model_draws_total$yhat_total), by=8000)
    median_os_list_weibull_total = c(median_os_list_weibull_total,
      median(stan_weibull_survival_model_draws_total[idx,]$yhat_total, na.rm = T))
    median_os_list_weibull_total_exp = c(median_os_list_weibull_total_exp,
      median(stan_weibull_survival_model_draws_total_exp[idx,]$yhat_total_exp, na.rm = T))
  }
  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp


  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * 8000]
  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * 8000]

  Data_survival$factor_neg = yhat_weibull_sort_average_total
  Data_survival$factor_pos = yhat_weibull_sort_average_total_exp
  Data_survival$factor_neg[Data_survival$factor_neg > 10000] = 10000
  Data_survival$factor_pos[Data_survival$factor_pos > 10000] = 10000

  traditional_fit = survfit(Surv(event = censor,
                                 time = time_palliative_final) ~ gene_select,
                                 data = Data_survival)
  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival$factor_neg)),
                                 time = factor_neg) ~ 1, 
                                 data = Data_survival)
  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival$factor_pos)),
                                 time = factor_pos) ~ 1, 
                                 data = Data_survival)
  
  g=c(g, list(ggsurvplot(
      fit = list(traditional_fit = traditional_fit,
                 simulation_fit_neg = simulation_fit_neg,
                 simulation_fit_pos = simulation_fit_pos),
      combine = TRUE,
      data = Data_survival,
      xlab = "Time from diagnosis to final observation (days)",
      ylab = "Survival Probability",
      censor = TRUE,
      conf.int = FALSE,
      pval = FALSE,
      risk.table = TRUE,
      risk.table.y.text = FALSE,
      tables.theme = clean_theme(), 
      legend = c(0.8,0.8),
      xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                      paste(mut_gene, "(+), Unadjusted"),
                      paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                      paste(mut_gene, "(+), Adjusted for Delayed Entry"))
    ) +   
    labs(title = paste(Cancer, " ", sum(traditional_fit[[1]]), " patients ",
                "Median OS ",
               as.integer(summary(traditional_fit)[[18]][[13]])," (",
               as.integer(summary(traditional_fit)[[18]][[15]]),"-",
               as.integer(summary(traditional_fit)[[18]][[17]]),") (mut(-), unadj.)/",
               as.integer(summary(traditional_fit)[[18]][[14]]), " (",
               as.integer(summary(traditional_fit)[[18]][[16]]),"-",
               as.integer(summary(traditional_fit)[[18]][[18]]),") (mut(+), unadj.)/\n",
               as.integer(median_os_summary_weibull_total[[2]]), 
               " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
               as.integer(median_os_summary_weibull_total[[3]]), ") (mut(-), adj.)/",
               as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
               as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
               as.integer(median_os_summary_weibull_total_exp[[3]]),
               ") (mut(+), adj.) days",
               sep=""),
      subtitle = paste("Survival difference with gene mutation: ",
            as.integer(median_os_summary_weibull_bias[[2]]), " (",
            as.integer(median_os_summary_weibull_bias[[1]]), "-", 
            as.integer(median_os_summary_weibull_bias[[3]]), 
            ") days, median (95% CI)",
            sep=""))))
    }
  }
  file.remove(paste(target_pathway[i], ".tmp", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_target_pathway_diagnosis_to_death.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

Data_summary = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) 

table_tmp = Data_summary %>%
  tbl_summary(
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_n() %>%
  modify_caption("**Table X. Patients with selected disease, diagnosis-death, Characteristics**") %>%
  bold_labels()
table_tmp

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_option,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment option exsisted**") %>%
  modify_caption("**Table 9. Diagnosis to final observation - Expert panel option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 10. Enrollment to final observation - Recommended treatment done**") %>%
  bold_labels()
}

Data_summary_treat = Data_survival %>% dplyr::select(
  症例.基本情報.性別.名称.,
  症例.基本情報.年齢,
  sampling_age,
  症例.背景情報.喫煙歴有無.名称.,
  症例.背景情報.喫煙年数,
  症例.背景情報.喫煙１日本数,
  症例.背景情報.アルコール多飲有無.名称.,
  症例.背景情報.ECOG.PS,
  症例.背景情報.重複がん有無.異なる臓器..名称.,
  症例.背景情報.多発がん有無.同一臓器..名称.,
  症例.検体情報.パネル.名称.,
  EP_option,
  EP_treat,
  enrollment,
  time_palliative_enroll,
  time_enroll_final,
  time_palliative_final,
  diagnosis) %>%
  dplyr::filter(EP_option == 1)

survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                          data=Data_survival)
survival_simple_ = survfit(Surv(time_enroll_final, censor)~EP_treat,
                          data=Data_survival)
if(survival_simple_[[1]][[1]] != survival_simple[[1]][[1]] & length(survival_simple_) == length(survival_simple)){
Data_summary_treat %>%
  tbl_summary(
    by = EP_treat,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 2,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_spanning_header(c("stat_1", "stat_2") ~
                           "**Treatment done**") %>%
  modify_caption("**Table 2''. Diagnosis to final observation - Recommended treatment done, only patients with treatment option**") %>%
  bold_labels()
}

survival_simple = survfit(Surv(time_enroll_final, censor)~diagnosis,
                          data=Data_survival)
if(length(Data_survival[,1]) != survival_simple[[1]][[1]]){
Data_summary %>%
  tbl_summary(
    by = diagnosis,
    statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                     "{mean} ({sd})",
                                     "{median} ({p25}, {p75})", 
                                     "{min}, {max}"),
                     all_categorical() ~ "{n} / {N} ({p}%)"),
    type = all_continuous() ~ "continuous2",
    digits = all_continuous() ~ 3,
    missing = "no") %>%
  add_p(pvalue_fun = ~style_pvalue(.x, digits = 3)) %>%
  add_overall() %>%
  add_n() %>%
  modify_header(label ~ "**Variable**") %>%
  modify_caption("**Table 11. Diagnosis to final observation - diagnosis**") %>%
  bold_labels()
}

Report_lolliplot_diagnosis = Report_lolliplot %>%
  dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_survival$C.CAT調査結果.基本項目.ハッシュID)
write.csv(Report_lolliplot_diagnosis, file = paste(OutDirName, Cancer, "_only_pts_with_diagnosis_to_death_info_", target_gene[[1]][[1]], ".csv", sep=""))

Data_major_gene = Data_report %>%
  dplyr::filter(C.CAT調査結果.エビデンス.エビデンスレベル == "F")
oncogenic_genes = Data_major_gene %>%
  dplyr::select(C.CAT調査結果.基本項目.ハッシュID, C.CAT調査結果.変異情報.マーカー) %>%
  dplyr::distinct() %>%
                dplyr::count(C.CAT調査結果.変異情報.マーカー) %>%
                dplyr::arrange(-n)
c_name_gene= c("gene_mutation", "All patients", "Selected disease with target gene mut", "Selected disease without target gene mut", "Other diseases")
tmp_no = oncogenic_genes
Data_case_tmp = Data_case %>%
  dplyr::mutate(diagnosis = case_when(
    C.CAT調査結果.基本項目.ハッシュID %in% Data_case_target$C.CAT調査結果.基本項目.ハッシュID &
    C.CAT調査結果.基本項目.ハッシュID %in% (Data_major_gene %>%
          dplyr::filter(C.CAT調査結果.変異情報.マーカー ==
            target_gene[[1]][[1]]))$C.CAT調査結果.基本項目.ハッシュID ~ "Selected disease with target gene mutation",
    C.CAT調査結果.基本項目.ハッシュID %in% Data_case_target$C.CAT調査結果.基本項目.ハッシュID ~ "Selected disease without target gene mutation",
    TRUE ~ "Other diseases"
  ))
disease_list = c("Selected disease with target gene mutation", "Selected disease without target gene mutation", "Other diseases")
if(length(unique(Data_case_tmp$diagnosis)) == 3){
  for(i in 1:3){
    tmp_no[,2] = 0
    ID_target_tmp = unique((Data_case_tmp %>% dplyr::filter(diagnosis == disease_list[[i]]))
                           $C.CAT調査結果.基本項目.ハッシュID) 
    tmp = Data_major_gene %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, C.CAT調査結果.変異情報.マーカー) %>%
        dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_target_tmp) %>%
        dplyr::distinct() %>%
                  dplyr::count(C.CAT調査結果.変異情報.マーカー)
    tmp_no$n = lapply(tmp_no[,1],
                                 function(k) {
      as.vector(tmp$n[match(k,
                tmp$C.CAT調査結果.変異情報.マーカー)])
    })
    tmp_no[is.na(tmp_no)] <- 0
    oncogenic_genes = cbind(oncogenic_genes, unlist(tmp_no$n))
  }
  colnames(oncogenic_genes) =c_name_gene
  
  DT::datatable(oncogenic_genes, filter = 'top', extensions = 'Scroller',
               caption = "Genes with oncogenic mutations in entire cohorts (patients)")
}

```

```{r Drug_duration , eval = (Drug_analysis == 1), echo = (Source_code == 1 & Drug_analysis == 1), warning=FALSE}

# Only 1st gene_pattern will be analyzed
target_gene_names = unlist(target_gene)[seq(1,length(unlist(target_gene)),2)]
ID_All = unique(Data_report$C.CAT調査結果.基本項目.ハッシュID)
g = NULL 
for(pattern in 1:length(gene_pattern)){
  ID_mut = NULL
  for(i in 1:length(gene_pattern[[pattern]])){
    ID_tmp = ID_All
    for(j in 1:length(target_gene)){
      if(bitwAnd(2^(j-1), gene_pattern[[pattern]][i]) > 0){
        if(target_gene[[j]][2] == "WT"){
          Data_tmp = Data_report %>% dplyr::filter(
            C.CAT調査結果.変異情報.マーカー == target_gene[[j]][1] &
            C.CAT調査結果.エビデンス.エビデンスレベル_WT == "WT")
          ID_tmp2 = Data_tmp$C.CAT調査結果.基本項目.ハッシュID
          Data_tmp = Data_report %>% dplyr::filter(
          C.CAT調査結果.変異情報.マーカー == target_gene[[j]][1])
          ID_tmp3 = setdiff(ID_All, Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
          ID_tmp = intersect(ID_tmp, union(ID_tmp2, ID_tmp3))
        } else{
          Data_tmp = Data_report %>% dplyr::filter(
            C.CAT調査結果.変異情報.マーカー == target_gene[[j]][1] &
              (C.CAT調査結果.エビデンス.エビデンスレベル_WT == target_gene[[j]][2] |
               C.CAT調査結果.変異情報.変異内容 == target_gene[[j]][2]))
          ID_tmp = intersect(ID_tmp, Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
        }
      }
    }
    ID_mut = union(ID_mut, ID_tmp)
  }
  
  mut_gene = ""
  for(l in 1:length(gene_pattern[[pattern]])){
    for(m in 1:length(target_gene_names)){
      if(bitwAnd(2^(m-1), gene_pattern[[pattern]][[l]]) > 0){
        mut_gene = paste(mut_gene,
                     target_gene[[m]][1], " ", target_gene[[m]][2],
                     " and ", sep="")
      }
    }
    mut_gene = substr(mut_gene, 1, nchar(mut_gene)-5)
    mut_gene = paste(mut_gene, " or ", sep="")
  }
  mut_gene = substr(mut_gene, 1, nchar(mut_gene)-4)
  
  tmp1 = (Data_case_target %>%
      dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% ID_mut) %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                    症例.基本情報.性別.名称.,
                    症例.基本情報.年齢,
                    症例.背景情報.喫煙年数,
                    症例.背景情報.喫煙１日本数,
                    症例.背景情報.アルコール多飲有無.名称.,
                    症例.背景情報.ECOG.PS,
                    症例.管理情報.登録日,
                    症例.がん種情報.食道.胃.腸.HER2.名称.,
                    症例.転帰情報.転帰.名称.,
                    症例.転帰情報.最終生存確認日,
                    症例.転帰情報.死亡日,
                    diagnosis) %>%
      dplyr::distinct()
    )
  tmp1$target_gene_mutation = "+"
  tmp2 = (Data_case_target %>%
      dplyr::filter(!C.CAT調査結果.基本項目.ハッシュID %in% ID_mut) %>%
      dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                    症例.基本情報.性別.名称.,
                    症例.基本情報.年齢,
                    症例.背景情報.喫煙年数,
                    症例.背景情報.喫煙１日本数,
                    症例.背景情報.アルコール多飲有無.名称.,
                    症例.背景情報.ECOG.PS,
                    症例.管理情報.登録日,
                    症例.がん種情報.食道.胃.腸.HER2.名称.,
                    症例.転帰情報.転帰.名称.,
                    症例.転帰情報.最終生存確認日,
                    症例.転帰情報.死亡日,
                    diagnosis) %>%
      dplyr::distinct()
    )
  tmp2$target_gene_mutation = "-"
  Data_mut = rbind(tmp1, tmp2)
  
  Data_mut = Data_mut %>%
    dplyr::mutate(final_observe = case_when(
      症例.転帰情報.最終生存確認日 == "          " ~ 症例.転帰情報.死亡日,
      症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
      症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
      TRUE ~ ""))
  Data_mut = Data_mut %>%
    dplyr::mutate(症例.管理情報.登録日 =
                      as.Date(Data_mut$症例.管理情報.登録日)) %>%
    dplyr::mutate(症例.転帰情報.死亡日 =
                      as.Date(Data_mut$症例.転帰情報.死亡日)) %>%
    dplyr::mutate(症例.転帰情報.最終生存確認日 =
                      as.Date(Data_mut$症例.転帰情報.最終生存確認日)) %>%
    dplyr::mutate(final_observe = as.Date(Data_mut$final_observe))
  
  tmp1 = Data_case_target %>%
    dplyr::filter(症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == selected_drug[[1]]) %>%
    dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                    c("その他", "緩和")) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                    症例.EP前レジメン情報.化学療法レジメン名称,
                    症例.EP前レジメン情報.投与開始日,
                    症例.EP前レジメン情報.投与終了日,
                    症例.EP前レジメン情報.最良総合効果.名称.) %>%
    dplyr::distinct()
  tmp1 = tmp1 %>%
    dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                      as.Date(tmp1$症例.EP前レジメン情報.投与開始日)) %>%
    dplyr::mutate(症例.EP前レジメン情報.投与終了日 =
                      as.Date(tmp1$症例.EP前レジメン情報.投与終了日))
  tmp1 = tmp1 %>%
    dplyr::mutate(Drug_length = 
                    as.integer(difftime(tmp1$症例.EP前レジメン情報.投与終了日,
                                        tmp1$症例.EP前レジメン情報.投与開始日,
                                        units = "days"))) %>%
    dplyr::arrange(症例.EP前レジメン情報.投与終了日) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.EP前レジメン情報.化学療法レジメン名称,
                  症例.EP前レジメン情報.最良総合効果.名称.,
                  Drug_length)
  
  colnames(tmp1) = c("C.CAT調査結果.基本項目.ハッシュID",
                     "レジメン",
                     "最良総合効果",
                     "Drug_length")
  tmp2 = Data_case_target %>%
    dplyr::filter(症例.EP後レジメン情報.薬剤名.YJ一般名.EN. == selected_drug[[1]]) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                    症例.EP後レジメン情報.化学療法レジメン名称,
                    症例.EP後レジメン情報.投与開始日,
                    症例.EP後レジメン情報.投与終了日,
                    症例.EP後レジメン情報.最良総合効果.名称.) %>%
    dplyr::distinct()
  tmp2 = tmp2 %>%
    dplyr::mutate(症例.EP後レジメン情報.投与開始日 =
                      as.Date(tmp2$症例.EP後レジメン情報.投与開始日)) %>%
    dplyr::mutate(症例.EP後レジメン情報.投与終了日 =
                      as.Date(tmp2$症例.EP後レジメン情報.投与終了日))
  tmp2 = tmp2 %>%
    dplyr::mutate(Drug_length = 
                    as.integer(difftime(tmp2$症例.EP後レジメン情報.投与終了日,
                                        tmp2$症例.EP後レジメン情報.投与開始日,
                                        units = "days"))) %>%
    dplyr::arrange(症例.EP後レジメン情報.投与開始日) %>%
    dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                  症例.EP後レジメン情報.化学療法レジメン名称,
                  症例.EP後レジメン情報.最良総合効果.名称.,
                  Drug_length)
  colnames(tmp2) = c("C.CAT調査結果.基本項目.ハッシュID",
                     "レジメン",
                     "最良総合効果",
                     "Drug_length")
  Data_drug = rbind(tmp1, tmp2)
  
  Data_drug = Data_drug %>%
    tidyr::replace_na(replace = list(Drug_length = 0))
  Data_drug_unique = Data_drug %>%
    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
  Summary_drug = left_join(Data_mut, Data_drug_unique, by="C.CAT調査結果.基本項目.ハッシュID")
  
  Summary_drug = Summary_drug %>%
    dplyr::mutate(censor = case_when(
      症例.転帰情報.転帰.名称. == "死亡" ~ 1,
      TRUE ~ 0
    )) %>%
    dplyr::mutate(
      OS_after_CGP = as.integer(difftime(Summary_drug$final_observe,
                                        Summary_drug$症例.管理情報.登録日,
                                        units = "days"))
    )
    if(SD_is_effective == 2){
      Summary_drug = Summary_drug %>%
        dplyr::mutate(effect = case_when(
          最良総合効果 == "PR" ~ 1,
          最良総合効果 == "CR" ~ 1,
          最良総合効果 == "SD" & Drug_length > LongSD_days ~ 1,
          最良総合効果 == "SD" & Drug_length <= LongSD_days ~ 0,
          最良総合効果 == "NE" ~ 0,
          最良総合効果 == "PD" ~ 0,
          TRUE ~ -1
        ))
    } else if(SD_is_effective == 1){
      Summary_drug = Summary_drug %>%
        dplyr::mutate(effect = case_when(
          最良総合効果 == "PR" ~ 1,
          最良総合効果 == "CR" ~ 1,
          最良総合効果 == "SD" & Drug_length > LongSD_days ~ 1,
          最良総合効果 == "SD" & Drug_length <= LongSD_days ~ 1,
          最良総合効果 == "NE" ~ 0,
          最良総合効果 == "PD" ~ 0,
          TRUE ~ -1
        ))
    } else{
      Summary_drug = Summary_drug %>%
        dplyr::mutate(effect = case_when(
          最良総合効果 == "PR" ~ 1,
          最良総合効果 == "CR" ~ 1,
          最良総合効果 == "SD" & Drug_length > LongSD_days ~ 0,
          最良総合効果 == "SD" & Drug_length <= LongSD_days ~ 0,
          最良総合効果 == "NE" ~ 0,
          最良総合効果 == "PD" ~ 0,
          TRUE ~ -1
        ))
    }
  
  WriteXLS(Summary_drug, ExcelFileName=
        paste(OutDirName, Cancer, "summary_", mut_gene, "_", selected_drug[[1]], ".xls", sep=""))
  Summary_drug$all_one = 1
 
  Summary_drug_duration = Summary_drug %>%
    tidyr::drop_na(Drug_length)
  Summary_drug_duration = Summary_drug_duration %>%
    dplyr::filter(Drug_length > 0)
  
  survival_simple = survfit(Surv(Drug_length, all_one) ~ target_gene_mutation,
                            data=Summary_drug_duration)
  #print(survival_simple)
  diff_0 = survdiff(Surv(Drug_length, all_one)~target_gene_mutation,
                    data=Summary_drug_duration, rho=0)
  diff_1 = survdiff(Surv(Drug_length, all_one)~target_gene_mutation,
                    data=Summary_drug_duration, rho=1)
  g=c(g, list(ggsurvplot(
    fit = survival_simple,
    combine = TRUE,
    data = Summary_drug_duration,
    xlab = paste("Time from", selected_drug[[1]], "start date (days)"),
    ylab = "Survival Probability",
    censor = TRUE,
    pval = FALSE,
    conf.int = FALSE,
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    tables.theme = clean_theme(),
    legend = c(0.7,0.9),
    xlim = c(0, max(Summary_drug_duration$Drug_length) * 1.05),
    break.x.by = ceiling(max(Summary_drug_duration$Drug_length) / 500) * 100
  ) + labs(title = paste(Cancer, ",",selected_drug[[1]],
                         "Drug time-to-treatment-failure,", mut_gene),
        subtitle = paste("Median OS(95%CI), mutation(-):", 
                         summary(survival_simple)[[18]][[13]], "(",
                         summary(survival_simple)[[18]][[15]], "-",
                         summary(survival_simple)[[18]][[17]], "), mutation(+):",
                         summary(survival_simple)[[18]][[14]], "(",
                         summary(survival_simple)[[18]][[16]], "-",
                         summary(survival_simple)[[18]][[18]], ")\n",
             "log-rank test, p-value:", format(
         1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),
         digits=3),
         ", Gehan-Wilcoxon test, p-value:", format(
         1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),
         digits=3), sep=""))))

  if(SD_is_effective == 2){
    title_name = "CR_PR_long_SD"
  } else if(SD_is_effective == 1){
    title_name = "CR_PR_SD"
  } else if(SD_is_effective == 0){
    title_name = "CR_PR"
  }
  
  Summary_drug_duration = Summary_drug %>%
    dplyr::filter(effect != -1)
  ans = fisher.test(matrix(c(
    length((Summary_drug_duration %>% dplyr::filter(target_gene_mutation == "+" & effect == 1))$effect),
    length((Summary_drug_duration %>% dplyr::filter(target_gene_mutation == "-" & effect == 1))$effect),
    length((Summary_drug_duration %>% dplyr::filter(target_gene_mutation == "+" & effect == 0))$effect),
    length((Summary_drug_duration %>% dplyr::filter(target_gene_mutation == "-" & effect == 0))$effect)), byrow=TRUE, ncol=2))


  ans <- glm(Summary_drug_duration$effect ~ target_gene_mutation,
             data=Summary_drug_duration, family=binomial)
  s.ans = summary(ans)
  coe <- s.ans$coefficient
  RR <- exp(coe[,1])
  RRlow <- exp(coe[,1]-1.96*coe[,2])
  RRup <- exp(coe[,1]+1.96*coe[,2])
  N <- nrow(data)
  aic <- AIC(ans)
  result <- cbind(coe,RR,RRlow,RRup,aic,N)
  colnames(result)[6:7] <- c("RR95%CI.low","RR95%CI.up")
  write.csv(result, file = paste(OutDirName, Cancer, "_", selected_drug[[1]],
    "_effective_indicates_", title_name, "_", mut_gene,
    "_logistic_regression.csv", sep=""))
}
#g
pdf(file = paste(OutDirName, Cancer, "_drug_time_to_treatment_failure.pdf", sep=""), width = 10, height = 10)
print(g, newpage = TRUE)
dev.off()

```

```{r PGPV_kosugi, eval = (PGPV_analysis == 1), echo = (Source_code == 1 & PGPV_analysis == 1), warning=FALSE}
PGPV_table = data.frame(readxl::read_xls(paste(Working, "/source/PGPV_kosugi.xls", sep="")))
rownames(PGPV_table) = PGPV_table$Gene
PGPV_table[,"Total_mutations"] = 0
PGPV_table[,"Validated_PGV"] = 0
PGPV_table[,"PGPV"] = 0
PGPV_table[,"PGPV_and_PGV"] = 0
List_PGPV = list.files(paste(Working, "/source/PGPV", sep=""))
for(i in List_PGPV){
  Organ =  str_split(i, "_")[[1]][[1]]
  PGPV_table[,Organ] = ""
  table_tmp = readxl::read_xls(paste(Working, "/source/PGPV/", i, sep=""))
  for(j in 1:length(table_tmp[,1][[1]])){
    if(any(PGPV_table[,1] == table_tmp[j,1][[1]])){
      PGPV_table[table_tmp[j,1][[1]],"Total_mutations"] =
        PGPV_table[table_tmp[j,1][[1]],"Total_mutations"] +
        table_tmp[j,]$`mutation counts`
      PGPV_table[table_tmp[j,1][[1]],"Validated_PGV"] = 
        PGPV_table[table_tmp[j,1][[1]],"Validated_PGV"] +
        table_tmp[j,]$`NCC validated PGV`
      PGPV_table[table_tmp[j,1][[1]],"PGPV"] = 
        PGPV_table[table_tmp[j,1][[1]],"PGPV"] +
        table_tmp[j,]$`NCC PGPV`
      PGPV_table[table_tmp[j,1][[1]],"PGPV_and_PGV"] = 
        PGPV_table[table_tmp[j,1][[1]],"PGPV_and_PGV"] +
        table_tmp[j,]$`NCC PGPV & PGV`
      PGPV_table[table_tmp[j,1][[1]],Organ] =
        paste(table_tmp[j,]$`mutation counts`,
              table_tmp[j,]$`NCC validated PGV`,
              table_tmp[j,]$`NCC PGPV`,
              table_tmp[j,]$`NCC PGPV & PGV`,
              sep="/")
    } else{
      PGPV_table[table_tmp[j,1][[1]],] = 0
      PGPV_table[table_tmp[j,1][[1]],1] = table_tmp[j,1][[1]]
      PGPV_table[table_tmp[j,1][[1]],"Total_mutations"] =
        PGPV_table[table_tmp[j,1][[1]],"Total_mutations"] +
        table_tmp[j,]$`mutation counts`
      PGPV_table[table_tmp[j,1][[1]],"Validated_PGV"] = 
        PGPV_table[table_tmp[j,1][[1]],"Validated_PGV"] +
        table_tmp[j,]$`NCC validated PGV`
      PGPV_table[table_tmp[j,1][[1]],"PGPV"] = 
        PGPV_table[table_tmp[j,1][[1]],"PGPV"] +
        table_tmp[j,]$`NCC PGPV`
      PGPV_table[table_tmp[j,1][[1]],"PGPV_and_PGV"] = 
        PGPV_table[table_tmp[j,1][[1]],"PGPV_and_PGV"] +
        table_tmp[j,]$`NCC PGPV & PGV`
      PGPV_table[table_tmp[j,1][[1]],Organ] =
        paste(table_tmp[j,]$`mutation counts`,
              table_tmp[j,]$`NCC validated PGV`,
              table_tmp[j,]$`NCC PGPV`,
              table_tmp[j,]$`NCC PGPV & PGV`,
              sep="/")
    }
  }
}
if(!file.exists(paste(Working, "/output/PGPV_", Sys.Date(), sep=""))){
  dir.create(paste(Working, "/output/PGPV_", Sys.Date(), sep=""))
}
write.xlsx(x = PGPV_table, file = paste(Working, "/output/PGPV_", Sys.Date(), "/PGPV.xlsx", sep=""))
```
